--- 
title: "Elements of Statistical Modeling for Experimental Biology"
author: "Copyright 2018 Jeffrey A. Walker"
date: "Draft: `r Sys.Date()`"
site: bookdown::bookdown_site
output: bookdown::gitbook
documentclass: book
bibliography: [book.bib, packages.bib]
biblio-style: apalike
link-citations: yes
github-repo: 
description: "A first course in statistical modeling for experimental biology researchers"
---

```{r bookdown, include=FALSE}
# automatically create a bib database for R packages
knitr::write_bib(c(
  .packages(), 'bookdown', 'knitr', 'rmarkdown'
), 'packages.bib')

library(DiagrammeR)
library(here)

```

# Preface {-}
*More cynically, one could also well ask "Why has medicine not adopted frequentist inference, even though everyone presents P-values and hypothesis tests?" My answer is: Because frequentist inference, like Bayesian inference, is not taught. Instead everyone gets taught a misleading pseudo-frequentism: a set of rituals and misinterpretations caricaturing frequentist inference, leading to all kinds of misunderstandings.* -- Sander Greenland

We use statistics to learn from data with uncertainty. Traditional introductory textbooks in biostatistics implicitly or explicitly train students and researchers to "discover by p-value" using hypothesis tests (Chapter \@ref(p-values)). Over the course of many chapters, the student learns to use a look-up table or flowchart to choose the correct "test" for the data at hand, compute a test statistic for their data, compute a *p*-value based on the test statistic, and compare the *p*-value to 0.05. Textbooks typically give very little guidance about what can be concluded if $p < 0.05$ or if $p > 0.05$, but many researchers conclude, incorrectly, they have "discovered" an effect if $p < 0.05$ but found "no effect" if $p > 0.05$.

This book is an introduction to the statistical analysis of data from biological experiments with a focus on the estimation of treatment effects and measures of the uncertainty of theses estimates. Instead of a flowchart of "which statistical test", this book emphasizes a **regression modeling** approach using linear models and extensions of linear models.

"What what? In my previous class I learned that regression was for data with a continuous independent variable and that *t*-tests and ANOVA were for data with categorical independent variables." No! This misconception has roots in the history of regression vs. ANOVA and is reinforced by how introductory biostatistics textbooks, and their instructors, *choose* to teach statistics. In this class, you were probably taught to follow a flowchart strategy -- something like

```{r flowchart, echo=FALSE, fig.cap = "A small flow chart demo of the flowchart strategy of statistical analysis. This chart covers a very small subset of potential paths that could be built from an introductory biostatistics textbook."}

generate_it <- FALSE
if(generate_it == TRUE){
  mermaid("
graph TB
    a[I want to compare group means] --> b{How many groups?}
    b --> c((only 2))
    b --> d((more then 2))
    c --> e{are my data normal?}
    e --> f((yes))
    e --> g((no))
    f --> h{are samples in group 1<br>independent of those in<br>group 2?}
    h --> i((yes))
    h --> j((no))
    i --> k[independent sample t-test]
    j --> l[paired t-test]

")
}else{
  image_path <- here::here("images","flowchart.png")
  knitr::include_graphics(image_path)
}

```

Compared to the flowchart stratgy, the advantages of the regression modeling strategy include

1. A unified aproach in place of a collection of seemingly unrelated tests. The unified approach is the *regression model*.
    
It has long been appreciated that classical regression, *t*-tests, ANOVA, and other methods are all variations of the equation for a line $Y = mX + b$ using slightly different notation 

\begin{equation}
Y = \beta_0 + \beta_1 X + \varepsilon
\end{equation}

Chapter 1 explains the meaning of this notation but the point to make here is that because all regression models are variations of this equation, a modeling strategy of learning or doing statistics is more coherent than a flowchart strategy. Generalizations of this basic equation include general linear models, generalized least squares models, linear mixed models, generalized linear models, generalized additive models, causal graphical models, multivariate models, and machine learning. This book is not a comprehensive source for any of these methods but an introduction to the common foundations of all these methods.

2. Estimates of effects and uncertainty are, ultimately, *far* more useful than *p*-values. For example, to build useful models on the effects of an increasingly acidified ocean on coral growth, we want to estimate the *direction* and *magnitude* of the effects at different levels of acidification and how these estimates change under different conditions. We can compare the magnitude to a prediction of the magnitude from a mechanistic model of growth. We can use a magnitude and uncertainty to make predictions about the future of coral reefs, under different scenarios of ocean acidification. We can use the estimated effects and uncertainty to model the consequences of the effects of acidification on coral growth on fish production or carbon cycling.

By contrast, researchers learn little from a hypothesis test -- that is, comparing *p* to 0.05. A *p*-value is a measure of compatibility between the data and the null hypothesis and, consequently, a pretty good -- but imperfect -- tool to dampen the frequency that we are fooled by randomness. Importantly, a *p*-value **is not a measure of the size of an effect**. At most, a small *p*-value gives a researcher some confidence in the existence and direction of an effect. But if we are investigating the effects of acidified ocean water on coral growth, it would be absurd to conclude from a *p*-value that pH does or does not affect growth. pH affects *everything* about cell biology.

*p*-values are neither necessary nor sufficient for good data analysis. Properly understood, a *p*-value is a useful tool in the data analysis toolkit. As stated above, the proper use of *p*-values dampens the frequency that we are fooled by randomness. Importantly, the estimation of effects and uncertainty and the computation of a *p*-value are not alternatives. Indeed, the *p*-value returned by many hypothesis tests are computed from the regression model used to estimate the effects. Throughout this text, statistical models are used to compute a *p*-value *in addition to* the estimates of effects and their uncertainty.

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**NHST Blues** -- The "discovery by p-value" strategy, or Null-Hypothesis Significance Testing (NHST), has been criticized by statisticians for many, many decades. Nevertheless, introductory biostatistics textbooks written by both biologists and statisticians continue to organize textbooks around a collection of hypothesis tests, with much less emphasis on estimation and uncertainty. The NHST strategy of learning or doing statistics is easy in that it requires little understanding of the statistical model underneath the tests and its assumptions, limitations, and behavior. The NHST strategy in combination with point-and-click software enables "mindless statistics"^[Gegenrezer] and encourages the belief that statistics is a tool like a word processor is a tool, afterall, a rigorous analysis of one's data requires little more than getting p-values and creating bar plots. Indeed, many PhD programs in the biosciences require no statistics coursework and the only training available to students is from the other graduate students and postdocs in the lab. As a consequence, the biological sciences literature is filled with error bars that imply data with negative values and p-values that have little relationship to the probability of the data under the null. More importantly for science, the reported statistics are often not doing for the study what the researchers and journal editors think they are doing.
</div> 

## Math

## R and programming

```{r, echo=FALSE}
# uses immersion in R language instead of building up.
```




<!--chapter:end:index.Rmd-->

# Part I: Getting Started {-}

<!--chapter:end:chapters/01-part_1-getting_started.Rmd-->

# Getting Started -- R Projects and R Markdown

A typical statistical modeling project will consist of:

1. importing data from Excel or text (.csv or .txt) files
2. cleaning data
3. initial exploratory plots
4. analysis
5. model checking
6. generating plots
7. generating tables
8. writing text to describe the project, the methods, the analysis, and the interpretation of the results (plots and tables)

The best practice for reproducible research is to use as few software tools for these steps as possible. Too many research projects are not reproducible because the data were cleaned in Excel, and then different parts of the data were separately imported into a GUI statistics software for analysis, and then output from the statistics software was transcribed to Excel to make a table. And other parts of the analysis are used to create a plot in some plotting software. And then the tables and plots are pasted into Microsoft Word to create a report. Any change at any step in this process will require the researcher to remember all the downstream parts that are dependent on the change and to re-do an analysis, or a table, or a plot. Or a team member on the project making asks about how a particular variable was transformed and 

R studio encourages best practices by creating a **project folder** that contains all project documents and implementing a version of markdown called R Markdown. An R Markdown document can explicitly link all parts of the workflow so that changes in earlier steps automatically flow into the later steps. At the completion of a project, a researcher can choose "run all" from the menu and the data are read, cleaned, analyzed, plotted, tabled, and put into a report with the text.

## R vs R Studio

R is a programming language. It runs under the hood. You never see it. To use R, you need another piece of software that provides a **user interface**. The software we will use for this is R Studio. R Studio is a slick (very slick) **graphical user interface** (GUI) for developing R projects.

## Download and install R and R studio

[Download R for your OS](https://cran.r-project.org){target="_blank"}

[Download R Studio Desktop](https://www.rstudio.com/products/rstudio/download/){target="_blank"}

If you need help installing R and R studio, here is [Andy Field's Installing R and RStudio video tutorial](https://www.youtube.com/watch?v=ZvPFKfNHBNQ){target="_blank"})

## Open R Studio and modify the workspace preference

1. Open R Studio
2. Click on R Studio > Preferences to
3. Click on General in the left menu
4. diable "Restore .RData into workspace at startup"
5. Click on the "Save workspace to .RData on exit" popup menu and choose "Never"

```{r workspace-preference, echo = FALSE}
image_path <- here::here("images", "workspace.png")
knitr::include_graphics(image_path)
```

What's going on here? The workspace contains the values of all the objects created by the R code that you've run in the working R session. Nothing good comes from this. You want to start each R session with a clean slate, a blank workspace. This means that when you start a new R session, you will need to re-run all your code chunks to start where you left-off at the close of your last R session. This seems tedious but, be warned, bad things will happen if you save the workspace from the last session and re-load this at startup. Trust me. Just don't do it.

## If you didn't modify the workspace preferences from the previous section, go back and do it

## R Markdown in a nutshell

In this text, we will write code to analyze data using R Markdown. R markdown is a version of Markdown. Markdown is tool for creating a document containing text (like microsoft Word), images, tables, and code that can be output, or **knitted**, to the three modern output formats: html (web pages), pdf (reports and documents), and microsoft word (okay, this isn't modern but it is widely used).

The R Markdown, or .Rmd, document contains three components: 1) a YAML header, which specifies formatting and styles for the knitted document, the code "chunks", which are blocks of code that do something, and the space before and after the code chunks which contains text and any output images or tables from the code chunks.

## Install R Markdown

[Directions for installing R Markdown](https://bookdown.org/yihui/rmarkdown/installation.html#installation){target="_blank"}

R Markdown can output pdf files. The mechanism for this is to first create a LaTeX ("la-tek") file. LaTeX is an amazing tool for creating professional pdf documents. You do not need PDF output for this text, but I encourage you to download and install the tinytex distribution, which was created especially for R Markdown in R Studio.

[The tinytex distribution is here](https://yihui.org/tinytex/).

## Importing Packages
The R scripts you write will include functions in packages that are not included in Base R. These packages need to be downloaded from an internet server to your computer. You only need to do this once (although you have to redo it each time you update R). But, each time you start a new R session, you  will need to load a package using the `library()` function. Now is a good time to import packages that we will use

Open R Studio and choose the menu item "Tools" > "Install Packages". In the "packages" input box, insert the names of packages to install the package. The names can be separated by spaces or commas, for example "data.table, emmeans, ggplot2". Make sure that "install dependencies" is clicked before you click "Install". Packages that we will use in this book are

1. Import and wrangling packages
* devtools -- we use this to install packages that are not on CRAN
* here -- we use to read from and write to the correct folder
* janitor -- we use the function clean_names from this package
* readxl -- elegant importing from microsoft Excel spreadsheets
* data.table - improves functionality of data frames
2. analysis packages
* emmeans -- we use this to compute modeled means and contrasts
* nlme -- we use this for gls models
* lme4 -- we use this for linear mixed models
* lmerTest -- we use this for inference with linear mixed models
* glmmTMB -- we use this for generalized linear models
* MASS -- we will use glm.nb from this package
* afex -- we use this for classic ANOVA
* car -- we use this for model checking and for classic ANOVA
* DHARMa -- we use this for model checking generlaized linear models
* insight -- we use this to learn about models
3. graphing packages
* ggplot2 -- we use this for plotting
* ggsci -- we use this for the color palettes
* ggthemes -- we use this for the colorblind palette
* ggpubr -- we use this to make ggplots a bit easier
* ggforce -- we use this for improved jitter plots
* dabestr -- we use this to make several plot types
* cowplot -- we use this to combine plots

Once these are installed, you don't need to do this again although there will be additional packages that you might install. You simply need to use the `library()` function at the start of a markdown script.

## Create an R Studio Project for this textbook

1. Create a project folder within the Documents folder (Mac OS) or My Documents folder (Windows OS). All files associated with this book will reside inside this folder. The name of the project folder should be something meaningful, such as "Applied Biostatistics" or the name of your class (for students in my Applied Biostatistics class, this folder could be named "BIO_413"). 
2. Within the project folder, create new folders named
    1. "Rmd" -- this is where your R markdown files are stored 
    2. "R" -- this is where additional R script files are stored
    3. "data" -- this is where data that we download from public archives are stored
    4. "output" -- this is where you will store fake data generated in this class
    5. "images" -- this is where image files are stored
3. Open R Studio and click the menu item File > New Project...
4. Choose "Existing Directory" and navigate to your project folder
5. Choose "Create Project"
6. Check that a ".Rproj" file is in your project folder
7. Download and move the file [ggplot_the_model.R](https://raw.githubusercontent.com/middleprofessor/applied-biostats/master/R/ggplot_the_model.R) into the R folder.

```{r project-organization-image, echo = FALSE, fig.cap = "Project folder with the .Rproj file and all main folders located at the first level of the project"}
image_path <- here::here("images", "project_organization.png")
knitr::include_graphics(image_path)
```
The project directory should look like that in Figure \@ref(fig:project-organization-image). Importantly, the project file ("Applied Biostatistics.Rproj") and the main folders are all located at the first level within the project folder.

**Bug alert** If your .Rproj file is somewhere else (on the desktop, in the data folder, etc.), bad things will happen.

## Working on a project, in a nutshell

1. Wake up, brush teeth, and open the project by double-clicking the .Rproj icon. Alternatively, open R Studio and then use the File > Open Project to open the project. **The name of the project will at the top-right of the R Studio window.** We *always* want to work within an open project and the first workflow guarantees this. If we Open R Studio and then open a .Rmd file, we could be working within another project or no project at all. Bad things will happen.
2. Run previous code chunks, in order (top to bottom). Write new code in code chunks and run. When we run code, we add R objects to the **workspace**. The workspace contains the values of all the objects created by the R code that has been run in the working session. When I save the .Rmd file, these values are not saved, only the text and code chunks in the R Markdown document. This is a feature, not a bug.
3. When we are finished with the session, quit R Studio. If you get a popup window asking if you want to save the workspace, click "No". Then immediately go back to the section "Open R Studio and modify the workspace preference" above and follow the directions.

## Create an R Markdown file for this Chapter

1. The top-left icon in R Studio is a little plus sign within a green circle. Click this and choose "R Markdown" from the pull-down menu.
2. Give the file a meaningful title like "Chapter 1 -- Organization"
3. Delete all text below the first code chunk, starting with the header "## R Markdown"

### Modify the yaml header

Replace "output: html_document" in the yaml header with the following in order to creat a table of content (toc) on the left side of the page and to enable code folding
 
```
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
```

### Modify the "setup" chunk

The setup chunk should look something like this

```{r intro-setup, message=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# wrangling packages
library(here)
library(janitor)
library(readxl)
library(data.table)

# analysis packages
library(MASS) # negative binomial and some other functions
library(nlme) # gls and some lmm

# graphing packages
library(ggsci) # color palettes
library(ggpubr) # publication quality plots
library(ggthemes) # colorblind palette
library(ggforce) # better jitter
library(cowplot) # combine plots

# Okabe & Ito palette
pal_okabe_ito <- colorblind_pal()(8)[2:8] # ggthemes

# use here from the here package
here <- here::here

# load functions used by this text written by me
source(here("R/ggplot_the_model.R"))

# use clean_names from the janitor package
clean_names <- janitor::clean_names

data_folder <- "data"
```

`echo = TRUE` tells knitr to display the code within a code chunk when the R markdown file is knitted. `knitr::opts_chunk$set(echo = TRUE)` sets `echo = TRUE` for all code chunks in the containing .Rmd file (not all R markdown files in the project!).

`pal_okabe_ito <- colorblind_pal()(8)[2:8]` assigns the Okabe-Ito colorblind palette to the object `pal_okabe_ito`. We use this to specify colorblind-safe colors when generating plots.

`here <- here::here` makes sure that `here` uses the function from the here package and not some other package. Huh? Let's back up -- R is an open source project and packages are written by independent programmers and scientists and not employees of some central company. When someone develops a package, they create functions that do stuff. Sometimes developers of different packages create functions that have the same name. There is a function name conflict if we load two packages  with the same name. Our R session will use the function of the last loaded package as the function assigned to the name. If we want the name to be used with the function in the previously loaded package, then we need to either re-order the library() statements, or simply re-assign the name to the function that we want. This is what `here <- here::here` does. This script takes the function `here` from the package "here" and assigns it to the object "here"!.

### Create a "fake-data" chunk

4. Let's play around with an R Markdown file. Create a new chunk and label it "fake-data". Insert the following R script and then click the chunk's run button

```{r intro-fake-data}
set.seed(4)
n <- 10
fake_data <- data.table(
    treatment = rep(c("cn", "tr"), each = n),
    neutrophil_count_exp1 = rnegbin(n*2, 
                                    mu = rep(c(10, 15), each = n),
                                    theta = 1),
    neutrophil_count_exp2 = rnegbin(n*2, 
                                    mu = rep(c(10, 20), each = n),
                                    theta = 1)
)
# View(fake_data)
```

This chunk creates fake neutrophil counts in two different experiments. The comment (\#) sign before `View(fake_data)` "comments out" the line of code, so it is not run. View the data by highlighting `View(fake_data)` and choosing "Run selected line(s)" from the Run menu.

### Create a "plot" chunk

5. Create a new chunk and label it "plot". Insert the following R script and then click the chunk's run button

```{r intro-plot}
gg_1 <- ggstripchart(data = fake_data,
                x = "treatment",
                y = "neutrophil_count_exp1",
                color = "treatment",
                palette = "jco",
                add = "mean_se",
                legend = "none") +
    ylab("Neutrophil Count (Exp. 1)") +
  stat_compare_means(method = "t.test",
                     label.y = 50,
                     label = "p.format") +
    NULL

gg_2<- ggstripchart(data = fake_data,
                x = "treatment",
                y = "neutrophil_count_exp2",
                color = "treatment",
                palette = "jco",
                add = "mean_se",
                legend = "none") +
  ylab("Neutrophil Count (Exp 2)") +
  stat_compare_means(method = "t.test",
                     label.y = 65,
                     label = "p.format") +
NULL

plot_grid(gg_1, gg_2, labels = "AUTO")
```

Each plot shows the mean count for each group, the standard error of the mean count, and the *p*-value from a *t*-test. This statistical analysis and plot are typical of those found in experimental biology journals. This text will teach alterntatives that implement better practices.

### Knit

6. Knit to an html file
7. If you've installed tinytex (or some other LaTeX distribution), knit to a pdf file
8. Knit to a word document








<!--chapter:end:chapters/04-setup.Rmd-->

# Analyzing experimental data with a linear model

```{r ask1-fig2i-setup, include=FALSE, echo = FALSE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = FALSE, warning=FALSE, message=FALSE)

library(here)
library(janitor)
library(readxl)
library(data.table)
library(stringr)

# analysis packages
library(emmeans)
library(car) # qqplot, spreadlevel
library(DHARMa)

# graphing packages
library(ggsci)
library(ggpubr)
library(ggthemes)
library(ggforce)
library(cowplot)
library(lazyWeave) #pvalstring

here <- here::here
data_path <- "data"

ggplotsci_path <- here("R", "ggplotsci.R")
source(ggplotsci_path)

# Okabe & Ito palette
pal_okabe_ito <- colorblind_pal()(8)[2:8] # ggthemes

```

```{r ask1-fig2i-import}
data_folder <- "ASK1 inhibits browning of white adipose tissue in obesity"
file_name <- "41467_2020_15483_MOESM4_ESM.xlsx"
file_path <- here(data_path, data_folder, file_name)
  
fig_2_sheet <- "Source Date_Figure 2"

range_2i <- "A265:G266"
treatment_levels <- c("ASK1F/F", "ASK1Δadipo")
fig2i <- read_excel(file_path,
                     sheet = fig_2_sheet,
                     range = range_2i,
                     col_names = FALSE) %>%
  transpose(make.names=1) %>%
  data.table() %>%
  melt(measure.vars = treatment_levels,
       variable.name = "treatment",
       value.name = "liver_tg") %>%
  na.omit()

fig2i[, treatment := factor(treatment, treatment_levels)]

# View(fig2i)
```

```{r ask1-fig2i-m1, echo = TRUE}
fig2i_m1 <- lm(liver_tg ~ treatment, data = fig2i)
```

```{r ask1-fig2i-emm, echo = TRUE}
fig2i_m1_emm <- emmeans(fig2i_m1, specs = "treatment")
fig2i_m1_pairs <- contrast(fig2i_m1_emm,
                            method = "revpairwise") %>%
  summary(infer = TRUE)
```


```{r ask1-fig2i-gg-prep, echo = TRUE}
fig2i_m1_emm_dt <- summary(fig2i_m1_emm) %>%
  data.table
fig2i_m1_pairs_dt <- data.table(fig2i_m1_pairs)
fig2i_m1_pairs_dt[ , p_pretty := pvalString(p.value)]
fig2i_m1_pairs_dt[, group1 := 1]
fig2i_m1_pairs_dt[, group2 := 2]
```

```{r ask1-fig2i-gg1}
paired_list <- list(c("ASK1F/F", "ASK1Δadipo"))
fig2i_gg1 <- ggstripchart(
  data = fig2i,
  x = "treatment",
  y = "liver_tg",
  color = "treatment",
  add = "mean_se",
  palette = pal_okabe_ito_2
) +
  ylab("Liver TG\n(µmol per g liver)") +
  theme(legend.position="none") +
  theme(axis.title.x=element_blank()) +
  stat_compare_means(method = "t.test",
                     comparisons = paired_list,
                     label = "p.signif") +
  NULL
#fig2i_gg1
```

```{r ask1-fig2i-ggplot_the_model, warning=FALSE}
ggplot_the_model(fig2i_m1,
                 fig2i_m1_emm,
                 fig2i_m1_pairs,
                           palette = pal_okabe_ito_2,
                 rel_heights = c(0.5,1),
                           legend_position = "none",
                 y_label = "Liver TG\n(µmol per g liver)",
                 effect_label = "Effect (µmol per g liver)")
```


```{r ask1-fig2i-gg2}
fig2i_gg <- ggplot(data = fig2i,
                    aes(x = treatment,
                        y = liver_tg,
                        color = treatment)) +
  
  # points
  geom_sina(alpha = 0.5) +
  
  # plot means and CI
  geom_errorbar(data = fig2i_m1_emm_dt,
                aes(y = emmean,
                    ymin = lower.CL,
                    ymax = upper.CL,
                    color = treatment),
                width = 0
  ) +
  
  geom_point(data = fig2i_m1_emm_dt,
             aes(y = emmean,
                 color = treatment),
             size = 3
  ) +
  
 # plot p-values (y positions are adjusted by eye)
  stat_pvalue_manual(fig2i_m1_pairs_dt,
                     label = "p_pretty",
                     y.position=c(95),
                     tip.length = 0.01) +
  
  # aesthetics
  ylab("Liver TG (µmol per g liver)") +
  scale_color_manual(values=pal_okabe_ito[5:6],
                     name = NULL) +
  theme_pubr() +
  theme(legend.position="none") +
  theme(axis.title.x=element_blank()) +
  
  NULL

fig2i_gg

```

## This text is about the estimation of treatment effects and the uncertainty in our estimates using linear models. This, raises the question, what is "an effect"?

This text has an unusual start -- an example analysis. This example is a goal or target; it's what you will be working towards as you learn from this text. The data for the analysis come from multiple experiments presented in Figure 2 in the article [ASK1 inhibits browning of white adipose tissue in obesity](https://www.nature.com/articles/s41467-020-15483-7){target="_blank"}. The analysis is in the last part of this chapter. The second part of this chapter is just enough biology to help you understand the biological importance of each experiment. The first part of this chapter uses one experiment from Figure 2 to outline what the statistical analysis of experimental data is all about. Much of this outline will be repeated in "An introduction to linear models" chapter.

The analysis in part 3 of this chapter is a set of experiments exploring the consequences of adispose-tissue specific deletion of the ASK1 signaling protein on multiple, adverse effects of a high-fat diet in mice, including weight gain, glucose intolerance, and increased liver triacylglycerol levels. Think of this as a template for organizing your own R Markdown documents. This document is a re-analysis of the experiments in Figure 2 in the article [ASK1 inhibits browning of white adipose tissue in obesity](https://www.nature.com/articles/s41467-020-15483-7){target="_blank"}, including generation of the publication-ready plots. I chose the data in Figure 2 of this paper because of the diversity of analyses and plot types. My analyses and plots differ slightly from those of the researchers because I implemented better practices -- the stuff of this text.

The goal of the experiments is to measure the **effect** of the adipose-specific ASK1 deletion. To understand what I mean by "an effect", and to understand how we can estimate an effect by **fiting a linear model** to data, let's look more closely at the analysis for Figure 2i.

For Figure 2i, the researchers want to know if "knocking out" the ASK1 gene in the adipose tissue cells lowers the liver triglyceride (TG) level in mice fed a high-fat diet. That is, is $\overline{y}_{ASK1Δadipo} < \overline{y}_{ASK1F/F}$, where $\overline{y}_{ASK1Δadipo}$ is the mean liver TG level of the knockout (Δ is the del operator and refers to a deletion in genetics) mice and $\overline{y}_{ASK1F/F}$ is the mean liver TG level of the control mice. The difference in the means, $\overline{y}_{ASK1Δadipo} - \overline{y}_{ASK1F/F}$, is **the effect** (of ASK1 deletion on liver TG levels).

The measured means in each group are computed from a random **sample** of mice. If we only cared about the six mice in each group in this experiment, then we would not need to fit a linear model to the data to estimate the effect, we could simply compute each group mean and subtract the control mean from the knockout mean. But we care more about these dozen mice because we are trying to discover something general about ASK1 regulation of TG levels in mice, generally (and even in mammals, and especially humans, generally). To make this leap of **inference**, we use a model to claim that each sample mean is an **estimate** of the respective **population** mean. Given this model, we can compute the **standard error** of each mean and the **standard error of the difference in means**. A standard error is a measure of the **sampling variance of a statistic** and, therefore, a measure of the precision of the estimate. The standard error, then, is a measure of **uncertainty** in the estimate. Here is how to think about precision and uncertainty: if we were to repeat this experiment many, many times, we would generate a long list of mean TG levels for the control mice and a long list of mean TG levels for the knockout mice. The less variable the means in each list, the more precise. By using a model, we do not need to repeat this experiment many times to get a standard error.

The model we are going to fit to the Figure 2i data is

\begin{align}
y_i &= \beta_0 + \beta_1 x_i + \varepsilon_i\\
\varepsilon_i &\sim N(0, \sigma^2)
\end{align}

This is a model of how the Figure 2i data were generated. In this model, $y_i$ is the liver TG level for some fictional (generated!) mouse (the *i* stands for the *i*th fictional mouse generated) and $x_i$ is a variable that indicates the condition of the ask1 gene in ficitional mouse *i*. For $x_i$, a value of 0 is given to mice with a functional ASK1 gene and a value of 1 is given to mice with a knocked out gene. 

$\beta_0$ is the "true" mean of the TG level in mice fed a high-fat diet and with a functional ASK1 gene. By "true", I mean the mean that would be computed if we were to measure TG on an infinite number of these mice. The observed mean of the ASK1F/F group is an estimate of $\beta_0$. The sum $\beta_0$ + $\beta_1$ is the true mean of the TG level in mice fed a high-fat diet but with a knocked out ASK1 gene. This means that $\beta_1$ is the true difference in the means, or the **true effect**. The observed difference in means between the ASK1Δadipo and ASK1F/F groups is an estimate of $\beta_1$. This difference is the estimated effect.

Notice that the sum $\beta_0 + \beta_1 x_i$ equals the true mean of the infinite set of normal mice if $x_i = 0$ and equals the true mean of the infinite set of ASK1 knockout mice if $x_i = 1$. $\varepsilon_i$ is the **error** for mouse *i*, which is the difference between the TG level for mouse *i* and the **expected** TG value for mouse *i*. The expected value for a mouse with a normal ASK1 gene is $\beta_0$. The expected value for a mouse with a knocked out ASK1 gene is $\beta_0 + \beta_1$. The second line of the model simply states that $\varepsilon_i$ is modeled as a random sample from a normal distribution with a mean of zero and a variance of $\sigma^2$.

By **fitting a model to the data** we estimate the parameters $\beta_0$, $\beta_1$ and $\sigma$. It is the estimation of $\sigma$ that allows us to compute a measure of our uncertainty (a standard error) of our estimates of the means ($\beta_0$ and $\beta_0 + \beta_1$) and of the difference in the means ($\beta_1$).



Let's fit this model to the Figure 2i data using R

Robust inference from the model (generalizing from sample to population, including measures of the uncertainty of our estimates) requires that our data approximates the kind of data we'd expect from the data generating model specified above. All rigorous analysis should use specific **model checks** to evaluate this. First, the "normality check" -- we use a **quantile-quantile (QQ) plot** to see if our data approximate what we'd see if we sampled from a normal distribution. This looks okay, in the sense that the observed data points (open circles) fall within the boundaries set by the dashed line. Inference is pretty robust to moderate departure from normal.

```{r ask1-fig2i-qq, echo = TRUE}
set.seed(1)
qqPlot(fig2i_m1, id=FALSE)
```

Second, the "homogeneity check" -- we use a **spread level plot** to see if there is some pattern to the variance, for example if the spread of residuals is noticeably bigger in one group than another, or if the spread increases with the fitted value. 
This looks pretty good. Given these checks, lets move on and look at the table of model coefficients

```{r ask1-fig2i-spread, echo = TRUE, warning=FALSE}
spreadLevelPlot(fig2i_m1, id=FALSE)
```


```{r ask1-fig2i-coef, echo = TRUE}
fig2i_m1 <- lm(liver_tg ~ treatment, data = fig2i)
fig2i_m1_coef <- cbind(coef(summary(fig2i_m1)),
                        confint(fig2i_m1))
knitr::kable(fig2i_m1_coef, digits = c(1, 2, 1, 3, 1, 1))
```

The two values in the column "Estimate" are the estimates of $\beta_0$ and $\beta_1$. The top value (61.5) is the mean of the control mice (the units are µmol/g). The mean of the knockout mice is the sum of the two values (39.9 µmol/g). And the effect of ASK1 deletion on TG levels is simply the second value (-21.6 µmol/g). The standard error of the effect is 7.05 µmol/g. We can use the standard error to compute a *t*-value (-3.1, in the column "t value"). A t-value is a **test statistic**. The probability ("p value") of the significance test is 0.012. This if the probability of sampling a *t*-value as large or larger than the observed *t*-value, if we were to sample from a null distribution of *t*-values (a distribution of sampled t values if the true value of $\beta_1$ was 0). We can also use the standard error to compute a **95% confidence interval of the effect**. The lower bound of this interval is -37.3 µmol/g and the upper bound is -5.9 µmol/g. A confidence interval is another way of communicating uncertainty, and the way advocated in this text. In a 95% confidence interval, 95% of similarly constructed intervals (from hypothetical sampling of six mice from the ASK1 normal population and six mice from the ASK1 knockout population) will contain the true mean. Another way to think about a confidence interval is, it is the range of true differences that are compatible with the data, where compatible means "not rejected" by a *t*-test (a *t*-test between the estimated effect and any number inside the interval would return a *p*-value greater than 0.05).

Here is how we might report this result in a paper:

Mean TG level in ASK1Δadipo mice on a high-fat diet was 21.6 µmol/g less than that in ASK1F/F mice on a high-fat diet (95% CI: -37.3, -5.9, $p = 0.012$).

And a plot for the paper:

```{r ask1-fig2i-explainer-plot, echo=FALSE, warning=FALSE, message=FALSE}
do_it <- TRUE
if(do_it == TRUE){
  fig2i_explainer <- fig2i_gg +
    geom_segment(aes(x = 1,
                 y = fig2i_m1_emm_dt[treatment == "ASK1F/F",
                                             emmean],
                 xend = 1,
                 yend = fig2i_m1_emm_dt[treatment == "ASK1Δadipo",
                                             emmean]),
               linetype = "dashed",
               color = "black") +
    geom_segment(aes(x = 1,
                     y = fig2i_m1_emm_dt[treatment == "ASK1Δadipo",
                                          emmean],
                     xend = 2,
                     yend = fig2i_m1_emm_dt[treatment == "ASK1Δadipo",
                                             emmean]),
                 linetype = "dashed",
                 color = "black") +
    geom_segment(aes(x = 1,
                     y = fig2i_m1_emm_dt[treatment == "ASK1F/F",
                                          emmean],
                     xend = 2,
                     yend = fig2i_m1_emm_dt[treatment == "ASK1Δadipo",
                                             emmean]),
                 color = "black") +
    annotate(geom = "text",
             label = "1 - 0 = 1",
             x = 1.5,
             y = 37) +
    annotate(geom = "text",
             label = bquote(bar(y)[2] - bar(y)[1]),
             x = 1.1,
             y = 50)
  
  
  
  #fig2i_explainer
  image_path <- here("output", "fig2i_explainer.Rds")
  saveRDS(fig2i_explainer, image_path)
}

```



<!--chapter:end:chapters/05-ask1_fig2i.Rmd-->

# Background physiology to the experiments in Figure 2 of "ASK1 inhibits browning of white adipose tissue in obesity" {-}

A little background on the subject of the article: white adipose tissue (WAT) is composed of adipose (fat) cells that function as energy storage cells. The energy is in the form of the fatty acids in the triacylglycerols, which form large lipid drops in the cell. The stored fatty acids are released from the WAT when other organs need energy. Mammalian brown adipose tissue (BAT) is composed of adipose cells that burn the stored fat to generate heat. This is enabled by the expression of the protein *uncoupling receptor 1* (UCP1) in the mitochondria. UCP1 uncouples the proton gradient across the inner mitochondrial membrane from ATP synthesis.

In response to adverse health consequences of obesity, including metabolic syndrome, researchers are investigating various ways to increase BAT, or stimulate BAT activity, or transform WAT cells into more BAT-like cells, by turning up expression of UCP1. The regulation of UPC1 in WAT is a potential drug target for obesity.

The researchers of the ASK1 study investigated the effects of an intracellular signaling protein (ASK1) on the browning of white adipose tissue. Previous research had suggested that 1) inflammation stimulates ASK1 activity and 2) increased ASK1 acitivty inhibits UPC1 expression (Figure \@ref(fig:ask1-mechanism). The experiments in Figure 2 of the ASK1 study follow this up and explore the question, if ASK1 is knocked out in the WAT cells, will this reverse the adverse effects of a high-fat diet, including weight gain, glucose intolerance, and liver triacylglycerol levels?

```{r ask1-mechanism, echo=FALSE, fig.cap="Inflammation to obesity stimulates ASK1 activity. ASK1 activity inhibits UCP1 expression."}
image_path <- here::here("images", "ASK1_UCP1.png")
knitr::include_graphics(image_path)
```

For the experiments in Figure 2, the researchers created mice in which the ASK1 gene was inhibited from being expressed (or "knocked out") in the white adipose tissue cells. The $ask1$ treatment has two levels: "ASK1Δadipo", which are the adipocyte-specific ASK1 knockout (KO) mice, and "ASK1F/F", which are the controls. For some of the experiments, the researchers split the mice in each $ask1$ treatment level and assigned these to either a Chow or a High Fat Diet (HFD). This experimental design is two-crossed factors, each with two levels, which I call a $2 \times 2$ factorial design in this text.

* Some of the plots are coded directly in this document. Others use functions from the chapter "Plotting functions". But, to use these in an R Markdown document, these functions have to be saved in a "R Script" file. This script file then needs to be read at the start of the R Markdown document. I named the script file "ggplotsci.R" and placed it in a folder called "R" at the level of the project (directly within the project folder).

* This example was written with the Bookdown style sheet (because its part of this book), which doesn't have one nice features of creating R Markdown documents for reports and manuscripts -- code folding. In an R Markdown document with code folding, a user can toggle between showing and hiding code. The html output with code folding is here.

 

<!--chapter:end:chapters/05-ask1_intro.Rmd-->

---
title: Analysis for Figure 2 of ASK1 inhibits browning of white adipose tissue in obesity
author: "Jeffrey A. Walker"
date: "9/2/2020"
output:
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
---

# Analyses for Figure 2 of "ASK1 inhibits browning of white adipose tissue in obesity" {-}

## Setup

Some plots in this document require the file "ggplotsci.R" within the "R" folder, within the project folder.

```{r target-setup, echo=TRUE, warning=FALSE, message=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      warning=FALSE,
                      message=FALSE)

# wrangling packages
library(here)
library(janitor)
library(readxl)
library(data.table)
library(stringr)

# analysis packages
library(emmeans)
library(car) # qqplot, spreadlevel
library(DHARMa)

# graphing packages
library(ggsci)
library(ggpubr)
library(ggforce)
library(cowplot)
library(lazyWeave) #pvalstring

here <- here::here
data_path <- "data"

ggplotsci_path <- here("R", "ggplotsci.R")
source(ggplotsci_path)
```

## Data source

Data source: [ASK1 inhibits browning of white adipose tissue in obesity](https://www.nature.com/articles/s41467-020-15483-7){target="_blank"} 

This chunk assigns the path to the Excel data file for all panels of Figure 2. The data for each panel are in a single sheet in the Excel file named "Source Date_Figure 2".

```{r target-file-path}
data_folder <- "ASK1 inhibits browning of white adipose tissue in obesity"
file_name <- "41467_2020_15483_MOESM4_ESM.xlsx"
file_path <- here(data_path, data_folder, file_name)
  
fig_2_sheet <- "Source Date_Figure 2"
```

## control the color palette

```{r}
fig_2_palette <- pal_okabe_ito[5:6]
# fig_2_palette <- pal_okabe_ito
#fig_2_palette <- pal_nature_mod

```

## useful functions

A function to import longitudinal data from Fig 2

```{r target-fig2-import-in-parts}
# function to read in parts of 2b
import_fig_2_part <- function(range_2){
  fig_2_part <- read_excel(file_path,
                          sheet = fig_2_sheet,
                          range = range_2,
                          col_names = TRUE) %>%
    data.table()
  group <- colnames(fig_2_part)[1]
  setnames(fig_2_part, old = group, new = "treatment")
  fig_2_part[, treatment := as.character(treatment)] # this was read as logical
  fig_2_part[, treatment := group] # assign treatment group
  fig_2_part[, mouse_id := paste(group, .I)]
  return(fig_2_part)
}

```

```{r target-fig2-AUC}
# script to compute various area under the curves (AUC) using trapezoidal method
# le Floch's "incremental" auc substracts the baseline value from all points.
# This can create some elements with negative area if post-baseline values are less
# than baseline value.
# Some researchers "correct" this by setting any(y - ybar < 0 to zero. Don't do this.

auc <- function(x, y, method="auc", average = FALSE){
  # method = "auc", auc computed using trapezoidal calc
  # method = "iauc" is an incremental AUC of Le Floch
  # method = "pos_iauc" is a "positive" incremental AUC of Le Floch but not Wolever
  # method = "post_0_auc" is AUC of post-time0 values
  # if average then divide area by duration
  if(method=="iauc"){y <- y - y[1]}
  if(method=="pos_iauc"){y[y < 0] <- 0}
  if(method=="post_0_auc"){
    x <- x[-1]
    y <- y[-1]
  }
  n <- length(x)
  area <- 0
  for(i in 2:n){
    area <- area + (x[i] - x[i-1])*(y[i-1] + y[i])
  }
  value <- area/2
  if(average == TRUE){
    value <- value/(x[length(x)] - x[1])
  }
  return(value)
}

```

## figure 2b -- effect of ASK1 deletion on growth (body weight)
### figure 2b -- import

```{r target-fig2b-import}
range_list <- c("A21:N41", "A43:N56", "A58:N110", "A112:N170")
fig_2b_wide <- data.table(NULL)
for(range_i in range_list){
  part <- import_fig_2_part(range_i)
  fig_2b_wide <- rbind(fig_2b_wide,
                       part)
}

fig_2b <- melt(fig_2b_wide,
               id.vars = c("treatment", "mouse_id"),
               variable.name = "week",
               value.name = "body_weight")
fig_2b[, week := as.numeric(as.character(week))]
fig_2b[, c("ask1", "diet") := tstrsplit(treatment, " ", fixed=TRUE)]
fig_2b[, week_f := factor(week)]

```

### figure 2b -- exploratory plots

```{r target-fig2b-initial-plot-1}
qplot(x = week,
      y = body_weight,
      data = fig_2b,
      color = treatment) +
  facet_grid(ask1~diet)
```

1. no obvious outliers
2. reduced growth rate at bigger size

```{r target-fig2b-initial-plot-2}
qplot(x = week,
      y = body_weight,
      data = fig_2b,
      color = treatment) +
  geom_smooth()
```

1. loess smooth. Growth in ASK1F/F + HFD clearly greater than other three treatment combinations.


```{r target-fig2b-analysis}

```

## Figure 2c -- Effect of ASK1 deletion on final body weight
### Figure 2c -- import

```{r target-fig2c-import}
range_2c <- "A173:BD176"
y_cols <- c("ASK1F/F chow", "ASK1Δadipo chow", "ASK1F/F HFD", "ASK1Δadipo HFD")
fig_2c_import <- read_excel(file_path,
                     sheet = fig_2_sheet,
                     range = range_2c,
                     col_names = FALSE) %>%
  transpose(make.names=1) %>%
  data.table() %>%
  melt(measure.vars = y_cols,
       variable.name = "treatment",
       value.name = "body_weight_gain") %>%
  na.omit()
fig_2c_import[, mouse_id := paste(treatment, .I, sep = "_"),
       by = treatment]
```

### Figure 2c -- check own computation of weight change v imported value

Note that three cases are missing from fig_2c import that are in fig_2b

```{r}
# change colnames to char
fig_2c <- copy(fig_2b_wide)
weeks <- unique(fig_2b[, week])
setnames(fig_2c,
         old = colnames(fig_2c),
         new = c("treatment", paste0("week_", weeks), "mouse_id"))
fig_2c[, weight_gain := week_12 - week_0]
fig_2c <- fig_2c[, .SD, .SDcols = c("treatment",
                                "week_0",
                                "week_12",
                                "weight_gain")]
fig_2c[, mouse_id := paste(treatment, .I, sep = "_"),
       by = treatment]

fig_2c[, c("ask1", "diet") := tstrsplit(treatment, " ", fixed=TRUE)]

fig_2c_check <- merge(fig_2c,
                fig_2c_import,
                by = c("mouse_id"),
                all = TRUE)

#View(fig_2c_check)


```

### Figure 2c -- exploratory plots

```{r}
qplot(x = treatment,
      y = weight_gain,
      data = fig_2c)
```

* no obvious outliers
* variance increases with mean, as expected from growth, suggests a multiplicative model. But start with simple lm.

### Figure 2c -- fit the model: m1 (lm)

```{r}
fig_2c_m1 <- lm(weight_gain ~ week_0 + ask1*diet, data = fig_2c)
```

### Figure 2c -- check the model: m1

```{r}
# check normality assumption
set.seed(1)
qqPlot(fig_2c_m1, id=FALSE)
```

```{r}
spreadLevelPlot(fig_2c_m1, id=FALSE)

```

* QQ indicates possible right skew but especially left side is squashed toward mean
* spread-level indicates variance increases with mean

For p-value, this may not be too severe but for intervals, best to account for this. Try gamma with log link (which makes biological sense for growth)

### Figure 2c -- fit the model: m2 (gamma glm)

```{r}
fig_2c_m2 <- glm(weight_gain ~ week_0 + ask1*diet,
          family = Gamma(link = "log"),
          data = fig_2c)
```

### Figure 2c -- check the model, m2

```{r}
set.seed(1)
fig_2c_m2_sim  <-  simulateResiduals(fig_2c_m2,  n=250)
plot(fig_2c_m2_sim) 

```

* well behaved QQ and spread-level

### Figure 2c -- inference from the model
```{r, message = FALSE}
coef_table <- cbind(coef(summary(fig_2c_m2)),
                    exp(confint(fig_2c_m2))) %>%
  data.table(keep.rownames = TRUE)
coef_table[, Estimate:=exp(Estimate)]
knitr::kable(coef_table, digits = c(0,2,2,2,4,2,2))
```

```{r}
fig_2c_m2_emm <- emmeans(fig_2c_m2, specs = c("diet", "ask1"),
                         type = "response")
fig_2c_m2_pairs <- contrast(fig_2c_m2_emm,
                     method = "revpairwise",
                     simple = "each",
                     combine = TRUE,
                     adjust = "none") %>%
  summary(infer = TRUE)
```

```{r}
fig_2c_m2_emm
```

```{r}
fig_2c_m2_pairs
```

* within ASK1 Cn, HFD mean is 1.6X chow mean
* within ASK1 KO, HFD mean is 1.2X chow mean
* within chow, ASK1 KO mean is 1.0X ASK1 Cn mean
* within HFD, ASK1 KO mean is 0.86X ASK1 Cn mean

```{r}
# same as interaction effect in coefficient table
contrast(fig_2c_m2_emm, interaction = "pairwise", by = NULL) %>%
  summary(infer = TRUE)
```
* the reduction in weight gain in the ASK1 KO mice compared to ASK1 CN is 0.785X. Notice that p > 0.05.

### Figure 2c -- plot the model

```{r, message=FALSE}
fig_2c_m2_emm_dt <- summary(fig_2c_m2_emm) %>%
  data.table
fig_2c_m2_pairs_dt <- data.table(fig_2c_m2_pairs)
fig_2c_m2_pairs_dt[ , p_pretty := pvalString(p.value)]

dodge_width <- 0.8 # separation between groups
# get x positions of brackets for p-values
# requires looking at table and mentally figuring out
# Chow is at x = 1 and HFD is at x = 2
fig_2c_m2_pairs_dt[, group1 := c(1-dodge_width/4,
                                 1+dodge_width/4,
                                 1-dodge_width/4,
                                 2-dodge_width/4)]
fig_2c_m2_pairs_dt[, group2 := c(2-dodge_width/4,
                                 2+dodge_width/4,
                                 1+dodge_width/4,
                                 2+dodge_width/4)]

pd <- position_dodge(width = dodge_width)
fig_2c_gg <- ggplot(data = fig_2c,
                    aes(x = diet,
                        y = weight_gain,
                        color = ask1)) +
  
  # points
  geom_sina(alpha = 0.5,
            position = pd) +
  
  # plot means and CI
  geom_errorbar(data = fig_2c_m2_emm_dt,
                aes(y = response,
                    ymin = asymp.LCL,
                    ymax = asymp.UCL,
                    color = ask1),
                width = 0,
                position = pd
  ) +
  
  geom_point(data = fig_2c_m2_emm_dt,
             aes(y = response,
                 color = ask1),
             size = 3,
             position = pd
  ) +
  
  # plot p-values (y positions are adjusted by eye)
  stat_pvalue_manual(fig_2c_m2_pairs_dt,
                     label = "p_pretty",
                     y.position=c(28.5, 31, 26, 26),
                     tip.length = 0.01) +
  
  # aesthetics
  ylab("Weight Gain") +
  scale_color_manual(values=fig_2_palette,
                     name = NULL) +
  theme_pubr() +
  theme(legend.position="top") +
  theme(axis.title.x=element_blank()) +
  
  NULL

fig_2c_gg
```

### Figure 2c -- report

Results could be reported using either:

(This is inconsistent with plot, if using this, the plot should reverse what factor is on the x-axis and what factor is the grouping (color) variable) Mean weight gain in ASK1F/F mice on HFD was 1.56 (95% CI: 1.33, 1.82, $p < 0.0001$) times that of ASK1F/F mice on chow while mean weight gain in ASK1Δadipo mice on HFD was only 1.22 (95% CI: 1.01, 1.47, $p = 0.036$) times that of ASK1Δadipo mice on chow. This reduction in weight gain in ASK1Δadipo mice compared to ASK1F/F control mice was 0.79 times (95% CI; 0.61, 1.00, $p = 0.0531$).

(This is consistent with the plot in that its comparing difference in the grouping factor within each level of the factor on the x-axis) Mean weight gain in ASK1Δadipo mice on chow was trivially larger (1.03 times) than that in ASK1F/F mice on chow (95% CI: 0.83, 1.27, $p = 0.81$) while mean weight gain in ASK1Δadipo mice on HFD was smaller (0.81 times) than that in ASK1F/F control mice on HFD (95% CI: 0.72 , 0.91, $p = 0.0003$). This reduction in weight gain in ASK1Δadipo mice compared to ASK1Δadipo mice is 0.79 times (95% CI; 0.61, 1.00, $p = 0.0531$).

**note to research team**. The big difference in *p*-values between weight difference on chow and weight difference on HFD might lead one to believe there is a "difference in this difference". Using a *p*-value = effect strategy, this is not supported.

## Figure 2d -- Effect of ASK1 KO on glucose tolerance (whole curve)
### Figure 2d -- Import

```{r}
range_list <- c("A179:H189", "A191:H199", "A201:H214", "A216:H230")
fig_2d_wide <- data.table(NULL)
for(range_i in range_list){
  part <- import_fig_2_part(range_i)
  fig_2d_wide <- rbind(fig_2d_wide,
                       part)
}
fig_2d_wide[, c("ask1", "diet") := tstrsplit(treatment, " ", fixed=TRUE)]

# melt
fig_2d <- melt(fig_2d_wide,
               id.vars = c("treatment",
                           "ask1",
                           "diet",
                           "mouse_id"),
               variable.name = "time",
               value.name = "glucose")
fig_2d[, time := as.numeric(as.character(time))]

# for plot only (not analysis!)
shift <- 2
fig_2d[treatment == "ASK1F/F chow", time_x := time - shift*1.5]
fig_2d[treatment == "ASK1Δadipo chow", time_x := time - shift*.5]
fig_2d[treatment == "ASK1F/F HFD", time_x := time + shift*.5]
fig_2d[treatment == "ASK1Δadipo HFD", time_x := time + shift*1.5]

```

### Figure 2d -- exploratory plots
```{r}

qplot(x = time_x, y = glucose, color = treatment, data = fig_2d) +
  geom_line(aes(group = mouse_id), alpha = 0.3)
```
* no obvious unplausible outliers but two mice w/ high values in "F/F HFD"
* similar at time zero (initial effect is trivial)

use AUC conditional on time 0 glucose

```{r}
qplot(x = time,
      y = glucose,
      data = fig_2d,
      color = treatment) +
  geom_smooth()

```

### Figure 2d -- fit the model

### Figure 2d -- check the model

### Figure 2d -- inference

### Figure 2d -- plot the model

## Figure 2e -- Effect of ASK1 deletion on glucose tolerance (summary measure)

The researchers did create a table to import but this analysis uses the mean post-baseline glucose amount as the response instead of the area under the curve of over the full 120 minutes. This mean is computed as the post-baseline area under the curve divided by the duration of time of the post-baseline measures (105 minutes). This analysis will use fig_2d_wide since there is only one a single Y variable per mouse.

### Figure 2e -- message the data

```{r}
# AUC of post-baseline values
# do this after melt as we don't need this in long format)
fig_2e <- fig_2d_wide
fig_2e[, glucose_0 := get("0")]

times <- c(0, 15, 30, 45, 60, 90, 120)
time_cols <- as.character(times)
Y <- fig_2e[, .SD, .SDcols = time_cols]
fig_2e[, glucose_mean := apply(Y, 1, auc,
                           x=times,
                           method = "post_0_auc",
                           average = TRUE)]
```

### Figure 2e -- exploratory plots

```{r}
qplot(x = treatment, y = glucose_mean, data = fig_2e)
```

### Figure 2e -- fit the model
```{r}
fig_2e_m1 <- lm(glucose_mean ~ glucose_0 + ask1*diet, data = fig_2e)
```

### Figure 2e -- check the model

```{r}
# check normality assumption
set.seed(1)
qqPlot(fig_2e_m1, id=FALSE)
```

```{r}
spreadLevelPlot(fig_2e_m1, id=FALSE)

```
### Figure 2e -- inference from the model

```{r}
fig_2e_m1_coef <- coef(summary(fig_2e_m1))
fig_2e_m1_coef
```
```{r}
fig_2e_m1_emm <- emmeans(fig_2e_m1, specs = c("diet", "ask1"))
fig_2e_m1_pairs <- contrast(fig_2e_m1_emm,
                     method = "revpairwise",
                     simple = "each",
                     combine = TRUE,
                     adjust = "none") %>%
  summary(infer = TRUE)
```

```{r}
fig_2e_m1_emm
```
```{r}
fig_2e_m1_pairs
```

### Figure 2e -- plot the model
```{r, message=FALSE}
fig_2e_gg <- ggplot_the_response(
  fig_2e_m1,
  fig_2e_m1_emm,
  fig_2e_m1_pairs,
  wrap_col=NULL,
  y_label = "Post-baseline glucose (mmol per l)",
  contrast_rows = "all",
  p_pos = c(2,3,1,1),
  palette = pal_okabe_ito_4,
)

fig_2e_gg

```

## Figure 2f -- Effect of ASK1 deletion  on glucose infusion rate
### Figure 2f -- import
```{r, message=FALSE}
range_2f <- "A239:I240"
treatment_levels <- c("ASK1F/F", "ASK1Δadipo")
fig_2f <- read_excel(file_path,
                     sheet = fig_2_sheet,
                     range = range_2f,
                     col_names = FALSE) %>%
  transpose(make.names=1) %>%
  data.table() %>%
  melt(measure.vars = treatment_levels,
       variable.name = "treatment",
       value.name = "glucose_infusion_rate") %>%
  na.omit()

fig_2f[, treatment := factor(treatment, treatment_levels)]
```

### Figure 2f -- exploratory plots

### Figure 2f -- fit the model

```{r}
fig_2f_m1 <- lm(glucose_infusion_rate ~ treatment, data = fig_2f)
```

### Figure 2f -- check the model

### Figure 2f -- inference

```{r}
fig_2f_m1_coef <- summary(fig_2f_m1) %>%
  coef()
fig_2f_m1_emm <- emmeans(fig_2f_m1, specs = "treatment")
fig_2f_m1_pairs <- contrast(fig_2f_m1_emm,
                            method = "revpairwise") %>%
  summary(infer = TRUE)
```

```{r}
fig_2f_m1_pairs
```

### Figure 2f -- plot the model

```{r}
fig_2f_m1_emm_dt <- summary(fig_2f_m1_emm) %>%
  data.table
fig_2f_m1_pairs_dt <- data.table(fig_2f_m1_pairs)
fig_2f_m1_pairs_dt[ , p_pretty := pvalString(p.value)]
fig_2f_m1_pairs_dt[, group1 := 1]
fig_2f_m1_pairs_dt[, group2 := 2]


fig_2f_gg <- ggplot(data = fig_2f,
                    aes(x = treatment,
                        y = glucose_infusion_rate,
                        color = treatment)) +
  
  # points
  geom_sina(alpha = 0.5) +
  
  # plot means and CI
  geom_errorbar(data = fig_2f_m1_emm_dt,
                aes(y = emmean,
                    ymin = lower.CL,
                    ymax = upper.CL,
                    color = treatment),
                width = 0
  ) +
  
  geom_point(data = fig_2f_m1_emm_dt,
             aes(y = emmean,
                 color = treatment),
             size = 3
  ) +
  
 # plot p-values (y positions are adjusted by eye)
  stat_pvalue_manual(fig_2f_m1_pairs_dt,
                     label = "p_pretty",
                     y.position=c(95),
                     tip.length = 0.01) +
  
  # aesthetics
  ylab("Glucose infusion rate") +
  scale_color_manual(values=fig_2_palette,
                     name = NULL) +
  theme_pubr() +
  theme(legend.position="none") +
  theme(axis.title.x=element_blank()) +
  
  NULL

fig_2f_gg
```

## Figure 2g -- Effect of ASK1 deletion on tissue-specific glucose uptake

### Figure 2g -- import
```{r}
range_list <- c("A244:G247", "A250:H253")
# import ASK1F/F
fig_2g_1 <- read_excel(file_path,
                     sheet = fig_2_sheet,
                     range = "A244:G247",
                     col_names = FALSE) %>%
  transpose(make.names=1) %>%
  data.table()
fig_2g_1[, treatment := "ASK1F/F"]

# import ASK1Δadipo
fig_2g_2 <- read_excel(file_path,
                     sheet = fig_2_sheet,
                     range = "A250:H253",
                     col_names = FALSE) %>%
  transpose(make.names=1) %>%
  data.table()
fig_2g_2[, treatment := "ASK1Δadipo"]

# combine
fig_2g <- rbind(fig_2g_1, fig_2g_2)


```

### Figure 2g -- exploratory plots

### Figure 2g -- fit the model

```{r}
# a more sophisticated would be a mixed model to dampen noise
fig_2g_m1_ingWAT <- lm(ingWAT ~ treatment, data = fig_2g)
fig_2g_m1_epiWAT <- lm(epiWAT ~ treatment, data = fig_2g)
fig_2g_m1_Muscle <- lm(Muscle ~ treatment, data = fig_2g)
fig_2g_m1_BAT <- lm(BAT ~ treatment, data = fig_2g)
```


### Figure 2g -- check the model

### Figure 2g -- inference
```{r}
fig_2g_infer <- function(m1){
  m1_emm <- emmeans(m1, specs = "treatment")
  m1_pairs <- contrast(m1_emm,
                              method = "revpairwise") %>%
    summary(infer = TRUE)
  return(list(emm = m1_emm,
              pairs = m1_pairs))
}

fig_2g_m1_emm_dt <- data.table(NULL)
fig_2g_m1_pairs_dt <- data.table(NULL)
m1_list <- list(fig_2g_m1_ingWAT, 
             fig_2g_m1_epiWAT,
             fig_2g_m1_Muscle,
             fig_2g_m1_BAT)
y_cols <- c("ingWAT", "epiWAT", "Muscle", "BAT")
for(i in 1:length(y_cols)){
  m1_infer <- fig_2g_infer(m1_list[[i]])

  m1_emm_dt <- summary(m1_infer$emm) %>%
    data.table
  fig_2g_m1_emm_dt <- rbind(fig_2g_m1_emm_dt,
                            data.table(tissue = y_cols[i], m1_emm_dt))
  m1_pairs_dt <- m1_infer$pairs %>%
    data.table
  fig_2g_m1_pairs_dt <- rbind(fig_2g_m1_pairs_dt,
                            data.table(tissue = y_cols[i], m1_pairs_dt))
}

```


```{r}
fig_2g_m1_pairs_dt
```

### Figure 2g -- plot the model

```{r, message = FALSE, warning=FALSE}
# melt fig_2g

fig_2g_long <- melt(fig_2g,
                    id.vars = "treatment",
                    variable.name = "tissue",
                    value.name = "glucose_uptake")

# change name of ASK1Δadipo label
fig_2g_long[treatment == "ASK1Δadipo", treatment := "ASK1-/-adipo"]
fig_2g_m1_emm_dt[treatment == "ASK1Δadipo", treatment := "ASK1-/-adipo"]

fig_2g_m1_pairs_dt[ , p_pretty := pvalString(p.value)]
fig_2g_m1_pairs_dt[, group1 := 1]
fig_2g_m1_pairs_dt[, group2 := 2]

fig_2g_plot <- function(tissue_i,
                        y_lab = FALSE, # title y-axis?
                        g_lab = FALSE  # add group label?
                        ){
  y_max <- max(fig_2g_long[tissue == tissue_i, glucose_uptake], na.rm=TRUE)
  y_min <- min(fig_2g_long[tissue == tissue_i, glucose_uptake], na.rm=TRUE)
  y_pos <- y_max + (y_max-y_min)*.05
  
  gg <- ggplot(data = fig_2g_long[tissue == tissue_i],
                      aes(x = treatment,
                          y = glucose_uptake,
                          color = treatment)) +
    
    # points
    geom_sina(alpha = 0.5) +
    
    # plot means and CI
    geom_errorbar(data = fig_2g_m1_emm_dt[tissue == tissue_i],
                  aes(y = emmean,
                      ymin = lower.CL,
                      ymax = upper.CL,
                      color = treatment),
                  width = 0
    ) +
    
    geom_point(data = fig_2g_m1_emm_dt[tissue == tissue_i],
               aes(y = emmean,
                   color = treatment),
               size = 3
    ) +
    
    # plot p-values (y positions are adjusted by eye)
    stat_pvalue_manual(fig_2g_m1_pairs_dt[tissue == tissue_i],
                       label = "p_pretty",
                       y.position=c(y_pos),
                       tip.length = 0.01) +
    
    # aesthetics
    ylab("Glucose Uptake") +
    scale_color_manual(values=fig_2_palette,
                       name = NULL) +
    ggtitle(tissue_i)+
    
    theme_pubr() +
    theme(legend.position="top") +
    theme(axis.title.x = element_blank(),
          axis.text.x = element_blank(),
          axis.ticks.x = element_blank()) +
    
    NULL
  
  if(y_lab == FALSE){
    gg <- gg + theme(axis.title.y = element_blank())
  }
  if(g_lab == FALSE){
    gg <- gg + theme(legend.position="none")
  }
  
  return(gg)
}

y_cols <- c("ingWAT", "epiWAT", "Muscle", "BAT")
legend <- get_legend(fig_2g_plot("ingWAT", y_lab = TRUE, g_lab = TRUE))
gg1 <- fig_2g_plot("ingWAT", y_lab = TRUE, )
gg2 <- fig_2g_plot("epiWAT")
gg3 <- fig_2g_plot("Muscle")
gg4 <- fig_2g_plot("BAT")
top_gg <- plot_grid(gg1, gg2, gg3, gg4,
                       nrow=1,
                       rel_widths = c(1.15, 1, 1.05, 1.1)) # by eye
plot_grid(top_gg, legend,
          nrow=2,
          rel_heights = c(1, 0.1))
```

## Figure 2h

## Figure 2i -- Effect of ASK1 deletion on liver TG

```{r}
range_2i <- "A265:G266"
treatment_levels <- c("ASK1F/F", "ASK1Δadipo")
fig_2i <- read_excel(file_path,
                     sheet = fig_2_sheet,
                     range = range_2i,
                     col_names = FALSE) %>%
  transpose(make.names=1) %>%
  data.table() %>%
  melt(measure.vars = treatment_levels,
       variable.name = "treatment",
       value.name = "liver_tg") %>%
  na.omit()

fig_2i[, treatment := factor(treatment, treatment_levels)]

# View(fig_2i)
```

### Figure 2i -- fit the model

```{r, echo = TRUE}
fig_2i_m1 <- lm(liver_tg ~ treatment, data = fig_2i)
```

### Figure 2i -- check the model

```{r, echo = TRUE}
set.seed(1)
qqPlot(fig_2i_m1, id=FALSE)
```


```{r, echo = TRUE}
spreadLevelPlot(fig_2i_m1, id=FALSE)
```

* The QQ plot looks okay, in the sense that the observed data points (open circles) fall within the boundaries set by the dashed line.
* spread level looks pretty good
* Fit normal model

### Figure 2i -- inference

```{r}
fig_2i_m1 <- lm(liver_tg ~ treatment, data = fig_2i)
fig_2i_m1_coef <- cbind(coef(summary(fig_2i_m1)),
                        confint(fig_2i_m1))
fig_2i_m1_emm <- emmeans(fig_2i_m1, specs = "treatment")
fig_2i_m1_pairs <- contrast(fig_2i_m1_emm,
                            method = "revpairwise") %>%
  summary(infer = TRUE)
```

```{r}
fig_2i_m1_pairs
```

### Figure 2i -- plot the model

```{r, echo = TRUE}

fig_2i_m1_emm_dt <- summary(fig_2i_m1_emm) %>%
  data.table
fig_2i_m1_pairs_dt <- data.table(fig_2i_m1_pairs)
fig_2i_m1_pairs_dt[ , p_pretty := pvalString(p.value)]
fig_2i_m1_pairs_dt[, group1 := 1]
fig_2i_m1_pairs_dt[, group2 := 2]


fig_2i_gg <- ggplot(data = fig_2i,
                    aes(x = treatment,
                        y = liver_tg,
                        color = treatment)) +
  
  # points
  geom_sina(alpha = 0.5) +
  
  # plot means and CI
  geom_errorbar(data = fig_2i_m1_emm_dt,
                aes(y = emmean,
                    ymin = lower.CL,
                    ymax = upper.CL,
                    color = treatment),
                width = 0
  ) +
  
  geom_point(data = fig_2i_m1_emm_dt,
             aes(y = emmean,
                 color = treatment),
             size = 3
  ) +
  
 # plot p-values (y positions are adjusted by eye)
  stat_pvalue_manual(fig_2i_m1_pairs_dt,
                     label = "p_pretty",
                     y.position=c(95),
                     tip.length = 0.01) +
  
  # aesthetics
  ylab("Liver TG (µmol per g liver)") +
  scale_color_manual(values=fig_2_palette,
                     name = NULL) +
  theme_pubr() +
  theme(legend.position="none") +
  theme(axis.title.x=element_blank()) +
  
  NULL

fig_2i_gg
```

### Figure 2i -- report the model

Mean TG level in ASK1Δadipo mice on a high-fat diet was 21.6 µmol/g less than that in ASK1F/F mice on a high-fat diet (95% CI: -37.3, -5.9, $p = 0.012$) (Figure xxx).

## Figure 2j


<!--chapter:end:chapters/05-ask1-r_markdown.Rmd-->

# Part III: R fundamentals {-}

<!--chapter:end:chapters/06-part_3_r_fundamentals.Rmd-->

# Data -- Reading, Wrangling, and Writing

```{r data-setup, echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(fig.width=6, fig.height=4) 

# import and wrangling packages
library(here) # here() creates the absolute path to the file
library(janitor) # clean_names to clean col labels of imported data
library(readxl) # import excel
library(data.table) # make data.frames data.tables

# analysis packages
library(emmeans) # get estimated marginal means and CIs, used for plot

# plotting packages
library(ggplot2) # ggplot environment
library(ggpubr) # publication ready plots
library(cowplot) # combine plots

here <- here::here # make sure ``here` uses `here::here`

# relative paths to project folders
data_folder <- "data" # path to data that are imported
output_folder <- "output" # path to data that are saved
image_folder <- "images"
```

Importing data into R can be a struggle for new R users and, unfortunately, most online "how to import" sources give easy but superficial methods that don't follow best practices for increasing reproducibility or do not allow flexible organization of files within a project.

(TL;DR -- use `here()` from the here package)

1. `df <- read.table(file="clipboard")` imports data copied to the clipboard, from an Excel/Sheets file or from an open text file. For this to be semi-reproducible, a comment specifying the filename, worksheet and range that was copied is necessary. More problematic (catastrophically so for reproducibility), is, how does a researcher know that they highlighted and copied the correct range in the Excel sheet? 
2. `df <- read.csv(file.choose())` opens the familiar "open file" dialog box, which lets the user navigate to the file of choice. For this to be semi-reproducible, a comment specifying the filename to import is necessary. The catastrophic problem (for reproducibility) is, how does a researcher know which file was actually opened during an R session? The researcher might think they opened "walker_maine_bee_data_clean_androscoggin.csv" but mistakenly opened "walker_maine_bee_data_clean_aroostook.csv".
3. `df <- read.table(file="my_data.txt")` and `df <- read_excel(file="my_data.xlsx")` are reproducible because the filename is explicitly specified. But, this method requires that "my_data" is physically located in the same folder as the file containing the R script (the .Rmd file in our case) and this violates the best practice of clean project organization with different folders for the different kinds of files (data, R scripts, images, manuscript text, etc.).
4. R Studio has an elegant import tool in the environment pane that opens a custom dialog box that allows the researcher to navigate to the file, and to specify what part of the file to import, such as the specific sheet and range for an Excel file. This has the same reproducibility issues as #1 and #2 but R Studio includes the equivalent script, which adds all relevant information for reproducility. One then simply copies and pastes this script into a code chunk and voila! The next time the script is run, the data can be imported from the script without using menus and dialog boxes. Except that..the script does not seem to take into account that the **working directory** of an R Markdown file is not the project folder but the folder containing the R Markdown file and so this two-step method fails. More personally, I'd prefer to run a chunk that quickly opens the data file instead of re-navigating through my file system and re-specifying the sheet and range every time I re-start the project in a new R session.

There are at least three solutions to the issues raised above, all requiring some understanding of **file paths** and directory structure in an operating system. A file such as "my_data.xlsx" has an **absolute** file path, which is the full address of the file (the filename is something like your house street number). The absolute file path of "my_data.xlsx" might be "/Users/jwalker/Documents/applied-biostatistics/data/my_data.xlsx". A **relative** file path is the file path from the **working directory**. In an R Studio project, the working directory is the project directory, which is the directory containing the .Rproj file. This will be the working directory of the console. *Importantly*, the working directory of an R Markdown code chunk is the folder containing the saved R Markdown file. An R Studio Notebook is an R Markdown file so the working directory of a notebook code chunk is the folder containing the saved notebook file. If an R Markdown file is located within the *rmd* folder, which is located within the project folder, then the relative file path to "my_file.xlsx" is "../data/my_file.xlsx". The ".." tells the file OS to move "up" into the parent directory (which is the project folder) and the "data" tells the file OS to move "down" into the data folder. These are put together into a single address using "/". The beauty of relative paths is that they remain the same -- and so do not break one's script -- if the project folder, and all of its contents including the *data* folder and the *rmd* folder, is moved to another location on the hard drive (say into a new "Research" folder). By contrast, the absolute file path changes, which breaks any old script.

The three solutions are

1. Create a **relative path** to the file using something like `file_path <- "../data/my_data.xlsx"`. This should *always* work but it fails on some computers. For example, if the project folder is on a Windows OS (but not Mac OS) desktop, the assigned relative address doesn't seem to look in the folder containing the file.
2. Create a setup chunk that reroutes the working directory to the project folder using the script

```{r, eval=FALSE}
# use this in a chuck called "setup" to force the working directory to be
# at the level of the project file.
knitr::opts_knit$set(root.dir = rprojroot::find_rstudio_root_file())
```

For this to work, the chunk has to be named “setup”, that is, the text inside the curly brackets at the top of the chunk should be “r setup”. Then, with this chunk, the relative file path is `file_path <- "../data/my_data.xlsx"` if "my_data.xlsx" is immediately inside the data folder which is immediately inside the project folder. This should work on any machine, and should work even if a project folder is moved.

3. Use the function `here()`. The most robust solution seems to be using the function `here()` from the here package. The function works something like this

```{r, eval=FALSE}
data_folder <- "data" # path to data that are imported
file_name <- "my_data.xlsx"
file_path <- here(data_folder, file_name) # paste together parts of the address
my_file <- read_excel(file = file_path)
```

`here()` creates an absolute path, but one that is created on the fly, and will change (or should change) correctly if the project folder is moved on the same machine, to a cloud drive, or to another machine altogether.

## Learning from this chapter

It will be easiest to learn from this chapter by starting with a clean R Markdown file for the chapter. Create a new R Markdown file and save it to the "Rmd" folder of your "applied biostats" project.

**Important:** import/export scripts will not work properly until the file is saved! Get in the habit of creating the file, saving it immediately, and saving it often.

1. Keep the first chunk that includes the script `knitr::opts_chunk$set(echo = TRUE)`
2. Delete all text below this chunk, starting with the header "## R Markdown"
3. Copy and paste this chunk into your setup chunk

```{r data-setup-2, eval=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# import and wrangling packages
library(here) # here() creates the absolute path to the file
library(janitor) # clean_names to clean col labels of imported data
library(readxl) # import excel
library(data.table) # use data.table for wrangling

# analysis packages
library(emmeans) # get estimated marginal means and CIs, used for plot

# plotting packages
library(ggplot2) # ggplot environment
library(ggpubr) # publication ready plots

here <- here::here # make sure ``here` uses `here::here`

# relative paths to project folders
data_folder <- "data" # path to data that are imported
output_folder <- "output" # path to data that are saved
image_folder <- "images"
```

Note on the script above
1. Be kind to the future you by loading only the packages necessary for the code in the R Markdown file that you are working on. If your default is to load everything, the future you will be confused why something was installed.
2. Be kind to the future you by commenting on why a package is loaded; usually this is a specific function from the package
3. `here <- here::here` is my favorite script ever. What is it doing? One can read this as "assign the function here from the here package to the object here" (this is reading the script right to left). Why do this? It turns out the multiple packages define a function called "here". If any of these packages are loaded after the here package, then `here` from here won't work -- it will be replaced by `here` from the more recently loaded package. To make sure that `here` uses the function from the here package, I simply reassign `here` from the here package to the object "here" *after* loading in all packages.

## Working in R {#data-working-in-r}
### Importing data
#### Excel file
The Excel dataset is from an experiment on the growth response of zebra finch chicks to an incubation call that presumably signals "hot environment" to the embryos ([Mariette, M.M. and Buchanan, K.L., 2016. Prenatal acoustic communication programs offspring for high posthatching temperatures in a songbird. Science, 353(6301), pp.812-814](http://science.sciencemag.org/content/353/6301/812){target="_blank"}). The source file is from the Dryad Repository here: 

**file name**: "allDatasetsMarietteBuchanan2016.xls"

**source**: https://datadryad.org/stash/dataset/doi:10.5061/dryad.v8969

Steps

1. Copy the title of the article, which is "Prenatal acoustic communication programs offspring for high post-hatching temperatures in a songbird"
2. Create a new folder within the "data" folder. Name the folder the title of the paper by pasting the name from the clipboard. This is the "data from" folder, since it contains the data from the publication with the title of the folder.
3. Download the .xls file into this folder

A .xls file is an old (pre 2007) Microsoft Excel file type. It is a binary file and can only be opened into a readable format with software that knows how to translate the proprietary format. The more modern Excel file type is .xlsx, which contains within it multiple xml components. An xml file is a text file, and so contains readable content, but the content is xml code to display something. In general, I am a big advocate of archiving stuff as text files (manuscripts, data, scripts, blog posts) because these will *always* be readable by future software. Microsoft Excel is not likely to die anytime soon and software that can read .xls and especially .xlsx files (again, .xlsx files are text files) is even less likely to disappear but we can feel even more confident if data are archived as text files. That said, a single microsoft excel file with multiple sheets is an efficient method for distributing data and the readxl package provides excellent tools for reading different sheets of a single .xls or .xlsx file.

The code below uses the function `read_excel()` from the package readxl. More about the amazing power of this package is the [tidyverse page](https://readxl.tidyverse.org){target="_blank"} and [chapter 11](http://r4ds.had.co.nz/data-import.html){target="_blank"} in the *R for Data Science* book.

```{r data-chick-xls}
folder <- "Prenatal acoustic communication programs offspring for high post-hatching temperatures in a songbird"
fn <- "allDatasetsMarietteBuchanan2016.xls"
file_path <- here(data_folder, folder, fn)

chick <- read_excel(file_path, 
                    sheet = "nestlingMass")
chick <- clean_names(chick) # clean column names
chick <- data.table(chick) # convert to data.table

# View(chick)
```

This text will consistently uses this protocol for storing and retrieving downloaded files. The final line of the chunk is commented out. I do this so that it does not run when the R Markdown sheet is knitted. But I can highlight "View(chick") with the cursor and "Run selected line(s)" from the "Run" menu and get a new tab with a spreadsheet like view of the imported data.

The first three lines in the script above creates the directory path to the file. This path includes three variables

1. `data_folder` -- assigned to "data" in the setup chunk. "data" is a folder within the project folder that contains (or will contain) all datasets for this text. The data come from many different published papers, so the data for each publication gets its own folder within "data".
2. `folder` -- the name of the "data from" folder within "data" containing the data files. In this text, these folder names will always be the name of the published paper.
3. `filename` -- the name of the file to read. There may be multiple data files within the publication's data folder.

These are all put together into the absolute path using the function `here()` from the here package. Take a look at the value of file_path to confirm.

The next three lines (starting with `chick <-`)

1. `read_excel` imports the data and assign this to a data.frame named `chick`
2. `clean_names` cleans the column names of `chick`
3. `data.table` converts `chick` to a data.table. A data.table is a data.frame with magical properties.

In steps 2 and 3, the functions take the data.frame and process it in some way and then assigned the processed data.frame to an object that has the same name (`chick`). This script can be made slightly more "elegant" using the "pipe" operator `%>%`.

```{r}
chick <- read_excel(file_path, 
                    sheet = "nestlingMass") %>% # import
  clean_names() %>% # clean the column names
  data.table() # convert to data.table
```

This is a single line of code containing three separate operations all piped together. A way to think about this is

1. The `read_excel` function imports the data from the file located at file_path and assigns this data to `chick`. The pipe operator then sends this to
2. the `clean_names` function, which cleans the column names of "chick". The pipe operator then sends this to
3. the `data.table` function, which convert the data.frame to a data.table.

A 3rd way to do this is with nested functions.

```{r}
chick <- data.table(clean_names(read_excel(file_path, 
                    sheet = "nestlingMass"))) # convert to data.table
```

Some R users think the piped code is more readable then the three separate functions. Maybe. I think we all agree that the nested functions are the least readable. I use pipes, but I think its worth using the three separate functions to learn what each is doing.

Let's back up to understand these steps, and especially the `clean_names` step. In your chunk with three separate lines of code, click on the first line, the one containing `read_excel`, so that the cursor is somewhere on this line. Then click on the "run" menu in the top right of the R Studio panel and choose the first item "Run selected line(s)". The file will be re-imported.

Look at the column names (or column headers in Excel lingo) of the imported data using `names` or `colnames` (yes, there are elebenty million ways to do anything in R) (`names` is very general in that it can be used to return the names of the parts of any list, while `colnames` is specific to matrix-like objects). Type this into the console, not into your R Markdown chunk:

```{r}
names(chick)
```

In general, it is bad practice to include spaces, parentheses, and special characters such as -, $ or ^, in the column names of a data frame because these increase handling costs later on. The best practice is to replace a blank with an underscore, for example `rank_in_nest`. Some coders separate words with a period (`rank.in.nest`). Others mash words together into a single word like this `rankinnest` but this should generally be avoided because the result can be hard to read. Finally, some coders use Caps to separate  words like this `RankInNest`. This is easier to read than simple concatenation but the underscore is the easiest to read.

The `clean_names` from the janitor package is a beautiful function to clean the column names of a data frame including replacing spaces with an underscore and stripping parentheses. The default clean includes changing any uppercase letter to lower case. Many coders like to work with all lowercase variable names to avoid having to hit the shift key. I am one of these.

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**Worst Practices** -- resist the temptation to change the column names in the data file, which reduces reproducibility. Leave original data files original. Always increase reproducibility!
</div>

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**colleague blues** -- Most researchers live in an Excel world and save data in a way that is efficient for computing stuff in Excel but not efficient for statistical analysis using R or other statistical computing software packages (with the exception of Graphpad Prism). Analyzing data will be much less frustrating if the data are saved in a format that facilitates analysis. Best practices for creating data files

1. https://www.youtube.com/watch?time_continue=309&v=Ry2xjTBtNFE -- An excellent video introduction to best practices for organizing data in a spreadsheet that will subsequently be analyzed by statistics software. 

2. Broman, K. W., & Woo, K. H. (2017). Data organization in spreadsheets (No. e3183v1). https://doi.org/10.7287/peerj.preprints.3183v1 -- An excelllent review of best practices for organizing data in a spreadsheet.
</div>

##### The read_excel function

`read_excel` is a beautifully flexible function because Excel. Data can be in different sheets and there can be different datasets within a single sheet. And, researchers tend to use Excel like a blackboard in that an Excel sheet often contains calculations such as means, standard deviations and t-tests. When using `read_excel` it is important to send the function enough information to read the correct data. For the chick data, if we simply used

```{r}
chick <- read_excel(file_path) %>%
  clean_names() %>%
  data.table()
```

without specifying the sheet, `read_excel` defaults to reading the first sheet ("OccurrenceIncubationCall"), which is not what we wanted. We can specify the exact range to important using the `range =` argument

```{r}
chick <- read_excel(file_path, 
                    sheet = "nestlingMass",
                    range = "A1:Q131") %>% # import
  clean_names() %>% # clean the column names
  data.table() # convert to data.table

```

This isn't necessary for these data because the "nestlingMass" sheet contains only a matrix of data and not extraneous information and the `read_excel` function is smart enough to figure this out. For many of the data sets in wet bench experimental biology, the range argument will be crucial because multiple datasets are archived on a single sheet.

##### Explore with plots

Just for fun, let's plot the data and reproduce something close to Fig. 2A and B. We are using the `qplot` function, which is from the ggplot2 package. qplots are quick plots -- something you want to do to quickly look at data but don't want to turn into a publication quality plot.

```{r data-plotchickdata}
qplot(x = nest_temperature_above_ambient,
      y = day13_mass,
      data = chick[playback_treatment == "treat"]) +
  geom_smooth(method = "lm")

qplot(x = nest_temperature_above_ambient,
      y = day13_mass,
      data = chick[playback_treatment == "cont"]) +
  geom_smooth(method = "lm")

```

Notes on the code to make these plots

1. the names of the columns to use as the x and y axes were not embedded into quotes (e.g. "nest_temperature_above_ambient"). Sometimes a column name has to be in quotes and sometimes not. In general, if the column name is sent to the function as a list (even if its a list with a single item), the names need to be in quotes. Regardless, remember this when you are debugging.
2. The first plot includes only the subset of data in which the value of `playback_treatment` is "treat". Similarly, the second plot includes only the subset of data in which the value of `playback_treatment` is "cont". I have sent a subset of the data to the plot function. There are elebenty million ways to subset data in R, I have done it the "data.table way".
3. each argument in the qplot function is on a separate line (created by adding a return after the comma) and the `geom_smooth` function is on a new line. This just makes the function more readable then not doing this. What do you think?

```{r, eval = FALSE}
qplot(x = nest_temperature_above_ambient, y = day13_mass, data = chick[playback_treatment == "treat"]) + geom_smooth(method = "lm")
```

4. I have included the name of each argument. This isn't necessary but it makes the function more readable *and* avoids potential bugs. The arguments without the argument names looks like this

```{r eval = FALSE}
qplot(nest_temperature_above_ambient, day13_mass, data = chick[playback_treatment == "treat"]) + geom_smooth(method = "lm")
```

#### Text file
The example dataset comes from an experiment on the effect of [neonicotinoid pesticides on bumble bee colony growth](http://science.sciencemag.org/content/early/2012/03/28/science.1215025){target="_blank"}.

**file name**: "Whitehorn, O'Connor, Wackers, Goulson (2012) Data from 'Neonicotinoid pesticide reduces bumblebee colony growth and queen production'.csv.csv"

Yes the name of the file has both single quotes in the file name and ".csv" as part of the file name so that, including the extension, the end of the name is ".csv.csv". This is not a good file name

**source**: https://datadryad.org//resource/doi:10.5061/dryad.1805c973

Steps

1. Copy the title of the paper title, which is "Neonicotinoid pesticide reduces bumblebee colony growth and queen production"
2. Create a new folder within "data". Name the folder the title of the paper by pasting from the clipboard. This is the "data from" folder, since it contains the data from the publication with the title of the folder.
3. Download the .csv file into this folder

A .csv file is a text file that is comma-delimted, which means that the entries of a row are separated by commas. A text file is readable by any text editor software and most other kinds of software. Datasets that are stored as text files are typically saved as either .csv (where the entries of a row are separated by commas) or .txt (where the entries are separated by tabs). The base R way to read a .csv file is using `read.csv`. The `read.table` function is more versatile, as the delimiter can be specified. The function `fread()` from the data.table package is fast, smart, and flexible. It is smart in the sense that it guesses what the delimter is. Unfortunately, because of spaces in the column labels for this file, fread guesses incorrectly (another reason why spaces in column labels should be avoided). To overcome this, the statement below specifies that the file contains a "header" (a line containing column labels)

```{r data-bee-csv}
folder <- "Neonicotinoid pesticide reduces bumblebee colony growth and queen production"
filename <- "Whitehorn, O'Connor, Wackers, Goulson (2012) Data from 'Neonicotinoid pesticide reduces bumblebee colony growth and queen production'.csv.csv"
file_path <- here(data_folder, folder, filename)
bee <- fread(file_path, header=TRUE) %>%
  clean_names()
```

Here, as with the import of the Excel file, the first three lines create the directory path to the file. There is no need to pipe `bee` to `data.table()` because `fread` automatically importants the data as a data.table. It does not need to be converted.

Here is a reproduction of Fig 2 from the journal article using the `ggbarplot` function from the `ggpubr` package.

```{r data-plot-bee-fig-2}
bee[, treatment := factor(treatment, c("Control", "Low", "High"))] # reorder factor levels

ggbarplot(data=bee,
          x="treatment", 
          y="new_queens", 
          add = "mean_se")
```


#### Troubleshooting file import

If you get an error that starts with "Error: `path` does not exist:" then R is not "seeing" your specified file given the path you've given it.

1. Make sure you loaded the package `here` in a "setup" chunk and that you have run the chunk
2. Make sure you have assigned `data_folder <- "data"` in the setup chunk and have run the chunk.
3. Make sure your "data" folder is *one level* inside your project folder. "one level" means it is not buried deeper inside other folders within the project folder.
4. Make sure your "data from" folder (the folder with the title of the publication) is one level inside your "data" folder
5. Make sure your data file is one level inside the correct "data from" folder.
6. **Bug alert** Make sure you have the name of the "data from ..." folder correct in your script. Do not type the name of the folder. Instead, go to the finder and highlight the folder containing the data file, copy the name, return to the R markdown script, type `folder <- ""` and then paste the clipboard (the name of the folder) in between the quote marks.
7. **Bug alert** Make sure the file name is correct in the script. As with the folder name, I go to the finder and copy the file name and paste it in place. In Windows use ctrl-a instead of ctrl-c to copy the full filename including the extension.

More generally, Humans are very good at understanding misspelled and OdDLy capitalized words but the R language (or any computer language) is very literal. R is **case sensitive** (some programming languages are not). "Prenatal acoustic communication", "Prenatal Acoustic Communication", and "prenatal acoustic communication" are all different values. Spelling AND capitalization have to be perfect, not simply close. Spelling includes *spaces*. A frequent bug is a file name typed as "Prenatal acoustic communication" when the actual name is "Prenatal acoustic  communication". Can you spot the bug? The original (what we need to copy) has two spaces between "acoustic" and "communication" while the incorrect copy has only one.

Spelling bugs are avoided by simply copying and pasting names of folders, names of files, column names of data frames, and level names of factors, which leads to a general rule of R scripting...

#### Rule number one in R scripting {# rule1}

**Always copy and paste any text that will be inserted into quotes**

Do not try to type it out. You have been warned.

## Data wrangling

Data archived in Excel spreadsheets, at least in wet-bench experimental biology projects, are generally not in a format this is readily analyzed in R, or any statistical software other than perhaps Graphpad Prism. Use these examples as templates for how to import and wrangle Excel-archived data in your project.

### Reshaping data -- Wide to long
#### Wide to long -- Adipsin data

Source: [Adipsin preserves beta cells in diabetic mice and associates with protection from type 2 diabetes in humans](https://www.nature.com/articles/s41591-019-0610-4){target="_blank"}

[Public source](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7256970/){target="_blank"} -- the Adipsin paper is behind a paywall. A public source of the paper from NIH is available.

[Link to source data](https://www.nature.com/articles/s41591-019-0610-4#Sec39){target="_blank"}

```{r, out.width="60%", echo=FALSE}
image_name <- "adipsin_1k_wide.png"
image_path <- here(image_folder, image_name)
knitr::include_graphics(image_path)
```

Fig. 1k of the Adipsin paper presents a bar plot of the glucose uptake in response to control (GFP) or [adipsin](https://en.wikipedia.org/wiki/Factor_D){target="_blank"} treatment. A screenshot of the Excel-archived data is shown above. The data are in **wide format**. In wide-format, the values of a single variable (here, this is glucose uptake level) are given in separate columns for each treatment level (group). The values for the GFP group are in Column A and the values for the Adipsin group are in Column B. Wide format is efficient for computations in a spreadsheet, such as computing means and standard deviations of columns of data, and for plotting.

For most statistical analyses of experimental data in R (and most statistics software), all values of a single variable should be in a single column. This is called **long format**. I've manually rearranged the data from the archived spread sheet into long format by stacking each group's values into a single column, shown in the screen capture below. All values of glucose uptake are in a single column. In long format, there needs to be a way to identify which values belong to which group and this is achieved here with column "treatment". In adition to the treatment column.

```{r, out.width="60%", echo=FALSE}
image_name <- "adipsin_1k_long.png"
image_path <- here(image_folder, image_name)
knitr::include_graphics(image_path)
```

The difference between wide and long also reflects how we think about statistical analysis. When we do a *t*-test to compare the means of glucose uptake between GFP and Adipsin groups, we might think we have two things: the set of glucose uptake values for the GFP group and the set of values for the Adipsin group. When we fit a linear model, we also have two things, the variable *treatment* containing treatment level assignment and the variable *glucose_uptake* containing the glucose uptake values. In wide format, there is nothing to suggest that *treatment* is a variable.

There are many functions to tidy data from wide to long. `melt` from the data.table package is especially useful. It is data.table's version of melt from the reshape2 package.

The major arguments of `data.table::melt` are

`melt(data, id.vars, measure.vars, variable.name, value.name)`

`melt` takes the data in the columns listed in `measure.vars` and stacks these into a single column named `value.name`. The names of the columns in `measure.vars` are the values of the elements in a new column named `variable.name`. The elements of any column in `id.vars` are repeated *p* times, where *p* is the number of columns that were stacked.

Let's melt the three different response variables of the adipsin data and `merge` them into a single data.table. There are several ways to combine data sets including `merge` and `cbind`. We'll compare these later.

```{r}
file_folder <- "Adipsin preserves beta cells in diabetic mice and associates with protection from type 2 diabetes in humans"
fn <- "41591_2019_610_MOESM3_ESM.xlsx"
file_path <- here(data_folder, file_folder, fn)

treatment_levels <- c("db/db-GFP", "db/db-Adipsin")

# as separate line
fig_1k_wide <- read_excel(file_path,
                    sheet = "Figure 1k",
                    range = "A3:B9")
fig_1k_wide <- data.table(fig_1k_wide)
fig_1k <- melt(fig_1k_wide,
               measure.vars = treatment_levels,
               variable.name = "treatment",
               value.name = "glucose_uptake")

# or piped -- which do you prefer?
fig_1k <- read_excel(file_path,
                    sheet = "Figure 1k",
                    range = "A3:B9") %>%
  data.table() %>%
  melt(measure.vars = treatment_levels,
       variable.name = "treatment",
       value.name = "glucose_uptake")

# View(fig_1k) # highlight without the comment sign and "run selected lines()" to view
```

A pretty-good-plot using the `ggpubr` package

```{r, warning=FALSE}
# put warning=FALSE into the chunk header to supress the warning

gg <- ggstripchart(x = "treatment",
                   y = "glucose_uptake",
                   add = "mean_se",
                   data = fig_1k)

gg
```


#### Wide to long -- Enteric nervous system data

Source: [Rolig, A. S., Mittge, E. K., Ganz, J., Troll, J. V., Melancon, E., Wiles, T. J., … Guillemin, K. (2017). The enteric nervous system promotes intestinal health by constraining microbiota composition. PLOS Biology, 15(2), e2000689. ](https://doi.org/10.1371/journal.pbio.2000689)

[Source data](https://doi.org/10.1371/journal.pbio.2000689.s008)

Let's import and reshape the data for figure 2d. Look at the excel file and the data in Fig. 2d. There is a single treament with four levels, but the authors have organized the data in each level in separate columns and used the column header as the level name.

```{r, out.width="60%", echo=FALSE}
image_name <- "enteric_fig2d_wide.png"
image_path <- here(image_folder, image_name)
knitr::include_graphics(image_path)
```

Let's melt the data from wide to long by stacking the four columns into a single column "neutrophil_count" and adding a treatment column identifying the group.

```{r}
folder <- "The enteric nervous system promotes intestinal health by constraining microbiota composition"
filename <- "journal.pbio.2000689.s008.xlsx"
file_path <- here(data_folder, folder, filename)

# figure 2D data
sheet_i <- "Figure 2"
range_i <- "F2:I24"
fig_2d_wide <- read_excel(file_path, sheet=sheet_i, range=range_i) %>%
  clean_names() %>%
  data.table()

# change column names by replacing without "_donor" in each name
# these new column names will become the levels of the treatment factor
new_colnames <- c("gf", "wt", "sox10", "iap_mo")
setnames(fig_2d_wide, old=colnames(fig_2d_wide), new=new_colnames)

# wide to long
fig_2d <- melt(fig_2d_wide, 
              measure.vars=colnames(fig_2d_wide), 
              variable.name="treatment", 
              value.name="neutrophil_count")

# omit empty rows
fig_2d <- na.omit(fig_2d)

# re-order factors
fig_2d[, treatment := factor(treatment,
                         levels = c("wt", "gf", "sox10", "iap_mo"))]

# View(fig_2d)
```

To learn (instead of just copy and modify), it's best to do this in steps and not run the whole chunk. At each step, look at the result using View. The script above includes three extra wrangling steps.

1. Changing column names in fig_2d_wide. The column names in wide format will become the treatment level names of the *treatment* factor after reshaping. It will be easier down the road if these names are shorter and the "_donor" in each name is redundant. The `setnames` function renames the column names.

2. For these data, the number of measures within the different treatments differs and, as a consequence, there are multiple cells with `NA` which indicates a missing value. `View(fig_2d_wide)` (this can be typed in the console) to see this. After reshaping to long format (fig_2d), the rows with missing values become empty rows -- there is no useful information in them (View this). To see this, re-run the lines of the chunk up to the line "# omit empty rows". The `na.omit` function deletes any row with missing values. Here, this deletes these information-less rows. Be very careful with `na.omit`. You do not want to delete rows of data that contain information you want.
3. For both analysis and plots, we want to compare values to the control level, which is named "wt" for the fig_2d data. That is, we want "wt" to be the *reference* level. To achieve this, the levels of the factor *treatment* need to be re-ordered using the `levels` argument. (note, I typically do not add "levels = ", but simply pass the list of levels)

#### Wide to long -- bee data

The example above is pretty easy, because the all columns in the original data frame are melted (stacked). Here is an example in which only a subset of columns are stacked. In addition, only a subset of the remaining columns are retained in the long format data frame. The data are from Panel A of supplement Fig. 8 (https://journals.plos.org/plosbiology/article/file?type=supplementary&id=info:doi/10.1371/journal.pbio.2003467.s019) from

Source: [Kešnerová, L., Mars, R.A., Ellegaard, K.M., Troilo, M., Sauer, U. and Engel, P., 2017. Disentangling metabolic functions of bacteria in the honey bee gut. PLoS biology, 15(12), p.e2003467.](https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.2003467)

[Source data](https://doi.org/10.1371/journal.pbio.2003467.s001)

```{r}
folder <- "Data from Disentangling metabolic functions of bacteria in the honey bee gut"
filename <- "journal.pbio.2003467.s001.xlsx"

# figure 2D data
sheet_i <- "S8 Fig"
range_i <- "A2:H12"
file_path <- here(data_folder, folder, filename)
fig_s8a_wide <- read_excel(file_path,
                      sheet=sheet_i,
                      range=range_i) %>%
  clean_names() %>%
  data.table()

# wide to long
stack_cols <- paste0("replicate", 1:5)
fig_s8a <- melt(fig_s8a_wide,
              id.vars = c("media", "time_h"),
              measure.vars = stack_cols, 
              variable.name = "Replicate", 
              value.name = "OD600") # measure of absorbance at 600nm
```

#### Wide to long -- stacking multiple sets of columns

This example comes from my lab, where a student measured sprint speed in each fish three times prior to treatment and three times following treatment. The wide format data looked something like this

```{r}
set.seed(1)
fd_wide <- data.table(fish_ID=paste0("fish",1:4),
                      treatment=rep(c("cn", "tr"), each=2),
                      length=rnorm(4, 12, 2),
                      pre_1=rnorm(4, 50, 5),
                      pre_2=rnorm(4, 50, 5),
                      pre_3=rnorm(4, 50, 5),
                      post_1=rnorm(4, 50, 5),
                      post_2=rnorm(4, 50, 5),
                      post_3=rnorm(4, 50, 5)
                      )
knitr::kable(fd_wide, digits=1)
```

To analyze the response (post-treatment sprint) adjusted for pre-treatment sprint, the three pre-treatment sprint measures need to be stacked into a single column and the three post-treatment measures need to be stacked into a single column. This is easy using `melt` from the data.table package.

```{r}
pre_cols <- paste("pre", 1:3, sep="_")
post_cols <- paste("post", 1:3, sep="_")
fd <- melt(fd_wide,
           id.vars=c("fish_ID", "treatment", "length"),
           measure.vars=list(pre_cols, post_cols),
           variable.name="Order",
           value.name=c("sprint_pre", "sprint_post"))
knitr::kable(fd, digits=1)
```

### Reshaping data -- Transpose (turning the columns into rows)
#### Transpose -- PI3K inhibitors data

Source: [Suppression of insulin feedback enhances the efficacy of PI3K inhibitors](https://www.nature.com/articles/s41586-018-0343-4)

[Source data](https://www.nature.com/articles/s41586-018-0343-4#Sec15)

Figure 3A of this publication is a plot of blood glucose level taken on the same individual mice from four treatment groups over six time periods. Data on a single variable such as blood glucose, taken on the same individual at multiple time points, are known as **longitudial data** but are often mistakenly called **repeated measures data**. There are mulitple ways to analyze longitudinal data, some goood, some less good. There are two reasonable ways to archive longitudinal data for analysis in R. The Excel-archived data for Figure 3A is neither. A screen capture of two of the four treatment groups is shown below.

```{r, out.width="60%", echo=FALSE}
image_name <- "pi3k-inhibitors.png"
image_path <- here(image_folder, image_name)
knitr::include_graphics(image_path)

```

In the archived data the individual mice are in columns. The measure at each time point is in rows. And the treatment group is in blocks. Typical data for analysis in R should have the individual mice in rows and each variable in columns (an exception in experimental biology is omics data, such as RNA expression levels. Many packages with functions to analyze these data have the genes on each row and the individual on each column). The Figure 3A data are turned on its side. We need to **transpose** the data, or rotate the matrix 90 degrees (make the columns rows and the rows columns) to turn the data into wide format. From this we can create a new data.table with the data in long format.

```{r}
folder <- "Suppression of insulin feedback enhances the efficacy of PI3K inhibitors"
filename <- "41586_2018_343_MOESM6_ESM.xlsx"
file_path <- here(data_folder, folder, filename)
pi3k_side <- read_excel(file_path,
                    sheet = "Figure 3A (Blood Glucose)",
                    range = "A2:U7",
                    col_names = FALSE) %>%
  data.table()

# give columns names as the treatment of each mouse
# verify n=5 per group
treatment_levels <- c("Chow", "Ketogenic", "Metformin", "SGLT2i")
colnames(pi3k_side) <- c("time",
                         rep(treatment_levels, each = 5))

# transpose
# keep colnames in "side" as values of treatment col in "wide"
# make values of "time" in "side" the colnames in "wide"
pi3k_wide <- transpose(pi3k_side,
                       keep.names = "treatment", 
                       make.names = "time")

# make a baseline column 
pi3k_wide[, glucose_0 := get("0")]

# make-up a mouse id for each mouse
pi3k_wide[, id := paste(treatment, 1:.N, sep = "_"), by = treatment]

# make treatement a factor with "chow" as reference
pi3k_wide[, treatment := factor(treatment, treatment_levels)]

# make a long version
pi3k_long <- melt(pi3k_wide,
                  id.vars = c("treatment", "id", "glucose_0"),
                  variable.name = "time",
                  value.name = "glucose")
```
Notes

1. Read the comments on the usage of the `keep.names` and `make.names` arguments of `transpose`. These are powerful.
2. pi3k_wide has column names that are times (in minutes). This presents wrangling problems (column names shouldn't be numbers. Here it is useful to create the long format data.table with a time column of numbers). For example, the code above creates copies the column "0" into a new column "glucose_0" using `glucose_0 := get("0")`. Had the code been `glucose_0 := "0"`, all values would be the character "0". Had the code been `glucose_0 := 0`, all values would be the number 0. `get` looks for the column with the name of whatever is inside the parentheses.

Let's do a quick plot to examine the data

```{r}
qplot(x = time,
      y = glucose,
      data = pi3k_long,
      color = treatment) +
  geom_line(aes(group = id))
```

### Combining data

Source [Bak, A.M., Vendelbo, M.H., Christensen, B., Viggers, R., Bibby, B.M., Rungby, J., Jørgensen, J.O.L., Møller, N. and Jessen, N., 2018. Prolonged fasting-induced metabolic signatures in human skeletal muscle of lean and obese men. PloS one, 13(9), p.e0200817.](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0200817)

[Source data](https://datadryad.org/stash/dataset/doi:10.5061/dryad.6121hj7)

The data are from a randomized **crossover** design where 18 men (9 lean and 9 obese) were measured for multiple metabolic markers at two times: 1) in a post-absorptive state after 12 hours overnight fast, and 2) in a prolonged fasting state after 72 hours of fasting. In addition, at each time point, metabolic markers were measured prior to and after an insulin infusion. Here, we want to reproduce values in Table 2, which are measures of mean blood insulin and metabolite levels after 12 hours and 72 hours fasting in both the lean and obese groups.

A difficulty for the analyst is that the response data are in the "Table 2" sheet but the variable containing the assignment to "lean" or "obese" group is in the "Table 1" sheet. To analyze these response, the two datasets need to be combined into a single data frame. The important consideration when combining data is that *like is matched with like*. For the fasting dataset, "like" is the subject id, and we have some data for each subject id in Table 1 and other data for the same subject ids in Table 2. This means that we essentially want to glue the columns of table 2 to the columns of table 1 in a way that insures that the correct data for each subject id is on the same row. This is a bit more complicated for these data because Table 1 contains 18 data rows, one for each subject id and Table 2 contains 36 data rows, 2 for each subject id, because each subject has data measured at 12 hours and at 72 hours.

```{r data-import-glucose, echo=FALSE}
folder <- "Prolonged fasting-induced metabolic signatures in human skeletal muscle of lean and obese men"
filename <- "datafiles.xlsx"
file_path <- here(data_folder, folder, filename)
# read only subject id and group assignment (lean or obese) in Table 1
table1 <- data.table(read_excel(file_path, sheet="Table 1", range="a1:b19"))
# read all of Table 2
table2 <- data.table(read_excel(file_path, sheet="Table 2", range="a1:z37"))

# clean column names
table1 <- clean_names(table1)
table2 <- clean_names(table2)

# add group from table1 to table2 by matching the column "id", which
# identifies the subject.
fasting <- merge(table2, table1, by="id")
```

### Subsetting data

It is common to see researchers create multiple subsets of data for further processing. This practice should be be discouraged because the same variables will be in multiple data frames and it can be hard to keep track of any processing of variables in the different datasets. Instead, subset the data at the level of analysis.

There are many ways to subset data in R. Experienced users tend to divide up into those using base R, those using the tidyverse packages, or those using data.table. Learn one well. This book uses data.table. Before outlining usage in data.table, let's back up a bit and review different *indexing* systems.

* In Excel, rows are specified (or "indexed") by numbers and columns by letters. Every cell has an address, for example C2 is the cell in the 2nd row and 3rd column. Notice that in Excel, the column part of the address comes before the row part.
* In statistics, it is extremely common to use a system where $x_{ij}$ is the value of the element in the *i*th row and *j*th column of the matrix **X**. Notice that in this notatin, the row index (*i*) comes before the column index (*j*).
* In programming languages, including R, it is extremely common to use a system where my_data[i, j] is the value of the element in the *i*th row and *j*th column of the matrix-like object named "my_data" (such as a data frame in R).
* data.table explicitly refers to the row index and column index as *i* and *j*.

#### Specifying a subset of rows ("observations" or "cases")

A subset of rows is specified using either a list of row numbers or 

In a data.table, a subset of rows is specified using either a list of row numbers or a combination of comparison operators (==, !=, >, <, >=, <=, %in%) and Boolean logic operators (&, |, ! -- these are "and", "or", "not") as *i*.

Let's use the pi3k_long data from above to explore this. First, the plot of plasma glucose for all individuals in each treatment group across all time points.

```{r}
qplot(x = time,
      y = glucose,
      data = pi3k_long,
      color = treatment) +
  geom_line(aes(group = id))
```

`pi3k_long[treatment == "Chow",])` is the subset of rows in which entries in the column "treatment" take the value "Chow" using the "is equal" ("==") operator

```{r}
qplot(x = time,
      y = glucose,
      data = pi3k_long[treatment == "Chow",],
      color = treatment) +
  geom_line(aes(group = id))

```

And the subset of rows in which entries in the column "treatment" take any value but "Chow" using the "not equal" operator ("!=").
 
```{r}
qplot(x = time,
      y = glucose,
      data = pi3k_long[treatment != "Chow",],
      color = treatment) +
  geom_line(aes(group = id))

```

The subset of rows in which entries in the column "treatment" take either the value "Chow" or the value "SGLT2i" by combining two "is equal" ("==") operators using the OR ("|") boolean operator

```{r}
qplot(x = time,
      y = glucose,
      data = pi3k_long[treatment == "Chow" | treatment == "SGLT2i",],
      color = treatment) +
  geom_line(aes(group = id))

```

The subset of rows in which entries in the column "time" take either the value "30" or the value "60" using the "in a list" operator (%in%). The values in the "time" column look like integers but are actually treatment levels (which act like string or character variables).

```{r}
qplot(x = time,
      y = glucose,
      data = pi3k_long[time  %in% c("30", "60"),],
      color = treatment) +
  geom_line(aes(group = id))

```

The subset of rows in which entries in the column "time_c" are less than or equal to 60 using the "less than or equal to" operator AND the value in the treatment column is in the list ("Chow", "SGLT2i"). The two comparisons are combined with the AND ("&") Boolean operator. 

```{r}
pi3k_long[, time_c := as.numeric(as.character(time))]
qplot(x = time,
      y = glucose,
      data = pi3k_long[time_c <= 30 & treatment %in% c("Chow", "SGLT2i"),],
      color = treatment) +
  geom_line(aes(group = id))

```

The same result as above but using different operators. I would describe this as, the subset of rows in which entries in the column "time_c" are less than or equal to 60 using the "less than or equal to" operator AND the value in the treatment column is either "Chow" OR "SGLT2i". The two comparisons are combined with the AND ("&") Boolean operator. The order of operations is determined by the parentheses, as with all algebra.

```{r}
pi3k_long[, time_c := as.numeric(as.character(time))]
qplot(x = time,
      y = glucose,
      data = pi3k_long[time_c <= 30 & (treatment == "Chow" | treatment == "SGLT2i"),],
      color = treatment) +
  geom_line(aes(group = id))

```


### Wrangling columns
#### Creating new columns that are functions of values in existing columnes
#### Change the reference level of a factor
#### Converting a single column with all combinations of a 2 x 2 factorial experiment into two columns, each containing the two levels of a factor

Source: [Tauriello, D., Palomo-Ponce, S., Stork, D. et al. TGFβ drives immune evasion in genetically reconstituted colon cancer metastasis. Nature 554, 538–543](2018) doi:10.1038/nature25492

[Source data](https://www.nature.com/articles/nature25492#Sec23)

filename: "41586_2018_BFnature25492_MOESM10_ESM.xlsx"

sheet: "Fig. 4h-tumours"

```{r, echo=FALSE}
folder <- "TGF-beta drives immune evasion in genetically reconstituted colon cancer metastasis"
fn <- "41586_2018_BFnature25492_MOESM10_ESM.xlsx"
file_path <- here('data', folder, fn)
tumor <- read_excel(file_path, sheet="Fig. 4h-tumours") %>%
  clean_names()  %>%
  data.table()
tumor[, name := factor(name)]

```

The analysis of the data in Fig. 4h specifies a single $X$ variable "Treatment" with four levels (or groups): "Con", "Gal", "aPD-L1", and "Gal+aPD-L1". These levels indicate that the design is actually **factorial** with two factors, each with two levels. The first factor has levels "no Gal" and "Gal". The second factor has levels "no aPD-L1", "aPD-L1". The single column Treatment "flattens" the  2 X 2 factorial design to a 4 x 1 design. In general, we would want to analyze an experiment like this as factorial model, because this allows us to make inferences about the *interaction effect* between the two factors. For these inferences, we need a standard error, or a confidence interval, or a *p*-value of the estimate, which we can easily get from the factorial model. In order to analyze the data with a factorial model, we need to create two new columns -- one column is the factor variable containing the two levels of Gal and one column is the factor variable containing the two levels of aPD-L1.

```{r}
gal_levels <- c("no Gal", "Gal")
tumor[, gal := ifelse(treatment == "Gal" | treatment == "Gal+aPD-L1",
                      gal_levels[2],
                      gal_levels[1])]

apd_levels <- c("no aPD-L1", "aPD-L1")
tumor[, apdl1 := ifelse(treatment == "aPD-L1" | treatment == "Gal+aPD-L1",
                      apd_levels[2],
                      apd_levels[1])]

# re-order factor levels
tumor[, gal:=factor(gal, gal_levels)]
tumor[, apdl1:=factor(apdl1, apd_levels)]

```

A way to check the results to make sure that our conversion is correct is to compute the sampel size for the 2 x 2 combinations, but include the original treatment column in the `by` list.

```{r}
tumor[!is.na(num_positive_per_mm), .(N=.N), by=.(treatment, gal, apdl1)]
```

That looks good.

**Bug alert** If you break Rule #1, and type in the treatment level "Gal+aPD-L1" as "Gal + aPD-L1", then you will get new columns containing junk.

```{r, echo=FALSE}
tumor2 <- copy(tumor)
gal_levels <- c("no Gal", "Gal")
tumor2[, gal := ifelse(treatment == "Gal" | treatment == "Gal + aPD-L1",
                      gal_levels[2],
                      gal_levels[1])]

apd_levels <- c("no aPD-L1", "aPD-L1")
tumor2[, apdl1 := ifelse(treatment == "aPD-L1" | treatment == "Gal + aPD-L1",
                      apd_levels[2],
                      apd_levels[1])]

# re-order factor levels
tumor2[, gal:=factor(gal, gal_levels)]
tumor2[, apdl1:=factor(apdl1, apd_levels)]
tumor2[!is.na(num_positive_per_mm), .(N=.N), by=.(treatment, gal, apdl1)]

```

Remember Rule #1. Always copy and paste any text that will be inserted into quotes. This is easily done here by typing `unique(tumor$treatment)` into the console. This function returns the unique values of the column "treatment" of the data.table "tumor".

> unique(tumor$treatment)
[1] "Con"        "Gal"        "aPD-L1"     "Gal+aPD-L1"

Now, copy the name of a level and paste into your code. Repeat until done.

### Missing data

Source: [Deletion of Cdkn1b in ACI rats leads to increased proliferation and pregnancy-associated changes in the mammary gland due to perturbed systemic endocrine environment](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1008002)

[Source data](https://doi.org/10.1371/journal.pgen.1008002.s008)

Supplement Figure 1F of this paper shows weight as a function of age class and genotype for the whole body and 8 organs. There are some missing weights in the Excel-archived data. These missing data are designated with a minus "-" sign. To import these data in correctly, use the `na = ` argument in the `read_excel` function.

```{r}
file_folder <- "Deletion of Cdkn1b in ACI rats leads to increased proliferation and pregnancy-associated changes in the mammary gland due to perturbed systemic endocrine environment"
file_name <- "journal.pgen.1008002.s008.xlsx"
file_path <- here(data_folder, file_folder, file_name)
  
fig_s1f <- read_excel(file_path,
                    sheet = "all weights",
                    range = "A2:K57",
                    na = "-",
                    col_names = TRUE) %>%
  clean_names() %>%
  data.table()

fig_s1f[, genotype := factor(genotype, c("+/+", "-/-"))]
fig_s1f[, age_class := ifelse(age_at_sac_wks <= 6.0, "4-6", "8+")]

# View(fig_s1f)
```

Notes

1. In R, a value of "NA" represents missing.
1. The default value for `na = ` is an empty (or blank) cell (not a space but a cell that is empty).
2. `na = ` accepts a list of strings, for example `na = c("", "-99", "--")` that will all be read as na.

#### Handling missing data
##### Many base R functions used for summary measures require NA handling 

```{r}
mean(fig_s1f[, ovary]) # returns "NA"
mean(fig_s1f[, ovary], na.rm = TRUE) # returns the mean

sd(fig_s1f[, ovary]) # returns "NA"
sd(fig_s1f[, ovary], na.rm = TRUE) # returns the mean

sum(fig_s1f[, ovary]) # returns "NA"
sum(fig_s1f[, ovary], na.rm = TRUE) # returns the mean

```
There are many ways to get the sample size for a particular variable. Be careful if using `length()` which counts NA as part of the vector of values.

##### The !is.na function is useful

```{r}
length(fig_s1f[, ovary])
length(fig_s1f[!is.na(ovary), ovary])
```
Notes

1. `!is.na(ovary)` is taking the subset of rows of fig_s1f for which the value of "ovary" is not NA (!is.na is read "not is.na")

This is especially useful if you are creating your own code uses counts. Here I create a table of means, standard error of the mean, and 95% CIs of the mean for each genotype group. But first, this script generates the wrong N for each group (since there are missing values), although the mean and SD are correct.

```{r}
fig_s1f[, .(mean = mean(spleen, na.rm = TRUE),
            n = .N,
            sd = sd(spleen, na.rm = TRUE)),
        by = genotype]

```

To compute the correct *n*, which will be necessary for computing the SE and the CI, use `!is.na`

```{r}
spleen_summary <- fig_s1f[!is.na(spleen), .(mean = mean(spleen),
            n = .N,
            sd = sd(spleen)),
        by = genotype]
spleen_summary[, se := sd/sqrt(n)]
spleen_summary[, lower := mean + se*qt(.025, (n-1))]
spleen_summary[, upper := mean + se*qt(.975, (n-1))]
spleen_summary
```

##### ggplot functions automatically handle missing values

with a useful warning.

```{r}
qplot(x = body_wt_g_sac,
      y = spleen,
      color = genotype,
      data = fig_s1f)
```

##### Regression model functions (lm, glm, gls, etc.) handle missing values by default

Missing data in regression model functions such as `lm` are handled using the argument `na.action = ` and the default is "na.omit", which omits any rows that contain a missing value in one or more of the model variables (it includes rows if these contain missing values only in the columns not included in the model). It's as if the user took the subset of data including only the columns containing the model variables and then deleted any row with missing values.

Here is the coefficient table of the fit model object that did not explictly tell the `lm` function how to handle missing data.

```{r}
m1 <- lm(spleen ~ body_wt_g_sac + genotype,
         data = fig_s1f)
coef(summary(m1))
```
Here is the coefficient table of the fit model object that did explicitly tell `lm` how to handle missing data, using the argument `na.action = "na.exclude"`. These coefficient tables are the same.

```{r}
m2 <- lm(spleen ~ body_wt_g_sac + genotype,
         data = fig_s1f,
         na.action = "na.exclude")
coef(summary(m2))
```

#### But...beware of fitted, predicted, or residual values from regression model functions unless you've explictly told the function how to handle missing values

**Use `na.action = "na.exclude"`** if you want to add the fitted (or predicted) values or residuals as new columns in the original data object (fig_sf1). Compare the length of the fitted values vector from models m1 (using the default "na.omit") and m2 (using the "na.exclude").

```{r}
length(fitted(m1))
length(fitted(m2))

```
There are 55 observations (rows in the data) but only 42 complete rows with no missing values. The vector of fitted values from m1 has 42 fitted values. The vector of fitted values from m2 has 55 elements, the 42 fitted values plus 13 NA elements.

This is important if we want to do something like add the fitted values (or residuals, or some function of these) to the original data object (fig_sf1). Here I compute the spleen weights adjusted to the mean body weight of the control ("+/+") group using the residuals from m1 and m2.
 
```{r}
mean_x_control <- mean(fig_s1f[genotype == "+/+", body_wt_g_sac])
b <- coef(m1)
fig_s1f[, spleen_adj_m1 := b[1] +
          b[2]*mean_x_control +
          b[3]*(as.integer(genotype)-1 +
          residuals(m1))]
fig_s1f[, spleen_adj_m2 := b[1] +
          b[2]*mean_x_control +
          b[3]*(as.integer(genotype)-1 +
          residuals(m2))]
# View(fig_s1f)
```

The computation of "spleen_adj_m1" returns a warning that the values of `residuals(m1)` were recycled (the first 42 elements of the new column were filled with the 42 residuals and the last 13 elements of the new column were filled with the first 13 residuals) -- after the first row of missing data, all of these computed adjusted values are wrong. Using `residuals(m2`), the adjusted values are matched to the correct row and the rows with missing variables do not have an adjusted value (because there is no residual to compute this).

## Saving data

For many projects, it is uncommon to save data. I might save simulated data if it takes a long time (tens of minutes to hours or even days) to generate these and I simply want to work with the simulated data in the future and not have to regenerate it. Or I might save processed data if it takes a long time to import and process and I want to analyze the processed data in the future and not have to re-import and process it.

If the data will only be used in this or future R projects, the data can be saved as an R object using `saveRDS()`

```{r data-rds}
outfile_name <- "Prenatal acoustic communication programs offspring for high post-hatching temperatures in a songbird.Rds"
save_file_path <- here(output_folder, outfile_name)
saveRDS(object = chick, file = save_file_path)

# to read this use
chick <- readRDS(save_file_path)
```

Reading a large .Rds file is very fast compared to reading the same data stored as a text file. However, if the data need to be imported into some other software, such as a spreadsheet, then save the data as a text file.

```{r data-save}
# save the data to output folder

# tab delimited
outfile_name <- "Prenatal acoustic communication programs offspring for high post-hatching temperatures in a songbird.txt"
save_file_path <- here(output_folder, outfile_name)
write.table(chick, save_file_path, sep="\t", quote=FALSE)

# comma delimited
outfile_name <- "Prenatal acoustic communication programs offspring for high post-hatching temperatures in a songbird.csv"
save_file_path <- here(output_folder, outfile_name)
write.table(chick, save_file_path, sep=",", quote=FALSE)

```

Look at your project directory to make sure the file is where it should be! We used `write.table()` to create a tab-delimited text file using `sep="\t"` to specify tabs to separate the row elements. "\t" is the standard character string for a tab. Check in your output folder and open the file in a text editor.

## Exercises

1. Import and pretty-good-plot the data for Figure 2i of the Adipsin paper. You will need to download and archive the Excel file for "Figure 2". Store this within the "Adipsin preserves beta cells..." folder.

```{r, echo=FALSE, message=FALSE, warning=FALSE}
file_folder <- "Adipsin preserves beta cells in diabetic mice and associates with protection from type 2 diabetes in humans"
fn <- "41591_2019_610_MOESM4_ESM.xlsx"
treatment_levels <- c("db/db-GFP", "db/db-Adipsin",	"WT")
file_path <- here(data_folder, file_folder, fn)

adipsin_fig2i <- read_excel(file_path,
                    sheet = "Fig 2i",
                    range = "A3:C15") %>%
  data.table() %>%
  melt(measure.vars = treatment_levels,
       variable.name = "treatment",
       value.name = "nkx6_1")
# View(fig2i)

```


* The data are the percent of cells staining for [NKX6.1](https://en.wikipedia.org/wiki/NKX6-1), which is a transcription factor protein that regulates beta cell development in the pancreas. Beta cells sense glucose levels in the blood and secrete insulin. Disruption of the insulin signaling system results in Diabetes mellitus.
* The data are in wide format, with each treatment group in a separate column. The data need to be melted into long format with a new column called "treatment".
* This will give you a pretty good plot of the data (if the data object is named "adipsin_fig2i")

```{r, warning=FALSE, message=FALSE}
ggstripchart(data = adipsin_fig2i,
             x = "treatment",
             y = "nkx6_1",
             add = "mean_se")
```


2. Import and quick pretty-good-plot the data for Figure 3b of the PI3K paper. You will need to download and archive the Excel file for "Figure 3". Store this within the "Suppression of insulin feedback enhances..." folder.

```{r, echo=FALSE}
folder <- "Suppression of insulin feedback enhances the efficacy of PI3K inhibitors"
filename <- "41586_2018_343_MOESM6_ESM.xlsx"
file_path <- here(data_folder, folder, filename)
pi3k_fig3b_wide <- read_excel(file_path,
                         sheet = "Figure 3B (c-Peptide)",
                         range = "A1:D6",
                         col_names = TRUE) %>%
  data.table()

treatment_levels <- colnames(pi3k_fig3b_wide)

pi3k_fig3b <- melt(data = pi3k_fig3b_wide,
                   measure.vars = treatment_levels,
                   variable.name = "treatment",
                   value.name = "c_peptide")

# View(pi3k_fig3b)

```


* The data are [c-peptide](https://en.wikipedia.org/wiki/C-peptide) levels in response to the treatments. C-peptide is cleaved from the pro-insulin polypeptide and circulates in the blood and is a marker of how much insulin is being produced by the beta cells of the pancreas. 
* The data are in wide format, with each treatment group in a separate column. The data need to be melted into long format with a new column called "treatment".
* Modify the code from exercise 1 to pretty-good-plot the data as in exercise 1.

```{r, echo=FALSE}
ggstripchart(data = pi3k_fig3b,
             x = "treatment",
             y = "c_peptide",
             add = "mean_se")

```


<!--chapter:end:chapters/06-data-reading_writing.Rmd-->

# Plotting Models

*So, along the lines of Sarah Susanka’s “Not So Big House,” Kolbert asks the group, “What would a Pretty Good House look like?”* -- Michael Maines^["The Pretty Good House - Finding the right balance between construction cost and energy performance". https://www.greenbuildingadvisor.com/article/the-pretty-good-house]

```{r plots-setup, echo=FALSE, message=FALSE}

library(here)
library(janitor)
library(readxl)
library(data.table)
library(stringr)
library(dplyr) # recode

# analysis packages
library(nlme)
library(emmeans)
library(car) # qqplot, spreadlevel
library(lmPerm)

# graphing packages
library(ggplot2) # ggplot environment
library(ggpubr) # publication ready plots
library(ggforce) # jitter
library(ggsci) # color palettes
library(cowplot) # combine plots
library(lazyWeave) # pretty pvalues
library(rstatix) # even prettier pvalues
library(kableExtra)

here <- here::here

file_path <- here("R/ggplot_the_model.R")
source(file_path)

clean_names <- janitor::clean_names

data_folder <- "data"

minus <- "\u2013"

```

Plots should be the focus of both the reader and researcher. Instead of mindless plotting, a researcher should ask a series of questions of every plot

1. What is the point of each element in a plot?
2. Are these the points that I most want to communicate?
3. Are there better practices for communicating these points?
4. Are the points that I want to communicate that are not covered by these elements?

The answer to these questions should inform what is and what is not plotted. The result is a pretty good plot. The idea of a pretty good plot is borrowed from the "pretty good house" concept that grew out of a collaborative group of builders and architects in Northern New England. The "pretty good house" combines best practices for building an earth friendly, high performance home at a reasonable cost. There is no pretty good house governing body that awards certificates of achievement but, instead, a set of metrics and a collection of building practices that can achieve these.

A typical pretty good plot contains some combination of

1. Modeled effects with confidence intervals. "Effects" are differences between groups in response to treatment -- the *raison d'etre* of an experiment.
2. Modeled means and confidence intervals.
3. Individual data points or a summary distribution of these.

## Pretty good plots show the model and the data

The data to introduce best practices in plotting come from Figure 2d and Figure 2e from "ASK1 inhibits browning of white adipose tissue in obesity", introduced in the introductor chapter (Analyzing experimental data with a linear model)

```{r plots-import-functions, echo=FALSE}
# function to read in parts of 2b
import_fig_2_part <- function(range_2){
  fig_2_part <- read_excel(file_path,
                          sheet = "Source Date_Figure 2",
                          range = range_2,
                          col_names = TRUE) %>%
    data.table()
  group <- colnames(fig_2_part)[1]
  setnames(fig_2_part, old = group, new = "treatment")
  fig_2_part[, treatment := as.character(treatment)] # this was read as logical
  fig_2_part[, treatment := group] # assign treatment group
  fig_2_part[, mouse_id := paste(group, .I)]
  return(fig_2_part)
}

```

```{r plots-AUC-functions, echo=FALSE}
# script to compute various area under the curves (AUC) using trapezoidal method
# le Floch's "incremental" auc substracts the baseline value from all points.
# This can create some elements with negative area if post-baseline values are less
# than baseline value.
# Some researchers "correct" this by setting any(y - ybar < 0 to zero. Don't do this.

auc <- function(x, y, method="auc", average = FALSE){
  # method = "auc", auc computed using trapezoidal calc
  # method = "iauc" is an incremental AUC of Le Floch
  # method = "pos_iauc" is a "positive" incremental AUC of Le Floch but not Wolever
  # method = "post_0_auc" is AUC of post-time0 values
  # if average then divide area by duration
  if(method=="iauc"){y <- y - y[1]}
  if(method=="pos_iauc"){y[y < 0] <- 0}
  if(method=="post_0_auc"){
    x <- x[-1]
    y <- y[-1]
  }
  n <- length(x)
  area <- 0
  for(i in 2:n){
    area <- area + (x[i] - x[i-1])*(y[i-1] + y[i])
  }
  value <- area/2
  if(average == TRUE){
    value <- value/(x[length(x)] - x[1])
  }
  return(value)
}

```


```{r plots-fig2e-file-path, echo=FALSE}
data_from <- "ASK1 inhibits browning of white adipose tissue in obesity"
file_name <- "41467_2020_15483_MOESM4_ESM.xlsx"
file_path <- here::here(data_folder, data_from, file_name)
  
fig_2_sheet <- "Source Date_Figure 2"
```

```{r plots-fig2e-import, echo=FALSE, message=FALSE, warning=FALSE}

range_list <- c("A179:H189", "A191:H199", "A201:H214", "A216:H230")
fig2e_wide <- data.table(NULL)
for(range_i in range_list){
  part <- import_fig_2_part(range_i)
  fig2e_wide <- rbind(fig2e_wide,
                       part)
}
fig2e_wide[, c("ask1", "diet") := tstrsplit(treatment, " ", fixed=TRUE)]

# melt
fig2e <- melt(fig2e_wide,
               id.vars = c("treatment",
                           "ask1",
                           "diet",
                           "mouse_id"),
               variable.name = "time",
               value.name = "glucose")
fig2e[, time := as.numeric(as.character(time))]

# for plot only (not analysis!)
shift <- 2
fig2e[treatment == "ASK1F/F chow", time_x := time - shift*1.5]
fig2e[treatment == "ASK1Δadipo chow", time_x := time - shift*.5]
fig2e[treatment == "ASK1F/F HFD", time_x := time + shift*.5]
fig2e[treatment == "ASK1Δadipo HFD", time_x := time + shift*1.5]
```

```{r plots-fig2e-auc, echo=FALSE}
# AUC of post-baseline values
# do this after melt as we don't need this in long format)
fig2e_wide[, glucose_0 := get("0")]

times <- c(0, 15, 30, 45, 60, 90, 120)
time_cols <- as.character(times)
Y <- fig2e_wide[, .SD, .SDcols = time_cols]
fig2e_wide[, glucose_auc := apply(Y, 1, auc,
                           x = times,
                           method = "auc",
                           average = FALSE)]
fig2e_wide[, glucose_mean := apply(Y, 1, auc,
                           x = times,
                           method = "post_0_auc",
                           average = TRUE)]

# View(fig2e_wide)
```

```{r plots-fig2e-fit-model, echo=FALSE}
# fit the model
fig2e_m1 <- lm(glucose_mean ~ glucose_0 + ask1*diet,
               data = fig2e_wide,
               na.action = "na.exclude")

# coef table
fig2e_m1_coef <- cbind(coef(summary(fig2e_m1)),
                       confint(fig2e_m1))

# calculate adjusted values at mean of chow/ASK1F/F
ycols <- c("ask1", "diet", "glucose_0", "glucose_mean")
new_data <- fig2e_wide[, .SD, .SDcols = ycols]
# mean covariate in reference group
glucose_0_cn_mean <- mean(fig2e_wide[treatment == "ASK1F/F chow", glucose_0])
new_data[, glucose_0 := glucose_0_cn_mean]
fig2e_wide[, glucose_mean_adj := predict(fig2e_m1, new_data)
            + residuals(fig2e_m1)]

# emm table
# the adjusted means are based on cov = mean(cov)
fig2e_m1_emm <- emmeans(fig2e_m1,
                        specs = c("ask1", "diet"))

# recalculate the adjusted means so that the adjusted mean of the
# reference group is the raw mean of the reference group
m1_offset <- mean(fig2e_wide[treatment == "ASK1F/F chow", glucose_mean_adj]) -
  data.table(summary(fig2e_m1_emm))[ask1 == "ASK1F/F" & diet == "chow", emmean]

fig2e_m1_emm <- emmeans(fig2e_m1,
                        specs = c("ask1", "diet"),
                        offset = m1_offset)

check_it <- FALSE
if(check_it == TRUE){
  fig2e_m1_emm %>%
    summary %>%
    data.table()
  fig2e_wide[, .(raw = mean(glucose_mean),
                 adj = mean(glucose_mean_adj)), by = c("ask1", "diet")]
}

# contrast table
fig2e_m1_pairs <- contrast(fig2e_m1_emm,
                           method = "revpairwise",
                           adjust = "none",
                           simple = "each",
                           combine = TRUE
                           ) %>%
  summary(infer = TRUE)


```

### Pretty good plot component 1: Modeled effects plot

```{r plots-prepare-fig2e-effects, echo = FALSE, message=FALSE}
# reverse order of rows so flipped plot has original order
fig2e_pairs_plot_order <- data.table(fig2e_m1_pairs)[rev(1:.N)]

# order "contrast" factor as in table!
# create contrast list by hand with copy and paste
contrast_list_1 <- c("ASK1Δadipo: HFD - chow",
                   "ASK1F/F: HFD - chow",
                   "HFD: ASK1Δadipo: ASK1Δadipo - ASK1F/F",
                   "chow: ASK1Δadipo - ASK1F/F")
contrast_list_2 <- c("ASK1Δ: HFD - chow",
                   "ASK1F/F: HFD - chow",
                   "HFD: ASK1Δ - ASK1F/F",
                   "chow: ASK1Δ - ASK1F/F")

fig2e_pairs_plot_order[, effect := factor(contrast_list_2,
                                      contrast_list_2)]


# add the interaction row
interaction_row <- fig2e_m1_coef["ask1ASK1Δadipo:dietHFD",
                                            c("Estimate", 
                                              "Pr(>|t|)",
                                              "2.5 %",
                                              "2.5 %")]
interaction_row <- data.table(effect = row.names(fig2e_m1_coef),
                              fig2e_m1_coef)[4]
interaction_row[1, effect := "Diet:ASK1 interaction"]
setnames(interaction_row, 
         old = c("Estimate",
                 "Std. Error",
                 "Pr(>|t|)",
                 "2.5 %",
                 "97.5 %"),
         new = c("estimate",
                 "SE",
                 "p.value",
                 "lower.CL",
                 "upper.CL"))

fig2e_pairs_plot_order <- rbindlist(list(interaction_row,
                                         fig2e_pairs_plot_order),
                                    fill = TRUE)

# make sure of order of factor levels for plot
level_order <- fig2e_pairs_plot_order[, effect]
fig2e_pairs_plot_order[, effect := factor(effect, level_order)]

# pretty p-values
fig2e_pairs_plot_order[, p_pretty := pvalString(p.value)]

# group columns for p-value brackets
fig2e_pairs_plot_order[, group1 := c(0, 4, 3, 4, 2)]
fig2e_pairs_plot_order[, group2 := c(0, 2, 1, 3, 1)]

```

```{r plots-fig2e-gg-effects, echo=FALSE, fig.cap = "Effects plot of glucose AUC data"}
# get bounds for y-axis
min_bound <- min(fig2e_pairs_plot_order[, lower.CL])
max_bound <- min(fig2e_pairs_plot_order[, upper.CL])
y_lo <- min(min_bound+min_bound*0.2,
            -max_bound)
y_hi <- max(max_bound + max_bound*0.2,
            -min_bound)
y_lims <- c(y_lo, y_hi)
y_lims <- c(-6, 9)

fig2e_gg_effect <- ggplot(data=fig2e_pairs_plot_order, 
                          aes(x = effect,
                              y = estimate)) +
  # confidence level of effect
  geom_errorbar(aes(ymin=lower.CL, 
                    ymax=upper.CL),
                width=0, 
                color="black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # zero effect
  geom_hline(yintercept=0, linetype = 2) +
  
  # p-value
  annotate(geom = "text",
           label = fig2e_pairs_plot_order$p_pretty,
           x = 1:5,
           y = 9) +
  
  # aesthetics
  scale_y_continuous(position="right") +

  coord_flip(ylim = y_lims) + 
  theme_pubr() +
  ylab("Effects (mmol/L)") +
  theme(axis.title.y = element_blank()) +
  NULL

fig2e_gg_effect
```

Biologists infer the biological consequences of a treatment by interpreting the magnitude and sign of treatment "effects", such as the differences in means among treatment levels. Why then do we mostly plot treatment level means, where effect magnitude and sign can only be inferred *indirectly*, by mentally computing differences in means? A pretty good plot *directly* communicates treatment effects and the uncertainty in the estimates of these effects using an **effects plot**.

Figure \@ref(plots-fig2e-gg-effects) is an effects plot of the linear model fit to the glucose tolerance data. The effects plot is "flipped". The y-axis is the categorical variable -- it contains the labels identifying the pair of groups in the contrast and the direction of the difference. In addition to the pairwise comparisons, I include the interaction effect on the y-axis. The x-axis is the continuous variable -- it contains the **simple effects**, which is the difference in means between the two groups identified by the y-axis labels. Additionally, the y-axis includes the estimate of the $diet \time genotype$ interaction effect. The bars are 95% confidence intervals of the effects (either simple effects or interaction effect), which is the range of values that are compatible with the observed data at the 95% level.

We can use the effects and CIs of the effects to evaluate the treatment effects. For example, when on a high fat diet (HFD), the mean, post-baseline plasma glucose level in the ASK1$\Delta$adipo is 3.5 mmol/L less than that for the control (ASK1F/F). Differences less than 5.3 mmol/L less than ASK1F/F levels or greater than 1.7 mmol/L less than ASK1F/F levels are not very compatible with the data. It is up to the research community to decide if 1.7 mmol/L or 3.5 mmol/L differences are physiologically meaningful effects.

### Pretty good plot component 2: Modeled mean and CI plot

```{r plots-fig2-gg-response-adjusted, echo=FALSE, fig.cap="Response plot"}
# use a data.table version of fig2e_m1_emm for plot
fig2e_m1_emm_dt <- summary(fig2e_m1_emm) %>%
  data.table

pd <- position_dodge(1)
fig2e_gg_response_adj <- ggplot(data = fig2e_wide,
                                aes(x = diet,
                                    y = glucose_mean_adj,
                                    color = ask1)) +
  
  # points
  geom_sina(alpha = 0.5) +
  
  # plot means and CI
  geom_errorbar(data = fig2e_m1_emm_dt,
                aes(y = emmean,
                    ymin = lower.CL,
                    ymax = upper.CL,
                    color = ask1),
                width = 0,
                position = pd
  ) +
  
  geom_point(data = fig2e_m1_emm_dt,
             aes(y = emmean,
                 color = ask1),
             size = 3,
                position = pd
  ) +

  # aesthetics
  ylab("Post-baseline glucose\nadjusted for baseline (mmol/L)") +
  scale_color_manual(values=pal_okabe_ito_blue,
                     name = NULL) +
  theme_pubr() +
#  theme(legend.position = c(0.25, 0.9)) +
  theme(legend.position = "bottom") +
  
  NULL

# add p-value brackets
fig2e_pairs_plot_order[, group1 := group1/2+0.25]
fig2e_pairs_plot_order[, group2 := group2/2+0.25]

fig2e_gg_response_adj_p <- fig2e_gg_response_adj +
  stat_pvalue_manual(fig2e_pairs_plot_order[2:5,],
                     label = "p_pretty", 
                     y.position=c(28, 26.5, 25, 25),
                     tip.length = 0.01) +
  theme(legend.position = "top") +
  theme(axis.title.x=element_blank()) +
  NULL

 fig2e_gg_response_adj_p
```

The response plot in Figure \@ref(fig:plots-fig2-gg-response-adjusted) "shows the model" -- by this I mean the plot shows the *modeled* means, represented by the large circles, the *modeled* 95% confidence intervals of each mean, represented by the error bars, and the *model-adjusted* individual response values, represented by the small colored dots. What do I mean by *modeled* means, modeled error intervals, and model-adjusted responses?

```{r plots-model-means-table, echo=FALSE}
# sample means and SE
fig2e_summary <- fig2e_wide[, .(N = .N,
                                 glucose_mean = mean(glucose_mean),
                                 sigma = sd(glucose_mean),
                                 SE = sd(glucose_mean)/sqrt(.N)),
                       by = c("ask1", "diet")]
setnames(fig2e_summary,
         old = c("glucose_mean", "sigma", "SE"),
         new = c("Sample mean", "Sample sigma", "Sample SE"))
fig2e_m1_emm_dt[, sigma := summary(fig2e_m1)$sigma]
ycols <- c("emmean", "sigma", "SE", "ask1", "diet")
fig2e_summary <- merge(fig2e_summary,
                       fig2e_m1_emm_dt[, .SD, .SDcols = ycols],
                       by = c("ask1", "diet"))
setnames(fig2e_summary,
         old = c("emmean", "sigma", "SE"),
         new = c("Model mean", "Model sigma", "Model SE"))
fig2e_summary <- doBy::orderBy(~ ask1 + diet, fig2e_summary)
knitr::kable(fig2e_summary, digits = c(1,1,2,2,2,2,2,2,2)) %>%
  kable_styling()
```

1. The modeled means and error intervals are estimated from the statistical model. Many published plots show the sample means and sample error intervals, which are computed within each group independently of the data in the other groups and are not adjusted for any covariates or for any hierarchical structure to the data.
2. A modeled mean will often be equal to the raw mean, but this will not always be the case. Here, the modeled means for the non-reference groups in Figure \@ref(fig:lots-fig2-gg-response-adjusted) do not equal the sample means because the modeled means are adjusted for the baseline measures of glucose (Table \@ref(tab:plots-model-means-table)) (specifically, the modeled means are conditional on the baseline being equal to the mean of the baseline in the reference group).
3. For most of the analyses in this text, modeled error intervals are not the same as the sample error intervals and are commonly conspicuously different. For the glucose tolerance data, the modeled error intervals are calculated from a pooled estimate of $\sigma$ while the sample error intervals are estimated from sample-specific estimates of $\sigma$ (Table \@ref(tab:plots-model-means-table)).
4. Model-adjusted responses are responses that are adjusted for covariates in the model. If there are no covariates in the model, the model-adjusted responses are the same as the raw response. In the glucose tolerance data, the model-adjusted responses are the modeled, individual response measures if all individuals had the same baseline glucose (the covariate).

Modeled means, error intervals, and responses are not commonly plotted but it is these values that are consistent with our inferences from the statistical model. There are many data sets in experimental biology where a plot of sample means, error intervals, and responses give a very distorted view of inference from the model.

The response plot in Figure \@ref(fig:plots-fig2-gg-response-adjusted) also "shows the data" by plotting response values as "jittered" dots. Showing the data

1. allows the reader to get a sense of the underlying sample size and distribution including outliers, which can be used to mentally model check the published statistical analysis. Adding a box plot, violin plot, or dot plot augments the communication of the distributions if there are enough data to justify the addition.
2. allows a reader to see the overlap in individual responses among groups and to evaluate the biological consequences of this overlap.

### Combining Effects and Modeled mean and CI plots -- an Effects and response plot.

```{r plots-fig2e-ggplot, echo=FALSE, message=FALSE, warning=FALSE, fig.cap="Effect of diet and ASK1 deletion on post-baseline glucose. Top: effects plot of 2 X 2 simple effects (difference in means) and of the diet X genotype interaction. Bars are 95% confidence intervals of the effects. Unadjusted p-values from the linear model are given. Bottom: response plot of the means and 95% confidence interval of each diet X genotype combination."}
fig2e_gg_response_adj <- fig2e_gg_response_adj +
  coord_flip() +
  theme(axis.title.y = element_blank())



fig2e_fig <- plot_grid(fig2e_gg_effect,
                       fig2e_gg_response_adj,
                       nrow = 2,
                       align = "v",
                       axis = "l",
                       rel_heights = c(0.7,1))
fig2e_fig
```

Combining the effects and response plots into a single plot is an easy solution to issues that arise if only one or the other is used. What are these issues?

While a response plot like that in Figure \@ref(fig:plots-fig2-gg-response-adjusted) is standard in biology, it fails to show the effects, and the uncertainty in the effects, *explicitly*. To infer the effects from the plot, a reader must perform mental math -- either compute the difference or the ratio between pairs of means. This mental math is easy enough if the comparisons are between individual treatment levels but much harder if the comparisons are between pooled sets of treatment levels, for example in a factorial experimental design. The mental math that is excessively difficult is the reconstruction of some kind of error interval of the contrasts, for example the 95% confidence intervals in Figure \@ref(plots-fig2e-gg-effects) and it is these intervals that are necessary for a researcher to infer the range of biological consequences that are compatible with the experiment's results. The inclusion of the *p*-values for all pairwise comparisons in a response plot gives the significance level of these contrasts, but of the kinds of summary results that we could present (contrasts, error intervals, *p*-values), the *p*-values are the least informative.

Effects plots are very uncommon in most of biology outside of meta-analysis and clinical medicine more generally. An effects plot alone fails to communicate anything about the sample size or conditional distribution of the data. Equally important, response values are often meaningful and researchers working in the field should be familiar with usual and unusual values. This can be useful for interpreting biological consequences of treatment effects but also for researchers and readers to asses the credibility of the data (for example, I have twice, once in my own data and once in a colleagues data, found mistakes in the measurement of an entire data set of response variable because the plotted values weren't credible).

### Some comments on plot components

1. Several recent criticisms of bar plots have advocated box plots or violin plots as alternatives. Box plots and violin plots are useful alternatives to jittered dots if there are sufficient data to capture the distribution but I wouldn't advocate replacing the plot of modeled means and confidence intervals with box or violin plots, as these communicate different things. More importantly, box and violin plots do not communicate the treatment effects.
2. Almost all plots in biology report the error bars that represent the sample standard error. As described above, sample standard error bars do not reflect the fit model and can be highly misleading, at least if interpreting as if they do reflect the model. Also, sample standard error bars can explicitly include absurd values or imply absurd confidence intervals. For example, I sometimes see standard error bars cross $y=0$ for a response that cannot be negative, such as a count. Even if the standard error bar doesn't cross zero, it is common to see standard error bars that imply (but do not explicitly show) 95% confidence intervals that cross zero, again for responses that cannot be negative. A standard error bar or confidence interval that crosses zero implies that negative means are compatible with the data. This is an absurd implication for responses that cannot have negative values (or are "bounded by" zero). Explicit or implicit error bars that cross zero are especially common for count responses with small means. *If* a researcher plots confidence intervals, these should be computed using a method that avoids absurd implications, such methods include the bootstrap and generalized linear models.
3. Significance stars are okay, the actual *p*-value is better, effects plots are best. Many researchers add star symbols to a plot indicating the level of significance of a particular paired comparison. Stars are okay in the sense that there is no inferential difference between $p = 0.015$ and $p = 0.045$. There's also no inferential difference between $p = 0.0085$ and $p = 0.015$, which highlights the weakness of -chotomizing any continuous variable. For this reason, a better, alternative would be to add the actual p-value (as above). A more serious criticism of stars is that it encourages researchers and readers to focus on statistical significance instead of effect size and uncertainty. A more valuable alternative, then, is to report the effects and uncertainty in an effects plot or a combined effects-and-response plot.

## Working in R

A reasonable goal of any research project should be a script to generate the final plots entirely within the R environment and not rely on external drawing software to add finishing features. This section covers some of the basics of using R packages to create plots. Later chapters cover some of the details that are specific to the analyses in that chapter.

[ggplot2](https://ggplot2.tidyverse.org) is one of the major plotting environments in R and the one that seems to have the strongest following, especially among new R users. ggplot2 has the ability to generate extremely personalized and finished plots. However, ggplot2 has a long learning curve and until one is pretty comfortable with its implementation of the grammar of graphics, creating a plot with multiple layers (mean points, error intervals, raw data points, p-values, text annotations) and modified aesthetics (axis text, point colors) can often require many hours of googling.

[ggpubr](https://cran.r-project.org/web/packages/ggpubr/index.html) is an extension to ggplot2 (it calls ggplot2 functions under the hood) and provides many canned functions for producing the kinds of ggplots that are published in biological journals. With one line of script, a researcher can generate a publishable plot.

**ggplot_the_model** and related functions in this chapter are my attempts to create a simple function for creating publication ready plots that highlight effect size and uncertainty.

Some of the basics for using ggpubr, ggplot2, and ggplot_the_model are outlined here. More specific examples are in each chapter.

### Source data

Data source: [ASK1 inhibits browning of white adipose tissue in obesity](https://www.nature.com/articles/s41467-020-15483-7){target="_blank"} 

The source data are that for Figure 2E. The response is $glucose\_auc$ the "area under the curve" of repeated measures of blood glucose during the 120 minutes of a glucose tolerance test. Glucose AUC is a measure of glucose tolerance, the higher the area, the higher the blood glucose over the two hours, and the worse the physiological response to a sudden rise in blood glucose. There are two treatment factor variables: 1) $Diet$, with levels "chow" and "HFD", where "chow" is normal mouse chow and "HFD" is a high fat diet, and 2) $ask1$, with levels "ASK1F/F" and "ASK1Δadipo" where "ASK1F/F" is the control level and "ASK1Δadipo" is the ASK1 adipose-deletion mouse described in Chapter 1.

#### Import

```{r plots-fig2e-import-view, echo=TRUE, message=FALSE, warning=FALSE}

data_from <- "ASK1 inhibits browning of white adipose tissue in obesity"
file_name <- "41467_2020_15483_MOESM4_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)
  
# the data are in "tranposed" format -- each row contains the n
# measures of a treatment level. Read, then transpose
# to make the treatment levels the columns
fig2e_wide <- read_excel(file_path,
                     sheet = "Source Date_Figure 2",
                     range = c("A233:O236"), # lot of NA
                     col_names = FALSE) %>%
  data.table %>%
  transpose(make.names = 1) # turn data onto 

# melt the four columns into a single "glucose_auc" column
# and create a new column containing treatment level.
y_cols <- colnames(fig2e_wide)

# melt
fig2e <- melt(fig2e_wide,
               measure.vars = y_cols,
               value.name = "glucose_auc",
               variable.name = "treatment")

# create two new columns that are the split of treatment
fig2e[, c("ask1", "diet") := tstrsplit(treatment,
                                        " ",
                                        fixed=TRUE)]

# since glucose_auc is the only response variable in this
# data.table, omit all rows with any NA
fig2e <- na.omit(fig2e)
# View(fig2e)
```

### How to plot the model

The steps throughout the text for plotting the model fit to experimental data are

1. fit the statistical model
2. use the fit model to estimate the modeled means and confidence limits using `emmeans` from the [emmeans package](https://cran.r-project.org/web/packages/emmeans/index.html).
3. use the `emmeans` object to estimate the contrasts of interests using the `contrast` function from emmeans.
4. Plot the individual points. If covariates are in the model, use the fit model from step 1 to plot the adjusted values of the points.
5. Use the values from step 2 to plot the modeled means and error intervals.
6. If including p-value brackets, use the values from step 3.

If you are using ggplot_the_model functions, then steps 4-6 are done for you.

Here, I fit a linear model to the fig2e data and compute the use the `emmeans` and `contrast` functions without comment. The details of these functions are in the chapters that follow. The fit model and construction of the plots is simplified from those above.

#### Fit the model

The response is the glucose AUC, which is the area under the curve of the data from the glucose tolerance test. The model is a factorial linear model with *ask1* genotype and *diet* as the two factors.

```{r plot-m1}
# glucose_auc is the AUC of the glucose tolerance curves computed using trapezoidal algorithm
m1 <- lm(glucose_auc ~ ask1*diet,
               data = fig2e)
```

#### Compute the modeled means table of estimated means and confidence intervals

Modeled means are computed by passing the model object (`m1`) to the `emmeans` function and specifying the columns containing the groups using the `specs = ` argument`.

```{r plot-m1-emm}
m1_emm <- emmeans(m1, specs = c("ask1", "diet"))
m1_emm
```

#### Compute the contrasts table of estimated effects with confidence intervals and p-values

Contrasts among levels, or combinations of levels, are computed by passing the emmeans object (`m1.emm`) to the `contrast` function. There are many important variations of this step. This text advocates computing planned comparisons, which requires expert knowledge and forthought. Here I limit the computation to the four simple effects (the effects of one factor in each of the levels of the other factor).

```{r plot-m1-pairs}
m1_simple <- contrast(m1_emm,
                     method = "revpairwise",
                     simple = "each",
                     combine = TRUE,
                     adjust = "none") %>%
  summary(infer = TRUE)
m1_simple
```

Notes

1. I often use `m1_pairs` as the name of the contrast table. Here I use `m1_simple` to remind me that I've limited the comparison to the four simple effects. If I only compute planned comparisons, I might use `m1_planned`.

### How to use the Plot the Model functions

The philosophy underneath these functions is to use the model fitted by the researcher to make the plots. The functions require information from three objects:

1. the data frame containing the modeled data
2. the modeled means and CIs from `emmeans`
3. the modeled effects, CIs and *p*-values from `emmeans::contrast`.

This philosophy strikes a balance between functions in which all of the statistical modeling is hidden and the researcher only sees the output and manually building ggplots. Actually, I strongly encourage researchers to learn how to build these plots and to not rely on canned functions, and I outline this after introducing the ggplot_the_model functions

These functions require the fit model object (`m1`), the emmeans object of modeled means (`m1_emm`) and the contrast object of modeled effects (`m1_simple`).

#### ggplot_the_response

For the response plot only, use `ggplot_the_response`.

```{r plots-ggplot_the_response}
m1_response_plot <- ggplot_the_response(
  fit = m1,
  fit_emm = m1_emm,
  fit_pairs = m1_simple,
  palette = pal_okabe_ito_blue,
  y_label = expression(paste("mmol ", l^-1, " min")),
  g_label = "none"
)

m1_response_plot
```
`ggplot_the_response` arguments:

**fit** model object from lm, lmer, nlme, glmmTMB

**fit_emm** object from 'emmeans'. Or, a data frame that looks like this object, with modeled factor variables in columns 1 and 2 (if a 2nd factor is in the model), a column of means with name "emmean", and columns of error intervals named "lower.CL" and "upper.CL"

**fit_pairs** object from `emmeans:contrast`. Or, a data frame that looks like this object.

**wrap_col = NULL** Not used at the moment

**x_label = "none"** A character variable used for the X-axis title.

**y_label = "Response (units)"** A character variable used for the Y-axis title. Use `expression(paste())` method for math.

**g_label = NULL** A character variable used for the grouping variable (the 2nd factor) title. Use "none" to remove

**dots = "sina"** controls the plotting of individual points. sina from ggforce package. Alternatives are "jitter" and "dotplot"

**dodge_width = 0.8** controls spacing between group means for models with a 2nd factor (the grouping variable)

**adjust = 0.5** controls spread of dots if using `dots = "sina"`

**contrast_rows = "all"** controls which rows of fit_pairs to use for *p*-value brackets. Use "none" to hide.

**p_pos = NULL** manual control of the y-coordinates for p-value brackets

**palette = pal_okabe_ito** allows control of the color palette. The default pal_okabe_ito palette is a color blind palette.

**legend_position = "top"** controls position of the legend for the grouping variable (the 2nd factor in a two-factor model)

**flip_horizontal = FALSE** controls the orientation of the axes.

**group_lines = FALSE** used for plotting lines connecting group means. Not yet implemented.

#### ggplot_the_effects

For the effects plot only, use `ggplot_the_effects`.

```{r plots-ggplot_the_effects}
m1_effects_plot <- ggplot_the_effects(
  fit = m1,
  fit_pairs = m1_simple,
  effect_label = expression(paste("Effect (mmol ", l^-1, " min)"))
)

m1_effects_plot
```

`ggplot_the_effects` arguments

**fit** model object from lm, lmer, nlme, glmmTMB

**fit_pairs** object from `emmeans:contrast`. Or, a data frame that looks like this object.

**contrast_rows = "all"** controls which rows of fit_pairs to include in plot.

**show_p = TRUE** controls show/hide of *p*-values

**effect_label = "Effect (units)"** character variable for the title of the effects axis title.

#### ggplot_the_model

For the combined response and effects plot, use `ggplot_the_model`.

```{r plots-ggplot_the_model}
m1_plot <- ggplot_the_model(
  fit = m1,
  fit_emm = m1_emm,
  fit_pairs = m1_simple,
  palette = pal_okabe_ito_blue,
  y_label = expression(paste("mmol ", l^-1, " min")),
  g_label = "none",
  effect_label = expression(paste("Effect (mmol ", l^-1, " min)"))
)

m1_plot
```
`ggplot_the_model` arguments

**fit** same as for `ggplot_the_response`

**fit_emm** same as for `ggplot_the_response`

**fit_pairs** same as for `ggplot_the_response`

**wrap_col = NULL** same as for `ggplot_the_response`

**x_label = "none"** same as for `ggplot_the_response`

**y_label = "Response (units)"** same as for `ggplot_the_response`

**g_label = NULL** same as for `ggplot_the_response`

**effect_label = "Effect (units)"** same as for `ggplot_the_effect`

**dots = "sina"** same as for `ggplot_the_response`

**dodge_width = 0.8** same as for `ggplot_the_response`

**adjust = 0.5** same as for `ggplot_the_response`

**contrast_rows = "all"** same as for `ggplot_the_response`

**p_pos = NULL** same as for `ggplot_the_response`

**palette = pal_okabe_ito** same as for `ggplot_the_response`

**legend_position = "bottom"** Except for default, same as for `ggplot_the_response`

**flip_horizontal = TRUE** Except for default, same as for `ggplot_the_response`

**group_lines = FALSE** used for plotting lines connecting group means. Not yet implemented.

**rel_heights = c(1,1)** used to control relative heights of the effects and response plot

#### ggplot_the_treatments

```{r plots-ggplot_the_treatments}
x_levels <- rbind(
  ASK1 = c("F/F", "Δadipo", "F/F", "Δadipo"),
  Diet = c("chow", "chow", "HFD", "HFD")
)

# this is the same code as above but hiding 
# legend position
m1_response_plot_base <- ggplot_the_response(
  fit = m1,
  fit_emm = m1_emm,
  fit_pairs = m1_simple,
  palette = pal_okabe_ito_blue,
  y_label = expression(paste("mmol ", l^-1, " min")),
  g_label = "none",
  legend_position = "none"
)

m1_response_plot2 <- ggplot_the_treatments(
  m1_response_plot_base,
  x_levels = x_levels,
  text_size = 3.5,
  rel_heights = c(1, 0.1)
)

m1_response_plot2
```

If you prefer plus and minus symbols, use `minus <- "\u2013"` for the minus sign instead of the hyphen "-"

```{r plots-ggplot_the_treatments-minus}
minus <- "\u2013" # good to put this in the setup chunk

x_levels <- rbind(
  Δadipo = c(minus, "+", minus, "+"),
  HFD = c(minus, minus, "+", "+")
)

m1_response_plot2 <- ggplot_the_treatments(
  m1_response_plot_base,
  x_levels = x_levels,
  text_size = 3.5,
  rel_heights = c(1, 0.1)
)

m1_response_plot2
```

### How to generate a Response Plot using ggpubr

Steps 1-3 were completed above.

#### Step 4: Plot the individual points {#plot-individual-points}

**Using ggplot**

I'm going to show how to create the initial, base plot of points using ggplot2 in order to outline **very briefly** how ggplot2 works.

```{r plot-ggplot-base}
m1_response <- ggplot(
  data = fig2e,
  aes(x = treatment, # these 2 lines define the axes
      y = glucose_auc,
      color = ask1 # define the grouping variable
  )) +
  
  # surprisingly, the code above doesn't plot anything
  # this adds the points as a layer
  geom_jitter(width = 0.2) +
  
  # change the title of the y-axis
  ylab(expression(paste("AUC (mmol ", l^-1, " min)"))) +
  
  # change the theme
  theme_pubr() +

  # these theme modifications need to be added after re-setting
  # the theme
  theme(
    legend.position = "none", # remove legend
    axis.title.x = element_blank() # remove the x-axis title
  ) +
  
  
  # change the colors palette for the points
  scale_color_manual(values = pal_okabe_ito_blue) +
  
  NULL

  
m1_response
```
Notes

1. The `ggplot` function requires a data frame passed to `data` containing the data to plot and an aesthetic (`aes`), which passes the column names that set the x and y axes. These column names must be in the data frame passed to `data`. `color = ` is an aesthetic that sets the grouping variable used to assign the different colors.
2. The x-axis is discrete but is numeric. The x-axis values are 1, 2, 3, 4. But instead of using these numbers as the labels for the x-axis values, ggplot uses the names of the groups (the four values of the column "treatment")

**Using ggpubr**

```{r plot-ggpubr-base}
m1_response <- ggstripchart(
  data = fig2e,
  x = "treatment",
  y = "glucose_auc",
  color = "ask1",
  xlab = "",
  ylab = expression(paste("AUC (mmol ", l^-1, " min)")),
  palette = pal_okabe_ito_blue,
  legend = "none"
  )

m1_response
```

#### Step 5: Plot the modeled means and 95% error intervals

To add points and error bars to `m1_response`, we need to tell ggplot the x-axis positions (or coordinates). These positions are the values of the "treatment" column in `fig2e`. The modeled means and and 95% CIs are in the `m1_emm` object but there is no "treatment" column, or any column with these values. We, therefore have to make this column before we can add the modeled means and CIs to the plot.

```{r plot-m1_emm_dt}
# convert m1_emm to a data.table
m1_emm_dt <- summary(m1_emm) %>%
  data.table()

# create treatment column
# make sure it matches values in the two factor columns
m1_emm_dt[, treatment := c("ASK1F/F chow",
                           "ASK1Δadipo chow",
                           "ASK1F/F HFD",
                           "ASK1Δadipo HFD")]
```

Now add the modeled means and CIs

```{r plot-mean-ci, warning=FALSE, message=FALSE}
m1_response <- m1_response + 

  # add layer containing means
  geom_point(data = m1_emm_dt,
             aes(y = emmean,
                 color = ask1), 
             size = 3) +
  
  # add layer containing error bars
  geom_errorbar(data = m1_emm_dt, 
                aes(y = emmean,
                    ymin = lower.CL, 
                    ymax = upper.CL,
                    color = ask1), 
                width = 0.05) +
  
  NULL

m1_response
```

Notes

1. `m1_response` generated by `ggpubr::stripchart` is a `ggplot2` object. This means modifications of the plot are implemented by adding these with the "+" sign.
2. The modeled means are in the column "emmean" in the data frame `m1_emm_dt`. We need to tell `geom_point` where to find the data using the `data =` argument. `geom_point()` (and other geoms) assumes that the points that we want to plot are defined by the same x and y column names used to create the plot -- if these don't exist, we need to state the x and y column names in the aesthetic function `aes`. Since we created a "treatment" column in `m1_emm_dt` that contains the x-axis coordiantes (1, 2, 3, 4), we do not need to tell ggplot where to find the x-values. But there is no "glucose_auc" column in `m1_emm_dt` so we need to tell `geom_point()` where to find the y values using `aes(y = "emmean")`.
3. Adding the modeled error intervals using `geom_errorbar` follows the same logic as adding the modeled means. Importantly, and interestingly, `y = emmean` has to be passed even though no information from this column is used to plot the error bars. 
4. Note that column names passed to a `ggpubr` function must be in quotes but column names passed to a `ggplot2` function cannot be in quotes

#### Step 6: Adding p-values

*p*-value brackets are added to a response plot using `stat_pvalue_manual` from the `ggpubr` package. This function needs a column of p-values, and a pair of columns that define the left and right x-axis positions of the bracket.

```{r plot-m1_simple_dt}
# convert m1_simple to a data.table
m1_simple_dt <- data.table(m1_simple)

# create group1 -- column containing x-position
# of the left side of the bracket
# need to look at m1_simple_dt to construct this.
# "ASK1F/F chow", "ASK1Δadipo chow" ,"ASK1F/F HFD", "ASK1Δadipo HFD" 

m1_simple_dt[, group1 := c("ASK1Δadipo chow",
                           "ASK1Δadipo HFD",
                           "ASK1F/F HFD",
                           "ASK1Δadipo HFD")]

# create group2 -- column containing x-position
# of the right side of the bracket
# need to look at m1_simple_dt to construct this.
# "ASK1F/F chow", "ASK1Δadipo chow" ,"ASK1F/F HFD", "ASK1Δadipo HFD" 

m1_simple_dt[, group2 := c("ASK1F/F chow",
                           "ASK1F/F HFD",
                           "ASK1F/F chow",
                           "ASK1Δadipo chow")]

m1_simple_dt[, p_rounded := p_round(p.value,
                                  digits = 2)]
m1_simple_dt[, p_pretty := p_format(p_rounded,
                                    digits = 2,
                                    accuracy = 1e-04,
                                    add.p = TRUE)]
```

Now add the p-values

```{r plot-manual-p-values}
# simply assigning this to a new plot with new name
# because I want to re-use the base plot in the next chunk
m1_response_p <- m1_response +
  stat_pvalue_manual(
    data = m1_simple_dt,
    label = "p_pretty",
    y.position = c(3300, 3300, 3450, 3600),
    tip.length = 0.01)

m1_response_p
```

Notes on adding *p*-values to the plot:

1. The `y.position` argument in `stat_pvalue_manual()` contains the position on the y-axis for the p-value brackets. I typically choose these values "by eye". Essentially, I look at the maximum y-value on the plot and then choose a value just above this for the first bracket. This may take some trial-and-error to position the brackets satisfactorily.
2. Use base R indexing to specify a subset. For example

```{r plot-manual-p-values-subset}
m1_response_p <- m1_response +
  stat_pvalue_manual(
    data = m1_simple_dt[c(2,4), ], # only rows 2, 4
    label = "p_pretty",
    y.position = c(3300, 3450),
    tip.length = 0.01)

m1_response_p

```

3. `ggpubr::stat_compare_means` automates the process somewhat but the function is too limited for statistics on anything but the simplest experiments. I don't advocate it's use. 

#### A variation for factorial models

The experiment in Fig2e has a factorial design and was analyzed (here, not in the original paper) using a factorial model. The factorial design can be represented in the plot by clustering the levels.

```{r plot-factorial-response-plot}
dodge_width = 0.4
jitter_width = 0.2
m1_simple_dt[, xmin := c(1-dodge_width/4,
                        2-dodge_width/4,
                        1-dodge_width/4,
                        1+dodge_width/4)]
m1_simple_dt[, xmax := c(1+dodge_width/4,
                        2+dodge_width/4,
                        2-dodge_width/4,
                        2+dodge_width/4)]

m1_response_fac <- ggstripchart(
  data = fig2e,
  x = "diet",
  y = "glucose_auc",
  color = "ask1",
  fill = "ask1",
  ylab = expression(paste("AUC (mmol ", l^-1, " min)")),
  palette = pal_okabe_ito_blue,
#  position = position_dodge(width = dodge_width)
  position = position_jitterdodge(dodge.width = dodge_width,
                                  jitter.width = jitter_width)
) +
  
  rremove("xlab") + #ggpubr function

  # add layer containing means
  geom_point(data = m1_emm_dt,
             aes(y = emmean,
                 color = ask1), 
             size = 3,
             position = position_dodge(width = dodge_width)) +
  
  # add layer containing error bars
  geom_errorbar(data = m1_emm_dt, 
                aes(y = emmean,
                    ymin = lower.CL, 
                    ymax = upper.CL,
                    color = ask1), 
                width = 0.05,
             position = position_dodge(width = dodge_width)) +
  
  # add p-value brackets
  stat_pvalue_manual(
    data = m1_simple_dt,
    label = "p_pretty",
    xmin = "xmin",
    xmax = "xmax",
    y.position = c(3300, 3300, 3450, 3600),
    tip.length = 0.01,
    size = 3) +
  
  NULL

m1_response_fac
```

#### How to add treatment combinations to a ggpubr plot

Many researchers in bench biology insert a grid of treatments below the x-axis of a plot. This is time consuming if we are using some external software to add to this. A much leaner workflow would be to add the treatment grid in the same step as generating the plot itself. Here is a kludgy way to do this using a ggpubr plot. A much more elegant method is described below

##### Variant 1 -- Fake axes

```{r plots-how to plot treatments}

use_this <- FALSE # if false use +/- symbols
if(use_this == TRUE){
  x_levels <- rbind(
    "ASK1:" = c("F/F", "Δadipo", "F/F", "Δadipo"),
    "Diet:" = c("chow", "chow", "HFD", "HFD")
  )
}else{
  minus <- "\u2013" # good to put this in the setup chunk
  x_levels <- rbind(
    "Δadipo:" = c(minus, "+", minus, "+"),
    "HFD: " = c(minus, minus, "+", "+")
  )
}


x_levels_text <- apply(x_levels, 2, paste0, collapse="\n")
x_levels_title <- paste(row.names(x_levels), collapse="\n")

x_axis_min <- 0.4
x_axis_max <- 4.5
y_plot_min <- 1200
y_axis_min <- 1350
y_axis_max <- 3500
y_breaks <- seq(1500, 3500, by = 500)
y_labels <- as.character(y_breaks)

m1_response_final <- m1_response_p +
  coord_cartesian(ylim = c(y_plot_min, y_axis_max)) +
  scale_y_continuous(breaks = y_breaks) +
  theme(
    axis.line = element_blank(), # remove both x & y axis lines
    axis.text.x = element_blank(), # remove x-axis labels
    axis.ticks.x = element_blank() # remove x-axis ticks
  ) +
  
  # add shortened y-axis line that doesn't extend to treatments
  geom_segment(aes(x = x_axis_min + 0.01,
                   y = y_axis_min,
                   xend = x_axis_min + 0.01,
                   yend = y_axis_max),
               size = 0.5) +
  
  # add x-axis line above treatments
  geom_segment(aes(x = x_axis_min,
                   y = y_axis_min,
                   xend = x_axis_max,
                   yend = y_axis_min),
               size = 0.5) +
  
  # add treatment combinations
  annotate(geom = "text",
           x = 1:4,
           y = y_plot_min,
           label = x_levels_text) +
  
  # add factor names
  annotate(geom = "text",
           x = x_axis_min,
           y = y_plot_min,
           label = x_levels_title,
           hjust = 0) +
  NULL
  
m1_response_final
```

Notes

1. `ggplot_the_treatments` should work with any ggplot, including those generated by ggpubr functions, with categorical values on the x-axis.
2. Here, I add the treatment combinations manually. The table of combinations is added inside the plot area (inside the axes). To make this look nice, and not weird, the x and y axes are removed and new lines are inserted to create new axis lines. The bottom of the new y-axis line starts above the treatment table. The new x-axis line is inserted above the treatment table.
3. To include the treatment combination (group) names, add this as the first row of `x_levels`.
4. Note that if the plot has a grid, this grid will extend into the area occupied by the treatment table using this method.

### How to generate a Response Plot with a grid of treatments using ggplot2

Above, I wrote a [short script for generating the base response plot](#plot-individual-points) using ggplot. In this plot, the x-axis variable ("treatment") is categorical and ggplot uses the integers 1, 2, 3, 4 as the coordinate values for the position of the four groups on the x-axis. Understanding this mapping is important for adding lines to a plot or annotating a plot with text, both of which requires x,y coordinates to position the feature. While the x-axis is discrete -- I cannot add a tick at x = 1.5 -- the horizontal dimension of the plot area itself is continuous and I can add points or lines or text at any x coordinate within the plot area. 

Here I generate the same plot using a continuous x-axis. This requires that I explicitly create a numeric column in the data with the x-axis position for each group. This is easy. The advantage of this is that I now have a continuous x-axis that is more manipulatable than a discrete x-axis. I use this to add treatment levels below the plot

#### First, wrangle the data

```{r plots-wrangle-for-continous-x}
# make sure treatment is a factor with the levels in the 
# desired order
fig2e[, treatment_i := as.integer(treatment)]

# convert m1_emm to a data.table
m1_emm_dt_i <- summary(m1_emm) %>%
  data.table()

# create treatment_i column
m1_emm_dt_i[, treatment_i := 1:4]

# convert m1_simple to a data.table
m1_simple_dt_i <- data.table(m1_simple)

# create group1 -- column containing x-position
# of the left side of the bracket
# need to look at m1_simple_dt to construct this.
# "ASK1F/F chow", "ASK1Δadipo chow" ,"ASK1F/F HFD", "ASK1Δadipo HFD" 
m1_simple_dt_i[, group1 := c(2, 4, 3, 4)]

# create group2 -- column containing x-position
# of the right side of the bracket
# need to look at m1_simple_dt to construct this.
# "ASK1F/F chow", "ASK1Δadipo chow" ,"ASK1F/F HFD", "ASK1Δadipo HFD" 
m1_simple_dt_i[, group2 := c(1, 3, 1, 2)]

m1_simple_dt_i[, p_rounded := p_round(p.value,
                                  digits = 2)]
m1_simple_dt_i[, p_pretty := p_format(p_rounded,
                                    digits = 2,
                                    accuracy = 1e-04,
                                    add.p = TRUE)]
```

Notes

1. The first line simply converts the categorical variable treatment into a numeric variable with values 1-4 assigned to the four treatment levels.
2. As above, `m1_emm_dt` is created and a column with the x-variable name is added
3. As above, `m1_simple_dt` is created and group1 and group2 columns with the x-axis positions of the two groups in the contrast are created. Here these columns are integers and not group names.

#### Second, generate the plot

```{r plots-ggplot-continuous-x}
m1_response_2 <- ggplot(
  data = fig2e,
  aes(x = treatment_i, # these 2 lines define the axes
      y = glucose_auc,
      color = ask1 # define the grouping variable
  )) +
  
  # add jittered points
  geom_jitter(width = 0.2) +
  
  # add layer containing means
  geom_point(data = m1_emm_dt_i,
             aes(y = emmean,
                 color = ask1), 
             size = 3) +
  
  # add layer containing error bars
  geom_errorbar(data = m1_emm_dt_i, 
                aes(y = emmean,
                    ymin = lower.CL, 
                    ymax = upper.CL,
                    color = ask1), 
                width = 0.05) +
  
  # add the p-value brackets
  stat_pvalue_manual(
    data = m1_simple_dt_i[c(2,4), ], # only rows 2, 4
    label = "p_pretty",
    y.position = c(3300, 3450),
    tip.length = 0.01) +
  
  # change the title of the y-axis
  ylab(expression(paste("AUC (mmol ", l^-1, " min)"))) +
  
  # change the theme
  theme_pubr() +

  # these theme modifications need to be added after re-setting
  # the theme
  theme(
    legend.position = "none", # remove legend
    axis.title.x = element_blank() # remove the x-axis title
  ) +
  
  # change the colors palette for the points
  scale_color_manual(values = pal_okabe_ito_blue) +
  
  NULL

  
m1_response_2
```

Notes

1. This code is exactly the same as [that above](#plot-individual-points) but uses an x-variable that is numeric instead of a factor.

#### Third, add the grid of treatments below the x-axis

```{r plots-method3-treatment-grid}
minus <- "\u2013" # good to put this in the setup chunk

# I've added x-axis group names in addition to the grid
x_levels <- rbind(
  c("", "Control", "Δadipo", "HFD", "Δadipo+HFD"),
  c("Δadipo:", minus, "+", minus, "+"),
  c("HFD: ", minus, minus, "+", "+")
)

x_breaks_text <- apply(x_levels, 2, paste0, collapse="\n")
x_breaks <- c(0.5, 1:4)

m1_response_2 <- m1_response_2 +
  coord_cartesian(xlim = c(0.5, 4.5)) +
  scale_x_continuous(breaks = x_breaks,
                     labels = x_breaks_text,
                     expand = c(0, 0)) +
  theme(axis.ticks.x = element_blank()) + # remove x-axis ticks
  
  NULL
        
m1_response_2
```
Notes

1. This is a good plot. A pretty good plot would include the effects.

### How to generate an Effects Plot

The effects plot is built using ggplot2 instead of ggpubr because the plot is "flipped" -- the y-axis is the categorical variable and the x-axis is the continuous variable. The base plot with the estimates (but not the error bars) could be made using ggpubr (see below) but the subsequent functions to modify the plot are awkward because of inconsistencies in the designation of the x and y-axis (old or new?).

```{r plots-how-to-plot-effects, fig.asp = 0.4, warning=FALSE}
# use the m1_simple_dt object created above

# create nice labels for the contrasts
m1_simple_dt[, contrast_pretty := c("ASK1Δadipo - Control",
                                    "ASK1Δadipo - HFD",
                                    "HFD - Control",
                                    "ASK1Δadipo+HFD - ASK1Δadipo")]
# make sure the levels of contrast_pretty are in the order of that in
# m1_simple_dt
m1_simple_dt[, contrast_pretty := factor(contrast_pretty,
                                         levels = contrast_pretty)]

m1_effects <- ggplot(data = m1_simple_dt,
                     aes(x = estimate,
                         y = contrast_pretty)) +
  
  geom_point() +
  
  xlab("Effects (mmol/L)") +

  # add error bars
  geom_errorbar(aes(x = estimate,
                    xmin = lower.CL,
                    xmax = upper.CL),
                width = 0.02) +
  
  # add zero effect line
  geom_vline(xintercept = 0,
             linetype = "dashed",
             color = pal_okabe_ito_blue[1]) +
  
  # add p-values
  annotate(geom = "text",
           x = m1_simple_dt$estimate + c(-30,0,0,0),
           y = 1:4 + 0.2,
           label = m1_simple_dt$p_pretty,
           size = 3) +
  
  theme_pubr() +
  rremove("ylab") + #remove y-axis, ggpubr function
  
  
  NULL

m1_effects
```

Notes

1. `c(-30,0,0,0)` is added to the x coordinate to shift the first p-value to the left of the zero-effect line.

**how would this be done with ggpubr?**

```{r plots-how-to-plot-effects-ggpubr, fig.asp = 0.4, warning=FALSE}
# use the modified m1_simple_dt object created in the previous chunk

# use ggpubr::ggerrorplot with no error
m1_effects_pubr <- ggerrorplot(data = m1_simple_dt,
                        y = "estimate",
                        x = "contrast_pretty",
                        ylab = "Effects (mmol/L)",
                        desc_stat = "mean", # mean only!
                        orientation = "horizontal") +
  
  # remove y-axis title
  rremove("ylab") + #ggpubr function
  
  # add error bars using ggplot function
  # note using original (not horizontal) axis designation
  geom_errorbar(aes(y = estimate,
                    ymin = lower.CL,
                    ymax = upper.CL),
                width = 0.02) +
  
  # add zero effect line using ggplot function
  # note using original (not horizontal) axis designation
  geom_hline(yintercept = 0,
             linetype = "dashed",
             color = pal_okabe_ito_blue[1]) +
  
  # add p-values
  # note using original (not horizontal) axis designation
  annotate(geom = "text",
           y = m1_simple_dt$estimate + c(-30,0,0,0),
           x = 1:4 + 0.3,
           label = m1_simple_dt$p_pretty,
           size = 3) +
  
  NULL

m1_effects_pubr
```

### How to combine the response and effects plots
#### Using the ggpubr response plot
The `cowplot::plotgrid()` function is used generally to arrange multiple plots into a single figure. Here I use it to combine the response and effects subplots into a single plot. The response plot is `m1_response` created using ggpubr.

```{r plots-combined-plot}
# modify the response and effects plots for consistency

# change group names for consistency
groups_pretty <- c("Control",
                   "ASK1Δadipo",
                   "HFD",
                   "ASK1Δadipo+HFD")
m1_bottom <- m1_response +
  scale_x_discrete(labels = groups_pretty) +
  coord_flip() # rotate to horizontal

m1_top <- m1_effects + # assign to new plot object
  scale_x_continuous(position="top") # move x-axis to "top"
  
m1_plot <- plot_grid(m1_top,
                     m1_bottom,
                     nrow = 2,
                     align = "v",
                     axis = "lr",
                     rel_heights = c(1, 1))

m1_plot
```
Notes

1. The response plot was flipped in this code. If you know that you want it in this orientation, simply build it with this orientation to avoid the kinds of axis designation ambiguities highlighted above with the ggpubr effects plot.
2. the `align` and `axis` arguments are used to force the plot areas to have the same width.
3. `rel_heights = c()` adjusts the relative heights of the top and bottom plot. This typically requires fiddling.
4. The placement of the p-values looks different than it does in the standalone effects plot. To improve the look in the combined plot, fiddle with the placement argument (the x and y positions) in the chunk that generates the effects plot.

#### Using a response plot with a treatment grid

Here, I re-build the response plot in a horizontal orientation

```{r plots-response-horizontal}
tab <- "\u0009"  # good to put this in the setup chunk
minus <- "\u2013" # good to put this in the setup chunk
minus_pad <- paste0(minus, "    ")
plus_pad <- "+    "

x_levels <- rbind(
  c(minus_pad, plus_pad, minus_pad, plus_pad),
  c(minus_pad, minus_pad, plus_pad, plus_pad)
)
x_breaks_text <- apply(x_levels, 2, paste0, collapse = "")
x_breaks_text <- c(x_breaks_text, "Δadipo HFD")
x_breaks <- c(1:4, 4.5)

m1_response_horiz <- ggplot(
  data = fig2e,
  aes(y = treatment_i, # these 2 lines define the axes
      x = glucose_auc,
      color = ask1 # define the grouping variable
  )) +
  
  # add jittered points
  geom_jitter(width = 0.2) +
  
  # add layer containing means
  geom_point(data = m1_emm_dt_i,
             aes(x = emmean,
                 color = ask1), 
             size = 3) +
  
  # add layer containing error bars
  geom_errorbar(data = m1_emm_dt_i, 
                aes(x = emmean,
                    xmin = lower.CL, 
                    xmax = upper.CL,
                    color = ask1), 
                width = 0.05) +
  
  # change the title of the y-axis
  xlab(expression(paste("AUC (mmol ", l^-1, " min)"))) +
  
  # change the theme
  theme_pubr() +

  # these theme modifications need to be added after re-setting
  # the theme
  theme(
    legend.position = "none", # remove legend
    axis.title.y = element_blank(), # remove the x-axis title
    axis.ticks.y = element_blank() # remove y-axis ticks
  ) +
  
  # change the colors palette for the points
  scale_color_manual(values = pal_okabe_ito_blue) +
  
  # add the treatment
  
  coord_cartesian(ylim = c(0.5, 4.5)) +
  scale_y_continuous(breaks = x_breaks,
                     labels = x_breaks_text,
                     expand = c(0, 0)) +
  NULL

m1_response_horiz
```

```{r plots-response-effects-plot-treatment}
m1_plot <- plot_grid(m1_top,
                     m1_response_horiz,
                     nrow = 2,
                     align = "v",
                     axis = "lr",
                     rel_heights = c(1, 1))

m1_plot
```

### How to add the interaction effect to response and effects plots

In the experiment for Figure 2E, a good question to ask is, is the effect of ASK1Δadipo conditional on diet? For example, a scenario where ASK1Δadipo lowers AUC about the same amount in both Chow and HFD mice (that is, the effect of ASK1Δadipo is not conditional on diet) has a different underlying biological explanation of the control of browning than a scenario where ASK1Δadipo lowers AUC only in HFD mice. To pursue this, we need an estimate of the interaction effect. In general, **if an experiment has a factorial design, we always want estimates of the interaction effects**

#### Adding interaction p-value to a response plot

Wrangle

```{r plots-ixn-response-wrangle}
m1_coef <- coef(summary(m1))
ixn_estimate <- m1_coef["ask1ASK1Δadipo:dietHFD", "Estimate"]
p_ixn <- m1_coef["ask1ASK1Δadipo:dietHFD", "Pr(>|t|)"] %>%
  p_round(digits = 3) %>%
  p_format(digits = 3, accuracy = 1e-04)
p_ixn_text <- paste0("interaction p = ", p_ixn)
```


```{r}
m1_response_ixn <- m1_response_fac +
  
  # add line connecting group means
  # comment out all lines to remove
  geom_line(data = m1_emm_dt,
            aes(y = emmean,
                group = ask1,
                color = ask1),
            position = position_dodge(width = dodge_width)) +
  
  # add p-value
  annotate(geom = "text",
           x = 1.5,
           y = 2300,
           label = p_ixn_text) +
  
  NULL

m1_response_ixn
```
#### Adding interaction effect to effects plot
##### Add the interaction effect to the contrast table

```{r plots-ixn-contrast}
# convert m1_simple to data.table
m1_simple_dt <- data.table(m1_simple)

# get interaction p-value using emmeans::contrast()
m1_ixn <- contrast(m1_emm,
                   interaction = c("revpairwise"),
                   by = NULL) %>%
  summary(infer = TRUE) %>%
  data.table()

setnames(m1_ixn,
         old = names(m1_ixn)[1:2], 
         new = names(m1_simple_dt)[1:2])
m1_ixn[, contrast := "ask1:diet"]

# note that column order does not need to be the same
m1_effects_dt <- rbind(m1_simple_dt,
                       m1_ixn)

m1_effects_dt[, contrast_pretty := c("Δadipo - Control",
                                     "Δadipo+HFD - HFD",
                                     "HFD - Control",
                                     "Δadipo+HFD - Δadipo",
                                     "Interaction")]
m1_effects_dt[, p_round := p_round(p.value, digits = 2)]
m1_effects_dt[, p_pretty := p_format(p_round,
                                     digits = 2,
                                    accuracy = 1e-04,
                                    add.p = TRUE)]

# don't forget this imperative step!
# otherwise your p-values and factor labels won't match!
m1_effects_dt[, contrast_pretty := factor(contrast_pretty,
                                          levels = contrast_pretty)]
```

```{r plots-ixn-effects-plot}
m1_effects <- ggplot(data = m1_effects_dt,
                     aes(x = estimate,
                         y = contrast_pretty)) +
  
  geom_point() +
  
  xlab("Effects (mmol/L)") +

  # add error bars
  geom_errorbar(aes(x = estimate,
                    xmin = lower.CL,
                    xmax = upper.CL),
                width = 0.02) +
  
  # add zero effect line
  geom_vline(xintercept = 0,
             linetype = "dashed",
             color = pal_okabe_ito_blue[1]) +
  
  # add p-values
  annotate(geom = "text",
           x = m1_effects_dt$estimate + c(-30,0,0,0,0),
           y = 1:5 + 0.2,
           label = m1_effects_dt$p_pretty,
           size = 3) +
  
  theme_pubr() +
  
  rremove("ylab") + #remove y-axis, ggpubr function

  NULL

m1_effects
```

```{r plots-ixn-combined-plot}
m1_top <- m1_effects +
  scale_x_continuous(position="top") # move x-axis to "top"

m1_plot <- plot_grid(m1_top,
                     m1_response_horiz,
                     nrow = 2,
                     align = "v",
                     axis = "lr",
                     rel_heights = c(1, 0.9))

m1_plot
```
Notes

1. This is a pretty good plot

<!--chapter:end:chapters/29-plotting_models.Rmd-->

# Part IV: Some Fundamentals of Statistical Modeling {-}

<!--chapter:end:chapters/10-part-II-statistics-fundamentals.Rmd-->

# Variability and Uncertainty (Standard Deviations, Standard Errors, Confidence Intervals) {#uncertainty}

```{r uncertain-setup, echo=FALSE}
library(here)
library(readxl)
library(janitor)
library(ggplot2)
library(data.table)

here <- here::here # plyr also defines a here function
data_path <- "data"
```

**Uncertainty** is the stuff of science. A major goal of statistics is measuring uncertainty. What do we mean by uncertainty? Uncertainty is the error in estimating a parameter, such as the mean of a sample, or the difference in means between two experimental treatments, or the predicted response given a certain change in conditions. Uncertainty is measured with a **variance** or its square root, which is a **standard deviation**. The standard deviation of a statistic is also (and more commonly) called a **standard error**.

Uncertainty emerges because of variability. In any introductory statistics class, students are introduced to two measures of variability, the "standard deviation" and the "standard error." These terms are absolutely fundamental to statistics -- they are the start of everything else. Yet, many biology researchers confuse these terms and certainly, introductory students do too.

When a research biologist uses the term "standard deviation," they are probably referring to the sample standard deviation which is a measure of the variability of a sample. When a research biologist uses the term "standard error," they are probably referring to the standard error of a mean, but it could be the standard error of another statistics, such as a difference between means or a regression slope. An important point to remember and understand is that all standard errors *are* standard deviations. This will make more sense soon.

## The sample standard deviation vs. the standard error of the mean

A standard deviation is the square root of the sampling variance.

### Sample standard deviation

The sample standard deviation is a measure of the variability of a sample. For example, were we to look at a histological section of skeletal muscle we would see that the diameter of the fibers (the muscle cells) is variable. We could use imaging software to measure the diameter of a sample of 100 cells and get a **distribution** like this

```{r histogram, echo=FALSE, message=FALSE}
n <- 100
set.seed(1) # set seed so that same fake data occurs each time.
fake_data <- data.table(Diameter=rnorm(n, mean=69.1, sd=3.1))
gg <- ggplot(data=fake_data, aes(x=Diameter)) +
  geom_histogram() +
  xlab("Diameter (µm)") +
  NULL
gg

```

The mean of this sample is `r round(mean(fake_data$Diameter),1)`µm and the standard deviation is `r round(sd(fake_data$Diameter),1)` µm. The standard deviation is the square root of the variance, and so computed by

\begin{equation}
s_y = \sqrt{\frac{\sum_{i=1}^n{(y_i - \overline{y})^2}}{n-1}}
(\#eq:variance)
\end{equation}

Memorize this equation. To understand the logic of this measure of variability, note that $y_i - \overline{y}$ is the **deviation** of the $i$th value from the sample mean, so the numerator is the sum of squared deviations. The numerator is a sum over $n$ items and the denominator is $n-1$ so the variance is (almost!) an averaged squared deviation. More variable samples will have bigger deviations and, therefore, bigger average squared deviations. Since the standard deviation is the square root of the variance, a standard deviation is the square root of an average squared deviation. This makes it similar in value to the averaged deviation (or average of the absolute values of the deviations since the average deviation is, by definition of a mean, zero).

#### Notes on the variance and standard deviation

1. Variances are additive but standard deviations are not. This means that the variance of the sum of two independent (uncorrelated) random variables is simply the sum of the variances of each of the variables. This is important for many statistical analyses.
2. The units of variance are the square of the original units, which is awkward for interpretation. The units of a standard deviation is the same as that of the original variable, and so is much easier to interpet.
3. For variables that are approximately normally distributed, we can map the standard deviation to the quantiles of the distribution. For example, 68% of the values are within one standard deviation of the mean, 95% of the values are within two standard deviations, and 99% of the values are within three standard deviations.

### Standard error of the mean
A standard error of a statistic is a measure of the precision of the statistic. The standard error of the mean is a measure of the precision of the estimate of the mean. The standard error of a difference in means is a measure of the precision of the estimate of the difference in means. The smaller the standard error, the more precise the estimate. The standard error of the mean (SEM) is computed as

\begin{equation}
SEM = \frac{s_y}{\sqrt{n}}
(\#eq:se)
\end{equation}

The SEM is often denoted $s_{\bar{y}}$ to indicate that it is a standard deviation of the mean ($\bar{y}$).

#### The standard error of the mean can be thought of as a standard deviation of an infinitely long column of re-sampled means

In what sense is a standard error a standard deviation? This is kinda weird. If we sample 100 cells in the slide of muscle tissue and compute the mean diameter, how can the mean have a standard deviation? There is only one value!

To understand how the SEM is a standard deviation, imagine that we sample $n$ values from $N(\mu, \sigma^2)$ (a normal distribution with mean $\mu$ and variance $\sigma^2$. The mean of our sample is an estimate of $\mu$ the standard deviation of sample is an estimate of $\sigma$) an infinite number of times and each time, we write down the mean of the new sample. The standard deviation of this infinitely long column of means is the standard error of the mean. Our observed SEM is an estimate of this true value because our observed  standard deviation is an estimate of $\sigma$.

#### A standard deviation can be computed for any statistic -- these are all standard errors. 

The SEM is only one kind of standard error. A standard deviation can be computed for any statistic -- these are all standard errors. For some statistics, such as the mean, the standard error can be computed directly using an equation, such as that for the SEM (equation \@ref(eq:se)). For other statistics, a computer intensive method known as the **bootstrap** is necessary to compute a standard error. We will return to the bootstrap in Section \@ref(bootstrap).

#### Notes on standard errors

1. The units of a standard error are the units of the measured variable.
2. A standard error is proportional to sample variability (the sample standard deviation, $s_y$) and inversely proportional to sample size ($n$). Sample variability is a function of both natural variation (there really is variation in diameter among fibers in the quadriceps muscle) and measurement error (imaging software with higher resolution can measure a diameter with less error). Since the SEM is a measure of the precision of estimating a mean, this means this precision will increase (or the SEM will decrease) if 1) an investigator uses methods that reduce measurement error and 2) an investigator computes the mean from a larger sample.
3. This last point (the SEM decreases with sample size) seems obvious when looking at equation \@ref(eq:se), since $n$ is in the denominator. Of course $n$ is also in the denominator of equation \@ref(eq:variance) for the sample standard deviation but the standard deviation does not decrease as sample size increases. First this wouldn't make any sense -- variability is variability. A sample of 10,000 cell diameters should be no more variable than a sample of 100 cell diameters (think about if you agree with this or not). Second, this should also be obvious from equation \@ref(eq:variance). The standard deviation is the square root of an average and averages don't increase with the number of things summed since both the the numerator (a sum) and denominator increase with $n$. 

## Using Google Sheets to generate fake data to explore the standard error
In statistics we are interested in estimated parameters of a **population** using measures from a **sample**. The goal in this section is to use Google Sheets (or Microsoft Excel) to use fake data to discover the behavior of sampling and to gain some intuition about uncertainty using standard errors.

### Steps
1. Open Google Sheets
2. In cell A1 type "mu". mu is the greek letter $\mu$ and is very common notation for the poplation value (the TRUE value!) of the mean of some hypothetical measure. In cell B1, insert some number as the value of $\mu$. Any number! It can be negative or positive.
3. In cell A2 type "sigma". sigma is the greek letter $\sigma$. $\sigma^2$ is very common (universal!) notation for the population (TRUE) variance of some measure or parameter. Notice that the true (population) values of the mean and variance are greek letters. This is pretty standard in statistics. In cell B2, insert some positive number (standard deviations are the positive square roots of the variance).
4. In cell A8 type the number 1
5. In cell A9 insert the equation "=A8 + 1". What is this equation doing? It is adding the number 1 to to the value in the cell above, so the resulting value should be 2.
6. In Cell B8, insert the equation "=normsinv(rand())*\$B\$2 + \$B\$1". The first part of the equation creates a random normal variable with mean 0 and standard deviation 1. multiplication and addition transform this to a random normal variable with mean $\mu$ and standard deviation $\sigma$ (the values you set in cells B1 and B2).
7. copy cell B8 and paste into cell B9. Now Higlight cells A9:B9 and copy the equations down to row 107. You now have 100 random variables sampled from a infinite population with mean $\mu$ and standard deviation $\sigma$.
8. In cell A4 write "mean 10". In cell B4 insert the equation "=average(B8:B17)". The resulting value is the **sample mean** of the first 10 random variables you created. Is the mean close to $\mu$?
9. In cell A5 write "sd 10". In cell B5 insert the equation "stdev(B8:B17)". The result is the **sample standard deviation** of the first 10 random variables. Is this close to $\sigma$?
10. In cell A6 write "mean 100". In cell B6 insert the equation "=average(B8:B107)". The resulting value is the **sample mean** of the all 100 random variables you created. Is this mean closer to $\mu$ than mean 10?
11. In cell A7 write "sd 100". In cell B7 insert the equation "=stdev(B8:B107)". The resulting value is the **sample standard deviation** of the all 100 random variables you created. Is this SD closer to $\sigma$ than sd 10?

The sample standard deviation is a measure of the variability of the sample. The more spread out the sample (the further each value is from the mean), the bigger the sample standard deviation. The sample standard deviation is most often simply known as "The" standard deviation, which is a bit misleading since there are many kinds of standard deviations!

Remember that your computed mean and standard deviations are estimates computed from a sample. They are estimates of the true values $\mu$ and $\sigma$. Explore the behavior of the sample mean and standard deviation by re-calculating the spreadsheet. In Excel, a spreadsheet is re-calculated by simultaneously pressing the command and equal key. In Google, command-R recalculates but is painfully slow. Instead, if using Google Sheets, just type the number 1 into a blank cell, and the sheet recalculates quickly. Do it again. And again.

Each time you re-calculate, a new set of random numbers are generated and the new means and standard deviations are computed. Compare mean 10 and mean 100 each re-calculation. Notice that these estimates are variable. They change with each re-calculation. How variable is mean 10 compared to mean 100? The variability of the estimate of the mean is a measure of **uncertainty** in the estimate. Are we more uncertain with mean 10 or with mean 100? This variability is measured by a standard deviation. This **standard deviation of the mean** is also called the **standard error of the mean**. Many researchers are loose with terms and use "The" standard error to mean the standard error of the mean, even though there are many kinds of standard errors. In general, "standard error"" is abbreviated as "SE." Sometimes "standard error of the mean" is specifically abbreviated to "SEM."

The standard error of the mean is a measure of the precision in estimating the mean. The smaller the value the more precise the estimate. The standard error of the mean *is* a standard deviation of the mean. This is kinda weird. If we sample a population one time and compute a mean, how can the mean have a standard deviation? There is only one value! And we compute this value using the sample standard deviation:  $SEM = \frac{SD}{\sqrt{N}}$. To understand how the SEM is a standard deviation, Imagine recalculating the spread sheet an infinite number of times and each time, you write down the newly computed mean. The standard error of the mean is the standard deviation of this infinitely long column of means.

## Using R to generate fake data to explore the standard error

note that I use "standard deviation" to refer to the sample standard deviation and "standard error" to refer to the standard error of the mean (again, we can compute standard errors as a standard deviation of any kind of estimate)

### part I
In the exercise above, you used Google Sheets to generate $p$ columns of fake data. Each column had $n$ elements, so the matrix of fake data was $n \times m$ (it is standard in  most fields to specify a matrix as rows by columns). This is *much* easier to do in R and how much grows exponentially as the size of the matrix grows.

To start, we just generate a $n \times p$ matrix of normal random numbers.

```{r fake-data-for-mean, sd, se}
# R script to gain some intuition about standard deviation (sd) and standard error (se)
# you will probably need to install ggplot2 using library(ggplot2) 
n <- 6 # sample size
p <- 100 # number of columns of fake data to generate
fake_data <- matrix(rnorm(n*p, mean=0, sd=1), nrow=n, ncol=p) # create a matrix
```

the 3rd line is the cool thing about R. In one line I'm creating a dataset with $n$ rows and $p$ columns. Each column is a sample of the standard normal distribution which by definition has mean zero and standard deviation of 1. But, and this is important, any sample from this distribution will not have exactly mean zero and standard deviation of 1, because it's a sample, the mean and standard deviation will have some small errror from the truth. The line has two parts to it: first I'm using the function "rnorm" (for random normal) to create a vector of n*m random, normal deviates (draws from the random normal distribution) and then I'm organizing these into a matrix (using the function "matrix")

To compute the vector of means, standard deviations, and standard errors for each column of `fake_data`, use the `apply()` function.

```{r}
means <- apply(fake_data,2,mean) # the apply function is super useful
sds <- apply(fake_data,2,sd)
sems <- sds/sqrt(n)
```

`apply()` is a workhorse in many R scripts and is often used in R scripts in place of a for-loop (see below) because it takes fewer lines of code.

The SEM is the standard deviation of the mean, so let's see if the standard deviation of the means is close to the true standard error. We sampled from a normal distribution with SD=1 so the true standard is

```{r}
1/sqrt(n)
```

and the standard deviation of the $p$ means is 

```{r}
sd(means)
```

Questions

1. how close is `sd(means)` to the true SE?
2. change p above to 1000. Now how close is sd(means) to the true SE?
3. change p above to 10,000. Now how close is sd(means) to the true SE?

### part II - means

This is a visualization of the spread, or variability, of the sampled means
```{r}
qplot(means)
```

Compute the mean of the means

```{r}
mean(means)
```

Questions

1.  Remember that the true mean is zero. How close, in general, are the sampled means to the true mean. How variable are the means? How is this quantified?
2. change n to 100, then replot. Are the means, in general, closer to the true mean? How variable are the means now?
3. Is the mean estimated with $n=100$ closer to the truth, in general, then the mean estimated with $n=6$?
4. Redo with $n=10000$

### part III - how do SD and SE change as sample size (n) increases?

```{r}
mean(sds)
```

Questions

1. what is the mean of the standard deviations when n=6 (set p=1000)
2. what is the mean of the standard deviations when n=100 (set p=1000)
3. when n = 1000? (set p=1000)
4. when n = 10000? (set p=1000)
5. how does the mean of the standard deviations change as n increases (does it get smaller? or stay about the same size)
6. repeat the above with SEM

```{r}
mean(sems)

```

Congratulations, you have just done a Monte Carlo simulation!

### Part IV -- Generating fake data with for-loops

A **for-loop** is used to iterate a computation.

```{r}
n <- 6 # sample size
n_iter <- 10^5 # number of iterations of loop (equivalent to p)
means <- numeric(n_iter)
sds <- numeric(n_iter)
sems <- numeric(n_iter)
for(i in 1:n_iter){
  y <- rnorm(n) # mean=0 and sd=1 are default so not necessary to specify
  means[i] <- mean(y)
  sds[i] <- sd(y)
  sems[i] <- sd(y)/sqrt(n)
}
sd(means)
mean(sems)
```

Questions

1. What do `sd(means)` and `mean(sems)` converge to as `n_iter` is increased from 100 to 1000 to 10,000?
2. Do they converge to the same number?
3. Should they?
4. What is the correct number?

Question number 4 is asking what is E(SEM), the "expected standard error of the mean". There is a very easy formula to compute this. What is it?

```{r echo=FALSE, eval=FALSE}
n <- 6 # sample size
n_iter <- 10^5 # number of iterations of loop (equivalent to p)
means <- numeric(n_iter)
sems <- numeric(n_iter)
sems2 <- numeric(n_iter)
for(i in 1:n_iter){
  y <- rnorm(n) # mean=0 and sd=1 are default so not necessary to specify
  means[i] <- mean(y)
  sems[i] <- sd(y)/sqrt(n)
  sems2[i] <- sems[i]^2
}
sd(means)
mean(sems)
sqrt(mean(sems2))
1/sqrt(n)
```

## Bootstrapped standard errors {#bootstrap}

The bootstrap is certainly one of the most valuable tools invented in modern statistics. But, it's not only a useful tool for applied statistics, it's a useful tool for understanding statistics. Playing with a parametric bootstrap will almost certainly induce an "aha, so that's what statisticians mean by ..." moment.

To understand the bootstrap, let's review a standard error. A *parametric* standard error of a mean is the *expected* standard deviation of an infinite number of means. A standard error of any statistic is the *expected* standard deviation of that statistic. I highlight *expected* to emphasize that parametric standard errors assume a certain distribution (not necessarily a Normal distribution, although the equation for the SEM in Equation \@ref(eq:se) assumes a normal distribution if the standard deviation is computed as in Equation \@ref(eq:sd)).

A bootstrapped standard error of a statistic is the **empirical** standard deviation of the statistic from a finite number of *samples*. The basic algorithm for a bootstrap is (here "the statistic" is the mean of the sample)

1. sample $n$ values from a probability distribution
2. compute the mean
3. repeat step 1 and 2 many times
4. for a bootstrapped standard error, compute the standard deviation of the set of means saved from each iteration of steps 1 and 2.

The probability distribution can come from two sources:

1. A **parametric bootstrap** uses samples from a parametric probability distribution, such as a Normal distribution or a poisson distribution (remember, these are "parametric" because the distribution is completely described by a set of parameters). A good question is why bother? In general, one would use a parametric bootstrap for a statistic for which there is no formula for the standard error, but the underlying data come from a parametric probability distribution.
2. A **non-parametric** bootstrap uses *resamples from the data*. The data are resampled *with replacement*. "Resample with replacement" means to sample $n$ times from the full set of observed values. If we were to do this manually, we would i) write down each value of the original sample on its own piece of paper and throw all pieces into a hat. ii) pick a paper from the hat, add its value to sample $i$, and return the paper to the hat. iii) repeat step ii $n$ times, where $n$ is the original sample size. The new sample contains some values multiple times (papers that were picked out of the hat more than once) and is missing some values (papers that were not picked out in any of the $n$ picks). A good question is, why bother? A non-parametric bootstrap assumes no specific *parametric* probability distribution but it does assume the distributio of the observed sample is a good approximation of the true population distribution (in which case, the probability of picking a certain value is a good approximation to the true probability).

### An example of bootstrapped standard errors using vole data

Let's use the vole data to explore the bootstrap and "resampling". The data are archived at Dryad Repository. Use the script in Section \@ref(vole-data) to wrangle the data into a usable format.

1. URL: https://datadryad.org//resource/doi:10.5061/dryad.31cc4
2. file: RSBL-2013-0432 vole data.xlsx
3. sheet: COLD VOLES LIFESPAN

The data are the measured lifespans of the short-tailed field vole (*Microtus agrestis*) under three different experimental treatments: vitamin E supplementation, vitamin C supplementation, and control (no vitamin supplementation). Vitamins C and E are antioxidants, which are thought to be protective of basic cell function since they bind to the cell-damaging reactive oxygen species that result from cell metabolism.

```{r import_voles_10, echo=FALSE, warning=FALSE, message=FALSE}
folder <- "Deleterious consequences of antioxidant supplementation on lifespan in a wild-derived mammal"
filename <- "RSBL-2013-0432 vole data.xlsx"
file_path <- here(data_path, folder, filename)
vole <- read_excel(file_path, sheet="COLD VOLES LIFESPAN", range="a2:d98") %>%
  clean_names() %>%
  data.table()
# create treatment column
# in rows where control=1, set treatment value to "control"
vole[control==1, treatment:="control"]
# in rows where vitamin_e=1, set treatment value to "vitamin_e"
vole[vitamin_e==1, treatment:="vitamin_E"]
# in rows where vitamin_c=1, set treatment value to "vitamin_c"
vole[vitamin_c==1, treatment:="vitamin_C"]

# change column "lifespan_days" to "lifespan"
setnames(vole, old="lifespan_days", "lifespan")

```

Let's compute the standard error of the mean of the control group lifespan using both a parametric and a nonparametric bootstrap. To implement the algorithm above using easy-to-understand code, I'll first extract the set of lifespan values for the control group and assign it to its own variable.

```{r uncertain-subset-voles}
control_voles <- vole[treatment=="control", lifespan]
```

`[treatment=="control", ]` indexes the rows (that is, returns the row numbers) that satisfy the condtion `treatment = "control"`. Or, put another way, it selects the **subset** of rows that contain the value "control" in the column "treatment". `[, lifespan]` indexes the column labeled "lifespan". Combined, these two indices extract the values of the column "lifespan" in the subset of rows that contain the value "control" in the column "treatment". The resulting vector of values is assigned to the variable "control_voles".

#### parametric bootstrap

```{r uncertain-parametric-bootstrap}
# we'll use these as parameters for parametric bootstrap
n <- length(control_voles)
mu <- mean(control_voles)
sigma <- sd(control_voles)

n_iter <- 1000 # number of bootstrap iterations, or p
means <- numeric(n_iter) # we will save the means each iteration to this

for(iter in 1:n_iter){ # this line sets up the number of iterations, p
  fake_sample <- rnorm(n, mean=mu, sd=sigma)
  means[iter] <- mean(fake_sample)
}
(se_para_boot <- sd(means))
```

#### non-parametric bootstrap

```{r bootstrap-Vole}
n_iter <- 1000 # number of bootstrap iterations, or p
means <- numeric(n_iter) # we will save the means each iteration to this
inc <- 1:n # inc indexes the elements to sample. By setting inc to 1:n prior to the loop, the first mean that is computed is the observed mean
for(iter in 1:n_iter){ # this line sets up the number of iterations, p
  means[iter] <- mean(control_voles[inc]) # inc is the set of rows to include in the computation of the mean.
  inc <- sample(1:n, replace=TRUE) # re-sample for the next iteration
}
(se_np_boot <- sd(means))
```

The parametric bootstrapped SEM is `r round(se_para_boot, 2)`. The non-parametric bootstrapped SEM is `r round(se_np_boot, 2)`. Run these several times to get a sense how much variation there is in the bootstrapped estimate of the SEM given the number of iterations. Compute the **parametric** standard error using equation \@ref(eq:se) and compare to the bootstrapped values.

```{r parametric-se, echo=FALSE}
se_para <- sd(control_voles)/sqrt(n)
```


## Confidence Interval

Here I introduce a **confidence interval** of a sample mean but the concept is easily generalized to any parameter. The mean of the Control voles is `r round(mean(control_voles), 1)` and the SE of the mean is `r round(se_para, 2)`. The SE is used to construct the lower and upper boundary of a "1 - $\alpha$" confidence interval using `lower <- mean(x) + qt(alpha/2, df = n-1)*se(x)` and `upper <- mean(x) + qt(1-(alpha/2), df = n-1)*se(x)`.

```{r}
(lower <- mean(control_voles) + qt(0.05/2, df=(n-1))*sd(control_voles)/sqrt(n))
(upper <- mean(control_voles) + qt(1 - 0.05/2, df=(n-1))*sd(control_voles)/sqrt(n))
```

The function `qt` maps a probability to a *t*-value -- this is the opposite of a *t* test, which maps a *t*-value to a probability. Sending $\alpha/2$ and $1 - \alpha/2$ to `qt` returns the bounds of the confidence interval on a standardized scale. Multiplying these bounds by the standard error of the control vole lifespan pops the bounds onto the scale of the control vole lifespans. 

We can check our manual computation with the linear model

```{r}
confint(lm(control_voles ~ 1))
```

### Interpretation of a confidence interval
Okay, so what *is* a confidence interval? A confidence interval of the mean is a measure of the uncertainty in the estimate of the mean. A 95% confidence interval has a 95\% probability (in the sense of long-run frequency) of containing the true mean. It is not correct to state that "there is a 95% probability that the true mean lies within the interval". These sound the same but they are two different probabilities. The first (correct interpretation) is a probability of a procedure -- if we re-do this procedure (sample data, compute the mean, and compute a 95% CI), 95% of these CIs will contain the true mean. The second (incorrect interpretation) is a probability that a parameter ($\mu$, the true mean) lies within some range. The second (incorrect) interepretation of the CI is correct only if we also assume that *any* value of the mean is equally probable (Greenland xxx), an assumption that is absurd for almost any data. 

Perhaps a more useful interpretation of a confidence interval is, a confidence interval contains the range of true means that are compatible with the data, in the sense that a $t$-test would not reject the null hypothesis of a difference between the estimate and any value within the interval (this interpretation does not imply anything about the true value) (Greenland xxx). The "compatibility" interpretation is very useful because it implies that values outside of the interval are less compatible with the data.

```{r echo = FALSE}
# @richarddmorey: Saying that there is evidence against true hypotheses, in the frequentist sense. Building simple models of the world and asking “if this were true, would I often find strong evidence against it?”

# @richarddmorey: But I think of frequentist stats more as a collection of thought experiments, and I know of few scientists that view freq. *probability* useless as a model.

# consider all possible values of statistic relative to observed value. As we move away, there is stronger evidence that that value is not true, or less compatability. The CI is the range of possible values of the statistic that we the data isn't very strong evidence against the value, and so that value is "compatabile" with the data. Compatability is a continuous concept but a CI is a trichotomy. A better representation of this concept is a p-curve.
```

Let's look at the confidence intervals of all three vole groups in light of the "compatibility" interpretation.

```{r}
vole_ci <- vole[, .(lifespan = mean(lifespan),
                    lo = mean(lifespan) + sd(lifespan)/sqrt(.N)*qt(.025, (.N-1)),
                    up = mean(lifespan) + sd(lifespan)/sqrt(.N)*qt(.975, (.N-1)),
                    N = .N),
                by = .(treatment)]
ggplot(data=vole_ci, aes(x=treatment, y=lifespan)) +
  geom_point() +
  geom_errorbar(aes(x=treatment, ymin=lo, ymax=up), 
                width=0.1) +
  NULL


```


<!--chapter:end:chapters/15-uncertainty_sd_and_se.Rmd-->

# P-values

*A general perception of a “replication crisis” may thus reflect failure to recognize that statistical tests not only test hypotheses, but countless assumptions and the entire environment in which research takes place. Because of all the uncertain and unknown assumptions that underpin statistical inferences, we should treat inferential statistics as highly unstable local descriptions of relations between assumptions and data, rather than as providing generalizable inferences about hypotheses or models. And that means we should treat statistical results as being much more incomplete and uncertain than is currently the norm.*^[Amrhein, V., Trafimow, D. and Greenland, S., 2019. Inferential statistics as descriptive statistics: There is no replication crisis if we don’t expect replication. The American Statistician, 73(sup1), pp.262-270.]

```{r pvalue-setup, echo=FALSE, warning=FALSE, message=FALSE}

# wrangling packages
library(here)
library(janitor)
library(readxl)
library(data.table)
library(stringr)

# analysis packages
library(emmeans)
library(mvtnorm)

# graphing packages
library(ggsci)
library(ggpubr)
library(ggforce)
library(ggthemes)
library(cowplot)
library(lazyWeave) #pvalstring

here <- here::here
data_folder <- "data"

source_path <- here("R", "ggplotsci.R")
source(source_path)

# Okabe & Ito palette
pal_okabe_ito <- colorblind_pal()(8)[2:8] # ggthemes
```

A *p*-value *is a measure of the compatibility between observed data and the null model*. Here, "compatibility" is a probability, specifically, the probability of sampling a test-statistic as or more extreme than the observed test statistic, if all the assumptions used to compute the *p*-value are true.

To deconstruct what this means, and the implications of the meaning, let's use the experiment from Figure 2i in the study on the browning of white adipose tissue in mice that was introduced in Chapter \@ref(analyzing-experimental-data-with-a-linear-model) ([Analyzing experimental data with a linear model]).

Data source: [ASK1 inhibits browning of white adipose tissue in obesity](https://www.nature.com/articles/s41467-020-15483-7)

```{r p-value-fig_2i, echo=FALSE, message=FALSE, warning=FALSE}
data_from <- "ASK1 inhibits browning of white adipose tissue in obesity"
file_name <- "41467_2020_15483_MOESM4_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)
  
fig_2_sheet <- "Source Date_Figure 2"

range_2i <- "A265:G266"
treatment_levels <- c("ASK1F/F", "ASK1Δadipo")
fig_2i <- read_excel(file_path,
                     sheet = fig_2_sheet,
                     range = range_2i,
                     col_names = FALSE) %>%
  transpose(make.names=1) %>%
  data.table() %>%
  melt(measure.vars = treatment_levels,
       variable.name = "treatment",
       value.name = "liver_tg") %>%
  na.omit()

fig_2i[, treatment := factor(treatment, treatment_levels)]

# View(fig_2i)

```

```{r p-value-fig_2i_m1, echo = FALSE}
fig_2i_m1 <- lm(liver_tg ~ treatment, data = fig_2i)
```

A linear model with `liver_tg` as the response variable and `treatment` as the single $X$-varaiable was fit to the data. Figure \@ref(p-value-fig2i-fig) is the plot of the modeled means, 95% confidence intervals of the mean, and *p*-value of the significance test of the effect of treatment on liver triacylglycerol.

```{r p-value-fig2i-fig, echo=FALSE, message=FALSE, warning=FALSE, fig.cap = "UCP1 expression, relative to the mean level in the control group. Mean (circle) and 95% confidence interval (line) are shown. Unadjusted p-values are from linear model with sh-RNA treatment and LPS treatment fully crossed."}
fig_2i_m1_emm <- emmeans(fig_2i_m1, specs = "treatment")
fig_2i_m1_pairs <- contrast(fig_2i_m1_emm,
                            method = "revpairwise") %>%
  summary(infer = TRUE)

gg1 <- ggplot_the_response(
  fig_2i_m1,
  fig_2i_m1_emm,
  fig_2i_m1_pairs,
  y_label = "Liver TG (µmol per g liver)",
  contrast_rows = "all",
  palette = pal_okabe_ito_4,
)

gg1
```

```{r pvalue-fig_2i_m1_coef, echo=FALSE}
fig_2i_m1_coef_p1 <- summary(fig_2i_m1) %>%
  coef()
fig_2i_m1_coef_p2  <- confint(fig_2i_m1)
fig_2i_m1_coef <- cbind(fig_2i_m1_coef_p1,
                        fig_2i_m1_coef_p2)
```

The coefficients of the model, and the standard error, 95% confidence interval, test-statistic, and *p*-value of each coefficient are shown in Table \@ref(tab:pvalue-coef-table) Recall from Chapter \@ref(analyzing-experimental-data-with-a-linear-model) that, with this model, the coefficient for the intercept term is the mean `liver_tg` for the control group, which is the estimate of the true mean for a mice with functional ASK1 protein. And the coefficient for the treatmentASK1Δadipo term is the difference in means between the knockout and control group, which is the estimate for the true effect of knocking out ASK1 in the adipose tissue. The *p*-value for this "effect" term is 0.012. How do we interpret this number?

```{r pvalue-coef-table, echo=FALSE}
#colnames(fig_2i_m1_coef) <- c("Estimate", "SE", "t", "p", "2.5 %", "97.5 %")
knitr::kable(fig_2i_m1_coef, digits = c(1, 2, 1, 3, 1, 1))
```

## A *p*-value is the probability of sampling a value as or more extreme than the test statistic if sampling from a null distribution

The test statistic in the table above is a *t*-value. For this specific model, this *t*-value is precisely the *t*-value one would get if they executed a classical Student *t*-test on the two groups of liver TG values. Importantly, this is not generally true. For many of the models in this text, a *t*-value is computed but this is *not* the *t*-value that would be computed in a classical *t*-test.

When we do a *t*-test, we get a *p*-value. The probability returned in a *t*-test is $p = \mathrm{prob}(t \ge t_{obs} | H_0)$. Read this as "the p-value is the probability of observing a t-value that is greater than or equal to the observed t-value, given the null model is true." Probability, in this text, is a long run frequency of sampling. The specific probability associated with the effect of treatment on liver TG is the long-run frequency of observing a *t*-value as big or bigger than the observed *t*-value (the one you actually got with your data) if the null is true. Let's parse this into "long run frequency of observing a *t*-value as big or bigger than the observed *t*-value" and "null is true".

A thought experiment: You open a google sheet and insert 12 standard, normal random deviates (so the true mean is zero and the true variance is one) in Column A, rows 1-12. You arbitrarily assign the first six values (rows 1-6) to treatment A and the second six values (rows 7-12) to treatment B. You use the space immediately below these data to compute the mean of treatment A, the mean of treatment B, the difference in means (A - B), and a *t*-value. Unfortunately, google sheets doesn't have a *t*-value function so you'd have to compute this yourself. Or not, since this is a thought experiment. Now "fill right" or copy and paste these functions into 999 new columns. You now have 1000 *t*-tests. The expected value of the difference in means is zero (why?) but the actual values will form a normal distribution about zero. Most will be close to zero (either in the negative or positive direction) but some will be further from zero. The expected *t*-value will also be zero (why?) and the distribution of these 1000 *t*-values will look normal but the tails are a little fuller. This row of *t*-values is a null distribution, because in generating the data we used the exact same formula for the values assigned to A and the values assigned to B. Now think of a *t*-value in your head, say 0.72 (remember that *t*-values will largely range from about -3 to +3 although the theoretical range is $-\infty$ to $+\infty$. What is the probability of observing a *t* of 0.72 *or bigger* if the null is true? Look at the row of *t*-values! Count the number of $t \ge 0.72$ and then divide by the total number of *t*-values in the row (1000) and you have a probability computed as a frequency. But remember the frequentist definition is the long run frequency, or the expected frequency at the limit (when you've generated not 1000 or even 1,000,000 but an infinite number of columns and *t*-values).

Some asides to the thought experiment: First, why "as big or bigger" and not just the probability of the value itself? The reason is that the probability of finding the exact *t* is 1/infinity, which doesn't do us much good. So instead we compute the probability of finding *t* as big, or bigger, than our observed *t*. Second, the *t*-test probability described above is a "one-tail probability". Because a difference can be both in the positive direction and the negative direction, we usually want to count all the $t \ge 0.72$ and the $t \le -0.72$ and then add these two counts to compute the frequency of *as extreme or more extreme* values. This is called a "two-tailed probability" because we find extremes at both tails of the distribution. Third, we don't really count $t \ge 0.72$ but take advantage of the beautiful mathematical properties of the theoretical *t* distribution, which allows us to compute the frequentist probability (expected long range frequency) given the *t*-value and the degrees of freedom using the *t*-distribution.

Now what do I mean with the phrase "null is true"? Most people equate "null is true" with "no difference in means" but the phrase entails much more than this. Effectively, the phrase is short for "all assumptions used to compute" the *p*-value (see the Sander Greenland quote at the start of this chapter). A *p*-value is based on modeling the real data with a theoretical sample in which all the values were randomly sampled from the same distribution and the assignment of the individual values to treatment was random. Random sampling from the same distribution has three important consequences. First, random assignment to treatment group means that the expected means of each group are the same, or put differently, the expected difference in means between the assigned groups is zero. Second, random assignment to treatment *also* means that the expected variances of the two groups are equal. And third, random sampling means that the values of each point are independent -- we cannot predict the value of one point knowing information about any other point. **Here is what is super important about this**: a low *p*-value is more likely if *any* one of these consequences is untrue about our data. A low *p*-value could arise from a difference in true means, or it could arise from a difference in true variances, or it could arise if the $Y$ values are not independent of each other. This is why we need certain assumptions to make a *p*-value meaningful for empirical data. By assuming independent error and homogenous (equal) variances in our two samples, a low *p*-value is evidence of unequal means.

Let's summarize: A pretty good definition of a *p*-value is: the long-run frequency of observing a test-statistic as large or larger than the observed statistic, if the null were true. A more succinct way to state this is

\begin{equation}
p = \mathrm{prob}(t \ge t_o | H_o)
\end{equation}

where *t* is a hypothetically sampled *t*-value from a null distribution, $t_o$ is the observed *t*-value, and $H_o$ is the null hypothesis. Part of the null hypothesis is the expected value of the parameter estimated is usually (but not always) zero -- this can be called the nil null. For example, if there is no ASK1 deletion effect on liver TG levels, then the expected difference between the means of the control and knockout mice is zero. Or, 

\begin{equation}
\mathrm{E}(\bar{Y}_{knockout} - \bar{Y}_{control} | H_o) = 0.0
\end{equation}

## Pump your intuition -- Creating a null distribution

The mean `liver_tg` in the knockout treatment is `r round(abs(fig_2i_m1_coef["treatmentASK1Δadipo","Estimate"]), 1)` µmol less than the mean `liver_tg` in the control treatment. This is the measured effect, or the **observed differences in means**. How confident are we in this effect? Certainly, if the researchers did the experiment with *two* control (ASK1F/F) treatment groups, they would measure some difference in their means simply because of finite sampling (more specifically, the many, many random effects that contribute to liver TG levels will differ between the two control groups). So let's reframe the question: are the observed differences unusually large compared to a distribution of differences that would occur if there were no effect? That is, if the ``null were true''. To answer this, we compare our observed difference to this **null distribution**. This comparison gives the probability (a long-run frequency) of "sampling" a random difference from the null distribution of differences that is as large, or larger, than the observed difference.

What is a null distribution? It is the distribution of a statistic (such as a difference in means, or better, a *t*-value) if the null were true. Here, I hope to pump your intuition by generating a null distribution that is relevant to the ASK1 liver TG data. See if you can understand the script before reading the explanation below.

```{r pvalue-d-null, message=FALSE, warning=FALSE, fig.cap="Null distribution for the difference in means of two samples from the same, inifinitely large population with a true mean and standard deviation equal to the observed mean and standard deviation of the ASK1 liver TG data."}
seed <- 1
n_iter <- 10^5 # number of iterations

mu <- mean(fig_2i[treatment == "ASK1F/F", liver_tg]) 
sigma <- sd(fig_2i[treatment == "ASK1F/F", liver_tg])

n <- nrow((fig_2i[treatment == "ASK1F/F",]))

sample1 <- matrix(rnorm(n*n_iter, mean=mu, sd=sigma), nrow=n) # 100,000 samples (each size n)
sample2 <- matrix(rnorm(n*n_iter, mean=mu, sd=sigma), nrow=n) # 100,000 samples

d_null <- apply(sample2, 2, mean) - apply(sample1, 2, mean)

qplot(d_null)
```

What have we done above? We've simulated an infinitely large population of mice that have a distribution of liver TG levels similar to that of the mice assigned to the control (ASK1F/F) group. The true mean ($\mu$) and standard deviation ($\sigma$) of the simulated TG level are equal to the observed mean and standard deviation of the TG levels of the control mice Then, the script:

1. randomly sample `r n` values from this population of simulated lifespans and assign to sample1. We sample `r n` values because that is the sample size of our control in the experiment.
2. randomly sample `r n` values from this population of simulated lifespans and assign to sample2.
3. compute the difference $\bar{Y}_{sample2} - \bar{Y}_{sample1}$.
4. repeat 1-3 100,000 times, each time saving the difference in means.
5. plot the distribution of the 100,000 differences using a histogram

The distribution of the differences is a null distribution. Notice that the mode of the null distribution is at zero, and the mean (`r round(mean(d_null), 5)`) is close to zero (if we had set $n$ to infinity, the mean would be precisely zero). *The expected difference between the means of two random samples from the same population is, of course, zero*. Don't gloss over this statement if that is not obvious. The tails extend out to a little more than +20 and -20. What this means is that it would be uncommon to randomly sample a value from this distribution of differences *as or more extreme* than our observed difference, -21.6. By "more extreme", I mean any value more negative than -21.6 or more postive than 21.6. So it would be uncommon to sample a value from this distribution whose *absolute value* is as or more extreme than 21.6. How uncommon would this be?

```{r pvalue-d-null-p}
diff_obs <- fig_2i_m1_coef["treatmentASK1Δadipo", "Estimate"]
null_diff_extreme <- which(abs(d_null) > abs(diff_obs))
n_extreme <- length(null_diff_extreme)
(p_d_null = n_extreme/n_iter)
```

In the 100,000 runs, only `r n_extreme` generated data with an absolute difference as large or larger than 21.6 (an "absolute difference" is the absoute value of the difference). The frequency of differences as large or larger than our observed difference is `r p_d_null`. This frequency is the *probability* of sampling a difference as or more extreme than the observed difference "under the null". It is a *p*-value, but it is not the *p*-value in the coefficient table. This is because the *p*-value in the coefficient table is computed from a distribution of *t*-values, not raw differences. This raises the question, what is a *t*-distribution, and a *t*-value, more generally?

## A null distribution of *t*-values -- the *t* distribution
A *t*-test is a test of differences between two values. These could be

1. the difference between the means of two samples (a "two-sample" *t*-test)
2. the difference between a mean of a sample and some pre-specified value (a "one-sample" *t*-test)
3. the difference between a coefficient from a linear model and zero

A *t*-test compares an observed *t*-value to a *t*-distribution. The null distribution introduced above was a distribution of mean differences under the null. A distribution of mean differences under the null is very specific to the mean and standard deviation of the population modeled and the sample size of the experiment. This isn't generally useful, since it will be unique to every study (at least it wasn't generally useful prior to the time of fast computers. One could, and some statisticians do, compute *p*-values using the algorithm above). A *t*-distribution is a way of transforming a null distribution of mean differences, which is unique to the study, into a distribution that is a function of sample size only. 

A *t*-distribution is a distribution of *t*-values under the null, where a *t*-value is a difference *standardized by its standard error*. For a two-sample *t*-test, this is

\begin{equation}
t = \frac{\bar{y}_2 - \bar{y}_1}{SE_{\bar{y}_2 - \bar{y}_1}}
(\#eq:t)
\end{equation}

The numerator is **the effect** while the denominator is the precision of the estimate. Like many test statistics, a *t*-value is a signal-to-noise ratio -- the effect is the signal and the SE of the difference is the noise.

A *t* distribution looks like a standard, normal distribution, except the tails are heavy, meaning there are more large-ish values than the normal. Like the standard normal distribution, large *t*-values are unlikely under the null and, therefore, a large *t* has a low probability -- or *p*-value -- under the null.

Looking at the equation for the two-sample *t*-test above, it is easy to see that three features of an experiment are associated with large *t* and small *p*-values: 1) big effect size (the numerator of the equation), 2) small sample standard deviations (which results in small standard errors of the difference, the denominator of equation \@ref(eq:t), and 3) large sample size (which results in small standard errors of the difference). As a quick-and-dirty generalization, absolute t-values greater than 3 are uncommon if the null is true.

The *p*-value for a *t*-test comes from comparing the observed *t* to a null *t* distribution and "counting" the values that are more extreme than the observed *t*. The *p*-value is the relative frequency of these more extreme values (relative to the total number of *t*-values in the distribution). I have "counting" in quotes because nothing is really counted -- there are an infinite number of *t*-values in the *t*-distribution. Instead, the t-distribution function is integrated to compute the fraction of the total area under the curve with *t*-values more extreme than the observed value. In a **two-tailed test**, this fraction includes both tails (positive *t*-values more positive than $|t|_{observed}$ and negative *t*-values more negative than $-|t|_{observed}$.

Let's modify the simulation of a null distribution of mean differences to generate a null distribution of *t*-values. I show the script, but don't just cut and paste the code. Spend time thinking about what the each line does. Explore it by copying parts and pasting into console.

```{r pvalue-t-null, echo=TRUE, message=FALSE, warning=FALSE, fig.cap="Null distribution of t-values. The simulation generated 10,000 t-tests with a true null."}

seed <- 1
n_iter <- 10^5 # number of iterations

mu <- mean(fig_2i[treatment == "ASK1F/F", liver_tg]) 
sigma <- sd(fig_2i[treatment == "ASK1F/F", liver_tg])

n <- nrow((fig_2i[treatment == "ASK1F/F",]))

sample1 <- matrix(rnorm(n*n_iter, mean=mu, sd=sigma), nrow=n) # 100,000 samples (each size n)
sample2 <- matrix(rnorm(n*n_iter, mean=mu, sd=sigma), nrow=n) # 100,000 samples

#way no. 1 - compute the t-tests manually
mean_diffs <- apply(sample2, 2, mean) - apply(sample1, 2, mean) # what is the apply function returning?
se_mean_diffs <- sqrt(apply(sample2, 2, sd)^2/n + apply(sample1, 2, sd)^2/n)
t_dis <- mean_diffs/se_mean_diffs

#way no.2 - compute the t-tests using the linear model
fake_data <- rbind(sample1, sample2)
treatment <- rep(c("control", "knockout"), each = n)
t_dis2 <- numeric(n_iter)
for(iter in 1:n_iter){
  y <- fake_data[, iter]
  fake_m1 <- lm(y ~ treatment)
  t_dis2[iter] <- coef(summary(fake_m1))["treatmentknockout", "t value"]
}


# plot the null distribution of t-values
qplot(t_dis2)

```

Now let's use this null distribution of *t*-values to compute a *p*-value

```{r pvalue-t-null-p}

# what is the p-value?
# the p-value is the number of t-values in t_dis that are as large
# or larger than the observed t. Large, negative t-values
# are as unlikely under the null as large, positive t-values.
# To account for this, we want to use absolute values in our counts
# this is a "two-tail test"

# first assign the observed t-value
t_obs <- fig_2i_m1_coef["treatmentASK1Δadipo", "t value"]


# now count the number of t-values in t_dis as big or bigger than this
# include the observed value as one of these (so add 1 to the count)
count <- sum(abs(t_dis) >= abs(t_obs))

# the p-value is the frequency of t_dis >= t_obs, so divide
# count by the total number of t-values in the distribution.
# Again add one since the observed value counts as a sample
(p_ASK1Δadipo <- count/(n_iter))

```

Hey that looks pretty good! Compare this to the *p*-value in the coefficient table above.

A *p*-value can be computed by counting the number of simulated *t*-values, *including the observed value*, that are equal to or more extreme (in either the positive or negative direction) than the observed *t*. Including the observed *t*, there are `r as.integer(count)` values that are more extreme than that observed. An approximate measure of *p* is this count divided by 100,001 (why is 1 added to the denominator?), which is `r p_ASK1Δadipo`. This simulation-based *p*-value is very (very!) close to that computed from the observed *t*-test.

## P-values from the perspective of permutation

A very intuitive way to think about *p*-values as a frequency is **random permutation**. A permutation is a re-arrangement of items. If there is an effect of ASK1 deletion on liver TG, then the arrangement of the values in the `treatment` column matters. If there is no effect of ASK1 deletion on liver TG, then the arrangement of the values in the `treatment` column does not matter.

Think about the structure of the liver TG data: there are two columns, `treatment`, which contains the assigned treatment, and `liver_tg`. The values in the `treatment` column were randomly assigned prior to the start of the experiment. If there is a negative effect of ASK1 deletion on liver TG, then assginment matters -- the values in the `liver_tg` column for the ASK1Δadipo rows will be smaller than, on average, the values in the ASK1F/F rows. That is, a specific value of a`liver_tg` is what it is because of the value of `treatement` in the same row. Assignment to ASK1F/F or ASK1Δadipo changes the expected value of `liver_tg`. But, if there were no true effect, then assignment to ASK1F/F or ASK1Δadipo does not change the expected value of `liver_tg`. The expected value of every cell in the `liver_tg` column would be the same regardless of what is in the `treatment` column.

In our thought experiment, let's leave the values in the `treatment` column be, and just randomly re-arrange or permute the values in the `liver_tg` column. What is the new expected diference in liver TG between the rows assigned to ASK1F/F and the rows assigned to ASK1Δadipo? The expected difference is Zero. Because the `liver_tg` values were randomly re-arranged, they *cannot* be caused by treatment assignment.

A permutation is a random re-arrangement of values in a column. Consider the many thousands of permutations of the values in the `liver_tg` column. A difference in means can be computed from each of these permuations and a distribution of differences can be generated. Is the observed difference extreme relative to the other values in this distribution? This is a permutation test -- it compares an observed statistic to a distribution of the statistic computed over many thousands of permutations.

Let's create a script for a permutation test

```{r pvalue-d-dist-perm, warning=FALSE, message=FALSE}

set.seed(1)
n_iter <- 5000 # number of random permutations

y <- fig_2i[, liver_tg]
x <- fig_2i[, treatment]

d_dist_perm <- numeric(n_iter)

for(iter in 1:n_iter){
  xbar1 <- mean(y[x == "ASK1F/F"])
  xbar2 <- mean(y[x == "ASK1Δadipo"])
  
  d_dist_perm[iter] <- xbar2 - xbar1
  
  # permute y
  y <- sample(y, replace=FALSE)
  # note that, when i=1, the first "permutation" is the original arrangement
}

qplot(d_dist_perm)
```

From this distribution of distances generated by random permuation of the response, we can compute a permutation *p*-value.

```{rpvalue-d-dist-perm-p }
(p_permute <- sum(abs(d_dist_perm) >= abs(d_dist_perm[1]))/n_iter)
```

## Parametric vs. non-parametric statistics

A statistic such as the difference in mean liver TG between ASK1Δadipo and ASK1F/F groups does not have "a" *p*-value. A *p*-value is the probability of observing an event *given a model of how the event was generated*. For the *p*-value in the coefficient table above, the event is sampling a *t*-value from a modeled *t* distribution that is as or more extreme than the observed *t*-value. The model generating the null distribution of *t*-values includes random sampling from a distribution that is defined by specific parameters (in this case, a mean and a variance), these parameters define the location and shape of the distribution of values that could be sampled. A *p*-value computed from a distribution that is defined by a set of parameters is a **parametric** *p*-value.

For the *p*-value computed using the permutation test, the event is the probability of of computing a difference of means *from a randomly permuted set of $Y$* as or more extreme than the observed difference of means. The distribution of differences from the permutated $Y$ data sets was not generated by any of the known distributions (normal, Poisson, binomial, etc.) given a specific value of parameters. Consequently, the permutation *p*-value is **non-parametric**.

The validity of all *p*-values depends on a set of model assumptions, which differ from model to model. The permutation *p*-value has fewer assumptions than a parametric *p*-value because no distribution is assumed (the permutation *p*-value is **distribution-free**).

## frequentist probability and the interpretation of p-values

### Background
There are at least three different meanings of **probability**.

1. **subjective probability** is the probability that an individual assigns to an event based on prior knowledge and the kinds of information considered reliable evidence. For example, if I asked a sample of students, what is the probability that a 30c homeopathic medicine could clear a *Streptococcus* infection from your respiratory system, their answers would differ because of variation in their knowledge of basic science, including chemistry and physics, their knowledge of what homeopathic medicines are, and how they weight different kinds of evidence.

2. **classical probability** is simply one divided by the number of possible unique events. For example, with a six-sided die, there are six possible unique events. The probability of rolling a 2 is $\frac{1}{6}$ and the probability of rolling an odd number is $\frac{1}{2}$.

3. **frequentist probability** is based on the concept of **long run frequency**. If I roll a die 10 times, the frequency of rolling a 2 will be approximately $\frac{1}{6}$. If I roll the die 100 times, the frequency of rolling a two will be closer to $\frac{1}{6}$. If I roll the die 1000 times, the frequency of rolling the die will be even closer to $\frac{1}{6}$. So the frequentist definition is the expected frequency given an infinite number of rolls. For events with continous outcomes, a frequentist probability is the long run frquency of *observing an outcome equal to or more extreme that that observed*.

### This book covers frequentist approaches to statistical modeling and when a probability arises, such as the *p*-value of a test statistic, this will be a frequentist probability.

When we do a *t*-test, we get a *p*-value. There are several ways to think about this probability. The most compact way is $P(data | null)$, which is literally read as the probability of the data given the null (or "conditional" on the null), but is really short for *the probability of the data, or something more extreme than the data, given that the null hypothesis is true*. The "probability of the data" is kinda vague. More specifically, we mean the probability of some statistic about the data such as the difference in means between group A and group B or the *t*-value associated with this difference. So, a bit more formally, the probability returned in a *t*-test is $\mathrm{prob}(t \ge t_{obs} | H_0)$. This is the long run frequency of observing a *t*-value as big or bigger than the observed *t*-value (the one you actually got with your data) if the null is true. Let's parse this into "long run frequency of observing a *t*-value as big or bigger than the observed *t*-value" and "null is true".

A thought experiment: You open a google sheet and insert 12 standard, normal random deviates (so the true mean is zero and the true variance is one) in Column A, rows 1-12. You arbitrarily assign the first six values (rows 1-6) to treatment A and the second six values (rows 7-12) to treatment B. You use the space immediately below these data to compute the mean of treatment A, the mean of treatment B, the difference in means (A - B), and a *t*-value. Unfortunately, google sheets doesn't have a *t*-value function so you'd have to compute this yourself. Or not, since this is a thought experiment. Now ``fill right'' or copy and paste these functions into 999 new columns. You now have 1000 *t* tests. The expected value of the difference in means is zero (why?) but the actual values will form a normal distribution about zero. Most will be close to zero (either in the negative or positive direction) but some will be further from zero. The expected *t*-value will also be zero (why?) and the distribution of these 1000 *t*-values will look normal but the tails are a little fuller. This row of *t*-values is a null distribution, because in generating the data we used the exact same formula for the values assigned to A and the values assigned to B. Now think of a *t*-value in your head, say 0.72 (remember that *t*-values will largely range from about -3 to +3 although the theoretical range is $-\infty$ to $+\infty$. What is the probability of observing a *t* of 0.72 *or bigger* if the null is true? Look at the row of *t*-values! Count the number of $t \ge 0.72$ and then divide by the total number of *t*-values in the row (1000) and you have a probability computed as a frequency. But remember the frequentist definition is the long run frequency, or the expected frequency at the limit (when you've generated not 1000 or even 1,000,000 but an infinite number of columns and *t*-values).

Some asides to the thought experiment: First, why "as big or bigger" and not just the probability of the value itself? The reason is that the probability of finding the exact *t* is 1/infinity, which doesn't do us much good. So instead we compute the probability of finding *t* as big, or bigger, than our observed *t*. Second, the *t*-test probability described above is a "one-tail probability". Because a difference can be both in the positive direction and the negative direction, we usually want to count all the $t \ge 0.72$ and the $t \le -0.72$ and then add these two counts to compute the frequency of *as extreme or more extreme* values. This is called a "two-tailed probability" because we find extremes at both tails of the distribution. Third, we don't really count $t \ge 0.72$ but take advantage of the beautiful mathematical properties of the theoretical *t* distribution, which allows us to compute the frequentist probability (expected long range frequency) given the *t*-value and the degrees of freedom using the *t*-distribution.

Now what do I mean with the phrase "null is true"? Most people equate "null is true" with ``no difference in means'' but the phrase entails much more than this. Effectively, the phrase means that the *p*-value is based on modeling the real data with a theoretical sample in which all the points were randomly sampled from the same distribution and that the assignment of the individual points to treatment was random. This model means the theoretical sample has three properties: First, random assignment to treatment after sampling from the same distribution means that the expected means are the same, or put differently, the expected difference in means between the assigned groups is zero. Second, random assignment to treatment after sampling from the same distribution *also* means that the expected variances of the two groups are equal. And third, random sampling means that the values of each point are independent -- we cannot predict the value of one point knowing information about any other point. **Here is what is super important about this**: if we get a really low *p*-value, any one of these consequences may be untrue about our data, for example it could be that the true means of the two treatment groups really are different, or it could mean it is the variances that differ between the two groups, or it could mean that the data (or technically, the errors) are not independent of each other. This is why we need certain assumptions to make a *p*-value meaningful for empirical data. By assuming independent error and homogenous (equal) variances in our two samples, a low *p*-value is evidence of unequal means.

### Two interpretations of the *p*-value

Since we want to be working scientists who want to use *p*-values as a tool, we need to know how to interpret (or use) the *p*-value to make reasonable inferences and how to avoid mis-interpreting the *p*-value and making unreasonable or even incorrect inferences. Ronald Fisher, the inventor of frequentist statistics, developed an interpretation of the *p*-value that is probably most useful for academic and applied research programs. Neyman and Pearson (Neyman-Pearson) gave the *p*-value a different interpretation, one that is probably most useful for industrial quality control. Today's biology researchers use an interpretation that is an odd hybrid of the two, which often leads to silly inference. Regardless, understanding the distinction between Fisher and Neyman-Pearson will inform how we write up our results in a manuscript. I'll describe these in the context of the two-sample *t*-test.

#### Fisher's interpretation

Fisher was working in the context of an agricultural experiments, the goal of which was to discover better agricultural practices. Does this new fertilizer work better than our old fertilizer? This is the context of much of modern biosciences and clinical medicine. Fisher thought of *p* as evidence against the null; the smaller the *p* the stronger the evidence that the two sampling distributions differ, which, in an experimental context, implies a treatment effect. If an experiment results in a large *p*-value, we can move on and test other fertilizers. If an experiment results in a small *p*-value, we want to pursue this new fertilizer more. Do more experiments! Fisher never thought of a single experiment as definitive. The decision to move on or pursue is only partly informed by the *p*-value and Fisher offered no rule about what *p*-value lies on the threshold of this decision. When pressed, Fisher might say that $p=0.05$ is a reasonable threshold. 

#### Neyman-Pearson interpretation

Neyman-Pearson thought of *p* as the necessary and sufficient information *to make a decision* between accepting the null (or at least not rejecting the null) or rejecting the null and accepting an alternative hypothesis. This decision balances two sorts of errors: Type I (false positives), which they called $\alpha$, and Type II (false negatives), which they called $\beta$. A false positive means the null was rejected but there really is no effect. A false negative means that the null was not rejected but there actually is an effect. $\alpha$ is set by the experimenter and is the long-term frequency (or "rate") of false positives **when the null is true** that the experimenters are willing to accept. This is easily understood in the context of manufacturing. I've just made a batch of beer that I now need to ship. I sample 10 cans and test the quality against a norm. If $p < \alpha$, we reject the null in favor of the alternative -- something may be wrong with the batch, it differs from the norm. We throw the beer away. If $p > \alpha$, we do not reject the null, nor the beer! We ship it.

After setting $\alpha$, the experimenter designs the experiment to achieve an acceptable rate of $\beta$. Since $\beta$ is the false negative rate then $1-\beta$ is the rate of not making a false negative error, that is, the rate of rejecting the null when there really is an effect. This is called the **power** of the experiment. An experiment with high power will have a low probability of a Type II error. An experiment with low power will have a high probability of a Type II error. Power is partly determined by sample size, the bigger the sample the smaller the *p*-value, all other things equal (think about why in the context of the formula for the *t*-value). Power is a function of error variance, both the natural variance and the component added because of measurement error (think about why in the context of the formula for the *t*-value). Power is also a function of $\alpha$. If we set a low $\alpha$ (say, $\alpha=0.01$), the test is conservative. We are more likely to fail to reject the null even if the null is false. A researcher can increase power by increasing sample size, using clever strategies to reduce measurement error, or increasing alpha.

An experimenter sets $\alpha$, computes the sample size needed to achieve a certain level of power ($1-\beta$), and then does the experiment. A thoughtful researcher will set $\alpha$ after considering and weighing the pros and cons of different levels of $\alpha$. If false positives have costly consequences (expense, time, deleterious side-effects), then set $\alpha$ to a low value, such as 0.01 or 0.001. For example, if an initial screen has identified a previously unknown candidate that potentially functions in the focal system of the researcher, then a researcher might decide to set a low $\alpha$ (0.001) in the initial tests of this candidate to avoid devoting time, personel, and expense to chasing a phantom (a false-positive candidate). If false positives have trivial consequences, then set $\alpha$ to a high value, such as 0.05, or 0.1, or even 0.2. For example, if the initial tests of a candidate in a functional system are cheap and fast to construct, then a researcher might choose to set a high $\alpha$ for the screen that identifies candidates. False positive candidates don't cost the lab much effort to identify them as false, but missing positive candidates because of a small $\alpha$ (which results in low power) at the screen stage costs the researcher the discovery of a potentially exciting coponent of the functional system.

In Fisher's interpretation, there is no $\alpha$, no $\beta$, no alternative hypothesis, and no sharp decision rule. Instead, in Fisher, *p* is a continuous measure of evidence against the null and its value is interpreted subjectively by an informed and knowledgeable expert using additional information to make decisions. Neyman-Pearson rejected Fisher's conception of *p* as evidence against the null arguing that a single experimental *p*-value is too noisy without embedding it into a more formal system of of decision making that maintains long-term type I error rates at $\alpha$, given a certain power. In Neyman-Pearson, *p* is compared to a threshold, $\alpha$ and this alone makes the decision. In Neyman-Pearson, *p* **is not treated as continuous information**. $p=0.00000001$ is no more evidence to use to reject the null than $p=0.049$.

### NHST

Most biology researchers today interpret *p* using a combination of Fisher and Neyman-Pearson concepts in what has become known as Null Hypothesis Significance Testing (NHST).

1. Nearly all papers in biology either explicitly state something like "P values < 0.05 were considered to be statistically significant" or implicitly use 0.05 as the "level of significance" ($\alpha$). Comparing a *p*-value to a pre-defined $\alpha$ is Neyman-Pearson.
2. Unlike Neyman-Pearson, there is no evidence that researchers are thoughtfully considering the level of $\alpha$ for each experiment. Instead, researchers mindlessly choose $\alpha=0.05$ because this is what everyone else uses.
3. Unlike Neyman-Pearson, but somewhat in the spirit of Fisher, researchers, journals, and textbooks, advocate trichotomizing a statistically significant *p* into "significance bins" -- three asterisks for $p < 0.001$, two asterisks for $0.001 < p < 0.01$, and one asterisk for $0.01 < p < 0.05$). This is not Neyman-Pearson. Again, Neyman-Pearson developed a system to control the long-run frequency of Type I error, which is controled by a strict use of $\alpha$. If an observed *p*-value is in the *** bin or the * bin is meaningless in a system using Neyman-Pearson. There is only "accept" ($p >= \alpha$) or "reject" ($p < \alpha$).
4. Many researchers report exact *p*-values when $p < 0.05$ but "n.s." (not significant) when $p > 0.05$. Reporting exact *p*-values is Fisher. Reporting n.s. is Neyman-Pearson.
5. Many researchers further polychomotomize the *p*-value space just above 0.05 by using language such as "marginally significant". If Neyman-Pearson and Fisher got together and spawned a love-child, this would be it.

## Some major misconceptions of the *p*-value

Setting the type I error rate $\alpha$ to 0.05 is so pervasive that I'm going to simply use "0.05" instead of "alpha" in discussing misconceptions.

### Misconception: *p* is the probability that the null is true *and* $1-p$ is probability that the alternative is true

Many researchers believe that if $p > 0.05$ then "there is no effect." A frequentist hypothesis test cannot show that an effect doesn't exist, only that the null has a low probablity of producing a test statistic as extreme or more extreme than the observed effect.

Many researchers believe that if $p < 0.05$ then "there is an effect." Again, a frequentist hypothesis test cannot show that an effect exists, only that the null has a low probablity of producing a test statistic as extreme or more extreme than the observed effect.

1. The statement "There is no effect of predators on feeding behavior" is not a valid conclusion of a frequentist hypothesis test.
2. The statement "We found no effect of predators on feeding behavior" is misleading because a frequentist hypothesis test can neither find an effect nor find no effect.

The two errors above are gross misconceptions that are pervasive in the biology literature. A more subtle issue is the belief that a low *p*-value shows that the researcher's explanatory hypothesis is correct. For example, researchers believe the result "the prey fish fed 14.2 (95% CI: 9.2, 19.2) minutes shorter in the presence of the predator fish" confirms their hypothesis that prey modulate feeding duration as a function of their ability to assess the risk of predation. Some alternative explanations:

1. The predator fish also competes with the prey fish for the prey fish's food and with less food the prey fish spends less time feeding because it gives up when food density drops below some amount.
2. The predator fish is introduced to the prey tank by hand and odorant molecules from the researcher's hands are detected by the prey and the prey reduces feeding duration because of these odorants.

Importantly, no single experiment confirms an explanatory hypothesis. Instead, alternative explanations require multiple experiments with different controls to "rigrously probe" the preferred hypothesis.

### Misconception: a *p*-value is repeatable
Many researchers believe that a *p*-value is a precise measure -- that if the experiment were replicated, a similar *p* would result. This belief requires at least two misconceptions. First, if the null were true, then *any* *p*-value is equally likely. $p=0.00137$ is just as likely as $p=0.492$. In other words, if the null were true, the *p*-value is not replicable at all! Second, the *p*-value is highly dependent on the sample, and can be highly variable among replications, but there is no true *p*-value, so there can be no estimate or standard error. Let's explore these. 

#### What is the distribution of *p*-values under the null?

I often ask students, "if the null were true, what is the most likely *p*-value?" or "if the null were true, what kind of *p*-values would we expect, that is what is the expected distribution". A common answer is $p=0.5$ is the most likely value and something like a normal curve, except the tails abruptly stop at 0 and 1, is the expected distribution.

```{r uniform, echo=FALSE}
set.seed(2)
beta <- 0.0
n <- 10
m <- 10^4
x <- matrix(rnorm(n*2*m), nrow=n*2)
x[(n+1):(n*2),] <- x[(n+1):(n*2),] + beta
pvalue <- numeric(m)
for(j in 1:m){
  pvalue[j] <- t.test(x[1:n, j], x[(n+1):(2*n), j], var.equal=TRUE)$p.value
}
qplot(pvalue, breaks=seq(0,1,by=0.05))
```

#### The incredible inconsistency of the *p*-value

```{r dance-setup, echo=FALSE}
set.seed(1)
power <- 0.4
n <- 10
m <- 15
beta <- power.t.test(n=n, sd=1, sig.level=0.05, power=power)$delta
x <- matrix(rnorm(n*2*m), nrow=n*2)
x[(n+1):(n*2), ] <- x[(n+1):(n*2), ] + beta
colnames(x) <- paste("Exp",1:m, sep=" ")
treatment <- rep(c("A", "B"), each=n)
cnames <- c("Estimate", "lower", "upper", "p")
res <- matrix(NA, nrow=m, ncol=length(cnames))
colnames(res) <- cnames
for(j in 1:m){
  tres <- t.test(x[,j] ~ treatment, var.equal=TRUE)
  res[j, "Estimate"] <- tres$estimate[1]-tres$estimate[2]
  res[j, "lower"] <- tres$conf.int[1]
  res[j, "upper"] <- tres$conf.int[2]
  res[j, "p"] <- round(tres$p.value, 3)
}
 exp <- c(paste("Exp 0",1:9, sep=""), paste("Exp", 10:m, sep=" "))
 res <- data.table(Exp=exp, res)
 p.lt.05 <- sum(res[,p] <=0.05)
# gg <- ggplot(res, aes(x=Exp, y=Estimate)) +
#   geom_point() +
#   geom_errorbar(aes(ymin=lower, ymax=upper)) +
#   annotate("text", x=1:m, y=res[, upper] + 0.25, label=res[, p]) +
#   coord_flip() +
#   NULL 
# gg

```

How replicable is the conclusion of an experiment if the *p*-value for a *t*-test is 0.03? If our conclusion is based on $p < 0.05$, then the conclusion is not very replicable. The simulation below shows the results of 15 replicates of an experiment with true power of 40%. There are five "significant" results (one less than expected) but several replicates have very high *p*-values.

```{r dance-fig, echo=FALSE, message=FALSE, warning=FALSE, fig.cap="Variability of *p*-values when the power is 0.4"}

gg <- ggplot(res, aes(x=Exp, y=p)) +
  geom_point() +
  annotate("text", x=1:m, y=res[, p] + 0.05, label=res[, p]) +
  coord_flip() +
  NULL
gg

```

### Misconception: 0.05 is the lifetime rate of false discoveries

An important and widespread misconception is that if a researcher consistently uses $\alpha=0.05$, then the frequency of incorrectly concluding an effect exists, or "discovering" an effect, over the lifetime of the researcher, will be 5%. This is incorrect. $\alpha$ is the rate of false positive if the null hypothesis is true, so our lifetime "false discovery" rate could only be 5% if everything we ever tested has no true effect! More generally, the **false discovery rate** is the frequency of false positives divided by the frequency of positives (the sum of false and true positives). This differs from the the Type I error rate, which is the frequency of false positives divided by the frequency of tests *in which the null is true*.

Imagine we test

1. 1000 null hypotheses over a lifetime
2. 60% are true nulls, this means there are 600 true nulls and 400 true effects
3. alpha is 5%. This means we expect to find $p \le 0.05$ 30 times ($0.05 \times 600$) when the null is true
4. power is 25%. This means we expect to find $p \le 0.05$ 100 times ($0.25 \times 400$) when the null is false
5. We have made $30 + 100=130$ "discoveries" (all experiments with $p \le 0.05$), but
6. 30 of the 130, or 23%, are "false discoveries". This is the false discovery rate.

Think about this. If the null is never true, you cannot have a false discovery--every $p \le 0.05$ is a true discovery (the false discovery rate is 0%). And if the null is always true, every $p < 0.05$ is a false discovery (the false discovery rate is 100%).

### Misconception: a low *p*-value indicates an important effect
Many researchers write results as if they believe that a small *p*-value means the effect is big or important. This may misconception may arise because of the ubiquitous use of "significant" to indicate a small p-value and "very" or "extremely" or "wicked" significant to indicate a really small p-value. Regardless, this is a misconception. A small p-value will usually result when there is high power (but can occur even if power is low) and power is a function of effect size, variability (the standard deviation), and sample size. A small *p* could result from a large effect size but can also result with a small effect size if the sample size is big enough.

This is easy to simulate (see script below). Let's model the effect of the genotype of a gene on height

```{r prediction-with-small-p, eval=TRUE}
set.seed(1)
rho <- 0.5
n <- 10^4
genotype <- c("+/+", "+/-", "-/-")
Sigma <- diag(2)
Sigma[1,2] <- Sigma[2,1] <- rho
X <- rmvnorm(n, mean=c(0,0), sigma=Sigma)
colnames(X) <- c("X1", "X2")
beta <- c(0.05, 0.05)
y <- X%*%beta + rnorm(n)
fit <- lm(y ~ X)
coefficients(summary(fit))
```


### Misconception: a low *p*-value indicates high model fit or high predictive capacity
On page 606, of Lock et al "Statistics: Unlocking the Power of Data", the authors state in item D "The p-value from the ANOVA table is 0.000 so the model as a whole is effective at predicting grade point averages." This is incorrect.  A p-value is not a measure of the predictive capability of a model because the p-value is a function of the signal, noise (unmodeled error), and *sample size* while predictive ability is a function of just the signal:noise ratio. If the signal:noise ratio is tiny, the predictive ability is small but the p-value can be tiny if the sample size is large. This is easy to simulate (see script below). The whole-model p-value is exceptionally small (0.00001002) but the relative predictive ability, measured by the $R^2$, is near zero (0.002).
 
```{r prediction-with-small-p-2, eval=TRUE}
set.seed(1)
rho <- 0.5
n <- 10^4
Sigma <- diag(2)
Sigma[1,2] <- Sigma[2,1] <- rho
X <- rmvnorm(n, mean=c(0,0), sigma=Sigma)
colnames(X) <- c("X1", "X2")
beta <- c(0.05, -0.05)
y <- X%*%beta + rnorm(n)
fit <- lm(y ~ X)
summary(fit)
```

## What the *p*-value does not mean

1. *p* is not the probability of the null being true. More formally, this probability is $Prob(null | data)$ but our *p*-value is $P(data | null)$. These are not the same. $P(null | data)$ is the probability of the null being true given the data. $P(data | null)$ is the probability of our data, or something more extreme than our data, conditional on a true null.
2. $1-p$ is not the probability of the alternative
3. *p* is not a measure of effect size.
4. *p* in one experiment is not the same level of evidence against the null as in another experiment
5. *p* is not a great indicator of which is more likely, H0 or H1.
6. If one treatment level has $p < 0.05$ and another treatment level has $p > 0.05$, this is not evidence that the treatment levels have different effects on the outcome.


## Recommendations

1. Simply report the exact *p*-value, along with a CI of the estimate.
* *P*-values are noisy, there is little reason to report more than two significant digits (report "$p = 0.011$" not "$p = 0.0108$") although some journals recommend more than two significant digits.
* For high *p*-values, report "$p = 0.23$" not "$p = n.s.$".
* For small *p*-values, there is little reason to report more than one significant digit (report "$p = 0.0002$" not "$p = 0.00018$").
* For really small p-values, there is little reason to report the exact *p*-value (report "$p < 0.0001$" and not "$p = 2.365E-11$"). Recognize that "really small" is entirely arbitrary. Rafael Irizarry suggested that *p*-values less than something like the probability of being killed by lightning strike should be reported as "$p < m$", where $m$ is the probability of being killed by lightning strike^[https://twitter.com/rafalab/status/1310610623898808320]. According Google University, this is 0.00000142 in one year or 0.00033 in one lifetime. This text will use $p < ls$" for *p*-values less than 0.0001 -- the lifetime probability of being killed by lightning strike in someone that spends too much time in doors analyzing data.
2. If $p < 0.05$ (or some other $\alpha$) do not report this as "significant" -- in fact, avoid the word "significant". In the english language, "significant" implies big or important. Small *p*-values can result even with trivially small effects if $n$ is big or sample variation is small. The phrase "ASK1 knockout had a significant effect on reducing liver TG ($p = 0.011$)" is
* potentially misleading, if we interpret "significant" to mean "having a large effect on the regulation of liver TG",
* wrong, if we interpret "significant" to mean "there is an ASK1 knockout effect".
A low *p*-value is evidence that the effect of ASK1 knockout is not zero, but I would wager that knocking out any gene expressed in white adipose cells will have *some* effect (however small) on liver TG.
3. If a decision needs to be made ("do we devote time, expense, and personel to pursue this further?"), then a *p*-value is a useful tool. If *p* is smaller than say 0.001, this is pretty good evidence that the data is not a fluke of sampling, **as long as we are justifiably confident in all the assumptions that went into computing this *p*-value**. A replicate experiment with a small *p*-value is better evidence. If *p* is closer to 0.01 or 0.05, this is only weak evidence of a fluke because of the sampling variability of *p*. A replicate experiment with a small *p*-value is *much* better evidence.

### Primary sources for recommendations

1. "". [Statistical tests, P values, confidence intervals, and power: a guide to misinterpretations](https://link.springer.com/article/10.1007/s10654-016-0149-3).
2. "Q: Why do so many colleges and grad schools teach p = 0.05? A: Because that's still what the scientific community and journal editors use. Q: Why do so many people still use p = 0.05? A: Because that's what they were taught in college or grad school." -- [ASA Statement on Statistical Significance and P-Values](https://amstat.tandfonline.com/doi/full/10.1080/00031305.2016.1154108)
3. "We then discuss our own proposal, which is to abandon statistical significance. We recommend dropping the NHST paradigm—and the p-value thresholds intrinsic to it—as the default statistical paradigm for research, publication, and discovery in the biomedical and social sciences." -- [Abandon Statistical Significance](https://www.tandfonline.com/doi/abs/10.1080/00031305.2018.1527253)
4. " We conclude, based on our review of the articles in this special issue and the broader literature, that it is time to stop using the term “statistically significant” entirely. Nor should variants such as “significantly different,” “p < 0.05,” and “nonsignificant” survive, whether expressed in words, by asterisks in a table, or in some other way." -- [Moving to a World Beyond “p < 0.05”](https://www.tandfonline.com/doi/full/10.1080/00031305.2019.1583913)
5. "We agree, and call for the entire concept of statistical significance to be abandoned."--[Scientists rise up against statistical significance](https://www.nature.com/articles/d41586-019-00857-9)

## Problems

Problem 1 -- simulate the distribution of *p* under the null. There are many ways to do this but a straightforard approach is to

1. Create a $2n \times m$ matrix of random normal deviates with mean 0 and sd 1
2. Do a *t*-test on each column, with the first $n$ values assigned to one group and the remaining $n$ values assigned to the second group. Save the *p*-value from each.
3. Plot a histogram of the *p*-values.
4. What is the distribution? What is the most likely value of *p*?

Problem 2 -- simulate power. Again, many ways to do this but following up on Problem 1.
1. Create a $2n \times m$ matrix of random normal deviates with mean 0 and sd 1
2. Add an effect to the first $n$ values of each column. Things to think about
    a. what is a good effect size to add? The effect/sd ratio, known as Cohen's d, is a relative (or standardized) measure of effect size. Cohen suggest 0.2, 0.5, and 0.8 as small, medium, and large standardized effects.
    b. should the same effect be added to each individual? Yes! It is the random component that captures the individual variation in the response.
3. Do a *t*-test on each column of the matrix, using the first $n$ values in group 1 and the remaining $n$ values in group 2. Save the p-values for each.
4. Compute the power, the relative frequency $p \le 0.05$.
5. Repeat with different values of $n$, effect size, and sd, but only vary one at a time. How does power vary with these three parameters?



<!--chapter:end:chapters/17-p-values.Rmd-->

# Errors in inference

```{r error-setup, echo=FALSE, message=FALSE}
# wrangling packages
library(here)
library(janitor)
library(readxl)
library(data.table)
library(stringr)

# analysis packages
library(emmeans)
library(mvtnorm)
library(lmPerm)
library(MASS)

# graphing packages
library(ggsci)
library(ggpubr)
library(ggforce)
library(cowplot)
library(lazyWeave) #pvalstring

here <- here::here
data_path <- "data"

source_path <- here("R", "ggplotsci.R")
source(source_path)
```

## Classical NHST concepts of wrong

As described in chapter (p-values), two types of error occur in classical Neyman-Pearson hypothesis testing, and in the NHST version that dominates modern practice. **Type I** error occurs when the null hypothesis is true but the *p*-value of the test is less than $\alpha$. This is a *false positive*, where a *positive* is a test that rejects the null. **Type II** error occurs when the null hypothesis is false but the *p*-value of the test is greater than $\alpha$. This is a *false negative*, where a *negative* is a test that accepts (or fails to reject) the null. **Power** is not an error but the frequency of true, positive tests (or the frequency of avoiding Type II error). $\alpha$ is not an error but the rate of Type I error that a researcher is willing to accept. Ideally, a researcher sets $\alpha$ based on an evaluation of the pros and cons of Type I and Type II error for the specific experiment. In practice, researchers follow the completely arbitary practice of setting $\alpha = 0.05$.

Why should a researcher care about $\alpha$ and power? Typically, most researchers don't give $\alpha$ much thought. And power is considered only in the context of calculating a sample size for an experiment for a grant proposal. But researchers should care about rates of Type I error and power because these (and similar concepts) can help guide decisions about which model to fit to a specific dataset. 

### Type I error

In classical Neyman-Pearson hypothesis testing, an important property of a hypothesis test is the **size of a test**, which may include an entire procedure that culminates in a hypothesis test. "Size" is a weird name for the probability of rejecting the null when the null is true. Size is not $\alpha$. $\alpha$ is the *nominal value* -- it's what a researcher wants. Size is the *actual value* -- it's what a researcher gets.

It would probably come as a surpise to most researches to learn that the size of some common tests used with data that look like the researcher's data is not 0.05. "used with data that look like the researcher's data" is important here -- a *t*-test doesn't have one size. With data that conform to the assumptions (independence, homogeneity, normality), the size of a *t*-test is $\alpha$. But with any violation, especially when the sample size differs between groups, the size of the *t*-test can move away from $\alpha$. A test that has a size that is less than $\alpha$ is "conservative" (fewer nulls are rejected than we think, so the *status quo* is more often maintained). A test that has a size that is greater than $\alpha$ is "liberal" (more nulls are rejected than we think, so the *status quo* is less often maintained). More conservative tests reduce power. More liberal tests *artificially* increase power and increase our rate of false rejection, which can mean "false discovery" if *p*-values are used as the arbiter of discovery.

#### Size example 1: the size of a *t*-test vs. a permutation test, when the data meet the assumptions

```{r, message=FALSE}
set.seed(1)
n <- 10
n_iter <- 10000
p_t <- numeric(n_iter)
p_perm <- numeric(n_iter)

treatment <- rep(c("cn", "tr"), each = n)
for(iter in 1:n_iter){
  sample_1 <- rnorm(n, mean = 10, sd = 1)
  sample_2 <- rnorm(n, mean = 10, sd = 1)
  y <- c(sample_1, sample_2)
  m1 <- lm(y ~ treatment) # no data statement necessary because both variables in workspace
  p_t[iter] <- coef(summary(m1))["treatmenttr", "Pr(>|t|)"]
  
  m2 <- lmp(y ~ treatment,
            perm = "Prob",
            settings = FALSE)
  p_perm[iter] <- coef(summary(m2))["treatment1", "Pr(Prob)"]
}
size_t <- sum(p_t < 0.05)/n_iter
size_perm <- sum(p_perm < 0.05)/n_iter
size_table <- data.table(Method = c("lm", "perm"),
                         Size = c(size_t, size_perm))
knitr::kable(size_table, digits = 4)
```

#### Size example 2: the size of a *t*-test vs. a permutation test, when the data have a right skewed distribution

```{r}
set.seed(1)
n <- 10
n_iter <- 10000
p_t <- numeric(n_iter)
p_perm <- numeric(n_iter)

treatment <- rep(c("cn", "tr"), each = n)
for(iter in 1:n_iter){
  #  qplot(rnegbin(n = 10^4, mu = 100, theta = 1))
  sample_1 <- rnegbin(n, mu = 100, theta = 1)
  sample_2 <- rnegbin(n, mu = 100, theta = 1)
  y <- c(sample_1, sample_2)
  # qplot(x=treatment, y = y)
  m1 <- lm(y ~ treatment) # no data statement necessary because both variables in workspace
  p_t[iter] <- coef(summary(m1))["treatmenttr", "Pr(>|t|)"]
  
  m2 <- lmp(y ~ treatment,
            perm = "Prob",
            settings = FALSE)
  p_perm[iter] <- coef(summary(m2))["treatment1", "Pr(Prob)"]
}
size_t <- sum(p_t < 0.05)/n_iter
size_perm <- sum(p_perm < 0.05)/n_iter
size_table <- data.table(Method = c("lm", "perm"),
                         Size = c(size_t, size_perm))
knitr::kable(size_table, digits = 4)
```

#### Size example 3: the size of a *t*-test vs. a permutation test, when the data have heterogenous variance and the sample size is unequal

```{r}
set.seed(1)
n1 <- 10
n2 <- n1/2
n_iter <- 10000
p_t <- numeric(n_iter)
p_perm <- numeric(n_iter)

treatment <- rep(c("cn", "tr"), times = c(n1, n2))
for(iter in 1:n_iter){
  #  qplot(rnegbin(n = 10^4, mu = 100, theta = 1))
  sample_1 <- rnorm(n1, mean = 10, sd = 0.5)
  sample_2 <- rnorm(n2, mean = 10, sd = 1)
  y <- c(sample_1, sample_2)
  # qplot(x=treatment, y = y)
  m1 <- lm(y ~ treatment) # no data statement necessary because both variables in workspace
  p_t[iter] <- coef(summary(m1))["treatmenttr", "Pr(>|t|)"]
  
  m2 <- lmp(y ~ treatment,
            perm = "Prob",
            settings = FALSE)
  p_perm[iter] <- coef(summary(m2))["treatment1", "Pr(Prob)"]
}
size_t <- sum(p_t < 0.05)/n_iter
size_perm <- sum(p_perm < 0.05)/n_iter
size_table <- data.table(Method = c("lm", "perm"),
                         Size = c(size_t, size_perm))
knitr::kable(size_table, digits = 4)
```

### Power

In classical Neyman-Pearson hypothesis testing, an important property of a hypothesis test is the **power of a test**. "Power" is the probability of rejecting the null when the null is false. A common way to think about power is, power is a test's ability to "detect" an effect if it exists. This makes sense using Neyman-Pearson but not Fisher (Using Fisher, a *p*-value is not a detector of an effect -- a reasoning brain is). Using Fisher, we could say that power is the sensitivity of a test (it takes less sample to provide the same signal).

#### Power example 1: the power of a *t*-test vs. a permutation test, when the data meet the assumptions

```{r, message=FALSE}
set.seed(1)
n <- 10
n_iter <- 10000
p_t <- numeric(n_iter)
p_perm <- numeric(n_iter)

treatment <- rep(c("cn", "tr"), each = n)
for(iter in 1:n_iter){
  sample_1 <- rnorm(n, mean = 10, sd = 1)
  sample_2 <- rnorm(n, mean = 11, sd = 1)
  y <- c(sample_1, sample_2)
  m1 <- lm(y ~ treatment) # no data statement necessary because both variables in workspace
  p_t[iter] <- coef(summary(m1))["treatmenttr", "Pr(>|t|)"]
  
  m2 <- lmp(y ~ treatment,
            perm = "Prob",
            settings = FALSE)
  p_perm[iter] <- coef(summary(m2))["treatment1", "Pr(Prob)"]
}
power_t <- sum(p_t < 0.05)/n_iter
power_perm <- sum(p_perm < 0.05)/n_iter
power_table_normal <- data.table(Method = c("lm", "perm"),
                         Power = c(power_t, power_perm))
knitr::kable(power_table_normal, digits = 3)
```

#### Power example 2: the power of a *t*-test vs. a permutation test, when the data look like typical count data

```{r}
set.seed(1)
n <- 10
n_iter <- 10000
p_t <- numeric(n_iter)
p_perm <- numeric(n_iter)

treatment <- rep(c("cn", "tr"), each = n)

for(iter in 1:n_iter){
  #  qplot(rnegbin(n = 10^4, mu = 100, theta = 1))
  sample_1 <- rnegbin(n, mu = 100, theta = 1)
  sample_2 <- rnegbin(n, mu = 300, theta = 1)
  y <- c(sample_1, sample_2)
  # qplot(x=treatment, y = y)
  m1 <- lm(y ~ treatment) # no data statement necessary because both variables in workspace
  p_t[iter] <- coef(summary(m1))["treatmenttr", "Pr(>|t|)"]
  
  m2 <- lmp(y ~ treatment,
            perm = "Prob",
            settings = FALSE)
  p_perm[iter] <- coef(summary(m2))["treatment1", "Pr(Prob)"]
}
power_t <- sum(p_t < 0.05)/n_iter
power_perm <- sum(p_perm < 0.05)/n_iter
power_table_count <- data.table(Method = c("lm", "perm"),
                         Power = c(power_t, power_perm))
knitr::kable(power_table_count, digits = 3)
```

## A non-Neyman-Pearson concept of power

Size and power are concepts specific to the Neyman-Pearson hypothesis testing framework. Size and power also have limited (or no) use in a research program in which the null hypothesis is never (or rarely) strictly true. That said, the concept of size and power are useful. For example, what if we framed power as the distribution of *p*-values instead of the frequency of *p*-values less than $\alpha$.

Table \@ref(tab:error-power-dis) shows the *p*-value at the 10th, 25th, 50th, 75th, and 90th percentile of the set of *p*-values computed in Power Example 2 above (count data). The *n*th percentile is the value in an ordered set of numbers in which *n* % are less than the value and 100 - *n*% are greater than the value. The 50th percentile is the median. The table shows that at all percentiles except the 90th, the permutation *p*-value is smaller than the *t*-test *p*-value. And, importantly, the value at 75% for both is ~ 0.12. This means that for experiments that generate data something like the fake data generated in Power Example 2, the permutation test is more sensistive to the incompatibility between the null model and the data than the *t*-test, except in the random samples when both methods fail.

```{r error-power-dis}
quantile_list <- c(0.1, 0.25, 0.5, 0.75, 0.9)
percentiles_t <- quantile(p_t, quantile_list)
percentiles_perm <- quantile(p_perm, quantile_list)

alt_power_table <- data.table(method = c("t-test", "permutation"),
                              (rbind(percentiles_t,
                                     percentiles_perm)))
knitr::kable(alt_power_table, digits = c(1, 4, 3, 3, 2, 2))
```

### Estimation error
### Coverage

This text advocates reporting a confidence interval with each reported effect size. An important property of an **estimator** is **coverage probability**, often shortened to "coverage".

### Type S error

Instead of framing the "size" concept as the rate of Type I error, what if we framed this as the rate that an estimate is in the correct direction (meaning, the sign of an effect is the same as the true value). And, 



### Type M error




<!--chapter:end:chapters/18-error.Rmd-->

# Part V: Introduction to Linear Models  {-}

<!--chapter:end:chapters/20-part-iii-linear-model.Rmd-->

# An introduction to linear models

```{r libraries, echo=FALSE, message=FALSE, warning=FALSE}

library(here)
library(janitor)
library(readxl)
library(data.table)
library(stringr)

# analysis packages
library(emmeans)
library(car) # qqplot, spreadlevel
library(DHARMa)

# graphing packages
library(ggsci)
library(ggpubr)
library(ggthemes)
library(ggforce)
library(cowplot)
library(lazyWeave) #pvalstring

here <- here::here
data_folder <- "data"

ggplotsci_path <- here("R", "ggplotsci.R")

# Okabe & Ito palette
pal_okabe_ito <- colorblind_pal()(8)[2:8] # ggthemes

source(ggplotsci_path)

```

All students are familiar with the idea of a linear model from learning the equation of a line, which is

\begin{equation}
Y = mX + b
(\#eq:line)
\end{equation}

where $m$ is the slope of the line and $b$ is the $Y$-intercept. It is useful to think of equation \@ref(eq:line) as a function that maps values of $X$ to values of $Y$. Using this function, if we input some value of $X$, we always get the same value of Y as the output.

```{r lineCode, echo=FALSE}
line2 <- function(x){
  return(m*x+b)
}
```

A linear model is a function, like that in equation \@ref(eq:line), that is fit to a set of data, often to model a process that generated the data or something like the data. The line in Figure \@ref(fig:line)A is just that, a line, but the line in Figure \@ref(fig:line)B is a linear model fit to the data in Figure \@ref(fig:line)B.

```{r line, echo=FALSE, message=FALSE, fig.cap="A line vs. a linear model. (A) the line $y=-3.48X + 105.7$ is drawn. (B) A linear model fit to the data. The model coefficients are numerically equal to the slope and intercept of the line in A."}
set.seed(1)
b0 <- 105.8
b1 <- -3.5
sigma <- 1.0
x <- rep(10:15, each=3)
n <- length(x)
y <- b0 + b1*x + rnorm(n)*sigma
b_vec <- coefficients(lm(y~x))
b <- b_vec[1]
m <- b_vec[2]

dt <- data.table(X=x, Y=y)
gg1 <- ggplot(data=dt, aes(x=X, y=Y)) +
  geom_smooth(method='lm', se=FALSE) +
  theme_minimal()

gg2 <- gg1 +
  geom_point()

gg <- plot_grid(gg1, gg2, labels = c("A", "B"))
gg

```

## Two specifications of a linear model

### The "error draw" specification

In introductory textbooks, a linear model is typically specified using an error-draw scheme.

\begin{align}
Y &= \beta_0 + \beta_1 X + \varepsilon\\
\varepsilon &\sim N(0, \sigma^2)
(\#eq:error-spec)
\end{align}

The first line of this specification has two components: the **linear predictor** $Y = \beta_0 + \beta_1 X$ and the **error**  $\varepsilon$. The linear predictor component looks like the equation for a line except that 1) $\beta_0$ is used for the intercept and $\beta_1$ for the slope and 2) the intercept term precedes the slope term. This re-labeling and re-arrangement make the notation for a linear model more flexible for more complicated linear models. For example $Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + \varepsilon$ is a model where $Y$ is a function of two $X$ variables.

The linear predictor is the deterministic or **systematic** part of the specification. As with the equation for a line, the linear predictor component of a linear model is a function that maps a specific value of $X$ to a *unique* value of $Y$. This mapped value is the **expected value**, or expectation, given a specific input value of $X$. The expectation is often written as $\mathrm{E}[Y|X]$, which is read as "the expected value of $Y$ given $X$", where "given X" means a specific value of X. This text will often use the word **conditional** in place of "given". For example, I would read $\mathrm{E}[Y|X]$ as "the expected value of $Y$ conditional on $X$". It is important to recognize that $\mathrm{E}[Y|X]$ is a **conditional mean** -- it is the mean value of $Y$ when we observe that $X$ has some specific value $x$ (that is $X = x$).

The second line of the specification \@ref(eq:error-spec) is read as "epsilon is distributed as Normal with mean zero and variance sigma squared". This line explicitly specifies the distribution of the error component of line 1. The error component of a linear model is a random "draw" from a normal distribution with mean zero and variance $\sigma^2$. The second line shows that the error component of the first line is **stochastic**. Using the error-model specification, we can think of any measurement of $Y$ as an expected value plus some random value sampled from a normal distribution with a specified variance. Because the stochastic part of this specification draws an "error" from a population, I refer to this as the **error-draw** specification of the linear model.

### The "conditional draw" specification

A second way of specifying a linear model is using a conditional-draw scheme.

\begin{align}
y_i &\sim N(\mu_i, \sigma^2)\\
\mathrm{E}(Y|X) &= \mu\\
\mu_i &= \beta_0 + \beta_1 x_i
(\#eq:draw-spec)
\end{align}

The first line states that the response variable $Y$ is a random variable independently drawn from a normal distribution with mean $\mu$ and variance $\sigma^2$. This first line is the **stochastic** part of the statistical model. The second line simply states that $\mu$ (the greek letter "mu") from the first line is the conditional mean (or expectation). The third line is the liner predictor, which states how $\mu_i$ is generated given that $X=x_i$. Again, the linear predictor is the **systematic** (or deterministic) part of the statistical model. It is systematic because the same value of $x_i$ will always generate the same $\mu_i$. With the conditional-draw specification, we can think of *a measurement* ($y_i$) as a random draw from the specified distribution. Because it is $Y$ and not some "error" that is drawn from the specified distribution, I refer to this as the **conditional-draw** specification of the linear model.

### Comparing the error-draw and conditional-draw ways of specifying the linear model

These two ways of specifying the model encourage slightly different ways of thinking about how the data (the response varible $Y$) were generated. The error-draw specification "generates" data by 1) constructing what $y_i$ "should be" given $x_i$ (this is the conditional expection), then 2) adding some error $e_i$ drawn from a normal distribution with mean zero and some specified variance. The conditional-draw specification "generates" data by 1) constructing what $y_i$ "should be" given $x_i$, then 2) drawing a random variable from some specified distribution *whose mean is this expectation*. This random draw is not "error" but the measured value $y_i$. For the error draw generation, we need only one hat of random numbers, but for the conditional draw generation, we need a hat for each value of $x_i$.

Here is a short script that generates data by implementing both the error-draw and conditional-draw specifications. See if you can follow the logic of the code and match it to the meaning of these two ways of specifying a linear model.

```{r intro-linear-models-fake-data}
n <- 5
b_0 <- 10.0
b_1 <- 1.2
sigma <- 0.4
x <- 1:n
y_expected <- b_0 + b_1*x

# error-draw. Note that the n draws are all from the same distribution
set.seed(1)
y_error_draw <- y_expected + rnorm(n, mean = 0, sd = sigma)

# conditional-draw. Note that the n draws are each from a different
# distribution because each has a different mean.
set.seed(1)
y_conditional_draw <- rnorm(n, mean = y_expected, sd = sigma)

data.table(X = x,
           "Y (error draw)" = y_error_draw,
           "Y (conditional draw)" = y_conditional_draw)
```

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
What's that code? `rnorm()` is a pseudorandom number generator that simulates random draws from a normal distribution with the specified mean and variance. The algorithm to generate the numbers is entirely deterministic -- the numbers are not truly random but are "pseudorandom". The list of numbers returned closely approximates a set of true, random numbers. The sequence of numbers returned is determined by the "seed", which can be set with the `set.seed()` function (R will use an internal seed if not set by the user).
</div> 

The error-draw specification is not very useful for thinking about data generation for data analyzed by generalized linear models, which are models that allow one to specify distribution families other than Normal (such as the binomial, Poisson, and Gamma families). In fact, thinking about a model as a predictor plus error can lead to the misconception that, in a generalized linear model, the error (or residuals from the fit) has a distribution from the non-Normal distribution modeled. This cannot be true because the distributions modeled using generalized linear models (other than the Normal) do not have negative values (some residuals must have negative values since the mean of the residuals is zero). Introductory biostatistics textbooks typically only introduce the error-draw specification because introductory textbooks recommend data transformation or non-parametric tests if the data are not approximately normal. This is unfortunate because generalized linear models are extremely useful for real biological data.

Although a linear model (or statistical model more generally) is a model of a data-generating process, linear models are not typically used to actually generate any data. Instead, when we use a linear model to understand something about a real dataset, we think of our data as one realization of a process that generates data like ours. A linear model is a model of that process. That said, it is incredibly useful to use linear models to create fake datasets for at least two reasons: to probe our understanding of statistical modeling generally and, more specifically, to check that a model actually creates data like that in the real dataset that we are analyzing.

### ANOVA notation of a linear model

Many textbooks treat ANOVA differently from regression and express a linear model as an ANOVA model (and generally do not use the phrase "linear model"). ANOVA models are all variations of 

\begin{equation}
y_{ij} &= \mu + \tau_{i} + \varepsilon_{ij}
(\#eq:anova-spec)
\end{equation}

Unlike the error and conditional draw specifications above, the ANOVA model doesn't have a linear predictor in the form of a regression equation (or the equation for a line) -- that is, there are neither $X$ variables nor coefficients ($\beta$). Instead, the ANOVA model is made up of a linear combination of means and deviations from means. In \@ref(eq:anova-spec), $\mu$ is the grand mean (the mean of the means of the groups), $\tau_i$ is the deviation of the mean of group $i$ from the grand mean (these are the effects), and $\varepsilon_{ij}$ is the deviation (or error) of individual $j$ from the mean of group $i$. Traditional ANOVA computes effects and the statistics for inference by computing means and deviations from means. Modern linear models compute effects and the statistics for inference by solving for the coefficients of the regression model. These are two ways of doing the same thing. That said, I don't find the ANOVA model very useful for understanding the biological consequences of an experiment. I want to know the effect of a treatment relative to one or more controls and not the deviation (effect) of a control and of a treatment relative to the grand mean.

## A linear model can be fit to data with continuous, discrete, or categorical $X$ variables

In the linear model fit to the data in Figure \@ref(fig:line)B, the $X$ variable is **continuous**, which can take any real number between the minimum $X$ and maximum $X$ in the data. For biological data, most variables that are continuous are positive, real numbers (a zero is not physically possible but could be recorded in the data if the true value is less than the minimum measurable amount). One exception is a **composition** (the fraction of a total), which can be zero. Negative values can occur with variables in which negative represent a direction (work, electrical potential) or a rate. **Discrete** variables are numeric but limited to certain real numbers. Most biological variables that are discrete are counts, and can be zero, but not negative. **Categorical variables** are non-numeric descriptions of a measure. Many of the categorical variables in this text will be the experimentally controlled treatment variable of interest (the variable $treatment$ containing the values "wild type" and "knockout") but some are measured covariates (the variable $sex$ containing the values "female" and "male").

### Fitting linear models to experimental data in which the $X$ variable is continuous or discrete

A linear model fit to data with a numeric (continous or discrete) $X$ is classical **regression** and the result is typically communicated by a regression line. The experiment introduced in Chapter \@ref(linear-models-with-a-single-continuous-x) [Linear models with a single, continuous *X*] is a good example. In this experiment, the researchers designed an experiment to measure the effect of warming on the timing of photosynthetic activity. Temperature was experimentally controlled at one of five settings (0, 2.25, 4.5, 6.75, or 9 °C above ambient temperature) within twelve, large enclosures. The response variable in the illustrated example is Autumn "green-down", which is the day of year (DOY) of the transition to loss of photosynthesis. The intercept and slope parameters of the regression line (Figure \@ref(fig:intro-linear-models-cont-slope)) are the coefficients of the linear model. The slope (4.98 days per 1 °C added warmth) estimates the **effect** of warming on green-down DOY. What is not often appreciated at the introductory biostatistics level is that **the slope is a difference in conditional means**. Any point on a regression line is the expected value of $Y$ at a specified value of $X$, that is, the conditional mean $\mathrm{E}(Y|X)$. **The slope is the difference in expected values for a pair of points that differ in $X$ by one unit**.

\begin{equation}
b_1 = \mathrm{E}(Y|X=x+1) - \mathrm{E}(Y|X=x+1)
\end{equation}

I show this in Figure \@ref(fig:intro-linear-models-cont-slope) using the points on the regression line at $x = 5$ and $x = 6$. Thinking about a regression coefficient as a difference in conditional means is especially useful for understanding the coefficients of a categorical $X$ variable, as described below.  

```{r intro-linear-models-cont-slope, echo=FALSE, warning = FALSE, message = FALSE, fig.cap="Illustration of the slope in a linear model with a numeric X. The slope (the coefficient of X) is the difference in expected value for any two X that are one unit apart. This is illustrated for the points on the line at x = 5 and x = 6."}

# image_path <- here("output", "fig_2c_explainer.Rds")
# fig_2c_explainer <- readRDS(image_path)
image_path <- here("output", "fig_2c_explainer.png")
fig_2c_explainer <- knitr::include_graphics(image_path)
fig_2c_explainer

```

### Fitting linear models to experimental data in which the $X$ variable is categorical {#slope}

Linear models can be fit to experimental data in which the $X$ variable is categorical -- this is the focus of this text! For the model fit to the data in Figure \@ref(fig:line)B, the coefficient of $X$ is the slope of the line. Perhaps surprisingly, 1) we can fit a model like equation \@ref(eq:error-spec) to data in which the $X$ variable is categorical and 2) the coefficient of $X$ *is* a slope. How is this possible? The slope of a line is $\frac{y_2 - y_1}{x_2 - x_1}$ where $(x_1, y_1)$ and $(x_2, y_2)$ are the graph coordinates of any two points on the line. What is the denominator of the slope function $(x_2 - x_1)$ when $X$ is categorical?

The solution to using a linear model with categorical $X$ is to recode the factor levels into numbers. An example of this was outlined in Chapter \@ref(analyzing-experimental-data-with-a-linear-model) ([Analyzing experimental data with a linear model]). The value of $X$ for individual mouse *i* is a number that indicates the treatment assignment -- a value of 0 is given to mice with a functional ASK1 gene and a value of 1 is given to mice with a knocked out gene. The regression line goes through the two group means (Figure \@ref(fig:intro-linear-models-cat-slope)). With the (0, 1) coding, $\overline{x}_{ASK1Δadipo} - \overline{x}_{ASK1F/F} = 1$, so the denominator of the slope is equal to one and the slope is simply equal to the numerator $\overline{y}_{ASK1Δadipo} - \overline{y}_{ASK1F/F}$. The coefficient (which is a slope!) *is* the difference in conditional means.

```{r intro-linear-models-cat-slope, echo=FALSE, warning = FALSE, message = FALSE, fig.cap="Illustration of the slope in a linear model with categorical X. The slope (the coefficient of X) is the difference in conditional means."}
set.seed(3)
image_path <- here("output", "fig_2i_explainer.Rds")
fig_2i_explainer <- readRDS(image_path)
fig_2i_explainer
```

## Statistical models are used for prediction, explanation, and description

Researchers typically use statistical models to understand relationships between one or more $Y$ variables and one or more $X$ variables. These relationships include

1. Descriptive modeling. Sometimes a researcher merely wants to describe the relationship between $Y$ and a set of $X$ variables, perhaps to discover patterns. For example, the arrival of a spring migrant bird ($Y$) as a function of sex ($X_1$) and age ($X_2$) might show that males and younger individuals arrive earlier. Importantly, if another $X$ variable is added to the model (or one dropped), the coefficients, and therefore, the precise description, will change. That is, the interpretation of a coefficient as a descriptor is *conditional* on the other covariates ($X$ variables) in the model. In a descriptive model, there is no implication of causal effects and the goal is not prediction. Nevertheless, it is very hard for humans to discuss a descriptive model without using causal language, which probably means that it is hard for us to think of these models as *mere description*. Like natural history, descriptive models are useful as patterns in want of an explanation, using more explicit causal models including experiments.

2. Predictive modeling. Predictive modeling is very common in applied research. For example, fisheries researchers might model the relationship between population density and habitat variables to predict which subset of ponds in a region are most suitable for brook trout (*Salvelinus fontinalis*) reintroduction. The goal is to build a model with minimal prediction error, which is the error between predicted and actual values for a future sample. In predictive modeling, the $X$ ("predictor") variables are largely instrumental -- how these are related to $Y$ is not a goal of the modeling, although sometimes an investigator may be interested in the relative importance among the $X$ for predicting $Y$ (for example, collecting the data may be time consuming, or expensive, or enviromentally destructive, so know which subset of $X$ are most important for predicting $Y$ is a useful strategy).

3. Explanatory (causal) modeling. Very often, researchers are explicitly interested in *how* the $X$ variables are causally related to $Y$. The fisheries researchers that want to reintroduce trout may want to develop and manage a set of ponds to maintain healthy trout populations. This active management requires intervention to change habitat traits in a direction, and with a magnitude, to cause the desired response. This model is predictive -- a specific change in $X$ predicts a specific response in $Y$ -- because the coefficients of the model provide knowledge on how the system functions -- how changes in the inputs *cause* change in the output. Causal interpretation of model coefficients requires a set of strong assumptions about the $X$ variables in the model. These assumptions are typically met in **experimental designs** but not **observational designs**.

With observational designs, biologists are often not very explicit about which of these is the goal of the modeling and use a combination of descriptive, predictive, and causal language to describe and discuss results. Many papers read as if the researchers intend explanatory inference but because of norms within the biology community, mask this intention with "predictive" language. Here, I advocate embracing explicit, explanatory modeling by being very transparent about the model's goal and assumptions.

## What do we call the $X$ and $Y$ variables?

The inputs to a linear model (the $X$ variables) have many names. In this text, the $X$ variables are typically

* **treatment variables** -- this term makes sense only for categorical variables and is often used for variables that are a **factor** containing the treatment assignment (for example "control" and "knockout")
* **factor variables** (or simply, **factors**) -- again, this term makes sense only for categorical variables
* **covariates** -- this term is usually used for the non-focal $X$ variables in a statistical model.

A linear model is a regression model and in regression modeling, the $X$ variables are typically called

* **independent variables** (often shortened to IV) -- "independent" in the sense that in a statistical model at least, the $X$ are not a function of $Y$. 
* **predictor variables** (or simply, "predictors") -- this makes the most sense in prediction models.
* **explanatory variables** -- this term is usually applied in observational designs and is best used if the explicit goal is causal modeling.

In this text, the output of a linear model (the $Y$ variable or variables if the model is multivariate) will most often be calle either of

* **response variable** (or simply, "response")
* **outcome variable** (or simply, "outcome")

These terms have a causal connotation in everyday english. These terms are often used in regression modeling with observational data, even if the model is not explicitly causal. On other term, common in introductory textbooks, is

* **dependent variable** -- "dependent" in the sense that in a statistical model at least, the $Y$ is a function of the $X$.

## Modeling strategy

A "best practice" sequence of steps used throughout this text to analyze experimental data is

1. **examine the data** using **exploratory plots** to
* examine individual points and identify outliers that are likely due to data transcription errors or measurement blunders
* examine **outlier** points that are biologically plausible, but raise ref flags about undue influe on fit models. This information is used to inform the researcher on the strategy to handle outliers in the statistical analysis, including algorithms for excluding data or implementation of **robust** methods.
* provide useful information for initial model filtering (narrowing the list of potential models that are relevant to the question and data). Statistical modeling includes a diverse array of models, yet almost all methods used by researchers in biology, and all models in this book, are generalizations of the linear model specified in \@ref(eq:draw-spec). For some experiments, there may be multiple models that are relevant to the question and data. Model checking (step 3) can help decide which model to ultimately use.
2. **fit the model**, in order to estimate the model parameters and the uncertainty in these estimates.
3. **check the model**, which means to use a series of diagnostic plots and computations of model output to check that the fit model reasonably approximates the data. If the diagnostic plots suggest a poor approximation, then choose a different model and go back to step 2. 
4. **inference from the model**, which means to use the fit parameters to learn, with uncertainty, about the system, or to predict future observations, with uncertainty.
5. **plot the model**, which means to plot the data, which may be adjusted, and the estimated parameters (or other results dervived from the estimates) with their uncertainty.

Note that step 1 (exploratory plots) is *not* data mining, or exploring the data for patterns to test.

## Predictions from the model

For the linear model specified in Model \@ref(eq:error-spec), the fit model is

\begin{equation}
y_i = b_0 + b_1 x_i + e_i
(\#eq:y-i)
\end{equation}

where $b_0$ and $b_1$ are the coefficients of the fit model and the $e_i$ are the residuals of the fit model. We can use the coefficients and residuals to recover the $y_i$, although this would rarely be done. More commonly, we could use the coefficients to calculate conditional means (the mean conditional on a specified value of $X$).

\begin{equation}
\hat{y}_i = b_0 + b_1 x_i
(\#eq:yhat)
\end{equation}

The conditional means are typically called **fitted values**, if the $X$ are the $X$ used to fit the model, or **predicted values**, if the $X$ are new. "Predicted values" is often shortened to "the prediction".

## Inference from the model

If our goal is inference, we want to use the fit parameters to learn, with uncertainty, about the system. Using equation \@ref(eq:y-i), the coefficients $b_0$ and $b_1$ are **point estimates** of the true, generating parameters $\beta_0$ and $\beta_1$, the $e_i$ are estimates of $\varepsilon_i$ (the true, biological "noise"), and $\frac{\sum{e_i^2}}{N-2}$ is an estimate of the true, population variance $\sigma^2$ (this will be covered more in chapter xxx but you may recognize that $\frac{\sum{e_i^2}}{N-2}$ is the formula for a variance). And, using equation \@ref(eq:yhat), $\hat{y}_i$ is the point estimate of the parameter $\mu_i$ (the true mean conditional on $X=x_i$). Throughout this text, Greek letters refer to a theoretical parameter and Roman letters refer to point estimates.

Our uncertainty in the estimates of the parameters due to sampling is the standard error of the estimate. It is routine to report standard errors of means and coefficients of the model. While a standard error of the estimate of $\sigma$ is available, this is effectively never reported, at least in the experimental biology literature, presumably because the variance is thought of as a nuisance parameter (noise) and not something worthy of study. This is a pity. Certainly treatments can effect the variance in addition to the mean.

**Parametric** inference assumes that the response is drawn from some probability distribution (Normal, or Poisson, or Bernouli, etc.). Throughout this text, I emphasize reporting and interpreting point estimates and **interval estimates** of the point estimate. A **confidence interval** is a type of interval estimate. A confidence interval of a parameter is a measure of the uncertainty in the estimate. A 95% confidence interval has a 95\% probability (in the sense of long-run frequency) of containing the parameter. This probability is a property of the population of intervals that could be computed using the same sampling and measuring procedure. It is not correct, without further assumptions, to state that there is a 95% probability that the parameter lies within the interval. Perhaps a more useful interpretation is that the interval is a **compatability interval** in that it contains the range of estimates that are compatible with the data, in the sense that a $t$-test would not reject the null hypothesis of a difference between the estimate and any value within the interval (this interpretation does not imply anything about the true value).

Another kind of inference is a **significance test**, which is the computation of the probability of "seeing the data" or something more extreme than the data, given a specified null hypothesis. This probability is the **p-value**, which can be reported with the point estimate and confidence interval. There are some reasonable arguments made by very influential statisticians that *p*-values are not useful and lead researchers into a quagmire of misconceptions that impede good science. Nevertheless, the current methodology in most fields of Biology have developed in a way to become completely dependent on *p*-values. I think at this point, a *p*-value can be a useful, if imperfect tool in inference, and will show how to compute *p*-values throughout this text.

Somewhat related to a significance test is a hypothesis test, or a **Null-Hypothesis Signficance Test** (NHST), in which the $p$-value from a significance test is compared to a pre-specified error rate called $\alpha$. Hypothesis testing was developed as a formal means of decision making but this is rarely the use of NHST in experimental biology. For almost all applications of *p*-values that I see in the literature that I read in ecology, evolution, physiology, and wet-bench biology, comparing a $p$-value to $\alpha$ adds no value to the communication of the results.

### Assumptions for inference with a statistical model

1. The data were generated by a process that is "linear in the parameters", which means that the different components of the model are added together. This additive part of the model containing the parameters is the linear predictor in specifications \@ref(eq:error-spec) and \@ref(eq:draw-spec) above. For example, a cubic polynomial model

\begin{equation}
\mathrm{E}(Y|X) = \beta_0 + \beta_1 X + \beta_2 X^2 + \beta_3 X^3
\end{equation}

is a linear model, even though the function is non-linear, because the different components are added. Because a linear predictor is additive, it can be compactly defined using matrix algebra

\begin{equation}
\mathrm{E}(Y|X) = \mathbf{X}\boldsymbol{\beta}
\end{equation}

where $mathbf{X}$ is the **model matrix** and $\boldsymbol{\beta}$ is the vector of parameters. We discuss these more in chapter xxx.

A **Generalized Linear Model** (GLM) has the form $g(\mu_i) = \eta_i$ where $\eta$ (the Greek letter "eta") is the linear predictor

\begin{equation}
\eta = \mathbf{X}\boldsymbol{\beta} 
\end{equation}

GLMs are extensions of linear models. There are non-linear models that are not linear in the parameters, that is, the predictor is not a simple dot product of the model matrix and a vector of parameters. For example, the Michaelis-Menten model is a non-linear model

\begin{equation}
\mathrm{E}(Y|X)  = \frac{\beta_1 X}{\beta_2 + X}
\end{equation}

that is non-linear in the parameters because the parts are not added together. This text covers linear models and generalized linear models, but not non-linear models that are also non-linear in the parameters.

2. The draws from the probability distribution are **independent**. Independence implies **uncorrelated** $Y$ conditional on the $X$, that is, for any two $Y$ with the same value of $X$, we cannot predict the value of one given the value of the other. For example, in the ASK1 data above, "uncorrelated" implies that we cannot predict the glucose level of one mouse within a specific treatment combination given the glucose level of another mouse in that combination. For linear models, this assumption is often stated as "independent errors" (the $\varepsilon$ in model \@ref(eq:error-spec)) instead of independent observations.

There are lots of reasons that conditional responses might be correlated. In the mouse example, correlation within treatment group could arise if subsets of mice in a treatment group are siblings or are housed in the same cage. More generally, if there are measures both within and among experimental units (field sites or humans or rats) then we'd expect the measures within the same unit to err from the model in the same direction. Multiple measures within experimental units (a site or individual) creates "clustered" observations. Lack of independence or clustered observations can be modeled using models with **random effects**. These models go by many names including linear mixed models (common in Ecology), hierarchical models, multilevel models, and random effects models. A linear mixed model is a variation of model \@ref(eq:error-spec). This text introduces linear mixed models in chapter xxx.

Measures that are taken from sites that are closer together or measures taken closer in time or measures from more closely related biological species will tend to have more similar values than measures taken from sites that are further apart or from times that are further apart or from species that are less closely related. Space and time and phylogeny create **spatial and temporal and phylogenetic autocorrelation**. Correlated error due to space or time or phylogeny can be modeled with **Generalized Least Squares** (GLS) models. A GLS model is a variation of model \@ref(eq:error-spec).

### Specific assumptions for inference with a linear model

1. **Constant variance** or **homoskedasticity**. The most common way of thinking about this is the error term $\varepsilon$ has constant variance, which is a short way of saying that random draws of $\varepsilon$ in model \@ref(eq:error-spec) are all from the same (or **identical**) distribution. This is explicitly stated in the second line of model specification \@ref(eq:error-spec). If we were to think about this using model specification \@ref(eq:draw-spec), then homoskedasticity means that $\sigma$ in $N(\mu, \sigma)$ is constant for all observations (or that the *conditional* probability distributions are identical, where *conditional* would mean adjusted for $\mu$)

Many biological processes generate data in which the error is a function of the mean. For example, measures of biological variables that grow, such as lengths of body parts or population size, have variances that "grow" with the mean. Or, measures of counts, such as the number of cells damaged by toxin, the number of eggs in a nest, or the number of mRNA transcripts per cell have variances that are a function of the mean. Heteroskedastic error can be modeled with **Generalized Least Squares**, a generalization of the linear model, and with **Generalized Linear Models** (GLM), which are "extensions" of the classical linear model.

2. Normal or **Gaussian** probability distribution. As above, the most common way of thinking about this is the error term $\varepsilon$ is Normal. Using model specification \@ref(eq:draw-spec), we'd say the conditional probablity distribution of the response is normal. A normal probability distribution implies that 1) the response is continuous and 2) the conditional probability is symmetric around $mu_i$. If the conditional probability distribution has a long left or right tail it is **skewed** left or right. Counts (number of cells, number of eggs, number of mRNA transcripts) and binary responses (sucessful escape or sucessful infestation of host) are not continuous and often often have asymmetric probablity distributions that are skewed to the right and while sometimes both can be reasonably modeled using a linear model they are more often modeled using generalized linear models, which, again, is an extension of the linear model in equation \@ref(eq:draw-spec). A classical linear model is a specific case of a GLM.

A common misconception is that inference from a linear model assumes that the raw response variable is normally distributed. Both the error-draw and conditional-draw specifications of a linear model show precisely why this conception is wrong. Model \@ref(eq:lm) states explicitly that it is the error that has the normal distribution -- the distribution of $Y$ is a mix of the distribution of $X$ and the error. Model \@ref(eq:draw-spec) states that the conditional outcome has a normal distribution, that is, the distribution after adjusting for variation in $X$.

## "linear model, "regression model", or "statistical model"?

Statistical modeling terminology can be confusing. The $X$ variables in a statistical model may be quantitative (continuous or integers) or categorical (names or qualitative amounts) or some mix of the two. Linear models with all quantitative independent variables are often called "regression models." Linear models with all categorical independent variables are often called "ANOVA models." Linear models with a mix of quantitative and categorical variables are often called "ANCOVA models" if the focus is on one of the categorical $X$ or "regression models" if there tend to be many independent variables.

This confusion partly results from the history of the development of regression for the analysis of observational data and ANOVA for the analysis of experimental data. The math underneath classical regression (without categorical variables) is the linear model. The math underneath classical ANOVA is the computation of sums of squared deviations from a group mean, or "sums of squares". The basic output from a regression is a table of coefficients with standard errors. The basic ouput from ANOVA is an ANOVA table, containing the sums of squares along with mean-squares, *F*-ratios, and *p*-values. Because of these historical differences in usage, underlying math, and output, many textbooks in biostatistics are organized around regression "vs." ANOVA, presenting regression as if it is "for" observational studies and ANOVA as if it is "for" experiments.

It has been recognized for many decades that experiments can be analyzed using the technique of classical regression if the categorical variables are coded as numbers (again, this will be explained later) and that both regression and ANOVA are variations of a more general, linear model. Despite this, the "regression vs. ANOVA" way-of-thinking dominates the teaching of biostatistics.

To avoid misconceptions that arise from thinking of statistical analysis as "regression vs. ANOVA", I will use the term "linear model" as the general, umbrella term to cover everything in this book. By linear model, I mean any model that is linear in the parameters, including classical regression models, marginal models, linear mixed models, and generalized linear models. To avoid repetition, I'll also use "statistical model".


<!--chapter:end:chapters/12-introduction-to-statistical-modeling.Rmd-->

# Linear models with a single, continuous *X* ("regression") {#regression}

```{r continuous-x-setup, echo=FALSE, message=FALSE}
# import and wrangling packages
library(here) # here() creates the absolute path to the file
library(janitor) # clean_names to clean col labels of imported data
library(readxl) # import excel
library(data.table) # make data.frames data.tables

# analysis packages
library(emmeans) # get estimated marginal means and CIs, used for plot
library(car) # qqplot
library(Metrics) # rmse

# plotting packages
library(ggplot2) # ggplot environment
library(ggpubr) # publication ready plots
library(ggpubfigs) # color palettes
# execute this command only the first time you use ggpubfigs
# devtools::install_github("JLSteenwyk/ggpubfigs")
library(cowplot) # combine plots
library(lazyWeave) # pretty pvalues

here <- here::here # make sure ``here` uses `here::here`

data_folder <- "data"

# Okabe & Ito palette
ito_seven <- friendly_pal("ito_seven") # ggpubfigs
pal_okabe_ito <- ito_seven[c(6,5,3,7,1,2,4)] # order of Wilke

# unikn::seecol(ito_seven)
# unikn::seecol(pal_okabe_ito)
```

1. observation warming on phenology
2. experiment warming on phenology
3. using regression to compare longitudinal (dietary methionine)

## A linear model with a single, continuous *X* is classical "regression"
### Analysis of "green-down" data
To introduce some principles of modeling with a single continuous $X$ variable, I'll use a dataset from 

[Richardson, A.D., Hufkens, K., Milliman, T. et al. Ecosystem warming extends vegetation activity but heightens vulnerability to cold temperatures. Nature 560, 368–371 (2018).](https://doi.org/10.1038/s41586-018-0399-1){target="_blank"}

[Source data](https://www.nature.com/articles/s41586-018-0399-1#Sec15){target="_blank"}

The data are from a [long-term experiment on the effects of warming and CO2 on a high-carbon northern temperate peatland](https://mnspruce.ornl.gov){target="_blank"} and is the focal dataset of the study. The experiment involves 10 large, temperature and CO2 controlled enclosures. CO2 is set to 400 ppm in five enclosures and 900 ppm in five enclosures. Temperature of the five enclosures within each CO2 level is set to 0, 2.25, 4.5, 6.75, or 9 °C above ambient temperature. The multiple temperature levels is a **regression design**, which allows a researcher to measure non-linear effects. Read more about the experimental design and the [beautiful implementation](https://mnspruce.ornl.gov/design){target="_blank"}.

The question pursued is in this study is, what is the causal effect of warming on the timing (or phenology) of the transition into photosynthetic activity ("green-up") in the spring and of the transition out of photosynthetic activity ("green-down") in the fall? The researchers measured these transition dates, or Day of Year (DOY), using foliage color. Here, we focus on the transition out of photosynthesis or "green-down" DOY.

**Import the data**

```{r continuous-x-import, echo=FALSE}
data_from <- "Ecosystem warming extends vegetation activity but heightens vulnerability to cold temperatures"
file_name <- "41586_2018_399_MOESM3_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)
fig2 <- read_excel(file_path) %>% # import
  clean_names() %>% # clean the column names
  data.table() # convert to data.table
# View(fig2)

fig2c <- fig2[panel == "2c",]
# View(fig2c)
```

1. **Examine the data**

```{r continuous-x-vet, message=FALSE}
gg1 <- qplot(x = temperature,
      y = transition_date,
      data = fig2c) +
  geom_smooth(method = "lm")
gg2 <- qplot(x = temperature,
      y = transition_date,
      data = fig2c) +
  geom_smooth(method = "lm", formula = y ~ poly(x, 2))
gg3 <- qplot(x = temperature,
      y = transition_date,
      data = fig2c) +
  geom_smooth()
plot_grid(gg1, gg2, gg3, ncol=3)
```

No plot shows any obvious outlier that might be due to measurement blunders or curation error. The linear regression in the left-most plot clearly shows that a linear response is sufficient to capture the effect of temperature on day of green-down.

2. **choose a model**. Because the $X$ variable ($temperature$) was experimentally set to five levels, the data could reasonably be modeled using either a linear model with a categorical $X$ or a linear model with a continuous $X$. The advantage of modeling $temperature$ as a continuous variable is that there is only one effect, the slope of the regression line. If modeled as a categorical factor with five levels, there are, at a minimum, four interesting effects (the difference in means between each non-reference level and reference (temperature = 0) level). Also, for inference, modeling $temperature$ as a continuous variable increases power for hypothesis tests.

3. **fit the model**

```{r continuous-x-fit}
# Step 1: fit the model
m1 <- lm(transition_date ~ temperature, data = fig2c)
```

4. **check the model**

```{r continuous-x-check-qq}
# check normality assumption
set.seed(1)
qqPlot(m1, id=FALSE)
```

The Q-Q plot indicates the distribution of residuals is well within that expected for a normal sample and there is no cause for concern with inference.

```{r continuous-x-check-spread}
# check homogeneity assumption
spreadLevelPlot(m1, id=FALSE)
```

The spread-location plot shows no conspicuous trend in how the spread changes with the conditonal mean. There is no cause for concern with inference.

5. **inference from the model**

```{r continuous-x-inference}
m1_coeff <- summary(m1) %>%
  coef()
m1_confint <- confint(m1)
m1_coeff <- cbind(m1_coeff, m1_confint)
m1_coeff
```
The effect of added temperature on the day of green-down is `r round(m1_coeff["temperature", "Estimate"],2)` d per 1 °C (95% CI: `r round(m1_coeff["temperature", "2.5 %"],1)`, `r round(m1_coeff["temperature", "97.5 %"],1)`; p `r pvalString(m1_coeff["temperature", "Pr(>|t|)"])`).

6. **plot the model**

```{r continuous-x-plot-response, echo=FALSE}

m1_b <- coef(m1)
m1_ci <- confint(m1)
m1_b0_text <- paste0("Intercept: ",
                  round(m1_b[1],1),
                  " (",
                  round(m1_ci[1,1],1),
                  ", ",
                  round(m1_ci[1,2],1),
                  ") d")
m1_b1_text <- paste0("Slope: ",
                  round(m1_b[2],2),
                  " (",
                  round(m1_ci[2,1],2),
                  ", ",
                  round(m1_ci[2,2],2),
                  ") d per 1 °C")

# regression line
m1_x <- min(fig2c[, temperature])
m1_xend <- max(fig2c[, temperature])
m1_y <- m1_b[1] + m1_b[2]*m1_x
m1_yend <- m1_b[1] + m1_b[2]*m1_xend

fig2c_gg_response <- ggplot(data = fig2c,
             aes(x = temperature,
                y = transition_date)) +
  
  # regression line first, to not block point
  geom_segment(x = m1_x,
               y = m1_y,
               xend = m1_xend,
               yend = m1_yend) +
  geom_point(size = 3,
             color = pal_okabe_ito[1]) +

  scale_x_continuous(breaks = c(0, 2.25, 4.5, 6.75, 9)) +
  xlab("Plot temperature (ΔT, °C)") +
  ylab("Autumn green-down (DOY)") +
  theme_pubr() +
  NULL
# fig2c_gg_response

```

```{r  continuous-x-plot-effect, echo=FALSE}
m1_coeff_dt <- data.table(term = row.names(m1_coeff),
                          data.table(m1_coeff))[2,] %>%
  clean_names()
m1_coeff_dt[ , p_pretty := pvalString(pr_t)]

min_bound <- min(m1_coeff_dt[, x2_5_percent])
max_bound <- min(m1_coeff_dt[, x97_5_percent])

y_lo <- min(min_bound+min_bound*0.2,
            -max_bound)
y_hi <- max(max_bound + max_bound*0.2,
            -min_bound)
y_lims <- c(y_lo, y_hi)

fig2c_gg_effect <- ggplot(data=m1_coeff_dt, 
                          aes(x = term,
                              y = estimate)) +
  # confidence level of effect
  geom_errorbar(aes(ymin=x2_5_percent, 
                    ymax=x97_5_percent),
                width=0, 
                color="black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # zero effect
  geom_hline(yintercept=0, linetype = 2) +
  
  # p-value
  annotate(geom = "text",
           label = m1_coeff_dt$p_pretty,
           x = 1,
           y = 7.5) +
  
  # aesthetics
  scale_y_continuous(position="right") +
  scale_x_discrete(labels = "Temperature\neffect") +
  ylab("Effects (day/°C)") +
  coord_flip(ylim = y_lims) + 
  theme_pubr() +
  theme(axis.title.y = element_blank()) +
  NULL

#fig2c_gg_effect
```

```{r continuous-x-plot, echo=FALSE, warning=FALSE, fig.cap="Modification of the published Figure 2c showing the experimental effect of warming on the date of autumn green-down (the transition to fall foliage color) in a mixed shrub community. The bottom panel is a scatterplot. The regression line shows the expected value of Y (transition day of year) given a value of X (added temperature). The slope of the regression line is the estimate of the effect. The estimate and 95% confidence interval of the estimate are given in the top panel."}
fig3d_fig <- plot_grid(fig2c_gg_effect,
                       fig2c_gg_response,
                       nrow=2,
                       align = "v",
                       axis = "lr",
                       rel_heights = c(0.4,1))
fig3d_fig
```

```{r continuous-x-save, echo=FALSE, warning=FALSE, message=FALSE}
save_plot <- TRUE
if(save_plot == TRUE){
  b_0 <- coef(m1)[1]
  b_1 <- coef(m1)[2]
  fig_2c_explainer <- fig2c_gg_response +
    # add slope text
    geom_text(aes(label = m1_b1_text,
                  x = 0.5,
                  y = 335),
              hjust = "left") +
    
    # vertical line
    geom_segment(aes(x = 6,
                 y = b_0 + b_1*5,
                 xend = 6,
                 yend = b_0 + b_1*6),
               linetype = "dashed",
               color = "black") +
    annotate(geom = "text",
             label = bquote(bar(y)[2] - bar(y)[1]),
             x = 6.5,
             y = b_0 + b_1*5.5) +

        # horizontal line
    geom_segment(aes(x = 5,
                     y = b_0 + b_1*5,
                     xend = 6,
                     yend = b_0 + b_1*5),
                 linetype = "dashed",
                 color = "black") +
    annotate(geom = "text",
             label = "6 - 5 = 1",
             x = 5.5,
             y = b_0 + b_1*5 - 1.5) +
    
  NULL
  fig_2c_explainer
  
  image_path <- here("output", "fig_2c_explainer.Rds")
  saveRDS(fig_2c_explainer, image_path)
  image_path <- here("output", "fig_2c_explainer.png")
  ggsave(image_path)
}
```
7. Report the results

The modeled effect of added temperature is `r m1_b1_text` (\@ref(fig:continuous-x-plot)).

### Learning from the green-down example

Figure \@ref(fig:continuous-x-plot) is a **scatterplot** with the green-down DOY for the mixed-shrub community on the $Y$ axis and added temperature on the $X$ axis. The line through the data is a regression line, which is the expected value of *Y* (green-down DOY) given a specific value of *X* (added temperature). The slope of the line is the **effect** of added temperature on timing of green-down. The intercept of the regression line is the value of the response (day of green-down) when $X$ is equal to zero. Very often, this value is not of interest although the value should be reported to allow predictions from the model. Also very often, the value of the intercept is not meaningful because a value of $X = 0$ is far outside the range of measured $X$, or the value is absurd because it is impossible (for example, if investigating the effect of body weight on metabolic rate, the value $weight = 0$ is impossible).

The intercept and slope are the coefficients of the model fit to the data, which is

\begin{equation}
day_i = b_0 + b_1 temperature_i + e_i
(\#eq:continuous-x-fit)
\end{equation}

where *day* is the day of green-down, *temperature* is the added temperature, and *i* refers (or "indexes") the *i*th enclosure. This model completely reconstructs the day of green-down for all ten enclosures. For example, the day of green-down for enclosure 8 is

\begin{equation}
332 = 289.458750 + 4.982745 \times 6.73 + 9.00737
\end{equation}

The coefficients in the model are estimates of the parameters of the **generating model** fit to the data

\begin{align}
day &= \beta_0 + \beta_1 temperature + \varepsilon\\
\varepsilon &\sim N(0, \sigma^2)
(\#eq:continuous-x-m1)
\end{align}

A generating model of the data is used to make inference, for example, a measure of uncertainty in a prediction in the timing of green-down with future warming, or a measure of uncertainty about the effect of temperature on green-down.

### Using a regression model for "explanation" -- causal models

In this text, "explanatory" means "causal" and a goal of explanatory modeling is the estimate of causal effects using a **causal interpretation** of the linear model (=regression) coefficients. "What what? I learned in my stats 101 course that we cannot interpret regression coefficients causally".

Statisticians (and statistics textbooks) have been quite rigid that a regression coefficient has a descriptive (or "observational", see below) interpretation but not a causal interpretation. At the same time, statisticians (and statistics textbooks) do not seem to have any issue with interpeting the modeled effects from an experiment causally, since this is how they are interpreted. But if the modeled effects are simply coefficients from a linear (= regression) model, then this historical practice is muddled.

Part of this muddled history arises from the use of "regression" for models fit to observational data with one or more continuous $X$ variables and the use of "ANOVA" for models fit to experimental data with one or more categorical $X$. This separation seems to have blinded statisticians from working on the formal probabilistic statements underlying causal interpretations of effect estimates in ANOVA and the synthesis of these statements with the probabilistic statements in regression modeling. Two major approaches to developing formal, probabilistic statements of causal modeling in statistics are the **Rubin causal model** and the **do-operator** of Pearl. Despite the gigantic progress in these approaches, little to none of this has found its way into biostatistics textbooks. 

#### What a regression coefficient means

A linear ("regression") model coefficient, such as the coefficient for temperature, $\beta_1$, has two interpretations, an **observational** interpretation and a **causal interpretation**. To understand these interpretations, it's useful to remember that a predicted value from a regression model is a **conditional mean**

\begin{equation}
\textrm{E}[day|temperature] = \beta_0 + \beta_1 temperature
(\#eq:continuous-x-conditional-mean)
\end{equation}

In words "the expected value of day conditional on temperature is beta-knot plus beta-one times temperature". An **expected value** is a long run average -- if we had an infinite number of enclosures with $temperature=x$ (where $x$ is a specific value of added temperature, say 2.5 °C), the average $day$ of these enclosures is $\beta_0 + \beta_1 x$.

The parameter $\beta_1$ is a *difference* in conditional means.

\begin{equation}
\beta_1 = \textrm{E}[day|temperature = x+1] - \textrm{E}[day|temperature = x]
(\#eq:beta)
\end{equation}

In words, "beta-one is the expected value of day of green-down when the temperature equals x + 1 minus the expected value of day of green-down when the temperature equals x." A very short way to state this is "beta-one is a difference in conditional means".

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
tl;dr. Note that the "+ 1" in this definition is mere convenience. Since the slope of a line is $\frac{y_2 - y_1}{x_2 - x_1}$, where ($x_1$, $y_1$) and ($x_2$, $y_2$) are the coordinates of any two points on the line, it is convenient to choose two points that differ in $X$ by one unit, which makes the fraction equal to the numerator only. The numerator is a difference in conditional means. It is also why the units of a regression coefficient are "per unit of $X$ even if defined as a difference in two $Y$ values.
</div> 

The difference between observational and causal interpretations of $\beta_1$ depends on the "event" conditioned on in $\textrm{E}[day|temperature]$. Let's start with the causal interpretation, which is how we should think about the regression coefficients in the green-down experiment.

#### Causal interpretation -- conditional on "doing" X = x

In the causal interpretation of regression, $\textrm{E}[day|temperature]$ is conditioned on "doing" a real or hypothetical intervention in the system where we set the value of $temperature$ to a specific value $x$ ("$temperature=x$), while keeping everything else about the system the same. This can be stated explicitly as $\textrm{E}[day|\;do\;temperature = x]$. Using the do-operator, we can interpret $\beta_1$ as an **effect coefficient**.

\begin{equation}
\beta_1 = \textrm{E}[day|\;do\;temperature = x+1] - \textrm{E}[day|\;do\;temperature = x]
\end{equation}

In words, "beta-one is the expected value of day of green-down were we to set the temperature to x + 1, minus the expected value of day of green-down were we to set the temperature to x."

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
tl;dr. In the green-down experiment, the researchers didn't set the temperature in the intervened enclosures to the ambient temperature + 1 but to ambient + 2.25, ambient + 4.5, ambient + 6.75, and ambient + 9.0. Again (see tl;dr above), the + 1 is mere convenience in the definition.
</div> 

#### Observational interpretation -- conditional on "seeing" X = x.

In the observational interpretation of regression, $\textrm{E}[day|temperature]$ is conditioned on sampling data and "seeing" a value of $temperature$. We can state this explicitly as $\textrm{E}[day|\;see\;temperature]$. From this, we can interpret $\beta_1$ as an **observational coefficient**

\begin{equation}
\beta_1 = \textrm{E}[day|\;see\;temperature = x+1] - \textrm{E}[day|\;see \;temperature = x]
\end{equation}

In words, "beta-one is the expected value of day of green-down when see that temperature equals x + 1 minus the expected value of day of green-down when we see that temperature equals x." To understand what I mean by "observational", let's imagine that the green-down data do not come from an experiment in which the researchers intervened and set the added temperature to a specifc value but from ten sites that naturally vary in mean annual temperature. Or from a single site with 10 years of data, with some years warmer and some years colder. Data from this kind of study is **observational** -- the researcher didn't intervene and set the $X$ values but merely observed the $X$ values.

If we sample (or "see") a site with a mean annual temperature that is 2.5 °C above a reference value, the expected day of green-down is $\textrm{E}[day|temperature = 2.5 °C]$. That is, values near $\textrm{E}[day|temperature = 2.5 °C]$ are more probable than values away from $\textrm{E}[day|temperature = 2.5 °C]$. Or, if the only information that we have about this site is a mean annual temperature that is 2.5 °C above some reference, then our best prediction of the day of green-down would be $\textrm{E}[day|temperature = 2.5 °C]$. We do not claim that the 4.98 day delay in green-down is caused by the warmer temperature, only that this is the expected delay relative to the reference having seen the data.

The seeing interpretation of a regression coefficient is **descriptive**-- it is a description of a mathematical relationship. In this interpretation, the coefficient is not causal in the sense of what the expected **response** in $Y$ would be were we to intervene in the system and change $X$ from $x$ to $x+1$.

#### Omitted variable bias

What is the consequence of interpreting a regression coefficient causally instead of observationally?

```{r continuous-x-dag, echo=FALSE, fig.cap = "a Directed Acyclic (or causal) Graph of a hypothetical world where the day of green-down is caused by two, correlated environmental variables, temperature and moisture, and to a noise factor (U) that represents an unspecified set of additional variables that are not correlated to either temperature or moisture."}
dag_table <- data.table(node = c("X", "M", "Y", "U"),
                        x = c(1, 2, 2, 2),
                        y = c(1, 2, 1, 2))
text_size <- 8
gg <- ggplot(data = dag_table,
             aes(x = x, y = y)) +
  geom_point(color = "white") +
  coord_cartesian(xlim = c(0, 3),
                  ylim = c(0, 3)) +
  geom_text(aes(label = "temperature",
                x = 1,
                y = 1,
                hjust = "right"),
            size = text_size) +
  geom_text(aes(label = "moisture",
                x = 1.5,
                y = 2),
            size = text_size) +
  geom_text(aes(label = "green-down DOY",
                x = 2,
                y = 1,
                hjust = "left"),
            size = text_size) +
  geom_text(aes(label = "U",
                x = 2.5,
                y = 2,
                hjust = "center"),
            size = text_size) +
  geom_segment(aes(x = 1 + .05,
                   y = 1,
                   xend = 2 - .05,
                   yend = 1),
               arrow = arrow(length = unit(0.3,"cm"))) +
  geom_segment(aes(x = 1.5 +.05,
                   y = 2 - .15,
                   xend = 2 - .0,
                   yend = 1 + .1),
               arrow = arrow(length = unit(0.3,"cm"))) +
  geom_curve(aes(x = 0.5,
                 y = 1 + .1,
                 xend = 1.5 -.3,
                 yend = 2 - 0),
             arrow = arrow(ends = "both",
                           length = unit(0.3,"cm")),
             curvature = -0.3) +
  geom_segment(aes(x = 2.5,
                   y = 1.85,
                   xend = 2.5,
                   yend = 1.15),
               arrow = arrow(length = unit(0.3,"cm"))) +
  annotate(geom = "text",
           label = "rho",
           parse = TRUE,
           x = 0.8,
           y = 1.85,
           size = text_size) +
  annotate(geom = "text", # temp-doy
           label = "beta[1]",
           parse = TRUE,
           x = 1.5,
           y = 0.85,
           size = text_size) +
  annotate(geom = "text", # moisture-doy
           label = "beta[2]",
           parse = TRUE,
           x = 1.8,
           y = 1.65,
           size = text_size) +
   theme_void() +
  NULL
gg
```

Let's expand the thought experiment where we have an observational data set of green down dates. In this thought experiment, only two variables systematically affect green-down DOY. The first is the temperature that the plants experience; the effect of $temperature$ is $\beta_1$. The second is the soil moisture that the plants experience; the effect of $moisture$ is $\beta_2$. $temperature$ and $moisture$ are correlated with a value $\rho$. This causal model is graphically represented by the **causal graph** above.

Lets fit two linear models.

\begin{align}
(\mathcal{M}_1)\; day &= b_0 + b_1 \; temperature + b_2 \; moisture + \varepsilon\\
(\mathcal{M}_2)\; day &= b_0 + b_1 \; temperature + \varepsilon
\end{align}

$\mathcal{M}_1$ the linear model including both $temperature$ and $moisture$ as $X$ variables and $\mathcal{M}_2$ the linear model including only $temperature$ as an $X$ variable. Given the true causal model above, $b_1$ is an **unbiased** estimate of the true causal effect of temperature $\beta_1$ in $\mathcal{M}_1$ because the expected value of $b_1$ is $\beta_1$. By contrast, $b_1$ is a **biased** estimate of the true causal effect of temperature $\beta_1$ in $\mathcal{M}_2$. The expected value of $b_1$ in $\mathcal{M}_2$ is the true, causal effect plus a bias term. 

\begin{equation}
\mathrm{E}(b_1|\mathcal{M}_1) = \beta + \rho \alpha \frac{\sigma_{moisture}}{\sigma_{temperature}}
\end{equation}

This bias ($\rho \alpha \frac{\sigma_{moisture}}{\sigma_{temperature}}$) is **omitted variable bias**. The omitted variable $moisture$ is an unmeasured, **confounding variable**. A variable $X_2$ is a confounder for variable $X_1$ if $X_2$ has both a correlation with $X_1$ and a path to the response $Y$ that is not through $X_1$. With ommitted variable bias, as we sample more and more data, our estimate of the effect doesn't converge on the truth but the truth plus some bias.

```{r echo = FALSE, eval = FALSE}
# b = beta + r*alpha
s_day <- sd(fig2[, transition_date]) # sigma_Y
s_temp <- sd(fig2[, temperature]) # sigma_X
s_moisture <- s_temp/3 # sigma_U
beta_2 <- -6.3 # effect of moisture
b1 <- m1_b[2]
b1_prime <- b1*s_temp/s_day
  
rho <- -0.7
sigma_u <- s_moisture
sigma_x <- s_temp
sigma_y <- s_day
sigma_z <- 1
zeta_u_prime <- sign(rho)*sqrt(abs(rho))
zeta_x_prime <- sqrt(abs(rho))
zeta_u <- zeta_u_prime*sigma_u/sigma_z
zeta_x <- zeta_x_prime*sigma_x/sigma_z

alpha <- beta_2
alpha_prime <- alpha*sigma_u/sigma_y

beta_prime <- b1_prime - zeta_u_prime*zeta_x_prime*alpha_prime
beta <- beta_prime*sigma_y/sigma_x

n <- 10^5
z <- rnorm(n)
u <- zeta_u*z + sqrt(sigma_u^2 - zeta_u^2)*rnorm(n)
x <- zeta_x*z + sqrt(sigma_x^2 - zeta_x^2)*rnorm(n)

# sigma_x
# sd(x)
# sigma_u
# sd(u)
# cor(u,x)

u_prime <- u/sigma_u
x_prime <- x/sigma_x
e <- rnorm(n)

y <- (beta_prime*x_prime + alpha_prime*u_prime + sqrt(1 - (beta_prime^2 + alpha_prime^2 + 2*beta_prime*alpha_prime*rho))*e)*sigma_y
sigma_y
sd(y)

m1 <- lm(y ~ x + u)
m2 <- lm(y ~ x)

alpha
beta
coef(m1)
beta + rho*alpha*sigma_u/sigma_x
coef(m2)
# b' = b*s_x/s_y = xyx/xxy
# b' = beta' + rho*alpha'
# b = b'*sd(y)/sd(x)
# b = (beta' + rho*alpha')*sd(y)/sd(x)
# alpha'=alpha*sd(u)/sd(y)
# beta'=beta*sd(x)/sd(y)
# b = (beta*sd(x)/sd(y) + rho*alpha*sd(u)/sd(y))*sd(y)/sd(x)
# b = beta*sd(x)/sd(y)*sd(y)/sd(x) + rho*alpha*sd(u)/sd(y)*sd(y)/sd(x)
# b = beta + rho*alpha*sd(u)/sd(x)
```

#### Causal modeling with experimental versus observational data

Causal interpretation requires conditioning on "doing X=x". For the green-down data, "doing X=x" is achieved by the random treatment assignment of the enclosures. How does random treatment assignment achieve this? Look again at Figure \@ref(fig:continuous-x-dag). If the values of $treatment$ are randomly assigned, then, by definition, the expected value of the correlation between $treatment$ and $moisture$ is zero. In fact, the expected value of the correlation between $treatment$ and any measurable variable at the study site is zero. Given, this, the expected value of the regression coefficient for $temperature$ ($b_1$) is $\beta$ because $\rho=0$. That is, the estimate of the true effect is unbiased. It doesn't matter if $moisture$, or any other variable, is excluded from the model. (That said, we may want to include moisture or other variables in the model to increase precision of the causal effect. This is addressed in the chapter "Models with Covariates"). This is what an experiment does and why experiments are used to answer causal questions.

Can we use observational data for causal modeling? Yes, but the methods for this are outside of the scope of this text. The mathematical foundation for this is known as **path analysis**, which was developed by geneticist and evolutionary biologist Sewell Wright (I include this because this work has inspired much of how I think about biology and because he is both my academic grandfather and great-grandfather). Causal analysis of observational data in biology is rare outside of ecology and epidemiology. Good starting points for the modern development of causal analysis are [Hernán MA and Robins JM (2020)](https://www.hsph.harvard.edu/miguel-hernan/causal-inference-book/){target="_blank"} and [Burgess et al. (2018)](https://www.annualreviews.org/doi/abs/10.1146/annurev-genom-083117-021731){target="_blank"}. A gentle introduction is [Pearl and Mackenzie (2018)](https://www.basicbooks.com/titles/judea-pearl/the-book-of-why/9780465097609/){target="_blank"}.

### Using a regression model for prediction -- prediction models

The researchers in the green-down study also presented estimates of the effect of temperature on green-down using two observational datasets. Let's use the one in Extended Data Figure 3d to explore using a regression model for prediction. The data are taken from measurements of the day of green-down over an 86 year period at a single site. The response variable is $green\_down\_anomaly$ (the difference between the day of green down for the year and the mean of these days over the 86 years). The predictor variable is $autumn\_temp\_anomaly$ (the difference between the mean temperature over the year and the mean of these means).

```{r continuous-x-import-fig-s3d, echo=FALSE}
data_from <- "Ecosystem warming extends vegetation activity but heightens vulnerability to cold temperatures"
file_name <- "41586_2018_399_MOESM6_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)
efig_3d <- read_excel(file_path,
                   range = "J2:K87",
                   col_names = FALSE) %>% # header causing issues
  data.table() # convert to data.table
setnames(efig_3d, old = colnames(efig_3d), new = c("autumn_temp_anomaly", "green_down_anomaly"))

# View(efig_3d)
```

```{r continuous-x-pred-fig, echo=FALSE, message=FALSE}
m1 <- lm(green_down_anomaly ~ autumn_temp_anomaly, data = efig_3d)

# get x for drawing slope
minx <- min(efig_3d[,autumn_temp_anomaly])
maxx <- max(efig_3d[,autumn_temp_anomaly])
new_x <- seq(minx, maxx, length.out = 20)
new_data <- data.table(autumn_temp_anomaly = new_x)
new_data[, yhat := predict(m1, newdata = new_data)]
new_data[, conf_lwr := predict(m1, 
                           se.fit = TRUE,
                           interval = "confidence",
                           newdata = new_data)$fit[, "lwr"]]
new_data[, conf_upr := predict(m1, 
                           se.fit = TRUE,
                           interval = "confidence",
                           newdata = new_data)$fit[, "upr"]]
new_data[, pred_lwr := predict(m1, 
                           se.fit = TRUE,
                           interval = "prediction",
                           newdata = new_data)$fit[, "lwr"]]
new_data[, pred_upr := predict(m1, 
                           se.fit = TRUE,
                           interval = "prediction",
                           newdata = new_data)$fit[, "upr"]]

gg <- ggscatter(data = efig_3d,
          x = "autumn_temp_anomaly",
          y = "green_down_anomaly",
          color = "black",
          shape = 21,
          size = 3,
          add = "reg.line",
          add.params = list(color = "steelblue",
                            fill = "lightgray"),
          xlab = "Temperature Anomaly (°C)",
          ylab = "Day of Green-down Anomaly (DOY)") +

  geom_ribbon(data = new_data,
              aes(ymin = pred_lwr,
                  ymax = pred_upr,
                  y = yhat,
                  fill = "band"),
              fill = "gray",
              alpha = 0.3) +
  
  geom_ribbon(data = new_data,
              aes(ymin = conf_lwr,
                  ymax = conf_upr,
                  y = yhat,
                  fill = "band"), alpha = 0.3) +
  
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
           size = 4,
           label.y = 10) +
  
  scale_fill_manual(values = pal_okabe_ito) +

  theme(legend.position="none") +

  NULL  

gg
```

The plot in figure \@ref(fig:continuous-x-pred-fig) shows the regression line of a linear model fit to the data. Two statistics are given in the plot.

1. The *p*-value of the slope (the coefficient $b_1$ of $autumn\_temp\_anomaly$)
2. The $R^2$ of the model fit.

In addition, two intervals are shown.

1. 95% **confidence interval** of the expected values in blue. The width of this band is a measure of the precision of the estimates of expected values. 
2. 95% **prediction interval** of the predictions in gray. The width of this band is a measure of how well the model predicts.

It is important that researchers knows what each of these statistics and bands are, when to compute them and when to ignore them them, and what to say about each when communicating the results.

#### 95% CI and *p*-value

Both the 95% confidence interval of the expected values and the *p*-value are a function of the standard error of the slope coefficient $b_1$ and so are complimentary statistics. The *p*-value is the probability of sampling the null that results in a $t$-value as or more extreme than the observed *t* for the coefficient $b_1$. This null includes the hypothesis $\beta_1 = 0$. The 95% confidence interval of the expected values is the band containing expected values (mean $green\_down\_anomaly$ conditional on $autumn\_temp\_anomaly$) that are compatible with the data. There are a couple of useful ways of thinking about this band.

1. The band captures an infinite number of regression lines that are compatible with the data. Some are more steep and predict smaller expected values at the low end of $autumn\_temp\_anomaly$ and higher expected values at the high end of $autumn\_temp\_anomaly$. Others are less steep and predict higher expected values at the low end of $autumn\_temp\_anomaly$ and lower expected values at the high end of $autumn\_temp\_anomaly$.
2. The band captures the 95% CI of the conditional mean at every value of $X$. Consider the 95% CI of the conditional mean at the mean value of $X$. As we move away from this mean value (lower to higher $X$), the 95% CI of the conditional mean increases.

A 95% CI and *p*-value are useful statistics to report if the purpose is *causal modeling*, as in the example above using the experimental green-down data (where the 95% CI was not presented as a confidence band of the expected values but a CI of the estimate of the effect of added temperature). A 95% CI and *p*-value are also useful statistics to report if the purpose is *descriptive modeling*, simply wanting to know how the conditional mean of the response is related to an $X$ variable.

#### 95% prediction interval and $R^2$

Both the $R^2$ and the 95% prediction interval are a function of the *population* variability of $green\_down\_anomaly$ conditional on $autumn\_temp\_anomaly$ (the spread of points along a vertical axis about the regression line) and are complimentary statistics. Briefly,

1. the $R^2$ is a measure of the fraction of the variance in the response that is accounted by the model (some sources say "explained by the model" but this is an odd use of "explain"). It is a number between 0 and 1 and is a measure of "predictability" if the goal is prediction modeling.
2. The 95% prediction interval will contain a new observation 95% of the time. It provides bounds on a **prediction** -- given a new observation, there is a 95% probability that the interval at $x_{new}$ will contain $y_{new}$.

To understand $R^2$ and the 95% prediction interval a bit better, let's back up.

\begin{equation}
green\_down\_anomaly_i = b_0 + b_1 autumn\_temp\_anomaly_i + e_i
(\#eq:doy_fit)
\end{equation}

Model \@ref(eq:eq:doy_fit) recovers the measured value of $green\_down\_anomaly$ for any year, given the $autumn\_temp\_anomaly$ for the year. The equation includes the linear predictor ($b_0 + b_1 autumn\_temp\_anomaly_i$) and the **residual** from the predictor ($e_i$). The predictor part of \@ref(eq:doy_fit) is used to compute $\hat{y}$ ("y hat"). 

\begin{equation}
\hat{y}_i = b_0 + b_1 autumn\_temp\_anomaly_i
(\#eq:doy_hat)
\end{equation}

The $\hat{y}$ are **fitted values**, if the values are computed from the data that were used for the fit, or **predicted values** (or simply the **prediction**), if the values are computed from values of the predictor variables outside the set used to fit the model. For the purpose of plotting, generally, with models with categorical factors as $X$, I prefer either **modeled values** or **conditional means** to fitted values.

#### $R^2$

A good model accounts for a sizable fraction of the total variance in the response. This fraction is the $R^2$ value, which is given in `summary(m1)` and accessed directly with

```{r }
summary(m1)$r.squared
```
 
$R^2$ is the fraction of the total variance of $Y$ that is generated by the linear predictor.

\begin{equation}
R^2 = \frac{\mathrm{VAR}(\hat{y})}{\mathrm{VAR}(y)}
\end{equation}

```{r}
yhat <- fitted(m1)
y <- efig_3d[, green_down_anomaly]
var(yhat)/var(y)
```

$R^2$ will vary from zero (the model accounts for nothing) to one (the model accounts for everything). $R^2$ is often described as the fraction of total variation *explained* by the model" but the usage of "explain" is observational and not causal. Because of the ambiguous usage of "explain" in statistics, I prefer to avoid the word.

It can be useful sometimes to think of $R^2$ in terms of **residual error**, which is the variance of the residuals from the model. The larger the residual error, the smaller the $R^2$, or

\begin{equation}
R^2 = 1 - \frac{\mathrm{VAR}(e)}{\mathrm{VAR}(y)}
\end{equation}

```{r}
e <- residuals(m1)
y <- efig_3d[, green_down_anomaly]
1 - var(e)/var(y)
```

The smaller the residuals, the higher the $R^2$ and the closer the predicted values are to the measured values. The sum of the model variance and residual variance equals the total variance and, consequently, $R^2$ is a signal to signal + noise ratio. The noise in $R^2$ is the sampling variance of the individual measures. The noise in a *t*-value is the sampling variance of the parameter (for m1, this is the sampling variance of $b_1$). This is an important distinction because it means that *t* and $R^2$ are not related 1:1. In a simple model with only a single $X$, one might expect $R^2$ to be big if the *p*-value for the slope is tiny, but this isn't necessarily true because of the different meaning of noise in each. A study with a very large sample size $n$ can have a tiny *p*-value *and* a small $R^2$. A *p*-value is not a good indicator of predictability. $R^2$ is. This is explored more below.

#### Prediction interval
A good prediction model has high predictability, meaning the range of predicted values is narrow. A 95% CI is a reasonable range to communicate. For any decision making using prediction, it is better to look at numbers than a band on a plot.

```{r echo=FALSE}
future_data <- data.table(autumn_temp_anomaly = seq(0.5, 2.5, by = 0.5))
yhat <- predict(m1, 
                se.fit = TRUE,
                interval = "prediction",
                newdata = future_data)$fit
future_data <- cbind(future_data, yhat)
pretty_names <- c("Autumn Temp Anomaly (°C)", "Expected (days)", "2.5% (days)", "97.5% (days)")
setnames(future_data, old = colnames(future_data), new = pretty_names)
knitr::kable(future_data, digits = c(1, 1, 1, 1))
```

Given these data and the fit model, if we see a 2 °C increase in mean fall temperature, we expect the autumn green-down to extend 3.2 days. This expectation is an average. We'd expect 95% of actual values to range between -6.4 days and 12.8 days. This is a lot of variability. Any prediction has a reasonable probability of being over a week early or late. This may seem surprising given the *p*-value of the fit, which is 0.0006. But the *p*-value is not a reliable indicator of predictability. It is a statistic related to the blue band, not the gray band.

To understand this prediction interval a bit more, and why a *p*-value is not a good indicator of predictability, it's useful to understand what makes a prediction interval "wide". The width of the prediction interval is a function of two kinds of variability

1. The variance of the expected value at a specific value of $X$. This is the square of the standard error of $b_1$. The blue band is communicating the CI based on this variance. The *p*-value is related to the wideth of this band.
2. The variance of $Y$ at a specific value of $X$. This variance is $\sigma^2$. It is useful for learning to think about the equation for the estimate of this variance.

\begin{equation}
\hat\sigma^2 = \frac{\sum{e_i^2}}{N-2}
\end{equation}

Again, $e_i$ is the residual for the *i*th case. The denominator ($N-2$) is the degrees of freedom of the model. Computing $\hat\sigma^2$ manually helps to insure that we understand what is going on.

```{r}
summary(m1)$sigma # R's calculation of sigma hat
df <- summary(m1)$df[2] # R's calculation of df. Check that this is n-2!
sqrt(sum(e^2)/df)
```

Remember that an assumption of the linear models that we are working with at this point is, this variance is constant for all values of $X$, so we have a single $\sigma$. Later, we will cover linear models that model heterogeneity in this variance. $\sigma$ is a function of variability in the population -- it is the population standard deviation conditional on $X$.

```{r excised-text-324583, echo=FALSE, eval=FALSE}
# <div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
# The denominator of the equation for the variance of $Y$ conditional on $X$ is the degrees of freedom of the model used to estimate the expected value of $Y$ conditional on $X$, or $\mathrom{E}[Y|X]$. What does degrees of freedom mean? The model fit to the data has two estimates, $b_0$ (the intercept) and $b_1$ (the slope). Consequently, once we fit the model, only $n - 2$ $Y$ variables can vary. If we fit the model using three cases we have one degree of freedom. Given $x_1$, $y_1$ can be any value. It is "free to vary".
# 
# sum y = mean
# sum xtx_1 xty adds up
# </div> 
```

Importantly, predictability is a function of both these components of variability. As a consequence, it is $R^2$, and not the *p*-value, that is the indicator of predictability. In the observational green-down data, even if we had thousands of years of data, we would still have a pretty low $R^2$ because of the population variability of $green\_down\_anomaly$ at a given $autumn\_temp\_anomaly$.

#### Median Absolute Error and Root Mean Square Error are absolute measures of predictability

A *p*-value is not *at all* a good guide to predictability. $R^2$ is proportional to predictability but is not really useful in any absolute sense. If we want to predict the effect of warming on the day of green-down, I would like to have a measure of predictability in the units of green-down, which is days. The prediction interval gives this for any value of $X$. But what about an overall measure of predictability?

Three overall measures of predictability are

1. $\hat{\sigma}$, the estimate of $\sigma$. This is the standard deviation of the sample conditional on $X$.
2. $RMSE$, the root mean squared error. This is

\begin{align}
SSE &= \sum{(y_i - \hat{y}_i)^2}\\
MSE &= \frac{SSE}{N}\\
RMSE &= \sqrt{MSE}
\end{align}

$SSE$ ("sum of squared error") is the sum of the squared residuals ($y_i - \hat{y}_i$) of the model. $MSE$ ("mean squared error") is the mean of the squared errors. $RMSE$ is the square root of the mean squared error. $RMSE$ is almost equal to $\hat{\sigma}$. The difference is the denominator, which is $N$ in the computation of $RMSE$ and $df$ (the degrees of freedom of the model, which is $N$ minus the number of fit parameters) in the computation of $\hat{\sigma}$.

3. $MAE$, the mean absolute error. This is

\begin{equation}
MAE = \frac{1}{N}\sum{|y_i - \hat{y}_i|}
\end{equation}

```{r continuous-x-average-error, echo=FALSE}
sigma_hat <- summary(m1)$sigma
y_observed <- efig_3d[, green_down_anomaly]
y_predicted <- predict(m1)
root_mean_square_error <- rmse(y_observed, y_predicted)
mean_absolute_error <- mae(y_observed, y_predicted)
error_table <- data.table(
  sigma = sigma_hat,
  RMSE = root_mean_square_error,
  MAE = mean_absolute_error
)
knitr::kable(error_table, digits = 2)
```
If the goal of an analysis is prediction, one of these statistics should be reported. For the model fit to the observational green-down data in Extended Figure 3d, these three statistics are given in Table \@ref(table:continuous-x-average-error) above (to two decimal places simply to show numerical difference between $\hat{\sigma}$ and $RMSE$). All of these are measures of an "average" prediction error in the units of the response. The average error is either 4.7 or 3.6 days, depending on which statistic we report. Why the difference? Both $\hat{\sigma}$ and $RMSE$ are functions of squared error and so big differences between predicted and actual value are emphasized. If an error of 6 days is more than twice as bad than an error of 3 days, report $RMSE$. Why $RMSE$ and not $\hat{\sigma}$? Simply because researchers using prediction models are more familiar with $RMSE$. If an error of 6 days is *not* more than twice as bad than an error of 3 days, report $MAE$.

#### Prediction modeling is more sophisticated than presented here

For data where the response is a non-linear function of the predictor, or for data with many predictor variables, researchers will often build a model using a **model selection** method. Stepwise regression is a classical model selection method that is commonly taught in intro biostatistics and commonly used by researchers. Stepwise regression as a method of model selection has many well-known problems and should be avoided.

Some *excellent* books that are useful for building models and model selection are

1. [The Elements of Statistical Learning](https://web.stanford.edu/~hastie/ElemStatLearn/){target="_blank"}
2. [Regression and Other Stories](https://avehtari.github.io/ROS-Examples/){target="_blank"}
3. [Regression Modeling Strategies](http://hbiostat.org/doc/rms/4day.html){target="_blank"}

### Using a regression model for creating a new response variable -- comparing slopes of longitudinal data

In the study in this example, the researchers compared the growth rate of tumors in mice fed normal chow versus mice with a methionine restricted diet. Growth rate wasn't actually compared. Instead, the researchers used a *t*-test to compare the size of the tumor measured at six different days. A problem with multiple *t*-tests for this dataset is that the errors (residuals from the model) on one day are correlated with the errors from another day because of the repeated measures on each mouse. This correlation will inflate Type I error rate. 

Instead of six *t*-tests, a better strategy for these data is to use a regression to calculate a tumor growth rate for each mouse. There are sixteen mice so this is sixteen fit models. Here I use a "for loop" to fit the model to the data from a single mouse and use the slope ($b_1$) as the estimate of the tumor growth rate for that mouse.

```{r continuous-x-long-import, echo=FALSE}
data_from <- "Dietary methionine influences therapy in mouse cancer models and alters human metabolism"
file_name <- "41586_2019_1437_MOESM2_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)

fig1f_wide <- read_excel(file_path,
                    sheet = "f",
                    range = "B7:R12",
                    col_names = FALSE) %>%
  data.table() # convert to data.table

setnames(fig1f_wide,
         old = colnames(fig1f_wide),
         new = c("day",
                 paste0("Cn_", 1:8),
                 paste0("MR_", 1:8)))

fig1f_wide <- transpose(fig1f_wide, make.names = 1, keep.names = "id")
fig1f_wide[, treatment := factor(substr(id, 1, 2))]

days <- c(21, 25, 28, 30, 34, 39)
fig1f <- melt(fig1f_wide,
              id.vars <- c("treatment", "id"),
              measure.vars <- as.character(days),
              variable.name = "day",
              value.name = "volume")
fig1f[, day := as.numeric(as.character(day))]
# View(fig1f)
# qplot(x = day, y = volume, color = treatment, data = fig1f) + geom_line(aes(group = id))
```

Use a for-loop to estimate growth rate for each mouse. In each pass through the loop

1. the subset of fig1f (the data in long format) belonging to mouse *i* is created
2. the linear model `volume ~ day` is fit to the subset
3. the coefficient of day (the slope, $b_1$) is inserted in mouse *i*'s row in the column "growth" in the data.table "fig1f_wide".

At the end of the loop, we have the data.table fig1f_wide which has one row for each mouse, a column for the treatment factor (diet) and a column called "growth" containing each mouse's growth rate. There are also columns of tumor volume for each mouse on each day but these are ignored.

```{r continuous-x-long-get-slopes}
N <- nrow(fig1f_wide)
id_list <- fig1f_wide[, id]
for(i in 1:N){
  mouse_i_data <- fig1f[id == id_list[i]] # subset
  fit <- lm(volume ~ day, data = mouse_i_data)
  fig1f_wide[id == id_list[i], growth := coef(fit)[2]]
}
# View(fig1f_wide)
# qplot(x = treatment, y = growth, data = fig1f_wide)
# qplot(x = day, y = volume, color = treatment, data = fig1f) + geom_smooth(aes(group = id), method = "lm", se = FALSE)
```

Step 3. **fit the model**

```{r continuous-x-long-fit}
m1 <- lm(growth ~ treatment, data = fig1f_wide)
```

Step 5. **inference from the model**

```{r continuous-x-long-inference}
m1_coef <- summary(m1) %>%
  coef
m1_ci <- confint(m1)
(m1_coef_table <- cbind(m1_coef, m1_ci))
```

Step 6. **plot the model**

```{r continuous-x-long-plot-response, echo=FALSE, message=FALSE}
fig1f_gg_response <- ggplot(data = fig1f,
                      aes(x = day, y = volume, color = treatment)) +
  geom_point() +
  geom_smooth(aes(group = id),
              method = "lm",
              se = FALSE) +
  xlab("Day") +
  ylab(expression(Tumor~Volume~(mm^3))) +
  scale_color_manual(values = pal_okabe_ito) +
  theme_pubr() +
  theme(
    legend.position = c(.15, .98),
    legend.justification = c("right", "top"),
    legend.box.just = "right",
    legend.margin = margin(6, 6, 6, 6),
    legend.title = element_blank()
    ) +

  NULL

# fig1f_gg_response
```

```{r continuous-x-long-plot-effect, echo=FALSE}
m1_coeff_dt <- data.table(term = row.names(m1_coef_table),
                          data.table(m1_coef_table))[2,] %>%
  clean_names()
m1_coeff_dt[ , p_pretty := pvalString(pr_t)]

min_bound <- min(m1_coeff_dt[, x2_5_percent])
max_bound <- min(m1_coeff_dt[, x97_5_percent])

y_lo <- min(min_bound+min_bound*0.2,
            -max_bound)
y_hi <- max(max_bound + max_bound*0.2,
            -min_bound)
y_lims <- c(y_lo, y_hi)

fig1f_gg_effect <- ggplot(data=m1_coeff_dt, 
                          aes(x = term,
                              y = estimate)) +
  # confidence level of effect
  geom_errorbar(aes(ymin=x2_5_percent, 
                    ymax=x97_5_percent),
                width=0, 
                color="black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # zero effect
  geom_hline(yintercept=0, linetype = 2) +
  
  # p-value
  annotate(geom = "text",
           label = m1_coeff_dt$p_pretty,
           x = 1,
           y = 7.5) +
  
  # aesthetics
  scale_y_continuous(position="right") +
  scale_x_discrete(labels = "MR\neffect") +
  ylab(expression(Growth~(mm^3/day))) +
  coord_flip(ylim = y_lims) + 
  theme_pubr() +
  theme(axis.title.y = element_blank()) +
  
  NULL

#fig1f_gg_effect
```

```{r continuous-x-long-plot, echo=FALSE, message=FALSE}
fig1f_gg <- plot_grid(fig1f_gg_effect,
                      fig1f_gg_response,
                      nrow=2,
                      align = "v",
                      axis = "lr",
                      rel_heights = c(0.4,1))
fig1f_gg
```

### Using a regression model for for calibration

## Working in R
### Fitting the linear model {#fit-the-model}

A linear model is fit with the `lm` function, which is very flexible and will be a workhorse in much of this text.

```{r continuous-x-working-fit}
m1 <- lm(transition_date ~ temperature, data = fig2c)
```

m1 is an lm model object that contains many different kinds information, such as the model coefficients.

```{r coef-function}
coef(m1)
```

We'll return to others, but first, let's explore some of the flexibility of the `lm` function. Two arguments were sent to the function

1. the model formula `transition_date ~ temperature` with the form `Y ~ X`, where Y and X are names of columns in the data.
* The model formula itself can be assigned to a variable, which is useful when building functions. An example

```{r}
coef_table <- function(x, y, data){
  m1_form <- formula(paste(y, "~", x))
  m1 <- lm(m1_form, data = data)
  return(coef(summary(m1)))
}

coef_table(x = "temperature",
           y = "transition_date",
           data = fig2c)
```

* Both Y and X can also be column names embedded within a function, for example

```{r}
m2 <- lm(log(transition_date) ~ temperature, data = fig2c)
coef(m2)
```

or

```{r}
m3 <- lm(scale(transition_date) ~ scale(temperature), data = fig2c)
coef(m3)
```

2. The data frame (remember that a data.table is a data frame) containing the columns with the variable names in the model formula. A data argument is not necessary but it is usually the better way (an exception is when a researcher wants to create a matrix of Y variables or to construct their own model matrix)

type `?lm` into the console to see other arguments of the `lm` function. 

```{r}
x <- fig2[, temperature]
y <- fig2[, transition_date]
m4 <- lm(y ~ x)
coef(m4)
```

### Getting to know the linear model: the `summary` function

The `lm` function returns an `lm` object, which we've assigned to the name `m1`. `m1` contains lots of information about our fit of the linear model to the data. Most of the information that we want for most purposes can be retrieved with the `summary` function, which is a general-purpose R command the works with many R objects.

```{r summary-function}
summary(m1)
```

What is here:

**Call**. This is the model that was fit

**Residuals**. This is a summary of the distribution of the residuals. From this one can get a sense of the distribution (for inference, the model assumes a normal distribution with mean zero). More useful ways to examine this distribution will be introduced later in this chapter.

**Coefficients table**. This contains the linear model coefficients and their standard error and associated $t$ and $p$ values.

1. The column of values under "Estimate" are the coefficients of the fitted model (equation \@ref(eq:continuous-x-fit)). Here, `r coef(m1)[1]` is the intercept ($b_0$) and `r coef(m1)[2]` is the effect of $temperature$ ($b_1$).
2. The column of values under "Std. Error" are the standard errors of the coefficients.
3. The column of values under "t value" are the *t-statistics* for each coefficient. A *t*-value is a **signal to noise ratio**. The coefficient $b_1$ is the "signal" and the SE is the noise. Get used to thinking about this ratio. A *t*-value greater than about 3 indicates a "pretty good" signal relative to noise, while one much below than 2 is not.
4. The column of values under "Pr(>|t|)" is the *p*-value, which is the t-test of the estimate. What is the *p*-value a test of? The *p*-value tests the hypothesis "how probable are the data, or more extreme than than the data, if the true parameter is zero?". Formally $p = \mathrm{freq(t' \ge t|H_o)}$, where $t'$ is the hypothetical *t*-value, *t* is the observed $t$-value, and $H_o$ is the null hypothesis. We will return to *p*-values in Chapter xxx.

**Signif. codes**. Significance codes are extremely common in the wet bench experimental biology literature but do not have much to recommend. I'll return to these in the *p*-values chapter.

Beneath the Signif. codes are some model statistics which are useful

**Residual standard error** This is $\sqrt{\sum{e_i^2}/(n-2)}$, where $e_i$ are the residuals in the fitted model. "degrees of freedom" is the number of $e_i$ that are "allowed to vary" after fitting the parameters, so is the total sample size ($n$) minus the number of parameters in the model. The fit model has two fit parameters ($b_0$ and $b_1$ so the df is $n-2$. Note that this is the denominator in the residual standard error equation.

**Multiple R-squared**. This is an important but imperfect summary measure of the whole model that effectively measures how much of the total variance in the response variable "is explained by" the model. Its value lies between zero and 1. **It's a good measure to report in a manuscript**, especially for observational data.

**F-statistic and p-value**. These are statistics for the whole model (not the individual coefficients) and I just don't find these very useful.

### Inference -- the coefficient table

There are different ways of 

```{r coef_table}
# step by step

m1_summary <- summary(m1) # get summary
m1_coef_p1 <- coef(m1_summary) # extract coefficient table
m1_coef_p2 <- confint(m1) # get CIs
m1_coef <- cbind(m1_coef_p1, m1_coef_p2) # column bind the two

# this can be shortened (but still readable) using
m1_coef <- cbind(coef(summary(m1)),
                 confint(m1))
m1_coef
```

Note that the *p*-value for the coefficient for *temperature* is very small and we could conclude that the data are not compatible with a model of no temperature effect on day of green-down. But did we need a formal hypothesis test for this? We haven't learned much if we have only learned that the slope is "not likely to be exactly zero" (Temperature effects everything in biology). A far more important question is not *if* there is a relationship between *temperature* and *day* of green-down, but what is the sign and magnitude of the effect and what is the uncertainty in our estimate of this effect. For this, we we want the coefficient and its SE or confidence interval, both of which are in this table. Remember, our working definition of a confidence interval:

**A confidence interval contains the range of parameter values that are compatible with the data, in the sense that a $t$-test would not reject the null hypothesis of a difference between the estimate and any value within the interval**

A more textbook way of defining a CI is: A 95% CI of a parameter has a 95% probability of including the true value of the parameter. It does not mean that there is a 95% probability that the true value lies in the interval. This is a subtle but important difference. Here is a way of thinking about the proper meaning of the textbook definition: we don't know the true value of $\beta_1$ but we can 1) repeat the experiment or sampling, 2) re-estimate $\beta_1$, and 3) re-compute a 95% CI. If we do 1-3 many times, 95% of the CIs will include $\beta_1$ within the interval.

Confidence intervals are often interpreted like $p$-values. That is, the researcher looks to see if the CI overlaps with zero and if it does, concludes there is "no effect". First, this conclusion is not correct -- **the inability to find sufficient evidence for an effect does not mean there is no effect, it simply means there is insufficient evidence to conclude there is an effect**!

Second, what we want to use the CI for is to guide us about how big or small the effect might reasonably be, given the data. Again, A CI is a measure of parameter values that are "compatible" with the data. If our biological interpretations at the small-end and at the big-end of the interval's range radically differ, then we don't have enough *precision* in our analysis to reach an unambiguous conclusion.

### How good is our model? -- Model checking

There are two, quite different senses of what is meant by a good model.

1. How good is the model at predicting? Or, how much of the variance in the response (the stuff to be "explained") is accounted for by the model? This was described above.
2. How well do the data look like a sample from the modeled distribution? If not well, we should consider alternative models. This is **model checking**.

For inference, a good model generates data that look like the real data. If this is true, our fit model will have well-behaved residuals. There are several aspects of "well-behaved" and each is checked with a diagnostic plot. This **model checking** is covered in more detail in chapter xxx.

Inference from model \@ref(eq:continuous-x-m1) assumes the data were sampled from a normal distribution. To check this, use a quantile-quantile or **Q-Q plot**. The `qqPlot` function from the car package generates a more useful plot than that from Base R.

```{r}
set.seed(1)
qqPlot(m1, id=FALSE)

```

Approximately normal residuals will track the solid line and stay largely within the boundaries marked by the dashed lines. The residuals from m1 fit to the green-down data track the solid line and remain within the dashed lines, indicating adequate model fit.

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
Note that a formal test of normality is often recommended. Formal tests do not add value above a diagnostic check. Robustness of inference (for example, a *p*-value) is a function of the type and degree of "non-normalness", not of a p-value. For a small sample, there is not much power to test for normality, so samples from non-normal distributions pass the test ($p > 0.05$) and are deemed "normal". For large samples, samples from distributions that deviate slightly from normal fail the test ($p < 0.05$) and are deemed "not normal". Inference with many non-normal samples with large $n$ are very robust (meaning infernece is not likely to fool you with randomness).
</div> 

Inference from model \@ref(eq:continuous-x-m1) assumes homogeneity of the response conditional on $X$. For a continuous $X$, this means the residuals should have approximately equal variance at low, mid, and high values of $X$ (and everywhere in between). One can visually inspect the spread of points in the $Y$ direction across the groups for categorical $X$ or along the *X*-axis for continuous $X$. A useful method for checking how residual variance changes (usually increases) with the conditional mean of $Y$ is a **spread-location plot**. The `spreadLevelPlot(m1)` function from the car package is useful.

```{r}
spreadLevelPlot(m1)

```

The dashed blue line shows linear trends while the magenta line shows non-linear trends. For the green-down data, the dashed line is very flat while the magenta-line shows what looks like random fluctations. Taken together, the two lines indicate adequate model fit. 

### Plotting models with continuous *X*
#### Quick plot

`qplot` from the ggplot package is useful for initial examinations of the data

```{r, message=FALSE}
qplot(x = temperature, y = transition_date, data = fig2c) + geom_smooth()
```

Some variations of this quick plot to explore include

1. use `geom_smooth(method = "lm")`

#### ggpubr plots

ggpubr is a package with functions that automates the construction of publication-ready ggplots. `ggscatter` can generate a publishable plot with little effort.

```{r message=FALSE}
ggscatter(data = fig2c,
          x = "temperature",
          y = "transition_date",
          add = "reg.line",
          xlab = "Added Temperature (°C)",
          ylab = "Day of Green-down (DOY)")

```

It take only a little more effort to add useful modifications.

```{r message=FALSE}
ggscatter(data = fig2c,
          x = "temperature",
          y = "transition_date",
          color = "black",
          fill = "red",
          shape = 21,
          size = 3,
          add = "reg.line",
          add.params = list(color = "steelblue",
                            fill = "lightgray"),
          conf.int = TRUE,
          xlab = "Added Temperature (°C)",
          ylab = "Day of Green-down (DOY)") +
  
  stat_regline_equation(size = 4,
                        label.y = 340) +
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
           size = 4,
           label.y = 335) +

  NULL
```

Notes

@. The interval shown is the 95% confidence interval of the expected values, which is what we want to communicate with the experimental green-down data. Were we to want to use this as a prediction model, we would want the 95% prediction interval. I'm not sure that ggpubr can plot a prediction interval. To see how I plotted the prediction interval in Figure \@ref(fig:continuous-x-pred-fig) above, see the Hidden Code section below.
@. The arguments of the `ggscatter` function are typed in explicitly (`x = "temperarture"` and not just `"temperature"`). Each argument starts on a new line to increase readability.
@. The `+` after the `ggscatter` function adds additional layers to the plot. Each additional component is started on a new line to increase readability.
@. The first line of the `stat_cor` function (everything within `aes()`) plots the $R^2$ instead of the correlation coefficient $r$. Copy and pasting this whole line just works.
@. Comment out the line of the ggplot script starting with `stat_cor` and re-run (comment out by inserting a `#` at the front of the line. A consistent way to do this is to triple-click the line to highlight the line and then type command-shift-c on Mac OS). The script runs without error because `NULL` has been added as the final plot component. Adding "NULL" is a useful trick.

```{r}
ggscatter(data = efig_3d,
          x = "autumn_temp_anomaly",
          y = "green_down_anomaly",
          color = "black",
          fill = "red",
          shape = 21,
          size = 3,
          add = "reg.line",
          add.params = list(color = "steelblue",
                            fill = "lightgray"),
          conf.int = TRUE,
          xlab = "Added Temperature (°C)",
          ylab = "Day of Green-down (DOY)")
```

#### ggplots

Use `ggplot` from the ggplot2 package for full control of plots. See the Hidden Code below for how I generated the plots in Figure \@ref(fig:continuous-x-plot) above.

### Creating a table of predicted values and 95% prediction intervals

efig_3d is the data.table created from importing the data in Extended Data Figure 3d above. The fit model is

```{r}
efig3d_m1 <- lm(green_down_anomaly ~ autumn_temp_anomaly, data = efig_3d)
```

The `predict(efig3d_m1)` function is used to compute either fitted or predicted values, and either the confidence or prediction interval of these values.

Unless specified by `newdata`, the default x-values used to generate the y in `predict(efig3d_m1)` are the x-values in the data used to fit the model. The returned values are the expected value for each $x_i$ in the data. The argument `newdata` passes a data.frame (remember a data.table is a data.frame!) with new x-values. Since these x-values were not in the data used to fit the model, the returned $y_{hat}$ are predicted values.

The range of x-values in the data is

```{r}
range(efig_3d[,autumn_temp_anomaly])
```

Let's get predicted values for a value of $autumn\_temp\_anomaly = 2$.

```{r}
predict(efig3d_m1, newdata = data.table(autumn_temp_anomaly = 2))
```

And predicted values across the range of measured values

```{r}
new_dt <- data.table(autumn_temp_anomaly = c(-2, -1, 1, 2))
predict(efig3d_m1, newdata = new_dt)
```
Add 95% confidence intervals (this could be used to create the band for plotting)

```{r}
predict(efig3d_m1, 
        newdata = new_dt,
        interval = "confidence",
        se.fit = TRUE)$fit

```
Change to the 95% prediction intervals

```{r}
predict(efig3d_m1, 
        newdata = new_dt,
        interval = "prediction",
        se.fit = TRUE)$fit
```
Put this all together in a pretty table. I've used knitr's `kable` function but there are table packages in R that allow extensive control of the output.

```{r}
efig3d_m1 <- lm(green_down_anomaly ~ autumn_temp_anomaly, data = efig_3d)
new_dt <- data.table(autumn_temp_anomaly = c(-2, -1, 1, 2))
prediction_table <- predict(efig3d_m1, 
                            newdata = new_dt,
                            interval = "prediction",
                            se.fit = TRUE)$fit
prediction_table <- data.table(new_dt, prediction_table)
pretty_names <- c("Autumn Temp Anomaly", "Estimate", "2.5%", "97.5%")
setnames(prediction_table, old = colnames(prediction_table), new = pretty_names)
knitr::kable(prediction_table, digits = c(1, 1, 1, 1))
```


## Hidden code

### Import and plot of fig2c (ecosystem warming experimental) data

Import

```{r continuous-x-import-2, echo=TRUE, eval=FALSE}
data_from <- "Ecosystem warming extends vegetation activity but heightens vulnerability to cold temperatures"
file_name <- "41586_2018_399_MOESM3_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)
fig2 <- read_excel(file_path) %>% # import
  clean_names() %>% # clean the column names
  data.table() # convert to data.table
# View(fig2)

fig2c <- fig2[panel == "2c",]
# View(fig2c)
```

Creating the response plot (the bottom component)

```{r continuous-x-plot-response-2, echo=TRUE, eval=FALSE}
m1_b <- coef(m1)
m1_ci <- confint(m1)
m1_b0_text <- paste0("Intercept: ",
                  round(m1_b[1],1),
                  " (",
                  round(m1_ci[1,1],1),
                  ", ",
                  round(m1_ci[1,2],1),
                  ") d")
m1_b1_text <- paste0("Slope: ",
                  round(m1_b[2],2),
                  " (",
                  round(m1_ci[2,1],2),
                  ", ",
                  round(m1_ci[2,2],2),
                  ") d per 1 °C")

# regression line
m1_x <- min(fig2c[, temperature])
m1_xend <- max(fig2c[, temperature])
m1_y <- m1_b[1] + m1_b[2]*m1_x
m1_yend <- m1_b[1] + m1_b[2]*m1_xend

fig2c_gg_response <- ggplot(data = fig2c,
             aes(x = temperature,
                y = transition_date)) +
  
  # regression line first, to not block point
  geom_segment(x = m1_x,
               y = m1_y,
               xend = m1_xend,
               yend = m1_yend) +
  # create black edge to points
  # geom_point(size = 4,
  #            color = "black") +
  geom_point(size = 3,
             color = pal_okabe_ito[1]) +
   scale_x_continuous(breaks = c(0, 2.25, 4.5, 6.75, 9)) +
  xlab("Plot temperature (ΔT, °C)") +
  ylab("Autumn green-down (DOY)") +
  theme_pubr() +
  
  NULL

# fig2c_gg_response
```

Creating the effects plot (the top component)

```{r  continuous-x-plot-effect-2, echo=TRUE, eval=FALSE}
m1_coeff_dt <- data.table(term = row.names(m1_coeff),
                          data.table(m1_coeff))[2,] %>%
  clean_names()
m1_coeff_dt[ , p_pretty := pvalString(pr_t)]

min_bound <- min(m1_coeff_dt[, x2_5_percent])
max_bound <- min(m1_coeff_dt[, x97_5_percent])

y_lo <- min(min_bound+min_bound*0.2,
            -max_bound)
y_hi <- max(max_bound + max_bound*0.2,
            -min_bound)
y_lims <- c(y_lo, y_hi)

fig2c_gg_effect <- ggplot(data=m1_coeff_dt, 
                          aes(x = term,
                              y = estimate)) +
  # confidence level of effect
  geom_errorbar(aes(ymin=x2_5_percent, 
                    ymax=x97_5_percent),
                width=0, 
                color="black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # zero effect
  geom_hline(yintercept=0, linetype = 2) +
  
  # p-value
  annotate(geom = "text",
           label = m1_coeff_dt$p_pretty,
           x = 1,
           y = 7.5) +
  
  # aesthetics
  scale_y_continuous(position="right") +
  scale_x_discrete(labels = "Temperature\neffect") +
  ylab("Effects (day/°C)") +
  coord_flip(ylim = y_lims) + 
  theme_pubr() +
  theme(axis.title.y = element_blank()) +
  
  NULL

#fig2c_gg_effect
```

Combining the response and effects plots into single plot

```{r continuous-x-plot-2, echo=TRUE, eval=FALSE}
fig3d_fig <- plot_grid(fig2c_gg_effect,
                       fig2c_gg_response,
                       nrow=2,
                       align = "v",
                       axis = "lr",
                       rel_heights = c(0.4,1))
fig3d_fig
```

### Import and plot efig_3d (Ecosysem warming observational) data

Import

```{r , echo=TRUE, eval=FALSE}
data_from <- "Ecosystem warming extends vegetation activity but heightens vulnerability to cold temperatures"
file_name <- "41586_2018_399_MOESM6_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)
efig_3d <- read_excel(file_path,
                   range = "J2:K87",
                   col_names = FALSE) %>% # header causing issues
  data.table() # convert to data.table
setnames(efig_3d, old = colnames(efig_3d), new = c("autumn_temp_anomaly", "green_down_anomaly"))

# View(efig_3d)
```

Plot

```{r , echo=TRUE, eval=FALSE}
m1 <- lm(green_down_anomaly ~ autumn_temp_anomaly, data = efig_3d)

# get x for drawing slope
minx <- min(efig_3d[,autumn_temp_anomaly])
maxx <- max(efig_3d[,autumn_temp_anomaly])
new_x <- seq(minx, maxx, length.out = 20)
new_data <- data.table(autumn_temp_anomaly = new_x)
new_data[, yhat := predict(m1, newdata = new_data)]
new_data[, conf_lwr := predict(m1, 
                           se.fit = TRUE,
                           interval = "confidence",
                           newdata = new_data)$fit[, "lwr"]]
new_data[, conf_upr := predict(m1, 
                           se.fit = TRUE,
                           interval = "confidence",
                           newdata = new_data)$fit[, "upr"]]
new_data[, pred_lwr := predict(m1, 
                           se.fit = TRUE,
                           interval = "prediction",
                           newdata = new_data)$fit[, "lwr"]]
new_data[, pred_upr := predict(m1, 
                           se.fit = TRUE,
                           interval = "prediction",
                           newdata = new_data)$fit[, "upr"]]

gg <- ggscatter(data = efig_3d,
          x = "autumn_temp_anomaly",
          y = "green_down_anomaly",
          color = "black",
          shape = 21,
          size = 3,
          add = "reg.line",
          add.params = list(color = "steelblue",
                            fill = "lightgray"),
          xlab = "Temperature Anomaly (°C)",
          ylab = "Day of Green-down Anomaly (DOY)") +

  geom_ribbon(data = new_data,
              aes(ymin = pred_lwr,
                  ymax = pred_upr,
                  y = yhat,
                  fill = "band"),
              fill = "gray",
              alpha = 0.3) +
  
  geom_ribbon(data = new_data,
              aes(ymin = conf_lwr,
                  ymax = conf_upr,
                  y = yhat,
                  fill = "band"), alpha = 0.3) +
  
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
           size = 4,
           label.y = 10) +
  
  scale_fill_manual(values = pal_okabe_ito) +

  theme(legend.position="none") +

  NULL  

gg
```

### Import and plot of fig1f (methionine restriction) data

Import

```{r continuous-x-long-import-2, echo=TRUE, eval=FALSE}
data_from <- "Dietary methionine influences therapy in mouse cancer models and alters human metabolism"
file_name <- "41586_2019_1437_MOESM2_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)

fig1f_wide <- read_excel(file_path,
                    sheet = "f",
                    range = "B7:R12",
                    col_names = FALSE) %>%
  data.table() # convert to data.table

setnames(fig1f_wide,
         old = colnames(fig1f_wide),
         new = c("day",
                 paste0("Cn_", 1:8),
                 paste0("MR_", 1:8)))

fig1f_wide <- transpose(fig1f_wide, make.names = 1, keep.names = "id")
fig1f_wide[, treatment := factor(substr(id, 1, 2))]

days <- c(21, 25, 28, 30, 34, 39)
fig1f <- melt(fig1f_wide,
              id.vars <- c("treatment", "id"),
              measure.vars <- as.character(days),
              variable.name = "day",
              value.name = "volume")
fig1f[, day := as.numeric(as.character(day))]
# View(fig1f)
# qplot(x = day, y = volume, color = treatment, data = fig1f) + geom_line(aes(group = id))
```

Creating the response plot (the bottom component)

```{r continuous-x-long-plot-response-2, echo=TRUE, eval=FALSE}
fig1f_gg_response <- ggplot(data = fig1f,
                      aes(x = day, y = volume, color = treatment)) +
  geom_point() +
  geom_smooth(aes(group = id), method = "lm", se = FALSE) +
  xlab("Day") +
  ylab(expression(Tumor~Volume~(mm^3))) +
  scale_color_manual(values = pal_okabe_ito) +
  theme_pubr() +
  theme(
    legend.position = c(.15, .98),
    legend.justification = c("right", "top"),
    legend.box.just = "right",
    legend.margin = margin(6, 6, 6, 6),
    legend.title = element_blank()
    ) +

  NULL

# fig1f_gg_response
```

Creating the effects plot (the top component)

```{r continuous-x-long-plot-effect-2, echo=TRUE, eval=FALSE}
m1_coeff_dt <- data.table(term = row.names(m1_coef_table),
                          data.table(m1_coef_table))[2,] %>%
  clean_names()
m1_coeff_dt[ , p_pretty := pvalString(pr_t)]

min_bound <- min(m1_coeff_dt[, x2_5_percent])
max_bound <- min(m1_coeff_dt[, x97_5_percent])

y_lo <- min(min_bound+min_bound*0.2,
            -max_bound)
y_hi <- max(max_bound + max_bound*0.2,
            -min_bound)
y_lims <- c(y_lo, y_hi)

fig1f_gg_effect <- ggplot(data=m1_coeff_dt, 
                          aes(x = term,
                              y = estimate)) +
  # confidence level of effect
  geom_errorbar(aes(ymin=x2_5_percent, 
                    ymax=x97_5_percent),
                width=0, 
                color="black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # zero effect
  geom_hline(yintercept=0, linetype = 2) +
  
  # p-value
  annotate(geom = "text",
           label = m1_coeff_dt$p_pretty,
           x = 1,
           y = 7.5) +
  
  # aesthetics
  scale_y_continuous(position="right") +
  scale_x_discrete(labels = "MR\neffect") +
  ylab(expression(Growth~(mm^3/day))) +
  coord_flip(ylim = y_lims) + 
  theme_pubr() +
  theme(axis.title.y = element_blank()) +
  
  NULL

#fig1f_gg_effect
```

Combining the response and effects plots into single plot

```{r continuous-x-long-plot-2, echo=TRUE, eval=FALSE, message=FALSE}
fig1f_gg <- plot_grid(fig1f_gg_effect,
                      fig1f_gg_response,
                      nrow=2,
                      align = "v",
                      axis = "lr",
                      rel_heights = c(0.4,1))
fig1f_gg
```

## Try it
### A prediction model from the literature

The data come from the top, middle plot of Figure 1e of

[Parker, B.L., Calkin, A.C., Seldin, M.M., Keating, M.F., Tarling, E.J., Yang, P., Moody, S.C., Liu, Y., Zerenturk, E.J., Needham, E.J. and Miller, M.L., 2019. An integrative systems genetic analysis of mammalian lipid metabolism. Nature, 567(7747), pp.187-193.](https://idp.nature.com/authorize/casa?redirect_uri=https://www.nature.com/articles/s41586-019-0984-y&casa_token=9rpBLACc40cAAAAA:BzRKQHYdU9DOrzBS0f0lwrbp07MlfVqqL4yuyedE5b_xZ8geHuUCPFIDE3yQUBO-ibQLCrCI4t84aO8){target="_blank"}

[Public source](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6656374/){target="_blank"}

[Source data](https://www.nature.com/articles/s41586-019-0984-y#Sec35){target="_blank"}

The researchers built prediction models from a hybrid mouse diversity panel (HMDP) to predict liver lipid levels from measured plasma lipid levels in mice *and* in humans. The value of the predictor ($X$) variable for an individual is not a measured value of a single plasma lipid but the predicted value, or **score**, from the prediction model based on an entire panel of lipid measurements in that individual. The $Y$ variable for the individual is the total measured level across a family of lipids (ceramides, triacylglycerols, diacylglycerols) in the liver of the individual. The question is, how well does the prediction score predict the actual liver level?

1. Use the data for TG (triacylglycerols) (the top, middle plot of Figure 1e). The column D is the "score" from the prediction model using plasma lipid levels. This is the $X$ variable (the column header is "fitted.total.Cer"). The column E is the total measured liver TG, so is the $Y$ variable.
2. Fit the linear model y ~ x.
3. Create a publication-quality plot of liver TG (*Y*-axis) against score (*X*-axis) -- what the researchers labeled "fitted.total.Cer". Include $R^2$ on the plot.
4. Advanced -- add a 95% prediction interval to the plot (the template code for this is in the Hidden Code section for efig3d)
5. Create a table of expected liver TG and 95% prediction interval of liver TG of the score values (13.5, 14, 14.5, 15, 15.5, 16).
6. Comment on the predictability of liver TG using the plasma scores.

```{r continuous-x-import-prediction, echo=FALSE, eval=FALSE}
#1
data_from <- "An integrative systems genetic analysis of mammalian lipid metabolism"
file_name <- "41586_2019_984_MOESM4_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)
fig1e <- read_excel(file_path,
                    sheet = "1e") %>% # import
  clean_names() %>% # clean the column names
  data.table() # convert to data.table

# View(fig1e)

m1 <- lm(total_tg ~ fitted_total_cer_4, data = fig1e)

#2
summary(m1)$r.squared

#3
gg3 <- ggscatter(data = fig1e,
          x = "fitted_total_cer_4",
          y = "total_tg",
          # color = "black",
          # fill = "red",
          shape = 21,
          size = 3,
          add = "reg.line",
          add.params = list(color = "steelblue",
                            fill = "lightgray"),
          conf.int = FALSE,
          xlab = "Predicted",
          ylab = "Liver Total TG") +
  
  stat_cor(aes(label = paste(..rr.label.., ..p.label.., sep = "~`,`~")),
           size = 4,
           label.y = 17) +
  NULL
gg3

#4
# get x for drawing slope
minx <- min(fig1e[,fitted_total_cer_4])
maxx <- max(fig1e[,fitted_total_cer_4])
new_x <- seq(minx, maxx, length.out = 20)
new_data <- data.table(fitted_total_cer_4 = new_x)
new_data[, yhat := predict(m1, newdata = new_data)]
new_data[, pred_lwr := predict(m1, 
                           se.fit = TRUE,
                           interval = "prediction",
                           newdata = new_data)$fit[, "lwr"]]
new_data[, pred_upr := predict(m1, 
                           se.fit = TRUE,
                           interval = "prediction",
                           newdata = new_data)$fit[, "upr"]]

gg4 <- gg3 +
  geom_ribbon(data = new_data,
              aes(ymin = pred_lwr,
                  ymax = pred_upr,
                  y = yhat,
                  fill = "band"),
              fill = "gray",
              alpha = 0.3)
gg4

#5
tg_predictions <- data.table(
  fitted_total_cer_4 = c(13.5, 14, 14.5, 15, 15.5, 16))
tg_predictions <- cbind(
  tg_predictions,
  predict(m1,
          newdata = tg_predictions,
          interval = "prediction"))

median_fitted <- median(fig1e$fitted_total_cer_4)
predict(m1,
        newdata = data.frame(fitted_total_cer_4 = median_fitted),
        interval = "prediction")
range(fig1e$total_tg)

# The model is not going to be very useful for predicting true liver TG. This is evidenced by the wide prediction interval. For example, at the low end of the fitted ceramide level, at a predicted score of 14.5, which is near the middle of the predicted range, 95% of actual values should fall within the range 12.8 - 16.2, which nearly covers the entire range of true liver TG values.
```
`

## Intuition pumps

### Correlation and $R^2

In the code below

1. change the value of `n` to explore effect on the variability of the estimate. Look at this variability, and the magnitude of the SE, and the magnitude of the p-value.
2. change the value of `beta_1` to explore effect on what different slopes and correlations look like. Notice that if the variance is fixed (as in this simulation) the expected slope and expected correlation are equal. (Because I've fixed the variance in this simple simulation, this code will fail if abs(beta_1) >= 1).

```{r, message=FALSE}
n <- 100 # choose between 3 and 10^5
beta_1 <- 0.6 # choose a value between -0.99 and 0.99
x <- rnorm(n)
y <- beta_1*x + sqrt(1-beta_1^2)*rnorm(n)
m1 <- lm(y ~ x)
slope <- paste("b_1: ", round(coef(m1)[2], 3))
se <- paste("SE: ", round(coef(summary(m1))[2,2], 3))
r <- paste("r: ", round(cor(x,y), 3))
r2 <- paste("R^2: ", round(summary(m1)$r.squared, 3))
ggscatter(data = data.table(x=x, y=y), x = "x", y = "y",
          add = "reg.line") +
  ggtitle(label = paste(slope,se,r,r2,sep=";  "))

```


<!--chapter:end:chapters/22-continuous_X.Rmd-->

# Linear models with a single, categorical *X* ("t-tests" and "ANOVA") {#oneway}

```{r categorical-x-setup, echo=FALSE, warning=FALSE, message=FALSE}
library(here)
library(janitor)
library(readxl)
library(data.table)
library(stringr)

# analysis packages
library(nlme) # gls
library(lmerTest) # lmm
library(emmeans)
library(car) # qqplot, spreadlevel, Anova
library(afex) #aov4
library(lmPerm)
library(glmmTMB)

# graphing packages
library(ggplot2) # ggplot environment
library(ggpubr) # publication ready plots
library(ggforce) # jitter
library(ggsci) # color palettes
library(cowplot) # combine plots
library(lazyWeave) # pretty pvalues
library(broom)
library(knitr)
library(kableExtra)

here <- here::here

plot_path <- here("R/ggplot_the_model.R")
source(plot_path)

clean_names <- janitor::clean_names

data_folder <- "data"

```

## A linear model with a single, categorical *X* variable estimates the effects of the levels of *X* on the response.

To introduce a linear model with a single, categorical $X$ variable, I'll use data from a set of experiments designed to measure the effect of the lipid 12,13-diHOME on brown adipose tissue (BAT) thermoregulation and the mechanism of this effect.

[Lynes, M.D., Leiria, L.O., Lundh, M., Bartelt, A., Shamsi, F., Huang, T.L., Takahashi, H., Hirshman, M.F., Schlein, C., Lee, A. and Baer, L.A., 2017. The cold-induced lipokine 12, 13-diHOME promotes fatty acid transport into brown adipose tissue. Nature medicine, 23(5), pp.631-637.](https://www.nature.com/articles/nm.4297){target="_blank"}

[Public source](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5699924/pdf/nihms916046.pdf){target="_blank"}

[Data source](https://www.nature.com/articles/nm.4297#Sec14){target="_blank"}

Download the source data files and move to a new folder named "The cold-induced lipokine 12,13-diHOME promotes fatty acid transport into brown adipose tissue".

Cold temperature and the neurotransmitter/hormone norepinephrine are known to stimulate increased thermogenesis in BAT cells. In this project,  the researchers probed the question "what is the pathway that mediates the effect of cold-exposure on BAT thermogenesis?". In the "discovery" component of this project, the researchers measured plasma levels of 88 lipids with known signaling properties in humans exposed to one hour of both normal (20 °C) and cold temperature (14 °C) temperature. Of the 88 lipids, 12,13-diHOME had the largest response to the cold treatment. The researchers followed this up with experiments on mice.

### Example 1 -- two treatment levels ("groups")
Let's start with the experiment in Figure 3d, which was designed to measure the effect of 12,13-diHOME on plasma triglyceride level. If 12,13-diHOME stimulates BAT activity, then levels in the 12,13-diHOME mice should be less than levels in the control mice.

####  Step 1 -- Understand the experiment design

**response variable**: $\texttt{serum_tg}$, a continuous variable.

**treatment variable**: $\texttt{treatment}$, with levels: "Vehicle", "12,13-diHOME" (the control or "Vehicle" mice were injected with saline). Coded as a factor.

**design**: single, categorical *X*

#### Step 2 -- import

Open the data and, if necessary, wrangle into an analyzable format. The script to import these data is in the section Hidden code below.

```{r, echo=FALSE}
data_from <- "The cold-induced lipokine 12,13-diHOME promotes fatty acid transport into brown adipose tissue"
file_name <- "41591_2017_BFnm4297_MOESM3_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)
```

```{r, echo=FALSE}
# ignore the column with animal ID. Based on methods, I am inferring
# that the six mice in vehicle group *are different* from the
# six mice in the 1213 group.
range_3d <- "B3:C9" # assume independent
sheet_3d <- "Figure 3d"
col_names_3d <- c("Vehicle", "1213")
treatment_levels <- c("Vehicle", "12,13-diHOME")
fig_3d <- read_excel(file_path,
                     sheet = sheet_3d,
                     range = range_3d,
                     col_names = TRUE) %>%
  data.table() %>%
  melt(measure.vars = col_names_3d,
       variable.name = "treatment",
       value.name = "serum_tg")

# change group name of "1213"
fig_3d[treatment == "1213", treatment := "12,13-diHOME"]

# make treatment a factor with the order in "treatment_levels"
fig_3d[, treatment := factor(treatment, treatment_levels)]

#View(fig_3d)
```

#### Step 3 -- inspect the data

The second step is to examine the data to

1. get a sense of sample size and balance
2. check for biologically implausible outliers that suggest measurement failure, or transcription error (from a notebook, not in a cell)
3. assess outliers for outlier strategy or robust analysis
4. assess reasonable distributions and models for analysis.

```{r, echo=TRUE}
qplot(x = treatment, y = serum_tg, data = fig_3d)
```

There are no obviously implausible data points. A normal distribution is a good, reasonable start. This can be checked more thoroughly after fitting the model.

#### Step 4 -- fit the model

```{r, echo=TRUE}
fig3d_m1 <- lm(serum_tg ~ treatment, data = fig_3d)
```

#### Step 5 -- check the model

```{r, echo = TRUE}
set.seed(1)
qqPlot(fig3d_m1, id=FALSE)
```

The Q-Q plot indicates the distribution of residuals is well within that expected for a normal sample and there is no cause for concern with inference.

```{r, echo = TRUE, warning=FALSE}
spreadLevelPlot(fig3d_m1, id=FALSE)
```

The spread-location plot shows no conspicuous trend in how the spread changes with the conditonal mean. There is no cause for concern with inference.

#### Step 6 -- inference
##### coefficient table

```{r fig3d_m1_coef, echo=TRUE}
fig3d_m1_coef <- cbind(coef(summary(fig3d_m1)),
                        confint(fig3d_m1))
kable(fig3d_m1_coef, digits = c(1,2,1,4,1,1)) %>%
  kable_styling()
```

##### emmeans table

```{r, echo=TRUE}
fig3d_m1_emm <- emmeans(fig3d_m1, specs = "treatment")
kable(fig3d_m1_emm, digits = c(1,1,2,0,1,1)) %>%
  kable_styling()
```

##### contrasts table

```{r, echo=TRUE}
fig3d_m1_pairs <- contrast(fig3d_m1_emm,
                            method = "revpairwise") %>%
  summary(infer = TRUE)

kable(fig3d_m1_pairs, digits = c(1,1,1,0,1,1,2,4)) %>%
  kable_styling()
```

#### Step 6 -- plot the model

```{r ggplot_the_model, echo = TRUE}
ggplot_the_model(
  fig3d_m1,
  fig3d_m1_emm,
  fig3d_m1_pairs,
  legend_position = "none",
  y_label = "Serum TG (µg/dL)",
  effect_label = "Effects (µg/dL)",
  palette = pal_okabe_ito_blue,
  rel_heights = c(0.5,1)
)
```

#### Step 7 -- report the model results

1. Different ways of reporting the results without making claims that are not evidenced by the statistical analysis

"The estimated average effect of 12,13-diHOME on serum TG is 7.17 µg/dL (95% CI: -12.4, -1.9, $p = 0.012$)."

"Mean serum TG in mice with 12,13-diHOME (35.5 µg/dL, 95% CI: 31.7, 39.2) was 7.17 µg/dL less (95% CI: -12.4, -1.9, p = 0.012) than mean serum TG in control mice (42.6 µg/dL, 95% CI: 38.9, 46.3)."

2. Problematic reports

"12,13-diHOME significantly reduced serum TG ($p = 0.012$)" (1.  In English, the current use of "significant" means "large" or "important". Only knowledge of the physiological consequences of TG reduction over the range of the CI can be use as evidence of importance. 2. The statistics do not provide evidence that it is the serum TG that caused this reduction. While 7.17 µg/dL is the difference, some or all of the effect could be due to, in general, sampling, execution errors such as non-random treatment assignment, clustered caging, clustered application of assays, or non-blinded measurement of the response.

### Understanding the analysis with two treatment levels

The variable $\texttt{treatment}$ in the Figure 3d mouse experiment, is a single, categorical $X$ variable. In a linear model, categorical variables are called **factors**. $\texttt{treatment}$ can take two different values, "Vehicle" and "12,13-diHOME". The different values in a factor are the **factor levels** (or just "levels"). "Levels" is a strange usage of this word; a less formal name for levels is "groups". In a **Nominal** categorical factor, the levels have no units and are unordered, even if the variable is based on a numeric measurement. For example, I might design an experiment in which mice are randomly assigned to one of three treatments: one hour at 14 °C, one hour at 18 °C, or one hour at 26 °C. If I model this treatment as a nominal categorical factor, then I simply have three levels. While I would certainly choose to arrange these levels in a meaningful way in a plot, for the analysis itself, these levels have no units and there is no order. **Ordinal** categorical factors have levels that are ordered but there is no information on relative distance. The treatment at 18 °C is not more similar to 14 °C than to 26 °C. Nominal categorical factors is the default in R and how all factors are analyzed in this text.

#### Linear models are regression models

The linear model fit to the serum TG data is

\begin{align}
serum\_tg &= treatment + \varepsilon\\
\varepsilon &\sim N(0, \sigma^2)
(\#eq:lm-model-example-1)
\end{align}

This notation is potentially confusing because the variable $\texttt{treatment}$ is a factor containing the words "Vehicle" and "12,13-diHOME" and not numbers. The linear model in \@ref(eq:lm-model-example-1) can be specified using notation for a regression model using

\begin{align}
serum\_tg &= \beta_0 + \beta_1 treatment_{12,13-diHOME} + \varepsilon\\
\varepsilon &\sim N(0, \sigma^2)
(\#eq:reg-model-example-1)
\end{align}

Model \@ref(eq:reg-model-example-1) is a regression model where $treatment_{12,13-diHOME}$ is not the variable $\texttt{treatment}$, containing the words "Vehicle" or "12,13-diHOME" but a numeric variable that indicates membership in the group "12,13-diHOME". This variable contains the number 1 if the mouse belongs to "12,13-diHOME" and the number 0 if the mouse doesn't belong to "12,13-diHOME". $treatment_{12,13-diHOME}$ is known as an **indicator variable** because it indicates group membership. There are several ways of coding indicator variables and the way described here is called **dummy** or treatment coding. Dummy-coded indicator variables are sometimes called **dummy variables**.

The `lm` function creates indicator variables under the table, in something called the **model matrix**. 

```{r cat-x-dummy-coding, echo=TRUE}
X <- model.matrix(~ treatment, data = fig_3d)
N <- nrow(X)
X[1:N,]
```

The columns of the model matrix are the names of the **model terms** in the fit model. R names dummy variables by combining the names of the factor and the name of the level within the factor. So the $X$ variable that R creates in the model matrix for the fit linear model in model \@ref(eq:reg-model-example-1) is $treatment12,13-diHOME$. You can see these names as terms in the coefficient table of the fit model.

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
There are alternatives to dummy coding for creating indicator variables. Dummy coding is the default in R and it makes sense when thinking about experimental data with an obvious control level. I also like the interpretation of a "interaction effect" using Dummy coding. The classical coding for ANOVA is deviation effect coding, which creates coefficients that are deviations from the grand mean. In contrast to R, Deviation coding is the default in many statistical software packages including SAS, SPSS, and JMP. The method of coding can make a difference in an ANOVA table. Watch out for this -- I've found several published papers where the researchers used the default dummy coding but interpreted the ANOVA table as if they had used deviation coding. This is both getting ahead of ourselves and somewhat moot, because I don't advocate reporting ANOVA tables.
</div>

<br>

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
Recall from stats 101 that the slope of $X$ in the model $Y = b_0 + b_1 X$ is $b_1 = \frac{\textrm{COV}(X,Y)}{\textrm{VAR}(X)}$. This can be generalized using the equation

\begin{equation}
\mathbf{b} = (\mathbf{X}^\top \mathbf{X})^{-1} \mathbf{X}^\top \mathbf{y}
\end{equation}

where $\mathbf{X}$ is the model matrix containing a column for an intercept, columns for all indicator variables, and columns for all numeric covariates. $\mathbf{b}$ is a vector containing the model coefficients, including the intercept in the first element. The first part of the RHS ($\mathbf{X}^\top \mathbf{X}$) is a matrix of the "sums of squares and cross-products" of the columns of $\mathbf{X}$. Dividing each element of this matrix by $N-1$ gives us the covariance matrix of the $\mathbf{X}$, which contains the variances of the $X$ columns along the diagonal, so this component has the role of the denominator in the stats 101 equation. Matrix algebra doesn't do division, so the inverse of this matrix is multiplied by the second part. The second part or the RHS ($\mathbf{X}^\top \mathbf{y}$) is a vector containing the cross-products of each column of $\mathbf{X}$ with $\mathbf{y}$. Dividing each element of this vector by $N-1$ gives us the covariances of each $X$ with $y$, so this component has the role of the numerator in the stats 101 equation.
</div>

<br>

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
Self-learning. `lm` fits the model `y ~ X` where X is the model matrix. Fit the model using the standard formula and the model using the model matrix. The coefficient table should be the same.

`m1 <- lm(serum_tg ~ treatment, data = fig_3d)`

`X <- model.matrix(~ treatment, data = fig_3d)`

`m2 <- lm(serum_tg ~ X, data = fig_3d)`

`coef(summary(m1))`

`coef(summary(m2))`

</div>
 
```{r echo=FALSE, eval=FALSE}
m1 <- lm(serum_tg ~ treatment, data = fig_3d)
X <- model.matrix(~ treatment, data = fig_3d)
m2 <- lm(serum_tg ~ X, data = fig_3d)
coef(summary(m1))
coef(summary(m2))
```

#### The Estimates in the coefficient table are estimates of the parameters of the linear model fit to the data.

```{r cat-x-fig3d_m1_coef-kable, echo=FALSE}
fig3d_m1_coef %>%
  kable(digits = c(1,2,1,4,1,1)) %>%
  kable_styling()
```

The row names of the coefficient table are the column names of the model matrix. These are the model terms. There are two terms (two rows) because there are two parameters in the regression model \@ref(eq:reg-model-example-1). The values in the column $\texttt{Estimate}$ in the coefficient table are the estimates of the regression parameters $\beta_0$ and $\beta_1$. These estimates are the coefficients of the fit model, $b_0$ and $b_1$.

#### The coefficients of a linear model using dummy coding have a useful interpretation

```{r example1-coef-understanding, echo=FALSE}
new_table <- data.table(
  coefficient = c("$b_0$", "$b_1$"),
  parameter = c("$\\beta_0$", "$\\beta_1$"),
  "model term" = row.names(fig3d_m1_coef),
  interpretation = c("$\\overline{Vehicle}$",
                     "$\\overline{12,13\\;diHOME} - \\overline{Vehicle}$")
)
new_table %>%
  kable(escape = FALSE,
        caption = "Understanding model coefficients of a linear model with a single treatment variable with two groups. The means in the interpretation column are conditional means.") %>%
  kable_styling()
```

It is important to understand the interpretation of the coefficients of the fit linear model \@ref(eq:lm-model-example-1) (Table \@ref(tab:example1-coef-understanding)).

1. The coefficient $b_0$ is the is the conditional mean of the response for the reference level, which is "Vehicle". Remember that a conditional mean is the mean of a group that all have the same value for one or more $X$ variables.
2. The coefficient $b_1$ is the difference between the conditional means of the "12,13-diHOME" level and the reference ("Vehicle") level:

\begin{equation}
\mathrm{E}[serum\_tg|treatment = \texttt{"12,13-diHOME"}] - \mathrm{E}[serum\_tg|treatment = \texttt{"Vehicle"}]
\end{equation}

Because there are no additional covariates in model, this difference is equal to the difference between the sample means $\bar{Y}_{12,13-diHOME} - \bar{Y}_{Vehicle}$. The *direction* of this difference is important -- it is non-reference level minus the reference level.

The estimate $b_1$ is the **effect** that we are interested in. Specifically, it is the measured effect of 12,13-diHOME on serum TG. When we inject 12,13-diHOME, we find the mean serum TG decreased by -7.2 µg/dL relative to the mean serum TG in the mice that were injected with saline. Importantly, the reference level is not a property of an experiment but is set by whomever is analyzing the data. Since the non-reference estimates are differences in means, it often makes sense to set the "control" treatment level as the reference level.

Many beginners mistakenly memorize the coefficient $b_1$ to equal the mean of the non-reference group ("12,13-diHOME"). Don't do this. In a regression model, only $b_0$ is a mean. The coefficient $b_1$ in model \@ref(eq:reg-model-example-1) is a difference in means.

```{r cat-x-coef-fig, echo=FALSE, fig.cap="What the coefficients of a linear model with a single categorical X mean. The means of the two treatment levels for the serum TG data are shown with the large, filled circles and the dashed lines. The intercept ($b_0$) is the mean of the reference treatment level (\"Vehicle\"). The coefficient $b_1$ is the difference between the treatment level's mean and the reference mean. As with a linear model with a continuous $X$, the coefficient $b_1$ is an effect."}

b <- fig3d_m1_coef[, "Estimate"]
b0 <- b[1]
x_adj <- 0.25
y_adj <- 0

gg <- ggplot_the_response(
  fig3d_m1,
  fig3d_m1_emm,
  fig3d_m1_pairs,
  legend_position = "none",
  palette = pal_okabe_ito_blue,
  y_label = "Serum TG (µg/dL)",
) +
  
  # b0 dotted line
  geom_segment(aes(x = 0.8,
                   y = b[1],
                   xend = 2.25,
                   yend = b[1]),
               linetype = 2,
               color = "black") +
  # b0 arrow
  geom_segment(aes(x = 0.65,
                   y = b[1] + y_adj,
                   xend = 0.4,
                   yend = b[1]),
               color = "black",
               arrow = arrow(length  =  unit(0.05, "npc"),
                             ends = "last",type = "open")) +
  annotate("text",
           label = "b[0]",
           parse = TRUE,
           x = 0.725,
           y = (b[1]) + y_adj,
           size = 8) +
  
  # b1 dotted line
  geom_segment(aes(x = 0.8,
                   y = b[1] + b[2],
                   xend = 2.25,
                   yend = b[1] + b[2]),
               linetype = 2,
               color = "black") +
  
  # b1 arrow
  geom_segment(aes(x = 2 + x_adj,
                   y = b[1] + b[2] + 0.3,
                   xend = 2 + x_adj,
                   yend = b[1]),
               color = "black",
               arrow = arrow(length  =  unit(0.05, "npc"),
                             ends = "both",type = "open")) +
  annotate("text",
           label = "b[1]",
           parse = TRUE,
           x = 1.9 + x_adj,
           y = (b[1] + 0.5*b[2]),
           size = 8) +
  NULL


gg
```

A geometric interpretation of the coefficients is illustrated in Figure \@ref(fig:cat-x-coef-fig). $b_0$ is the conditional mean of the reference level ("Vehicle") and is an estimate of $\beta_0$, the true, conditional mean of the population. $b_1$ is the difference in the conditional means of the first non-reference level ("12,13-diHOME") and the reference level ("Vehicle") and is an estimate of $\beta_1$, the true difference in the conditional means of the population with and without the treatment 12,13-diHOME. 

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
tl;dr. What is a population? In the experimental biology examples in this text, we might consider the population as a very idealized, infinitely large set of mice, or fish, or fruit flies, or communities from which our sample is a reasonably representative subset. For the experiments in the 12,13-diHOME study, the population might be conceived of as the hypothetical, infinitely large set of 12-week-old, male, C57BL/6J mice, raised in the mouse facility at Joslin Diabetes Center. An even more abstract way to way to think about what the population could be is the infinitely large set of values that could generated by the linear model.
</div> 

#### Better know the coefficient table

```{r}
fig3d_m1_coef %>%
  kable(digits = c(1,2,1,4,1,1)) %>%
  kable_styling()
```

1. The $\texttt{(Intercept)}$ row contains the statistics for $b_0$ (the estimate of $\beta_0$). Remember that $b_0$ is the conditional mean of the reference treatment ("Vehicle").
2. The $\texttt{treatment12,13-diHOME}$ row contains the statistics for $b_1$ (the estimate of $\beta_1$). Remember that $b_1$ is the difference in conditional means of the groups "12,13-diHOME" and "Vehicle".
3. The column $\texttt{Estimate}$ contains the model coefficients, which are estimates of the parameters.
4. The column $\texttt{Std. Error}$ contains the model SEs of the coefficients. The SE of $\texttt{(Intercept)}$ is a standard error of a mean (SEM). The SE of $\texttt{treatment12,13-diHOME}$ is a standard error of a difference (SED).
5. The column $\texttt{t value}$ contains the test statistic of the coefficients. This value is the ratio $\frac{Estimate}{SE}$. For this model, we are only interested in the test statistic for $b_1$. Effectively, we will never be interested in the test statistic for $b_0$ because the mean of a group will never be zero.
6. The column $\texttt{Pr(>|t|)}$ contains the *p*-values for the test statistic of the coefficients. For this model, and all models in this text, we are only interested in the *p*-value for the non-intercept coefficients. 
7. The columns $\texttt{2.5 %}$ and $\texttt{97.5 %}$ contain the lower and upper limits of the 95% confidence interval of the estimate.

#### The emmeans table is a table of modeled means and inferential statistics

```{r fig3d_m1_emm_kable, echo=FALSE}
kable(summary(fig3d_m1_emm),
      digits = c(1,1,2,1,1,1),
      caption = "Estimated marginal means table for model fig3d_m1.") %>%
  kable_styling()
```

The table of marginal means for the model fit to the fig3d serum TG data (Model \@ref(eq:lm-model-example-1)) is given in Table \@ref(tab:fig3d_m1_emm_kable). The table of marginal means gives the **modeled* mean, standard error and confidence interval for all specified groups. There is no test-statistic with a *p*-value because there is no significance test. In this text, I'll refer to this table as the "emmeans table", since it is the output from the `emmeans` function  ("em" is the abbreviation for "estimated marginal"). I'll use "modeled means" to refer to the means themselves as these are the estimate of means from a fit linear model.

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
A **marginal mean** is the mean over a set of conditional means. For example, if a treatment factor has three levels, the conditional means are the means for each level and the marginal mean is the mean of the three means. Or, if the conditional means are the expected values given a continous covariate, the marginal mean is the expected value at the mean of covariate. The specified emmeans table of the fig3d data is not too exciting because it simply contains the conditional means -- the values are not marginalized over any $X$. Because the emmeans table contains different sorts of means (conditional, marginal, adjusted), this text will generally refer to the means in this table as "modeled means".
</div>

```{r cat-x-emmeans-table, echo = FALSE}
kable(summary(fig3d_m1_emm),
      digits = 5,
      caption = "The emmeans table contains modeled means, SE, and CIs.") %>%
  kable_styling()
```


```{r cat-x-summary-table, echo = FALSE}
# summary table
summary_table <- fig_3d[, .(mean = mean(serum_tg),
                            SE = sd(serum_tg)/sqrt(.N)),
                        by = treatment]
df <- nrow(fig_3d) - 2
summary_table[, lower.CL := mean + SE*qt(0.025, df)]
summary_table[, upper.CL := mean + SE*qt(0.975, df)]

kable(summary_table,
      digits = 5,
      caption = "A summary table contains sampled means, SE, and CIs.") %>%
  kable_styling()
```

1. The means in the emmeans table are modeled means. Here, and for many linear models, these will be equal to the sampled means. This will not be the case in models with one or more continuous covariates, or if marginal means are explicitly specified. 
2. Unlike the modeled means, the modeled standard errors and confidence intervals will, effectively, never equal sample standard errors and confidence intervals. In many models, plots using sample statistics can lead to very deceiving inference of differences between groups. This is why this text advocates plotting the model -- using modeled means and confidence intervals in plots.

It is exceptionally important to understand the difference between the means, SEs, and CIs in the emmeans table and the statistics of the same name in a summary table of the data.

1. The statistics in a summary table are **sampled** means, SEs, and CIs -- these statistics are computed for the group using *only the data in the group*.
2. To understand modeled SEs and CIs, recall that the standard error of a sample mean is $\frac{s}{\sqrt{n}}$, where $s$ is the sample standard deviation and $n$ is the sample size in the group. The computation of the SE in the emmeans table uses the same equation, except the numerator is not the sample standard deviation of the group but the model standard deviation $\hat{\sigma}$. As with the sample SE, the denominator for the modeled SE is the sample size $n$ for the group. Since the numerator of the modeled SE is the same for all groups, the modeled SE will be the same in all groups that have the same sample size, as seen in the marginal means table for the model fit to the Figure 3d data. This is not true for sampled SEs, since sampled standard deviations will always differ.

It may seem odd to use a common standard deviation in the computation of the modeled SEs. It is not. Remember that an assumption of the linear model is homogeneity of variances -- that all residuals $e_i$ are drawn from the same distribution ($N(0, \sigma^2)$) (a "single hat") regardless of group. The model standard deviation $\hat{\sigma}$ is the estimate of the square root of the variance of this distribution. Given this interpretation, it is useful to think of each sample standard deviation as an estimate of $\sigma$ (the linear model assumes that all differences among the sample standard deviations are due entirely to sampling). The model standard deviation is a more precise estimate of $\sigma$ since it is computed from a larger sample (all $N$ residuals).

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
The model standard deviation is called the "pooled" standard deviation in the ANOVA literature and is computed as a sample-size weighted average of the sample standard deviations.
</div>

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
The modeled standard error of the mean uses the estimate of $\sigma$ from the fit model. This estimate is

\begin{equation}
\hat{\sigma} = \sqrt{\frac{\sum{(y_i - \hat{y}_i)^2}}{df}}
\end{equation}

Create a code chunk that computes this. Recall that $(y_i - \hat{y}_i)$ is the set of residuals from the model, which can be extracted using `residuals(fit)` where "fit" is the fit model object. $df$ is the model degrees of freedom, which is $N-k$, where $N$ is the total sample size and $k$ is the number of parameters that are fit. This makes sense -- for the sample variance there is one parameter that is fit, the mean of the group. In model fig3d_m1, there are two parameters that are fit, the intercept and the coefficient of treatment12,13-diHOME.
</div>

```{r cat-x-explore-1, echo=FALSE, eval=FALSE}
summary(fig3d_m1)$sigma
df <- nrow(fig_3d) - 2
sqrt(sum(residuals(fig3d_m1)^2)/df)
```

#### Estimates of the effects are in the contrasts table

```{r fig3d_m1_pairs, echo = FALSE}
kable(fig3d_m1_pairs, digits = c(1,1,2,1,1,1,1,3)) %>%
  kable_styling()
```

1. This table is important for reporting treatment effects and CIs and for plotting the model. A contrast is a difference in means. With only two treatment levels, the table of contrasts doesn't give any more information than the coefficient table -- the single contrast is the coefficient $b_1$ in the coefficient table. Nevertheless, I advocate computing this table to stay consistent and because the script (or function) to plot the model uses this table and not the coefficient table.
2. The value in the column $\texttt{Estimate}$ is the mean of the non-reference group ("12,13-diHOME") minus the mean of the reference group ("Vehicle").
3. The value in the "SE" column is the standard error of the difference (SED), specifically the difference in the estimate column. This SE is computed using the model standard deviation $\sigma$.
4. The values in the "lower.CL" and "upper.CL" columns are the bounds of the 95% confidence interval of the estimate. Remember (from Chapter \@ref(variability)) to think of this interval as containing potential values of the true parameter (the true difference in means between the two groups) that are reasonably compatible with the data. Don't think of the interval as having 95% probability of containing the true effect. Remember that a confidence interval applies to the procedure and not a parameter -- 95% of the CIs from hypothetical, replicate experiments that meet all the assumptions used to compute the CI will include the true effect.
5. The columns "t.ratio" and "p.value" contains the *t* and *p* values of the significance (not hypothesis!) test of the estimate. The *t*-statistic is the ratio of the estimate to the SE of the estimate (use the console to confirm this given the values in the table). It is a signal (the estimate) to noise (SE of the estimate) ratio. The *p*-value is the probability of sampling from normal distribution with the observed standard deviation, randomly assigning the sampled values to either "Vehicle" or "12,13-diHOME", fitting the linear model, and observing a *t*-value as or more extreme than the observed *t*. A very small *p*-value is consistent with the experiment "not sampling from distributions with the same mean" -- meaning that adding a treatment affects the mean of the distribution. This is the logic used to infer a treatment effect. Unfortunately, it is also consistent with the experiment not approximating other conditions of the model, including non-random assignment, non-independence, non-normal conditional responses, and variance heterogeneity. It is up to the rigorous researcher to be sure that these other model conditions are approximated or "good enough" to use the *p*-value to infer a treatment effect on the mean.

#### *t* and *p* from the contrasts table -- when there are only two levels in $X$ -- are the same as *t* and *p* from a *t*-test

Compare

coefficient table:

```{r echo = TRUE}
m1 <- lm(serum_tg ~ treatment, data = fig_3d)
coef(summary(m1)) %>%
  kable() %>%
  kable_styling()
```
contrast table:

```{r echo = TRUE}
emmeans(m1, specs = "treatment") %>%
  contrast(method = "revpairwise") %>%
  kable() %>%
  kable_styling()
```

t-test:

```{r}
m2 <- t.test(fig_3d[treatment == "12,13-diHOME", serum_tg],
             fig_3d[treatment == "Vehicle", serum_tg],
             var.equal = TRUE)
glance(m2) %>% # glance is from the broom package
  kable() %>%
  kable_styling()
```

Notes

1. The default `t.test` in R is the Welch t-test for heterogenous variance. To compute the Student t-test, use `var.equal = TRUE`.
2. The "statistic" in the t-test output contains the *t*-value of the *t*-test. It is precisely the same as the *t*-statistic in the coefficient table and the contrast table.
3. The *p*-values in all three tables are precisely the same.

The *t* and *p* values for the *t*-test are the same as those for the linear model, because the *t*-test is a specific case of the linear model. Reasons to abandon classic *t*-tests and learn the linear modeling strategy include

1. A linear modeling strategy encourages researchers to think about the effect and uncertainty in the effect and not just a *p*-value.
2. The linear model is nearly infinitely flexible and expandible while the *t*-test is limited to a few variations.

There is rarely a reason to ever use the `t.test()` function. Throw the function away. Ignore web pages that teach you to use it. The *t*-test is easy to learn, which encourages its overuse. If your only tool is a t-test, every problem looks like a comparison between two-means.

### Example 2 -- three treatment levels ("groups")
#### Understand the experiment design
The data come from the experiment reported in Figure 2a of the 12,13-diHOME article described above. This experiment was designed to probe the hypothesis that 12,13-diHOME is a mediator of known stimulators of increased BAT activity (exposure to cold temperature and sympathetic nervous system activation). Mice were assigned to control (30 °C), one-hour exposure to 4 °C, or 30 minute norepinephrine (NE) treatment level (NE is the neurotransmitter of the sympathetic neurons targeting peripheral tissues).

**response variable**: $\texttt{diHOME}$, the serum concentration of 12,13-diHOME. a continuous variable.

**treatment variable**: $\texttt{treatment}$, with levels: "Control", "1 hour cold", "30 min NE". Coded as a factor.

1. "Control" -- the negative control. We expect diHOME to be low relative to the two treated conditions.
2. "1 hour cold" -- focal treatment 1. If diHOME is a mediator, then response should be relatively high.
3. "30 min NE" -- focal treatment 2. If diHOME is a mediator, then response should be relatively high.

**planned comparisons**

1. "1 hour cold" - "Control" -- If diHOME is a mediator of cold, then difference should be positive.
2. "30 min NE" - "Control" -- If diHOME is a mediator of NE, then difference should be positive.

**design**: single, categorical *X*

```{r cat-x-import-fig2a, echo=FALSE, message=FALSE}
file_name <- "41591_2017_BFnm4297_MOESM2_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)

# assuming mice are independent and not same mouse used for all three treatment
melt_col_names <- paste("Animal", 1:6)
fig2a <- read_excel(file_path,
                     sheet = "Fig 2a",
                     range = "A3:G6",
                     col_names = TRUE) %>%
  data.table() %>%
  melt(measure.vars = melt_col_names,
       variable.name = "id",
       value.name = "diHOME") # cannot start a variable with number
setnames(fig2a, old = colnames(fig2a)[1], new = "treatment")

treatment_order <- c("Control", "1 hour cold", "30 min NE")
fig2a[, treatment := factor(treatment, treatment_order)] # order levels

#View(fig2a)
```

#### fit the model {#fig2a-m1}

```{r, echo=TRUE}
fig2a_m1 <- lm(diHOME ~ treatment, data = fig2a)
```

#### check the model

```{r, echo = TRUE}
set.seed(1)
qqPlot(fig2a_m1)
```

The Q-Q plot indicates potential issues at the extreme quantiles, what is called "heavy tails". The two values are the extreme values in the "30 min NE" group. This could be the result of a small sample from a response with a larger variance.

```{r, echo = TRUE, warning=FALSE}
spreadLevelPlot(fig2a_m1, id=FALSE)
```

The combination of the raw residuals and the spread-level plot suggests heterogeneity but low confidence in anything given the small sample size.

#### Inference from the model
##### coefficient table

```{r, echo=TRUE}
fig2a_m1_coef <- cbind(coef(summary(fig2a_m1)),
                        confint(fig2a_m1))
fig2a_m1_coef %>%
  kable(digits = c(1,2,1,4,1,1)) %>%
  kable_styling()
```
##### emmeans table

```{r categorical-x-fig2a-lm-emmeans, echo=TRUE}
fig2a_m1_emm <- emmeans(fig2a_m1, specs = "treatment")

fig2a_m1_emm %>%
  kable(digits = c(1,1,2,0,1,1)) %>%
  kable_styling()
```

##### contrasts table

```{r, echo=TRUE}
fig2a_m1_planned <- contrast(fig2a_m1_emm,
                           method = "trt.vs.ctrl",
                           adjust = "none",
                           level = 0.95) %>%
  summary(infer = TRUE)

fig2a_m1_planned %>%
  kable(digits = c(1,1,1,0,1,1,2,4)) %>%
  kable_styling()
```

#### plot the model

```{r}
ggplot_the_model(
  fig2a_m1,
  fig2a_m1_emm,
  fig2a_m1_planned,
  legend_position = "none",
  y_label = "12,13-diHOME (pmol/mL)",
  effect_label = "Effects (pmol/mL)",
  palette = pal_okabe_ito_blue,
  rel_heights = c(0.5,1)
)
```

#### Report the model results

Compared to control mice (12.0 pmol/mL, 95% CI: 5.4, 18.6), mean serum 12,13-diHOME in mice exposed to one-hour cold (19.2 pmol/mL, 95% CI: 11.9, 26.4) was 7.1 pmol/mL higher (95% CI: -2.7, 16.9, $p = 0.14$) while mean Serum 12,13-diHOME in mice exposed to 30 minutes NE (26.8 pmol/mL, 95% CI: 20.2, 33.3) was 14.8 pmol/mL higher (95% CI: 5.4, 24.1 $p = 0.004$).

### Understanding the analysis with three (or more) treatment levels
#### Better know the coefficient table

The fit regression model for the data in Figure 2a is

\begin{equation}
diHOME_i = b_0 + b_1 treatment_{1\_hour\_cold,i} + b_2 treatment_{30\_min\_NE,i} + e_i
(\#eq:fit-dihome)
\end{equation}

The coefficients of the model are in the $\texttt{Estimate}$ column of the coefficient table.

```{r example2-coef-understand, echo=FALSE}
fig2a_m1_coef %>%
  kable(digits = c(1,1,1,0,1,1,2,4)) %>%
  kable_styling()
```

1. The $\texttt{(Intercept)}$ row contains the statistics for $b_0$ (the estimate of $\beta_0$). Here, $b_0$ is the mean of the reference group, which is "Control".
2. The $\texttt{treatment1 hour cold}$ row contains the statistics for $b_1$ (the estimate of $\beta_1$). Here, $b_1$ is the difference $\mathrm{E}[diHOME|treatment = \texttt{"1 hour cold"}] - \mathrm{E}[diHOME|treatment = \texttt{"Control"}]$. This difference in conditional means is equal to the difference in the sample means of the two groups *for this model* because there are no additional covariates in the model.
3. The $\texttt{treatment30 min NE}$ row contains the statistics for $b_2$ (the estimate of $\beta_2$). Here, $b_2$ is the difference $\mathrm{E}[diHOME|treatment = \texttt{"30 min NE"}] - \mathrm{E}[diHOME|treatment = \texttt{"Control"}]$. Do not make the mistake in thinking that the value in $\texttt{Estimate}$ is the mean of the "30 min NE" group.
4. The number of non-intercept coefficients generalizes to any number of levels of the factor variable. If there are $k$ levels of the factor, there are $k-1$ indicator variables, each with its own coefficient ($b_1$ through $b_{k-1}$) estimating the effect of that treatment level relative to the control (if using dummy coding).

Again --  Do not make the mistake in thinking that the values in $\texttt{Estimate}$ for the $\texttt{treatment1 hour cold}$ and $\texttt{treatment30 min NE}$ rows are the *means* of the "1 hour cold" and "30 min NE" groups. These coefficients are differences in means. And, to emphasize further understanding of these coefficients, both $b_1$ and $b_2$ are "slopes". Don't visualize this as a single line from the control mean through both non-control means. as two lines. Slopes is plural -- there are two regression lines. $b_1$ is the slope of the line from the control mean to the "1 hour cold" mean. $b_2$ is the slope of the line from the control mean to the "30 min NE" mean. The numerator of each slope is the difference between that group's mean and the control mean. The denominator of each slope is 1 (because each has the value 1 when the row is assigned to that group).

Two understand the names of the model terms, it's useful to recall the order of the factor levels of $\texttt{treatment}$, which is

```{r}
levels(fig2a$treatment) 
```

Given this ordering, the `lm` function creates a regression model with an intercept column for the "Control" group (because it is first in the list), an indicator variable for the "1 hour cold" group called `treatment1 hour cold`, and an indicator variable for the "30 min NE" group called `treatment30 min NE`. We can see these model names by peeking at the model matrix of the fit model

```{r}
fig2a_m1_X <- model.matrix(fig2a_m1)

head(fig2a_m1_X)
```

The column $\texttt{treatment1 hour cold}$ is a dummy-coded indicator variable containing the number 1, if the individual is in the "1 hour cold" group, or the number 0, otherwise. The column $\texttt{treatment30 min NE}$ is a dummy-coded indicator variable containing the number 1, if the individual is in the "30 min NE" group, or the number 0, otherwise.

The model coefficients, parameters, model term, and interpretation are summarized in the following table.

```{r echo = FALSE}
new_table <- data.table(
  coefficient = c("$b_0$", "$b_1$", "$b_2$"),
  parameter = c("$\\beta_0$", "$\\beta_1$", "$\\beta_2$"),
  "model term" = row.names(fig2a_m1_coef),
  interpretation = c("$\\overline{Control}$",
                     "$\\overline{1\\;hour\\;cold} - \\overline{Control}$",
                     "$\\overline{30\\;min\\;NE} - \\overline{Control}$")
)
new_table %>%
  kable(escape = FALSE,
        caption = "Understanding model coefficients of a linear model with a single treatment variable with three groups. The means in the interpreation column are conditional means.") %>%
  kable_styling()
```

#### The emmeans table

```{r example2-emm-understand, echo=FALSE}
fig2a_m1_emm%>%
  kable(digits = c(1,1,2,0,1,1)) %>%
  kable_styling()
```

This table is important for reporting means and CIs and for plotting the model. As in example 1, the values in the column "emmean" are the sample means of each group (what you would compute if you simply computed the mean for that group). Again, this is true for this model, but is not generally true.

And again, as in example 1, the SE for each mean is *not* the sample SE but the modeled SE -- it is computed from the residuals of the model and is an estimate of $\sigma$. *These are the SEs that you should report* because it is these SEs that are used to compute the *p*-value and CI that you report, that is, they tell the same "story". The SE for the "1 hour cold" group is a bit higher because the sample size $n$ for this group is smaller by 1.

#### The contrasts table

```{r example2-pairs-understand, echo = FALSE}
fig2a_m1_planned %>%
  kable(digits = c(1,1,2,0,1,1,1,4)) %>%
  kable_styling()
```

1. This table is important for reporting treatment effects and CIs and for plotting the model. A contrast is a difference in means.
2. The contrast table here has no more information than is in the coefficient table, but that is not generally true for models with treatment factors with more than two groups. In "Working in R" below, I show how to compute a contrast table with all pairwise comparisons (contrasts between all possible pairings of the groups) 
3. The column $\texttt{Contrast}$ contains the names of the contrasts. Note that the name gives the direction of the difference.
4. The values in the column $\texttt{estimate}$ are the contrasts. These are the differences in the conditional means of the groups identified in the $\texttt{Contrast}$ column. These are the **effects** that we are interested in.
5. The value in the "SE" column is the standard error of the difference (SED) of each contrast. This SE is computed using the model standard deviation $\sigma$.
6. The values in the "lower.CL" and "upper.CL" columns are the bounds of the 95% confidence interval of the estimate. Remember (from Chapter \@ref(variability)) to think of this interval as containing potential values of the true parameter (the true difference in means between the two groups) that are reasonably compatible with the data. Don't think of the interval as having 95% probability of containing the true effect. Remember that a confidence interval applies to the procedure and not a parameter -- 95% of the CIs from hypothetical, replicate experiments that meet all the assumptions used to compute the CI will include the true effect.
7. The columns "t.ratio" and "p.value" contains the *t* and *p* values of the significance (not hypothesis!) test of the estimate. The *t*-statistic is the ratio of the estimate to the SE of the estimate (use the console to confirm this given the values in the table). It is a signal (the estimate) to noise (SE of the estimate) ratio. The *p*-value is the probability of sampling from normal distribution with the observed standard deviation, randomly assigning the sampled values to the three groups (using the original sample sizes for each), fitting the linear model, and observing a *t*-value as or more extreme than the observed *t*. A very small *p*-value is consistent with the experiment "not sampling from distributions with the same mean" -- meaning that adding a treatment affects the mean of the distribution. This is the logic used to infer a treatment effect. Unfortunately, it is also consistent with the experiment not approximating other conditions of the model, including non-random assignment, non-independence, non-normal conditional responses, and variance heterogeneity. It is up to the rigorous researcher to be sure that these other model conditions are approximated or "good enough" to use the *p*-value to infer a treatment effect on the mean.

#### *t* and *p* from the contrasts table -- when there are more than two levels in $X$ -- are not the same as those from pairwise *t*-tests among pairs of groups {#oneway-ttest}

```{r fig2a_m1_pairs}
fig2a_m1_pairs <- contrast(fig2a_m1_emm,
                           method = "revpairwise",
                           adjust = "none") %>%
  summary(infer = TRUE)
```

The chunk above computes a contrast table that includes comparisons of all pairs of groups in the factor $\texttt{treatment}$ (this adds a 3rd comparison to the contrast table of planned comparisons above). The *t*-tests for the contrasts are derived from a single fit linear model.

In contrast to the analysis in the chunk above, researchers commonly fit separate *t*-tests for each pair of treatment levels.

```{r example2-t-tests-understand, echo=TRUE}
# classic t-test
test1 <- t.test(fig2a[treatment == "1 hour cold", diHOME],
                fig2a[treatment == "Control", diHOME],
                var.equal = TRUE)

test2 <- t.test(fig2a[treatment == "30 min NE", diHOME],
                fig2a[treatment == "Control", diHOME],
                var.equal = TRUE)

test3 <- t.test(fig2a[treatment == "30 min NE", diHOME],
                fig2a[treatment == "1 hour cold", diHOME],
                var.equal = TRUE)
```

Notes

1. Again, the default t.test in R is the Welch t-test for heterogenous variance. To compute the Student t-test, use `var.equal = TRUE`
2. To see the full `t.test` output, type "test1" into the console.

Compare the *t* and *p* values from the three independent tests with the *t* and *p*-values from the single linear model.

```{r example2-t-test-table, echo = FALSE}
y_cols <- c("contrast", "t.ratio", "p.value")
lm_t <- fig2a_m1_pairs[, y_cols]
colnames(lm_t)[2:3] <- c("t (lm)", "p (lm)")
t_t <- data.table(contrast =
                    c("1 hour cold - Control", 
                      "30 min NE - Control",
                      "30 min NE - 1 hour cold"),
                  "t (t-test)" = c(test1$statistic,
                                   test2$statistic,
                                   test3$statistic),
                  "p (t-test)" = c(test1$p.value,
                                   test2$p.value,
                                   test3$p.value))
merge(lm_t, t_t, by = "contrast") %>%
  kable() %>%
  kable_styling()
```

The *t* and *p*-values computed from three separate tests differ from the *t* and *p*-values computed from the single linear model shown in the contrasts table above. The values differ because the SE in the denominators used to compute the $t$-values differ. The linear model uses the same value of $\sigma$ to compute the SED (the denominator of *t*) for all three t-tests in the contrast table. Each separate t-test uses a different value of $\sigma$ to compute the SED. Which is correct? Neither -- they simply make different assumptions about the data generating model.

1. Most importantly, **never** do both methods, look at the *p*-values, and then convince yourself that the method with the *p*-values that match your hypothesis is the correct method. Human brains are very, very good at doing this. This is called p-hacking. When you p-hack, the interpretation of the *p*-value is effectively meaningless. P-hacking leads to irreproducible science.
2. In general, using the linear model is a better practice than the separate *t*-tests. The reason is the homogeneity of variance assumption. If we assume homogeneity of variances, then we can think of the sample standard deviation of all three groups as an estimate of $\sigma$. In the linear model, we use three groups to estimate $\sigma$ but in each separate *t*-test, we use only two groups. Consequently, our estimate of $\sigma$ in the linear model is more precise than that in the *t*-tests. While the difference can be large with any individual data set (it's pretty big with the fig2a data), the long-run advantage of using the linear model instead of separate *t*-tests is pretty small, especially with only three groups (the precision increases with more groups).
3. We can drop the homogeneity of variance assumption with either the linear model or the three, separate *t*-tests. This is outlined below in "Heterogeneity of variance". In this case, the *t* and *p*-values for the three comparisons are the same. Still, the linear model (that models heterogenity) is better practice than the separate *t*-tests because the linear model is much more flexible and expandable. 

#### The contrasts table -- when there are more than two levels in $X$ -- has multiple p-values. How to handle this "multiple testing" is highly controversial

**Multiple testing** is the practice of adjusting *p*-values and confidence intervals to account for the expected increase in the frequency of Type I error in a batch, or *family*, of tests. Multiple testing is a concept that exists because of Neyman-Pearson hypothesis testing strategy. If multiple tests are used to answer the same question then, these are in the same family. Issues surrounding multiple testing are fleshed out in more detail in Chapter xxx "Best Practices". Computing adjusted values is covered below in the "Working in R" section.

## Working in R
### Fit the model

```{r}
m1 <- lm(diHOME ~ treatment, data = fig2a)
```

1. As described in [Fitting the linear model](fit-the-model), `diHOME ~ treatment` is the model formula. All functions that will be used to fit models in this text will use a model formula. Many base R and package functions that are meant to be either "easy" or follow an ANOVA strategy use lists of dependent and independent variables instead of a model formula. We won't use these because they are not consistent with a linear modeling way of thinking about data analysis.
2. $\texttt{treatment}$ was specifically coded as a factor variable in the import and wrangle chunk and R will automatically create the correct indicator variables. If categorical variables on the RHS of the formula have not be converted to factors by the user, then R will treat character variables as factors and create the indicator variables. If the categorical variable is numeric (for example, the variable `time` might have levels 1, 2, 3 instead of "1 hour", "2 hours", "three hours"), R will treat the variable as numeric and not create the indicator variables. A user could use the formula `y ~ factor(time)` to force R to create the indicator variable. I prefer to explicitly create a version of the variable that is recoded as factor using something like `dt[, time_fac := factor(time, levels = c("1", "2", "3"))]`.

### Controlling the output in tables using the coefficient table as an example

```{r}
m1_coef <- cbind(coef(summary(m1)),
                 confint(m1))
m1_coef
```
The standard print of the `m1_coef` object is too long for the display and often wraps when printed on the display, which can make it hard to read. I use `knitr::kable` to print the table with fewer decimal places and `kableExtra::kable_styling` to make it a little prettier.

```{r}
# the row names are not part of the m1_coef object
# so there is no digit designation for this column

m1_coef %>% # pipe the m1_coef object to kable
  kable(digits = c(2,3,3,5,2,2)) %>% 
  kable_styling()

# explore other styles in the kableExtra package
```

### Using the emmeans function

```{r cat-x-m1_emm}
m1_emm <- emmeans(m1, specs = "treatment")
m1_emm
```

Notes

1. Note that printing the `emmeans` object displays useful information. Here the information, this information includes the confidence level used. If the object is printed using `kable` (as in the "Inference" and "Understanding" sections above), only the table is printed and the additional information is lost.
2. `emmeans` computes the modeled means of all combinations of the levels of the factor variables specified in `specs`.
3. If there are two factor variables in the model, and both are passed to `specs`, then the modeled means of all combinations of the levels of the two variables are computed. If only one factor variable is passed, then the marginal means (averaged over all levels of the missing factor) are computed. This will become more clear in the chapter "Models for two (or more) categorical X variables".
4. If there are continuous covariates in the model, the modeled means are computed at the average values of these covariates. These covariates do not need to be passed to the `specs` argument.
5. You can pass numeric and integer covariates to `specs` to control the value of the covariates used compute the modeled means. This is outlined in [Adding covariates to a linear model](chapter-covariates)

### Using the contrast function

```{r cat-x-using-contrast-1}
m1_planned <- contrast(m1_emm,
                       method = "trt.vs.ctrl",
                       adjust = "none",
                       level = 0.95) %>%
  summary(infer = TRUE)

m1_planned
```
Notes

1. Note that printing the `contrast` object displays useful information, including the confidence level used and the method of adjustment for multiple tests. If the object is printed using `kable` (as in the "Inference" and "Understanding" sections above), only the table is printed and the additional information is lost.
2. The `method` argument is used to control the set of contrasts that are computed. See below.
3. The `adjust` argument controls if and how to adjust for multiple tests. Each `method` has a default adjustment method. See below.
4. The `level` argument controls the percentile boundaries of the confidence interval. The default is 0.95. Including this argument with this value makes this level transparent.

#### method argument

The `method =` argument is used to control the set of contrasts that are computed. Type `help("contrast-methods")` into the console to see the list of available methods. Also, read the [comparisons and contrasts vignette](https://cran.r-project.org/web/packages/emmeans/vignettes/comparisons.html){target="_blank"} for more on `emmeans::contrast()`

1. `method = "trt.vs.ctrl"` computes all non-reference minus reference contrasts. This method was used in the "Inference" section because it gives the two contrasts of the planned comparisons identified in the "understand the experimental design" step.
2. The default adjustment for multiple tests is "dunnettx", which is [Dunnett's test](https://en.wikipedia.org/wiki/Dunnett%27s_test){target="_blank"}. "none" was specified in the inference section because both comparisons were planned.

```{r}
m1_tukey <- contrast(m1_emm,
                       method = "revpairwise",
                       adjust = "tukey",
                       level = 0.95) %>%
  summary(infer = TRUE)
m1_tukey
```
1. Both `method = "pairwise"` and `method = "revpairwise"` compute all pairwise comparisons. I prefer "revpairwise" because the contrasts that include the reference are in the direction non-reference minus reference.
2. The default adjustment for multiple tests is "tukey", which is the [Tukey's HSD method](https://en.wikipedia.org/wiki/Tukey%27s_range_test){target="_blank"}.

#### Adjustment for multiple tests

1. "none" -- no adjustment
2. "dunnettx" -- [Dunnett's test](https://en.wikipedia.org/wiki/Dunnett%27s_test){target="_blank"} is a method used when comparing all treatments to a single control.
3. "tukey" -- [Tukey's HSD method](https://en.wikipedia.org/wiki/Tukey%27s_range_test){target="_blank"} is a method used to compare all pairwise comparisons. One might use this in an exploratory experiment.
4. "bonferroni" -- [Bonferroni](https://en.wikipedia.org/wiki/Bonferroni_correction){target="_blank"} is a general purpose method to compare any set of multiple tests. The test is conservative. A better method is "holm"
5. "holm" -- [Holm-Bonferroni](https://en.wikipedia.org/wiki/Holm–Bonferroni_method){target="_blank"} is a general purpose method like the Bonferroni but is more powerful.
6. "fdr" -- controls the [false discovery rate](https://en.wikipedia.org/wiki/False_discovery_rate){target="_blank"} not the Type I error rate for a family of tests. One might use this in an exploratory experiment.
7. "mvt" -- based on the multivariate *t* distribution and using covariance structure of the variables. One might use this in an exploratory experiment.

#### Planned comparisons using custom contrasts

With any but purely exploratory experiment, we have certain contrasts that test the focal hypotheses. We care little to none about other contrasts. We can limit the contrasts computed using `emmeans::contrast()` by passing custom contrasts to `method`.

1. create a set of vectors with as many elements as there are rows in the emmeans table. Name each vector using the (or a) group name for the row in the emmeans table. For the vector created for row *j*, set the value of the *j*th element to 1 and set all other values to zero.
2. create a list of contrasts, with each set in the list the difference of two of the named vectors from step 1.

```{r}
# m1_emm # print in console to get row numbers
# set the mean as the row number from the emmeans table
cn <- c(1,0,0) # control is in row 1
cold <- c(0,1,0) # cold in in row 2
ne <- c(0,0,1) # ne is in row 3

# contrasts are the difference in the vectors created above
# the focal contrasts are in the understand the experimental
# design section
# 1. (cold - cn) 
# 2. (ne - cn)

planned_contrasts <- list(
  "Cold - Cn" = c(cold - cn),
  "NE - Cn" = c(ne - cn)
)
m1_planned <- contrast(m1_emm,
                       method = planned_contrasts,
                       adjust = "none"
) %>%
  summary(infer = TRUE)

m1_planned %>%
  kable(digits = c(1,1,2,0,1,1,1,4)) %>%
  kable_styling()
  
```

### How to generate ANOVA tables

ANOVA is Analysis of Variance. Researchers frequently use the term "ANOVA" as the name for an analysis of an experiment with single-factor with more than two groups. However, ANOVA is a general method of inference for complex experimental designs. ANOVA models and regression models are different ways of expressing the same underlying linear model.

```{r fig2a-anova-table, echo=FALSE, warning=FALSE, message=FALSE}
# .I is a data.table function that returns the row number
fig2a[, fake_id := paste("mouse", .I)]

m1_aov4 <- aov_4(diHOME ~ treatment + (1|fake_id),
                 data = fig2a)

m1_aov4_nice <- nice(m1_aov4)[1, -5]
colnames(m1_aov4_nice)[1] <- "Term"

m1_aov4_nice %>%
  kable(digits = c(0,1,1,3),
        caption = "ANOVA table for the Figure 2a (example 2) data.") %>%
  kable_styling()
```
An ANOVA table for a single factor with more than two groups has a single *p*-value for the treatment term. The single *p*-value is the probability of sampling a value of *F* as large or larger than the observed *F* under the null (no true effects of either treatment and all specifications of the generating model are true). There is not much we can do with this number - we want to estimate the effect sizes and their uncertainty and we don't get this from an ANOVA table. Many textbooks, websites, and colleagues suggest to 1) fit the ANOVA, 2) check the *F*, and, if $F < 0.05$, 3) do "tests after an ANOVA". These tests after an ANOVA are the planned comparisons and post-hoc tests described above using the linear model. In classical ANOVA, the initial computation of the means and sums of squares was a logical first step to the decomposition of these sums of squares to compute the contrasts. With modern linear models using regression, the ANOVA first step is unnecessary and not recommended.

If your PI, manager, thesis committee, or journal editor insists that you do ANOVA, and you cannot convince them otherwise, here is how to generate an ANOVA table in R. Note that even though we are generating that table, the computation of the contrast table and all inference from that is not part of the ANOVA.

#### The afex aov_4 function

The package afex was developed to make it much easier for researchers to generate ANOVA tables that look like those from other statistics packages including SAS, SPSS, JMP, and Graphpad Prism.

```{r aov_4, warning=FALSE, message = FALSE}
# .I is a data.table function that returns the row number
fig2a[, fake_id := paste("mouse", .I)]

m1_aov4 <- aov_4(diHOME ~ treatment + (1|fake_id),
                 data = fig2a)

anova(m1_aov4)
```

Notes

1. The afex package has three function names for generating the same ANOVA table and statistics -- here I'm using `aov_4` because this functions uses a linear model formula argument (specifically, that used in the lme4 package), which is consistent with the rest of this text.
2. The formula includes the addition of a random factor (`(1|id)`) even though there really is no random factor in this model. See Section \@ref(oneway-paired-t) below for a brief explanation of a random factor. The random factor (the factor variable "id" created in the line before the fit model line) identifies the individual mouse from which the response variable was measured. Because the response was only measured once on each individual mouse, "id" is not really a random factor but the addition of this in the model formula is necessary for the `aov_4` function to work.
3. It is easy to get an ANOVA table that you don't want in R. If you want an ANOVA table that matches one from Graphpad Prism or JMP or similar software, the best practice is using the ANOVA functions from the afex package. 
4. What do I mean by "an ANOVA table that you don't want"? With factorial ANOVA with unbalanced data, there are three ways to compute the sums of squares for the different terms of the ANOVA table. SAS termed these Type I, II, and III sums of squares and these names have stuck. Following SAS, almost all statistics packages use Type III as the default (or only) method for computing ANOVA tables. R uses Type I as the default. There are very good arguments for using Type II. This distinction is moot for single factor ANOVA or multi-factor ANOVA for balanced designs but is not moot for unbalanced multi-factor ANOVA or any ANOVA with covariates. If you want an ANOVA table from R to match what would be generated by Graphpad Prism or JMP (Type III), then the afex package is the best practice.

#### The car Anova function

The car package has the extremely useful `Anova` function although using it is a bit like doing brain surgery having only watched a youtube video.

```{r anova-car-Anova}
type3 <- list(treatment = contr.sum)
m1_type3 <- lm(diHOME ~ treatment,
                data = fig2a,
                contrasts = type3)
Anova(m1_type3, type="3")
```

Notes

1. `car::Anova` has arguments for reporting the Type III sum of squares. Again, this is not relevant to a single factor ANOVA with no covariates but to avoid making mistakes in the future, its good to know about best practices now. For this, I prefer the `afex::aov_4 function` above. But, this works also.
2. Background: The default model matrix in the `lm` function uses dummy (or treatment) coding. For a Type 3 SS ANOVA (the kind that matches that in Graphpad Prism or JMP), we need to tell `lm` to use sum (or deviation) coding.
3. Many online sites give you code to change this using 

```{r anova-options-contrasts}
options(contrasts = c("contr.sum", "contr.poly"))
```

If you're reading this book, you almost certainly don't want to do this because this code resets how R computes coefficients of linear models and SS of ANOVA tables. This will effect all future analysis until the contrasts are set to something else or a new R session is started. 
4. The best practice method for changing the contrasts in the model matrix is using the `contrasts` argument within the `lm` function, as in the code above to fit `m1_type3`. This is the safest practice because this sets the contrasts only for this specific fit.
5. The coefficients of `m1_type3` will be different from `m1`. The intercept will be the grand mean and the coefficients of the non-reference levels (the effects) will be their deviations from the grand mean. I don't find this definition of "effects" very useful for most experiments in biology.
5. The contrasts (differences in the means among pairs of groups) in the contrast table will be the same, regardless of the contrast coding.

**base R aov and anova**

```{r anova-aov}
m1_aov <- aov(diHOME ~ treatment, data = fig2a)
summary(m1_aov)
```

```{r anova-anova}
# same as m1 in the Example 2 section
m1 <- lm(diHOME ~ treatment, data = fig2a)
anova(m1)
```

Notes

1. Many introduction to statistics textbooks and websites use the base R `aov` function. I don't find this function useful given the afex package functions.
2. The base R `anova` is useful if you know what you are doing with it. 

## Issues in inference in models with a single, categorical $X$

### Lack of independence ("paired t-test") {#oneway-paired-t}

This is a preview of the chapter [Models with random effects -- Blocking and pseudoreplication](#lmm)

The data from the experiment for Figure 1b of the 12,13-diHOME article are the plasma concentrations of 12,13-diHOME in humans in response to either saline or one-hour cold challenge. The response variable ($\texttt{diHOME}$) violates the "independent" sampling assumption for inference because there are two measures of the response in each individual.

A "which test?" strategy points to a paired *t*-test in place of Student's *t*-test if there is a lack of independence between the two groups.

Here, I show empirically that a paired t-test is a special case of a *linear model with an added random factor*. The advantage of understanding this is, the linear model with added random factor is infinitely expandable while a paired t-test is limited to the comparison of two means. If you have more than two groups or a factorial design or a covariate, there is no elegant way to expand a paired t-test. 

A good way to think about the model generating the data is

\begin{equation}
diHome = (\beta_0 + \beta_{0_i}) + \beta_1 treatment\_cold + \varepsilon
\end{equation}

1. $\beta_0$ is the expected value of diHome in humans given a saline treatment.
2. $\beta_{0_i}$ is the expected difference between the expected value of diHome in human $i$ given a saline treatment and the expected value of diHome in humans given a saline treatment. $\beta_{0_i}$ is a kind of **random effect**
3. $\beta_0 + \beta_{0_i}$ is the expected value of diHome in human $i$ given a saline treatment. It is the **random intercept** for human $i$.

```{r fig1b-import, echo=FALSE}
data_from <- "The cold-induced lipokine 12,13-diHOME promotes fatty acid transport into brown adipose tissue"
file_name <- "41591_2017_BFnm4297_MOESM1_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)

fig1b <- read_excel(file_path,
                     sheet = "Fig 1 b thur c",
                     range = "A1:C20",
                     col_names = TRUE) %>%
  data.table() %>%
  clean_names() %>%
  na.omit() # get rid of blank row

setnames(fig1b,
         old = names(fig1b),
         new = c("sample", "diHOME", "bat_activity"))

fig1b[, id := substr(sample, 1, 6)]
fig1b[, treatment := ifelse(substr(sample, 8,8) ==
                              "C", "cold", "saline")]
fig1b[, treatment := factor(treatment, levels = c("saline", "cold"))]

# names(fig1b)
#View(fig1b)
```

#### Fit a linear model with added random factor

```{r fig1b-m1}
# fit the model
fig1b_m1 <- lmer(diHOME ~ treatment + (1|id), data = fig1b)
```

1. The added random factor is $\texttt{id}$, which contains the id of the human. $\texttt{id}$ didn't appear from nowhere, it is a column in the data.table fig1b.
2. We are now using the function `lmer()` from the lme4 package (loaded with package lmerTest)
3. `(1|id)` tells `lmer` to add $\texttt{id}$ as a random intercept.
4. $\texttt{id}$ is a **random factor**. $\texttt{treatment}$ if a **fixed factor**. Models with both fixed and random factors go by many names. In this text, I use "linear mixed model".

#### Inference 

Inference using the linear model with added random factor:

```{r fig1b_m1_pairs}
# estimated marginal means table
fig1b_m1_emm <- emmeans(fig1b_m1, specs = "treatment")

# contrasts table
fig1b_m1_pairs <- contrast(fig1b_m1_emm,
                     method = "revpairwise") %>%
  summary(infer = TRUE)

fig1b_m1_pairs %>%
  kable(digits = c(1,1,2,1,1,1,5,5)) %>%
  kable_styling()
```
Inference using paired t-test:

```{r fig1b_ttest}
fig1b_ttest <- t.test(
  x = fig1b[treatment == "cold", diHOME],
  y = fig1b[treatment == "saline", diHOME],
  paired = TRUE
)

fig1b_ttest
```
Notes

1. The *t* and *p* values are the same because the paired *t*-test is a special case of the linear model.

#### An example with more than two groups

Let's say each mice in the Figure 2A experiment was subjected to all three treatments so that we have three measures per mouse (I don't know that this isn't the case and the set up of the archived data is ambiguous). If this were the case, the response would violate the independence assumption.

The linear model (in regression notation) is

\begin{equation}
diHome = (\beta_0 + \beta_{0_i}) + \beta_1 treatment\_cold\ + \beta_2 treatment\_NE\ + \varepsilon 
\end{equation}

```{r fig2a-m2}
# fit the model
fig2a_m2 <- lmer(diHOME ~ treatment + (1|id), data = fig2a)
```

1. `fig2a_m2` is a linear mixed model. [fig2a_m1](#fig2a-m1) is a fixed effects model of the same data. Here, we wouldn't fit both. The researchers know if the response was measured for each treatment for each mouse or if different mice were used for each treatment. 

Inference using the linear model with added random factor:

```{r fig2a-m2_pairs}
# estimated marginal means table
fig2a_m2_emm <- emmeans(fig2a_m2, specs = "treatment")

# contrasts table
fig2a_m2_planned <- contrast(fig2a_m2_emm,
                     method = "trt.vs.ctrl",
                     adjust = "none") %>%
  summary(infer = TRUE)

fig2a_m2_planned %>%
  kable(digits = c(1,1,2,1,1,1,5,5)) %>%
  kable_styling()
```

Inference using two paired t-tests:

```{r fig2a_ttest_cold}
fig2a_ttest_cold <- t.test(
  x = fig2a[treatment == "1 hour cold", diHOME],
  y = fig2a[treatment == "Control", diHOME],
  paired = TRUE
)
fig2a_ttest_cold
```


```{r fig2a_ttest_ne}
fig2a_ttest_ne <- t.test(
  x = fig2a[treatment == "30 min NE", diHOME],
  y = fig2a[treatment == "Control", diHOME],
  paired = TRUE
)
fig2a_ttest_ne
```

Notes

1. **important** -- the *t*-test for the comparison of "1 hour cold" and "Control" excluded mouse 6 because of a missing "1 hour cold" measure. The linear mixed model was fit to all data, including the "Control" and "30 min NE" measures. The consequence is higher power/more precision in the linear mixed model.
2. The paired t-test *t* and *p* values are not equal to those from the linear mixed model. This is partly because of the missing data (note 1). Even if there were no missing data, these wouldn't be equal because the linear mixed model uses an estimate of $\sigma$ computed from the single model with all three groups to compute the standard errors. See the comparison of t-tests and the contrast table from a linear model in Section \@ref(ttests).
3. The results of the separate, paired t-tests and the linear mixed model differ in a way that might affect our inference about the system. Which is correct? Neither -- they simply make different assumptions about the data generating model.
* The linear model strategy has more power and precision but this advantage is small. The best reason to use linear models instead of separate *t*-tests is learning how to use linear models, and their extensions, gives you phenomenal cosmic power.
* Do not compute both separate paired *t*-tests **and** the linear models and then convince yourself that that the assumption of the method with the *p*-value that matches your hypothesis is correct. See the p-hacking discussion above.

### Heterogeneity of variances {#oneway-welch}

Heterogeneity of variance among treatment groups is a problem for inference, especially if the sample size is unequal among groups (statisticians tend to agree that heterogeneity is much more problematic than a non-normal response).

A "which test?" strategy points to a Welch's *t*-test in place of Student's *t*-test if there is heterogeneity of variances between treatment groups. A Welch t-test is infrequent in the experimental biology literature, perhaps because

1. it is poorly known and it doesn't occur to researchers to use a test that models heterogeneity of variances.
2. heterogeneity often arises in right-skewed data, which is often analyzed with a non-parametric test like the Mann-Whitney U test.

The Welch *t*-test is a special case of a linear model that explicitly models the within-group variance using **generalized least squares** (GLS). The 95% CI of a mean differences and *p*-values from the fit gls linear model and from Welch's t-test are the same. Advantages of using a linear modeling strategy is that a researcher uses the model to estimate effects (difference in means) and measures of uncertainty in the effects (standard errors or confidence intervals of the difference). Advantages of specifically using a GLS linear model is that it is easily expanded to analyze more complex designs including 1) more than one factor, 2) added covariates, 3) correlated residuals due to non-independence.

Some statisticians argue that researchers should *always* use a Welch t-test instead of Student's t-test. Given this logic, researchers should consider using GLS linear models for more complex experimental designs (added covariates, factorial) in place of classical ANCOVA and two-way ANOVA.

Modeling variance heterogeneity is the focus of Chapter \@ref(gls) so the account here is brief. Heterogeneity can be modeled using a generalized least squares linear model with the `gls` function. The `weights` argument is used to model the variances using each group's sample variance. In this example, I use the data from the Figure 1b experiment, which can be compared to the analysis of the same data in Example 2 above.

```{r fig2a_m3}
# gls fails with missing data
subdata <- fig2a[is.na(diHOME) == FALSE,] # omit rows with missing data
fig2a_m3 <- gls(diHOME ~ treatment,
                data = subdata,
                weights = varIdent(form = ~ 1 | treatment))
```

The model `fig2a_m3` uses variance computed in each group separately as the estimate of $\sigma$ for that group. The coefficient table of the GLS model is

```{r fig2a_m3_coef}
fig2a_m3_coef <- cbind(coef(summary(fig2a_m3)),
                       confint(fig2a_m3))

fig2a_m3_coef %>%
  kable(digits = c(4,4,4,6,4,4)) %>%
  kable_styling()
```
Notes

1. **Important** for reporting CIs and *p*-values. Unlike the linear model modeling homogenous variance, the CIs and *p*-values for the coefficients of $\texttt{treatment1 hour cold}$ and $\texttt{treatment30 min NE}$ are *not* the same as the *p*-values of these equivalent contrasts in the contrasts table (see below). The reason is, the computation of the CI and *p*-values in the two tables use two different degrees of freedom. Report the CI and *p*-values from the contrast table using the Satterthwaite df.

The modeled means and contrasts are computed as above for the `lm` object

```{r fig2a_m3_emm}
fig2a_m3_emm <- emmeans(fig2a_m3, specs="treatment")
fig2a_m3_emm
```

Notes

1. The SE of the means in this table are modeled SEs but are equal to the sample SE of the means, because this was specified in the GLS model.

```{r fig2a_m3_pairs}
fig2a_m3_pairs <-  contrast(fig2a_m3_emm,
                            method = "revpairwise",
                            adjust = "none") %>%
  summary(infer = TRUE)

fig2a_m3_pairs %>%
  kable(digits = c(1,4,4,1,4,4,4,5)) %>%
  kable_styling()
```
Notes

1. Compare the statistics for "1 hour cold - Control" and "30 min NE - Control" to the coefficients for $\texttt{treatment1 hour cold}$ and $\texttt{treatment30 min NE}$ in the coefficient table. The estimates, SE, and *t* are the same but the CIs and *p* values differ. The contrast function is using a different method ("satterthwaite") for computing the degrees of freedom and this results in a different value of the *t*-distribution tail area used to compute the CI and *p* value.

The 
```{r welch-t, echo=TRUE}
test1 <- t.test(fig2a[treatment == "1 hour cold", diHOME],
       fig2a[treatment == "Control", diHOME],
       var.equal = FALSE)

test2 <- t.test(fig2a[treatment == "30 min NE", diHOME],
       fig2a[treatment == "Control", diHOME],
       var.equal = FALSE)

test3 <- t.test(fig2a[treatment == "30 min NE", diHOME],
       fig2a[treatment == "1 hour cold", diHOME],
       var.equal = FALSE)
```

Notes

1. the default `t.test` is the Welch t-test. However, I've included `var.equal = FALSE` so the method is transparent.

Compare the contrast *p* values to the three Welch *t*-tests of all pairs of treatment levels in the fig2a experiment.

```{r welch-t-table, echo = FALSE}
y_cols <- c("contrast", "t.ratio", "p.value")
lm_t <- fig2a_m3_pairs[, y_cols]
colnames(lm_t)[2:3] <- c("t (gls lm)", "p (gls lm)")
t_t <- data.table(contrast =
                    c("1 hour cold - Control", 
                      "30 min NE - Control",
                      "30 min NE - 1 hour cold"),
                  "t (welch t)" = c(test1$statistic,
                                   test2$statistic,
                                   test3$statistic),
                  "p (welch t)" = c(test1$p.value,
                                   test2$p.value,
                                   test3$p.value))
merge(lm_t, t_t, by = "contrast") %>%
  kable() %>%
  kable_styling()
```

The *t* and *p*-values computed from the GLS linear model and from the three, pairwise Welch *t*-tests are the same (to about the 6th decimal place). They are the same because each is estimating $\sigma^2$ separately for each group and not as the pooled (among two groups for *t*-test or three groups for the linear model) estimate and because they use the same degrees of freedom to compute the *p*-value.

Let's summarize these comparisons

1. Inference from a linear model using homogenous variance (the `lm` function) and from a Student's *t*-test are the same if there are only two levels in the treatment variable.
2. Inference from a linear model using homogenous variance (the `lm` function) and from the series of pairwise, Student's *t*-tests differ when there are more than two levels in the treatment variable.
3. Inference from a GLS linear model using heterogenous variance (the `gls` function) and from a Welch *t*-test are the same regardless of the number of levels in the treatment variable.

Even though the linear model that models heterogeneity and the Welch *t*-test produce the same results, researchers should use the linear model because

1. A linear modeling strategy encourages researchers to think about the effect and uncertainty in the effect and not just a *p*-value.
2. The linear model is nearly infinitely flexible and expandible while the *t*-test has extremely limited flexibility (The Welch *t*-test is one way to expand the classical, Student's *t*-test).

### The conditional response isn't Normal

Of all the issues with inference from a linear model (including *t*-tests and ANOVA), non-normal "data" is the least worrisome if all we care about is *p*-values. Confidence intervals computed using the normal distribution on non-normal data can be awkward, for example, the CIs might include impossible values such as negative counts or percent of cells expressing a certain marker above 100 %.

Some alternative to a linear model assuming normal distribution include:

1. Count data -- count data tend to be right skewed and groups with higher mean counts tend to have larger variance.
* Generalized linear models using either a negative binomial or quasi-poisson distribution family are a good alternative to consider.
* Bootstrap confidence intervals are okay if the sample size is much larger (say, > 50) than typical in experimental biology. Permutation *p*-values are a good alternative regardless of sample size.
* A GLS linear model to account for any heterogeneity in variance that typically accompanies non-normal data. This doesn't address the non-normal distribution but, again, heterogeneity is typically a much larger problem than non-normality.
* Log transformation of a count response is controversial at best. log transformations can be pretty good at making the response look more like it was sampled from a normal distribution (and can make variances more similar). But...If the counts include zero, researchers have to add a kludge factor (typically equal to 1) to all counts prior to transformation. The value of the kludge factor is arbitrary and matters (adding .1 gives different results than adding 1). A log transformation (or any transformation) raises interpretation problems. For the log transformation, the effect is the difference between the means of the log-transformed counts. I don't know what the magnitude of an effect on a log scale means biologically. If we backtransform this, the effect is the ratio of the geometric means of the two groups. A ratio effect is easy enough to interpret but do we really want to model the geometric and not the arithmetic means? Finally, log transformation does not specifically address either the shape of the distribution or the heterogeneity that often comes with the non-normal shape. A generalized linear model can specifically model both the shape and the variance.
* non-parametric tests such as Mann-Whitney-Wilcoxan tests. Non-parametric tests are (almost) entirely focused on *p*-values and are effectively limited to pairwise comparisons (no added covariates, or factorial experiments). The Mann-Whitney-Wilcoxan is only a test of a difference in the median between two groups under the assumption of equal dispersion. Non-parametric tests were invented during the pre-computer age when *p*-values were computed using pencil and paper. Since at least the mid 1980s, there are better alternatives to non-parametric tests.
2. Fraction (proportion) data -- For example, the number of cells expressing a certain marker relative to all counted cells in the sample. Fraction data have hard bounds at 0 and 1, or at 0 and 100 if converted to a percent. Fraction data tend to be right skewed if the mean is closer to zero and left skewed if the mean is closer to the upper bound.
* Generalized linear model using a binomial distribution family with a logit link ("logistic regression") is a good alternative to consider. This model is used for Bernouili (success/fail) responses (for example the subject "lived" or "died"), where success is assigned the value 1 and fail is assigned the value 0. Using this model with proportion data is equivalent to assigning a 0 ("does not express marker") or 1 ("expresses marker") to all cells in the count.
* Bootstrap confidence intervals are okay if the sample size is much larger (say, > 50) than typical in experimental biology. Permutation *p*-values are a good alternative regardless of sample size.
* A GLS linear model to account for any heterogeneity in variance that typically accompanies non-normal data. This doesn't address the non-normal distribution but, again, heterogeneity is typically a much larger problem than non-normality.
* arcsin transformation. See [The arcsine is asinine: the analysis of proportions in ecology] (https://esajournals.onlinelibrary.wiley.com/doi/full/10.1890/10-0340.1)
* non-parametric tests such as Mann-Whitney-Wilcoxan tests. Non-parametric tests are (almost) entirely focused on *p*-values and are effectively limited to pairwise comparisons (no added covariates, or factorial experiments). The Mann-Whitney-Wilcoxan is only a test of a difference in the median between two groups under the assumption of equal dispersion. Non-parametric tests were invented during the pre-computer age when *p*-values were computed using pencil and paper. Since at least the mid 1980s, there are better alternatives to non-parametric tests.

```{r fig6f-import, echo=FALSE}
# need data_folder and data_from from earlier chunk
data_from <- "Exercise reduces inflammatory cell production and cardiovascular inflammation via instruction of hematopoietic progenitor cells"
file_name <- "41591_2019_633_MOESM8_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)

# assuming mice are independent and not same mouse used for all three treatment
melt_col_names <- c("Sedentary", "Exercise")
fig6f <- read_excel(file_path,
                     sheet = "Figure 6f",
                     range = "A7:B29",
                     col_names = TRUE) %>%
  data.table() %>%
  melt(measure.vars = melt_col_names,
       variable.name = "treatment",
       value.name = "neutrophils") %>%
  na.omit() # danger!

treatment_levels <- melt_col_names
fig6f[, treatment := factor(treatment,
                            levels = treatment_levels)]

# neutrophils is count/10^6
fig6f[, neutrophil_count := round(neutrophils*10^6, 0)]

fig6f[1, neutrophil_count]
#View(fig6f)
```

#### Linear model ("t-test") with non-normal count data

```{r}
m1 <- lm(neutrophil_count ~ treatment, data = fig6f)
m1_emm <- emmeans(m1, specs = "treatment")
m1_pairs <- contrast(m1_emm,
                     method = "trt.vs.ctrl") %>%
  summary(infer = TRUE)
```

#### GLM with negative binomial family with non-normal count data

```{r}
m2 <- glmmTMB(neutrophil_count ~ treatment,
              data = fig6f,
              family = nbinom2())
m2_emm <- emmeans(m2,
                  specs = "treatment",
                  type = "response")
m2_pairs <- contrast(m2_emm,
                     method = "trt.vs.ctrl",
                     type = "response") %>%
  summary(infer = TRUE)
```

```{r}
gg1 <- ggplot_the_model(
  m1,
  m1_emm,
  m1_pairs,
  legend_position = "none",
  y_label = "Neutrophil Count (/mL)",
  effect_label = "Effect (/mL)",
  palette = pal_okabe_ito_blue,
  rel_heights = c(0.75,1)
)

gg2 <- ggplot_the_model(
  m2,
  m2_emm,
  m2_pairs,
  legend_position = "none",
  y_label = "Neutrophil Count (/mL)",
  effect_label = "Effect (/mL)",
  palette = pal_okabe_ito_blue,
  rel_heights = c(0.75,1),
  effect_lims = c(0.35, 1.25)
)

plot_grid(gg1, gg2, nrow=2)
```


### Pre-post designs (change from baseline)

### Longitudinal designs

### Comparing responses normalized to a standard

### Comparing responses that are ratios

### Researcher degrees of freedom

```{r echo = FALSE, eval=FALSE}

ycols <- c("contrast", "p.value")
p_table <- fig2a_m1_pairs_dt[, .SD, .SDcols = ycols]
setnames(p_table, old = "p.value", new = "lm")
p_table <- cbind(p_table, "t test" = ttests[, "p"])
p_table <- cbind(p_table, "gls" = summary(fig2a_m2_pairs)[, "p.value"])
p_table <- cbind(p_table, "Welch" = welch_tests[, "p"])

# permutation
m3 <- lmp(diHOME ~ treatment,
                 data = subdata,
                 perm = "Exact")
coef(summary(m3))
# wilcoxan
p_table
```


Conspicuously, the *p*-value for the "1 hour cold - Control" contrast is 0.039, which is "significant" using the conventional 

## Hidden Code
### Importing and wrangling the fig_3d data for example 1

```{r, echo=TRUE}
data_folder <- "data"
data_from <- "The cold-induced lipokine 12,13-diHOME promotes fatty acid transport into brown adipose tissue"
```

```{r, echo=TRUE}
# need data_folder and data_from from earlier chunk
file_name <- "41591_2017_BFnm4297_MOESM3_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)

# ignore the column with animal ID. Based on methods, I am inferring
# that the six mice in vehicle group *are different* from the
# six mice in the 1213 group.
col_names_3d <- c("Vehicle", "1213")
treatment_levels <- c("Vehicle", "12,13-diHOME")
fig_3d <- read_excel(file_path,
                     sheet = "Figure 3d",
                     range = "B3:C9",
                     col_names = TRUE) %>%
  data.table() %>%
  melt(measure.vars = col_names_3d,
       variable.name = "treatment",
       value.name = "serum_tg")

# change group name of "1213"
fig_3d[treatment == "1213", treatment := "12,13-diHOME"]

# make treatment a factor with the order in "treatment_levels"
fig_3d[, treatment := factor(treatment, treatment_levels)]

#View(fig_3d)
```

### Importing and wrangling the fig2a data for example 2

```{r , echo=TRUE, message=FALSE}
# need data_folder and data_from from earlier chunk
file_name <- "41591_2017_BFnm4297_MOESM2_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)

# assuming mice are independent and not same mouse used for all three treatment
melt_col_names <- paste("Animal", 1:6)
fig2a <- read_excel(file_path,
                     sheet = "Fig 2a",
                     range = "A3:G6",
                     col_names = TRUE) %>%
  data.table() %>%
  melt(measure.vars = melt_col_names,
       variable.name = "id",
       value.name = "diHOME") # cannot start a variable with number
setnames(fig2a, old = colnames(fig2a)[1], new = "treatment")

treatment_order <- c("Control", "1 hour cold", "30 min NE")
fig2a[, treatment := factor(treatment, treatment_order)] # order levels

#View(fig2a)
```

### Importing and wrangling the fig6f data for non-normal example

```{r echo=TRUE}
# need data_folder from earlier chunk
data_from <- "Exercise reduces inflammatory cell production and cardiovascular inflammation via instruction of hematopoietic progenitor cells"
file_name <- "41591_2019_633_MOESM8_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)

# assuming mice are independent and not same mouse used for all three treatment
melt_col_names <- c("Sedentary", "Exercise")
fig6f <- read_excel(file_path,
                     sheet = "Figure 6f",
                     range = "A7:B29",
                     col_names = TRUE) %>%
  data.table() %>%
  melt(measure.vars = melt_col_names,
       variable.name = "treatment",
       value.name = "neutrophils") %>%
  na.omit() # danger!

treatment_levels <- melt_col_names
fig6f[, treatment := factor(treatment,
                            levels = treatment_levels)]

# neutrophils is count/10^6
fig6f[, neutrophil_count := round(neutrophils*10^6, 0)]

fig6f[1, neutrophil_count]
#View(fig6f)
```



<!--chapter:end:chapters/24-categorical_X.Rmd-->

# Model Checking

```{r model-check-setup, echo=FALSE, warning=FALSE, message=FALSE}
# import an messaging
library(here)
library(janitor)
library(readxl)
library(data.table)

# analysis
library(car)
library(MASS) # rnegbin

# graphics and output
library(ggpubr) # qq plot functions
library(ggpubfigs) # palette
library(qqplotr) ## qq plot functions
library(ggforce)
library(cowplot)

here <- here::here
data_folder <- "data"

# Okabe & Ito palette
ito_seven <- friendly_pal("ito_seven") # ggpubfigs
pal_okabe_ito <- ito_seven[c(6,5,3,7,1,2,4)] # order of Wilke

```

```{r model-check-quantilize, echo=FALSE}
# a function to transform a vector into quantiles
# not if the data are 1:n then the output is "rankits"
quantilize_1 <- function(x){
  # this is the ppoints(x) function
  m <- length(x)
  s <- trunc(rank(x))
  a <- ifelse(m <= 10, 3/8, 1/2)
  q <- (s-a)/(m + (1-a) - a)
  return(q)
}
```

## All statistical analyses should be followed by model checking 

We us a linear model to infer effects or predict future outcomes. Our inference is uncertain. Given the model assumptions, we can quantify this uncertainty with standard errors, and from these standard errors we can compute confidence intervals and *p*-values. It is good practice to use a series of **diagnostic plots**, diagnostic statistics, and simulation to check how well the data approximate the fit model and model assumptions. **Model checking** is used to both check our subjective confidence in the modeled estimates and uncertainty and to provide empirical evidence for subjective decision making in the analysis workflow.

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**NHST blues** -- Researchers are often encouraged by textbooks, colleagues, or the literature to test the assumptions of a *t*-test or ANOVA with formal hypothesis tests of distributions such as a Shapiro-Wilks test of normality or a Levine test of homogeneity. In this strategy, an alternative to the *t*-test/ANOVA is used if the distribution test's p-value is less than some cut-off (such as 0.05). Common alternatives include 1) transformations of the response to either make it more normal or the variances more homogenous, 2) implementation of alternative tests such as a Mann-Whitney-Wilcoxon (MWW) test for non-normal data or a Welch *t*-test/ANOVA for heterogenous variances. The logic of a test of normality or homogeneity before a *t*-test/ANOVA isn't consistent with frequentist thinking because the failure to reject a null hypothesis does not mean the null hypothesis is true. We shouldn't conclude that a sample is "normal" or that the variances are "homogenous" because a distributional test's *p*-value > 0.05. But, maybe we should of the distributional pre-test as an "objective" model check? The logic of this objective decision rule suffers from several issues. **First**, the subsequent *p*-value of the *t*test/ANOVA test is not valid because this *p*-value is the long-run frequency of a test-statistic as large or larger than the observed statistic conditional on the null -- not conditional on the subset of nulls with $p > 0.05$ in the distribution test. **Second**, real data are only approximately normal; with small $n$, it will be hard to reject the null of a normal distribution because of low power, but, as $n$ increses, a normality test will reject any real dataset. **Third**, and most importantly, our analysis should follow the logic of our goals. If our goal is the estimation of effects, we cannot get meaningful estimates from a non-parametric test (with a few exceptions) or a transformed response, as these methods are entirely about computing a "correct" *p*-value. Good alternatives to classic non-parametric tests and transformations are bootstrap estimates of confidence limits, permutation tests, and generalized linear models. 
</div> 

## Linear model assumptions

To faciliate explanation of assumptions of the linear model and extensions of the linear model, I will use both the error-draw and conditional-draw specifications of a linear model with a single $X$ variable.

error draw:
\begin{align}
y &= \beta_0 + \beta_1 x_i + \varepsilon_i\\
\varepsilon_i &\sim N(0, \sigma^2)
(\#eq:model-checking-spec1)
\end{align}

conditional draw:
\begin{align}
y_i &\sim N(\mu_i, \sigma^2)\\
\mathrm{E}(Y|X=x_i) &= \mu_i\\
\mu_i &= \beta_0 + \beta_1 x_i
(\#eq:model-checking-spec2)
\end{align}

This model generates random data using the set of rules specified in the model equations. To quantify uncertainty in our estimated parameters, including standard errors, confidence intervals, and *p*-values, we make the assumption that the data from the experiment is a random sample generated using these rules. 

The two rules specified in the model above (Model \@ref(eq:model-checking-spec)) are

1. The systematic component of data generation is $\beta_0 + \beta_1 X$.
* More generally, all linear models in this text specify systematic components that are *linear in the parameters*. Perhaps a better name for this is "additive in the parameters". Additive (or linear) simply means that we can add up the products of a parameter and an $X$ variable to get the conditional expectation $\mathrm{E}(Y|X)$.
* For observational inference, the rule $\mathrm{E}(Y| see\;X=x_i) = \mu_i$ is sufficient. For causal inference with data from an experiment, or with observational data and a well defined causal diagram, we would need to modify this to $\mathrm{E}(Y|do\; X=x_i) = \mu_i$.
2. The stochastic component of data generation is "IID Normal", where IID is **independent and identically distributed** and Normal refers to the Normal (or Gaussian) distribution. The IID assumption is common to all linear models. Again, for the purpose of this text, I define a "linear model" very broadly as a model that is linear in the parameters. This includes many extensions of the classical linear model including generalized least squares linear models, linear mixed models, generalized additive models, and generalized linear models. **Parametric** inference form all linear models requires the specification of a distribution family to sample. Families that we will use in this book include Normal, gamma, binomial, poisson, and negative binomial. This text will also cover **distribution free** methods of quantifying uncertainty using the bootstrap and permutation, which do not specify a sampling distribution family.

### A bit about IID
1. **Independent** means that the random draw for one case cannot be predicted from the random draw of any other case. A lack of independence creates *correlated error*. There are lots or reasons that errors might be correlated. If individuals are measured both within and among cages, or tanks, or plots, or field sites, then we'd expect the measures within the same unit (cage, tank, plot, site) to err from the model in the same direction because of environmental features shared by individuals within the unit but not by individuals in other units. Multiple measures within experimental units create "clusters" of error. Lack of independence or clustered error can be modeled using **generalized least squares (GLS)** models that directly model the structure of the error and with **random effects** models. Random effects models go by many names including linear mixed models (common in Ecology), hierarchical models, and multilevel models. Both GLS and random effects models are variations of linear models.

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**tl;dr** -- Measures taken within the same individual over time (*repeated measures*) are correlated and are common in all areas of biology. In ecology and evolutionary studies, measures that are taken from sites that are closer together or measures taken closer in time or measures from more closely related biological species will tend to have more similar error than measures taken from sites that are further apart or from times that are further apart or from species that are less closely related. Space and time and phylogeny create **spatial, temporal, and phylogenetic autocorrelation**.
</div> 

2. **Identical** means that all random draws at a given value of $X$ are from the same distribution. Using the error-draw specification of the model above, this can be restated as, the error-draw ($\varepsilon_i$) for every $i$ is from the same distribution $N(0, \sigma^2$). Using the conditional-draw specification, this can be restated as, the random-draw $y_i$ for every $i$ with the same expected value $\mu = \mu_i$ is from the same distribution $N(\mu_i, \sigma^2$). Understand the importance of this. Parametric inference using this model assumes that the sampling variance of $\mu$ *at a single value of $X$* is the same for *all values of $X$*. If $X$ is continuous, this means the spread of the points around the regression line is the same at all values of $X$ in the data. If $X$ is categorical, this means the spread of the points around the mean of a group is the same for all groups. A consequence of "identical", then, for all classical linear models, is the assumption of homogeneity (or **homoskedasticity**) of variance. If the sampling variance differs among the $X$, then the variances are heterogenous or **heteroskedastic**. Experimental treatments can affect the variance of the response in addition to the mean of the response. Heterogenous variance can be modeled using **Generalized Least Squares (GLS)** linear models. Many natural biological processes generate data in which the error is a function of the mean. For example, measures of biological variables that grow, such as size of body parts, have variances that "grow" with the mean. Or, measures of counts, such as the number of cells damaged by toxin, the number of eggs in a nest, or the number of mRNA transcripts per cell have variances that are a function of the mean. Both growth and count measures can sometimes be reasonably modeled using a linear model but more often, they are better modeled using a **Generalized Linear Model** (GLM).

## Diagnostic plots use the residuals from the model fit
### Residuals
A residual of a statistical model is $y_i - \hat{y}_i$. Remember that $\hat{y}_i$ is the predicted value of $Y$ when $X$ has the value $x_i$ (compactly written as $X=x_i$). And remember that $\hat{y}_i$ is the estimate of $\mu_i$. For linear models (but not generalized linear models), the residuals of the fit model are estimates of the $\varepsilon$ in equation \@ref(eq:model-checking-spec1). This *is not* true for generalized linear models because GLMs are not specified using \@ref(eq:model-checking-spec1).

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**Alert** A common misconception is that inference from a linear model assumes that the *response* (the measured $Y$) is IID Normal. This is wrong. Either specification of the linear model shows precisely why this conception is wrong. Model \@ref(eq:model-checking-spec1) explicitly shows that it is the error that has the normal distribution -- the distribution of $Y$ is a mix of the distribution of $X$ and that of the error. A more general way of thinking about the assumed distribution uses the specification in model \@ref(eq:model-checking-spec2), which shows that it is the *conditional* response that is assumed to be IID normal. Remember, a conditional response ($y_i$) is a random draw from the infinite set of responses at a given value of $X$.
</div> 

Let's look at the distribution of residuals versus the distribution of responses for a hypothetical experiment with a single, categorical $X$ variable (the experimental factor) with two levels ("Cn" for control and "Tr" for treatment). The true parameters are $\beta_0 = 10$ (the true mean for the control group, or $\mu_{0}$), $\beta_1=4$ (the difference between the true mean for the treatment minus the true mean for the control, or $\mu_1 - \mu_0$), and $\sigma = 2$ (the error standard deviation). 

```{r model-check-histogram, model-check-residuals1, echo=FALSE, fig.cap="Histogram of the (A) response, showing with modes near the true means of each group and (B) residuals, with a mode for both groups at zero."}

set.seed(20)
n <- 50
mu <- 10
sigma <- 2
beta <- sigma*2
fd <- data.table(
  Treatment = rep(c("Cn", "Tr"), each=n),
  Response = c(rnorm(n, mean = mu, sd = sigma),
               rnorm(n, mean = mu + beta, sd = sigma))
)

m1 <- lm(Response ~ Treatment, data=fd)
fd[, Residual := residuals(m1)]

bw <- 1
gg1 <- gghistogram(data=fd,
                   x = "Response",
                   color="Treatment",
                   fill="Treatment",
                   add = "mean",
                   rug = FALSE,
                   #bins=9,
                   binwidth = bw,
                   palette = pal_okabe_ito)
gg2 <- gghistogram(data=fd,
                   x = "Residual",
                   color="Treatment",
                   fill="Treatment",
                   add = "mean",
                   rug = FALSE,
                   #bins=9,
                   binwidth = bw,
                   palette = pal_okabe_ito)
plot_grid(gg1, gg2, labels="AUTO")

```

The plot above shows a histogram of the response (A) and residuals (B). In the plot of the response, the mode (the highest bar, or bin with the most cases) includes true mean for each group. And, as expected given $\beta_1=4$, the modes of the two groups are 4 units apart. It should be easy to see from this plot that the response does not have a normal distribution. Instead, it is distincly bimodal. But the distribution of the response *within* each level looks like these are drawn from a normal distribution -- and it should.  In the plot of the residuals, the values of both groups are shifted so that the mean of each group is at zero. The consequence of the shift is that the combined set of residuals does look like it is drawn from a Normal distribution.

The two plots suggest two different approaches for model checking. First, we could examine the responses within each level of the experimental factor. Or, second, we could examine the residuals of the fit model, ignoring that the residuals come from multiple groups. The first is inefficient because it requires as many checks as there are levels in the factor. The second requires a single check.

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**Alert** Some textbooks that recommend formal hypothesis tests of normality recommend the inefficient, multiple testing on each group separately. This isn't wrong, it's just more work than it needs to be and also suffers from "multiple testing".
</div> 

### A Normal Q-Q plot is used to check for characteristic departures from Normality 
A Normal Q-Q plot is a scatterplot of

1. **sample quantiles** on the *y* axis. The sample quantiles is the vector of $N$ residuals in rank order, from smallest (most negative) to largest (most positive). Sometimes this vector is standardized by dividing the residual by the standard deviation of the residuals (doing this makes no difference to the interpretation of the Q-Q plot).
2. **standard normal quantiles** on the *x* axis. This is the vector of standard, Normal quantiles given $N$ elements in the vector. "Standard Normal" means a normal distribution with mean zero and standard deviation ($\sigma$) one. A Normal quantile is the expected deviation given a probability. For example, if the probability is 0.025, the Normal quantile is `r qnorm(0.025)`. Check your understanding: 2.5% of the values in a Normal distribution with mean 0 and standard deviation one are more negative than `r qnorm(0.025)`. The Normal quantiles of a Normal Q-Q plot are computed for the set of $N$ values that evenly split the probability span from 0 to 1. For $N=20$, this would be

```{r}
p <- data.frame(quantile = qnorm(ppoints(1:20)))
row.names(p) <- ppoints(1:20)
p
```

A Normal Q-Q plot is not a test if the data are Normal. Instead, a Normal Q-Q plot is used to check for characteristic departures from Normality that are signatures of certain well-known distribution families. A researcher can look at a QQ-plot and reason that a departure is small and choose to fit a classic linear model using the Normal distribution. Or, a researcher can look at a QQ-plot and reason that the departure is large enough to fit a generalized linear model with a specific distribution family.

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**Stats 101** A quantile is the value of a distribution that is greater than $p$ percent of the values in the distribution. The 2.5% quantile of a uniform distribution from 0 to 1 is 0.025. The 2.5% quantile of a standard normal distribution is -1.96 (remember that 95% of the values in a standard normal distribution are between -1.96 and 1.96). The 50% quantile of a uniform distribution is 0.5 and the 50% quantile of a standard normal distribution is 0.0 (this is the median of the distribtion -- 50% of the values are smaller and 50% of the values are larger).
</div> 

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**Stats 201** A Q-Q plot more generally is a scatter plot of two vectors of quantiles either of which can come from a sample or a theoretical distribution. In the GLM chapter, the text will introduce Q-Q plots of residual quantiles transformed to have an expected uniform distribution. These are plotted against theoretical uniform quantiles from 0 to 1.
</div> 

#### Normal QQ-plot of the fake data generated above

```{r model-check-qq, echo=FALSE}
# fake data created above
# histogram
gg1_treatment <- gghistogram(data = fd,
                          x = "Response",
                          color="Treatment",
                          fill="Treatment",
                          add = "mean",
                          rug = FALSE,
                          bins=30,
                          palette = pal_okabe_ito) +
  theme_pubr(base_size = 8) +
#  theme(legend.position = "none") +
  NULL

gg2_treatment <- ggstripchart(data = fd,
                           x = "Treatment",
                           y = "Response",
                           color = "Treatment",
                           add = "boxplot",
                           alpha = 0.2,
                           palette = pal_okabe_ito) +
  theme_pubr(base_size = 8) +
#  theme(legend.position = "none") +
  NULL

fd[, Residual := residuals(lm(Response ~ Treatment))]
gg3_treatment <- ggqqplot(data = fd,
                          x = "Residual",
                          title="Normal") +
  theme_pubr(base_size = 8)

gg3_treatment
```

If the sampled distribution approximates a sample from a normal distribution, the scatter should fall along a line from the bottom, left to the top, right of the plot. The interpretation of a normal Q-Q plot is enhanced with a line of "expected values" of the sample quantiles if the sample residuals are drawn from a normal distribution. The closer the sample quantiles are to the line, the more closely the residuals approximate the expectation from a normal distribution. Because of sampling, the sampled values always deviate from the line, especially at the ends. The shaded gray area in the Q-Q plot in Figure \@ref(fig:model-check-qq) are the 95% confidence bands of the quantiles. A pattern of observed quantiles with some individual points outside of these boundaries indicates a sample that would be unusual if sampled from a Normal distribution.

Biological datasets frequently have departures on a Normal Q-Q plot that are characteristic of specific distribution families, including lognormal, binomial, poisson, negative binomial, gamma, and beta. It is useful to learn how to read a Normal Q-Q plot to help guide how to model your data.

What about the intepretation of the Q-Q plot in Figure \@ref(fig:model-check-qq)? At the small end of the distribution (bottom-left), the sample values are a bit more negative than expected, which means the left tail is a bit extended. At the large end (upper-right), the sample values are, a bit less positive than expected, which means the right tail is a bit shortened. This is a departure in the direction of a left skewed distribution. Should we fit a different model given these deviations? To guide us, we compare the quantiles to the 95% confidence band of the quantiles. Clearly the observed quantiles are within the range of quantiles that we'd expect if sampling from a Normal distribution.

### Mapping QQ-plot departures from Normality

Let's look at simulated samples drawn from non-normal distributions to identify their characteristic deviations. Each set of plots below shows

1. (left panel) A histogram of 10,000 random draws from the non-Normal distribution (blue). This histogram is superimposed over that of 10,000 random draws from a Normal distribution (orange) with the same mean and variance as that of the non-normal distribution.
2. (middle panel) Box plots and strip chart of a random subset ($N=1000$) of data in the left panel.
3. (right panel) Normal Q-Q plot of the non_Normal data only.

```{r model-check-normal, echo=FALSE}
# normal
n <- 10^4
mu <- 10
sigma <- 2

set.seed(1)

x_levels <- c("normal", "normal-2")

fd_normal <- data.table(
  distribution = factor(rep(x_levels, each = n), x_levels),
  response = c(rnorm(n, mean = mu, sd = sigma),
               rnorm(n, mean = mu, sd = sigma))
)
check_variance <- FALSE
if(check_variance == TRUE){
  fd_normal[, .(mean = mean(response),
               sd = sd(response),
               min = min(response)),
           by = distribution]
}

# histogram
gg1_normal <- gghistogram(data = fd_normal,
                          x = "response",
                          color="distribution",
                          fill="distribution",
                          add = "mean",
                          rug = FALSE,
                          bins=30,
                          palette = pal_okabe_ito) +
  theme_pubr(base_size = 8) +
  theme(legend.position = "none")

fd_subset <- fd_normal[, .SD[sample(.N,  10^3)],
                       by = distribution]
gg2_normal <- ggstripchart(data = fd_subset,
                           x = "distribution",
                           y = "response",
                           color = "distribution",
                           add = "boxplot",
                           alpha = 0.2,
                           palette = pal_okabe_ito) +
  theme_pubr(base_size = 8) +
  theme(legend.position = "none")

fd_subset <- fd_normal[distribution == x_levels[2],
                       .SD[sample(.N,  10^3)],
                       by = distribution]

fd_subset[, residual := residuals(lm(response ~ 1))]
gg3_normal <- ggqqplot(data = fd_subset,
                          x = "residual",
                          title="Normal") +
  theme_pubr(base_size = 8)

```

```{r model-check-lognormal, echo=FALSE, eval=FALSE}
n <- 10^4
mu <- 10
sigma <- 2
beta <- sigma*2

# lognormal
set.seed(1)
# see wikipedia for parameterization given expected mean and variance
location_cn <- log(mu^2/sqrt(sigma^2 + mu^2))
shape_cn <- sqrt(log(1 + sigma^2/mu^2))

# for effect using lognormal
# cv is constant in lognormal
cv <- sqrt(exp(shape_cn^2) - 1)
sigma_tr <- cv*(mu + beta)
location_tr <- log((mu + beta)^2/sqrt(sigma_tr^2 + (mu + beta)^2))
shape_tr <- sqrt(log(1 + sigma_tr^2/(mu + beta)^2))

x_levels <- c("normal", "lognormal")

fd_lognorm <- data.table(
  distribution = factor(rep(x_levels, each = n), x_levels),
  response = c(rnorm(n, mean = mu, sd = sigma),
               rlnorm(n, 
                      meanlog = location_cn,
                      sdlog = shape_cn)))
fd_lognorm[, .(mean = mean(response),
             var = sd(response)^2), by = distribution]

# histogram
gg1_lognormal <- gghistogram(data = fd_lognorm,
                             x = "response",
                             color="distribution",
                             fill="distribution",
                             add = "mean",
                             rug = FALSE,
                             bins=30,
                             palette = pal_okabe_ito) +
  theme(legend.position = "none")

fd_subset <- fd_lognorm[, .SD[sample(.N,  10^3)],
                       by = distribution]
gg2_lognormal <- ggstripchart(data = fd_subset,
                              x = "distribution",
                              y = "response",
                              color = "distribution",
                              add = "boxplot",
                              alpha = 0.2,
                              palette = pal_okabe_ito) +
  theme(legend.position = "none")

fd_subset <- fd_lognorm[distribution == x_levels[2],
                       .SD[sample(.N,  10^3)],
                       by = distribution]

fd_subset[, residual := residuals(lm(response ~ 1))]
gg3_lognormal <- ggqqplot(data = fd_subset,
                          x = "residual",
                          title="Lognormal")
```

```{r model-check-gamma, echo=FALSE, eval=TRUE}
# gamma
n <- 10^4
mu <- 10
set.seed(1)

# parameters shape = a and scale = s
# mean and variance are E(X) = a*s and Var(X) = a*s^2
scale_cn <- 0.8
shape_cn <- mu/scale_cn

sigma <- sqrt(shape_cn*scale_cn^2)

x_levels <- c("normal", "gamma")

fd_gamma <- data.table(
  distribution = factor(rep(x_levels, each = n), x_levels),
  response = c(rnorm(n, mean = mu, sd = sigma),
               rgamma(n, 
                      shape = shape_cn,
                      scale = scale_cn))
)
check_variance <- FALSE
if(check_variance == TRUE){
  fd_gamma[, .(mean = mean(response),
               sd = sd(response),
               min = min(response)),
           by = distribution]
}

# histogram
gg1_gamma <- gghistogram(data = fd_gamma,
                         x = "response",
                         color="distribution",
                         fill="distribution",
                         add = "mean",
                         rug = FALSE,
                         bins=30,
                         palette = pal_okabe_ito) +
  theme(legend.position = "none")

fd_subset <- fd_gamma[, .SD[sample(.N,  10^3)],
                       by = distribution]
gg2_gamma <- ggstripchart(data = fd_subset,
                          x = "distribution",
                          y = "response",
                          color = "distribution",
                          add = "boxplot",
                          alpha = 0.2,
                          palette = pal_okabe_ito) +
  theme(legend.position = "none")


fd_subset <- fd_gamma[distribution == x_levels[2],
                       .SD[sample(.N,  10^3)],
                       by = distribution]

fd_subset[, residual := residuals(lm(response ~ 1))]
gg3_gamma <- ggqqplot(data = fd_subset,
                          x = "residual",
                          title="Gamma")

```

```{r model-check-negbinom, echo=FALSE, eval=TRUE}
# neg binomial
n <- 10^4
mu <- 10
set.seed(1)

# mean mu and variance mu + mu^2/theta
theta = 4
sigma <- sqrt(mu + mu^2/theta)

x_levels <- c("normal", "right skew")

fd_negbin <- data.table(
  distribution = factor(rep(x_levels, each = n), x_levels),
  response = c(rnorm(n, mean = mu, sd = sigma),
               rnegbin(n, mu, theta = 2))
)
check_variance <- FALSE
if(check_variance == TRUE){
  fd_negbin[, .(mean = mean(response),
               sd = sd(response),
               min = min(response)),
           by = distribution]
}

# histogram
gg1_negbin <- gghistogram(data = fd_negbin,
                          x = "response",
                          color="distribution",
                          fill="distribution",
                          add = "mean",
                          rug = FALSE,
                          bins=30,
                          palette = pal_okabe_ito) +
  theme_pubr(base_size = 8) +
  theme(legend.position = "none")

fd_subset <- fd_negbin[, .SD[sample(.N,  10^3)],
                       by = distribution]
gg2_negbin <- ggstripchart(data = fd_subset,
                           x = "distribution",
                           y = "response",
                           color = "distribution",
                           add = "boxplot",
                           alpha = 0.2,
                           palette = pal_okabe_ito) +
  theme_pubr(base_size = 8) +
  theme(legend.position = "none")

fd_subset <- fd_negbin[distribution == x_levels[2],
                       .SD[sample(.N,  10^3)],
                       by = distribution]
fd_subset[, residual := residuals(lm(response ~ 1))]
gg3_negbin <- ggqqplot(data = fd_subset,
                          x = "residual",
                          title="Right skew") +
  theme_pubr(base_size = 8)

```


```{r model-check-t, echo=FALSE, eval=TRUE}
# t (heavy tail)
n <- 10^4
mu <- 10
set.seed(2)

# mean 0 (for n > 1) and variance n/(n-2) (for n > 2)
n_cn <- 6
df_cn <- n_cn - 2
sigma <- sqrt(n_cn/df_cn)

x_levels <- c("normal", "heavy tail")

fd_t <- data.table(
  distribution = factor(rep(x_levels, each = n), x_levels),
  response = c(rnorm(n, mean = mu, sd = sigma*2),
               rt(n, df = n_cn)*2 + mu)
)
check_variance <- FALSE
if(check_variance == TRUE){
  fd_t[, .(mean = mean(response),
               sd = sd(response),
               min = min(response)),
           by = distribution]
}

# histogram
gg1_t <- gghistogram(data = fd_t,
                     x = "response",
                     color="distribution",
                     fill="distribution",
                     add = "mean",
                     rug = FALSE,
                     bins=30,
                     palette = pal_okabe_ito) +
  theme_pubr(base_size = 8) +
  theme(legend.position = "none")

fd_subset <- fd_t[, .SD[sample(.N,  10^3)],
                       by = distribution]
gg2_t <- ggstripchart(data = fd_subset,
                      x = "distribution",
                      y = "response",
                      color = "distribution",
                      add = "boxplot",
                      alpha = 0.2,
                      palette = pal_okabe_ito) +
  theme_pubr(base_size = 8) +
  theme(legend.position = "none")

fd_subset <- fd_t[distribution == x_levels[2],
                       .SD[sample(.N,  10^3)],
                       by = distribution]

fd_subset[, residual := residuals(lm(response ~ 1))]
gg3_t <- ggqqplot(data = fd_subset,
                          x = "residual",
                          title="Heavy Tail") +
  theme_pubr(base_size = 8)

```


```{r model-check-beta-left, echo=FALSE, eval=TRUE}
# t (heavy tail)
n <- 10^4
mu <- 0.8
set.seed(1)

#shape1 = a and shape2 = b
# The mean is a/(a+b) and the variance is ab/((a+b)^2 (a+b+1)).
# b = a/mu - a
a <- 10
b <- a/mu - a
sigma <- sqrt(a*b/((a+b)^2 * (a+b+1)))
# a/(a+b)
x_levels <- c("normal", "left skew")

fd_beta <- data.table(
  distribution = factor(rep(x_levels, each = n), x_levels),
  response = c(rnorm(n, mean = mu, sd = sigma),
               rbeta(n, shape1 = a, shape2 = b))
)
check_variance <- FALSE
if(check_variance == TRUE){
  fd_beta[, .(mean = mean(response),
               sd = sd(response),
               min = min(response)),
           by = distribution]
}

# histogram
gg1_beta <- gghistogram(data = fd_beta,
                        x = "response",
                        color="distribution",
                        fill="distribution",
                        add = "mean",
                        rug = FALSE,
                        bins=30,
                        palette = pal_okabe_ito) +
  theme_pubr(base_size = 8) +
  theme(legend.position = "none")

fd_subset <- fd_beta[, .SD[sample(.N,  10^3)],
                       by = distribution]
gg2_beta <- ggstripchart(data = fd_subset,
                         x = "distribution",
                         y = "response",
                         color = "distribution",
                         add = "boxplot",
                         alpha = 0.2,
                         palette = pal_okabe_ito) +
  theme_pubr(base_size = 8) +
  theme(legend.position = "none")

fd_subset <- fd_beta[distribution == x_levels[2],
                       .SD[sample(.N,  10^3)],
                       by = distribution]

fd_subset[, residual := residuals(lm(response ~ 1))]
gg3_beta <- ggqqplot(data = fd_subset,
                          x = "residual",
                          title="Left skew") +
  theme_pubr(base_size = 8)

# gg1_beta
# gg2_beta
# gg3_beta
```

```{r model-check-plot-plot-normal, echo = FALSE}
gg_normal <- plot_grid(gg1_normal, gg2_normal, gg3_normal,
                       nrow = 1)
gg_right <- plot_grid(gg1_negbin, gg2_negbin, gg3_negbin,
                       nrow = 1)
gg_left <- plot_grid(gg1_beta, gg2_beta, gg3_beta,
                       nrow = 1)
gg_heavy <- plot_grid(gg1_t, gg2_t, gg3_t,
                       nrow = 1)
```

**Skewed-Right Q-Q**

```{r, echo = FALSE, out.width="90%", fig.asp = .4}
gg_right
```

The Normal Q-Q plot of a sample from a right-skewed distribution is characterized by sample quantiles at the high (right) end being more positive than the expected Normal quantiles. Often, the quantiles at the low (left) end are also less negative than the expected normal quantiles. The consequence is a concave up pattern.

The histograms in the left panel explain this pattern. The right tail of the skewed-right distribution extends further than the right tail of the Normal. It is easy to see from this that, if we rank the values of each distribution from small to large (these are the quantiles), the upper quantiles of the skewed-right distribution will be larger than the matching quantile of the Normal distribution. For example, the 99,990th quantile for the skewed-right distribution will be much more positive than the 99,990th quantile for the Normal distribution. The opposite occurs at the left tail, which extends further in the negative direction in the Normal than the skewed-right distribution.

The middle panel compares a boxplot and stripchart of samples from the two distributions to show what researchers should look for in their own publication-ready plots as well as the published plots of colleagues. The skewed-right plot exhibits several hallmarks of a skewed-right distribution including 1) a median line (the horizontal line within the box) that is closer to the 25th percentile line (the lower end of the box) than to the 75th percentile line (the upper end of the box), 2) a longer upper than lower whisker (the vertical lines extending out of the box), 3) more outliers above the upper whisker than below the lower whisker, and 4) a lengthened, upward smear of the scatter of points at the high end of the values, relative to the more compact smear at the low end of the values.

**Skewed-Left Q-Q**

```{r, echo = FALSE, out.width="90%", fig.asp = .4}
gg_left
```

The Normal Q-Q plot of a sample from a left-skewed distribution is characterized by sample quantiles at the low (left) end being more negative than the expected Normal quantiles. Often, the quantiles at the high (right) end are also less positive than the expected normal quantiles. The consequence is a concave down pattern.

The histograms in the left panel explain this pattern. The left tail of the skewed-left distribution extends further than the left tail of the Normal. It is easy to see from this that, if we rank the values of each distribution from small to large (these are the quantiles), the lower quantiles of the skewed-left distribution will be more negative than the matching quantile of the Normal distribution. For example, the 10th quantile for the skewed-left distribution will be much more negative than the 10th quantile for the Normal distribution. The opposite occurs at the right tail, which extends further in the positive direction in the Normal than the skewed-left distribution.

The skewed-left plot in the middle panel highlights several hallmarks of a skewed-left distribution including 1) a median line (the horizontal line within the box) that is closer to the 75th percentile line (the lower end of the box) than to the 25th percentile line (the upper end of the box), 2) a longer lower than upper whisker (the vertical lines extending out of the box), 3) more outliers below the lower whisker than above the upper whisker, and 4) a lengthened, downward smear of the scatter of points at the low end of the values, relative to the more compact smear at the upper end of the values.

**Heavy Tail Q-Q**

```{r, echo = FALSE, out.width="90%", fig.asp = .4}
gg_heavy
```

The Normal Q-Q plot of a sample from a heavy-tail distribution is characterized by sample quantiles at the low (left) end being more negative than the expected Normal quantiles and quantiles at the high (right) end that are more positive than the expected normal quantiles.

The histograms in the left panel explain this pattern. At each tail, the heavy-tail distribution has more **density** -- there are more values far from the mean -- compared to the Normal distribution. This is the origin of "heavy tail". It is easy to see from this that, if we rank the values of each distribution from small to large (these are the quantiles), the lower quantiles of the heavy tail distribution will be more negative than the matching quantile of the Normal distribution. For example, the 10th quantile for the heavy-tail distribution will be much more negative than the 10th quantile for the Normal distribution. Likewise, the upper quantiles of the heavy tail distribution will be more positive than the matching quantile of the Normal distribution. For example, the 99,990th quantile for the heavy-tail distribution will be much more positive than the 99,990th quantile for the Normal distribution.

The heavy-tail plot in the middle panel shows more boxplot outliers than in the Normal plot. This would be hard to recognize in a plot of real data.

#### Mapping characteristic departures on a Q-Q plot to specific distributions

1. Continuous response variables of length, area, weight, or duration will often look like samples from a continous probability distribution that is right-skewed, such as the **lognormal** or **gamma** distributions.
2. Count response variables will frequently look like samples from a discrete probability distribution that is right-skewed, such as the **poisson**, **quasi-poisson**, or **negative binomial** distributions.
3. Proportion (fraction of a whole) response variables will frequently look like samples from a continuous probability distribution bounded by 0 and 1, such as the **beta** distribution. Samples from a beta distribution can be left skewed, if the mean is near 1, right-skewed, if the mean is near zero, or symmetrical, if the mean is near 0.5.

#### Pump your intuition -- confidence bands of a Q-Q plot

In introducing the confidence bands of the Q-Q plot above, I stated "A pattern of observed quantiles with some individual points outside of these boundaries indicates a sample that would be unusual if sampled from a Normal distribution." Let's use a parametric bootstrap to explore this.

1. Sample $n$ values from a Normal distribution
2. Compute the sample quantiles by re-ordering the residuals of the sampled values from the sampled mean, from most negative to most positive.
3. Plot the quantiles against Normal quantiles for $n$ points.
4. Repeat steps 1-3 $n\_iter$ times, superimposing the new sample quantiles over all previous sample quantiles. This creates a band of all sample quantiles over $n\_iter$ iterations of sampling $n$ values from a Normal distribution.
5. At each value of the Normal quantile, compute the 95 percentile range of the sampled quantiles. Draw a ribbon inside these boundaries.

```{r model-check-pump-intuition, eval=FALSE}
n_iter <- 1000
n <- 20
normal_qq <- ppoints(n) %>%
  qnorm()
sample_qq <- numeric(n_iter*n)
inc <- 1:n
for(iter in 1:n_iter){
  y <- rnorm(n)
  y_res <- y - mean(y)
  sample_qq[inc] <- y_res[order(y_res)]
  inc <- inc + n
}

qq_data <- data.table(normal_qq = normal_qq,
                      sample_qq = sample_qq)

qq_ci <- qq_data[, .(median = median(sample_qq),
                     lower = quantile(sample_qq, 0.025),
                     upper = quantile(sample_qq, 0.975)),
                 by = normal_qq]

ggplot(data = qq_data,
       aes(x = normal_qq,
           y = sample_qq)) +
  geom_point(alpha = 0.2) +
  geom_ribbon(data = qq_ci,
              aes(ymin = lower,
                  ymax = upper,
                  y = median,
                  fill = "band"),
              fill = pal_okabe_ito[1],
              alpha = 0.3) +
  xlab("Normal Quantile") +
  ylab("Sample Quantile") +
  theme_grid() +
  
  NULL
```

### Model checking homoskedasticity

## Using R

Source: [Wellenstein, M.D., Coffelt, S.B., Duits, D.E., van Miltenburg, M.H., Slagter, M., de Rink, I., Henneman, L., Kas, S.M., Prekovic, S., Hau, C.S. and Vrijland, K., 2019. Loss of p53 triggers WNT-dependent systemic inflammation to drive breast cancer metastasis. Nature, 572(7770), pp.538-542.](https://www.nature.com/articles/s41586-019-1450-6){target="_blank"}

[Public source](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6707815/){target="_blank"}

[Data source](https://www.nature.com/articles/s41586-019-1450-6#Sec22){target="_blank"}

```{r echo=TRUE}
data_from <- "Loss of p53 triggers WNT-dependent systemic inflammation to drive breast cancer metastasis"
file_name <- "41586_2019_1450_MOESM3_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)
  
treatment_levels <- c("Trp53+/+", "Trp53-/-")
fig1f <- read_excel(file_path,
                     sheet = "Fig. 1f",
                     range = "A2:B23") %>%
  data.table() %>%
  melt(measure.vars = treatment_levels,
       variable.name = "treatment",
       value.name = "il1beta")
fig1f[, treatment := factor(treatment, treatment_levels)]

# be careful of the missing data. This can create mismatch between id and residual unless specified in lm

# head(fig1f)
```

Fit a linear model

```{r}
m1 <- lm(il1beta ~ treatment,
         data = fig1f)
```

```{r echo=FALSE, eval=FALSE}
m1 <- lm(il1beta ~ treatment,
         data = fig1f)
m1_res <- residuals(m1)
sigma_m1_res <- sd(m1_res)
m1_res_st <- m1_res/sigma_m1_res
n <- length(m1_res)

normal_qq <- ppoints(n) %>%
  qnorm()
sample_qq <- m1_res[order(m1_res)]

# mean + sd
parametric_slope <- sd(sample_qq)
parametric_intercept <- mean(sample_qq)

# quantile
m1_quantiles <- quantile(m1_res, c(.25, .75))
qnorm_quantiles <- qnorm( c(.25, .75))
m1_diff <- m1_quantiles[2] - m1_quantiles[1]
qnorm_diff <- qnorm_quantiles[2] - qnorm_quantiles[1] # = 1.349
quantile_slope <- m1_diff/qnorm_diff
quantile_intercept <- median(sample_qq)

# robust
qq_observed <- data.table(normal_qq = normal_qq,
                    sample_qq = sample_qq)
m2 <- rlm(sample_qq ~ normal_qq, data = qq_observed)
robust_intercept <- coef(m2)[1]
robust_slope <- coef(m2)[2]

# re-sample ribbon
resample_qq <- numeric(n_iter*n)
inc <- 1:n
for(iter in 1:n_iter){
  y <- rnorm(n, sd = sigma_m1_res)
  y_res <- y - mean(y)
  resample_qq[inc] <- y_res[order(y_res)]
  inc <- inc + n
}

qq_sim <- data.table(normal_qq = normal_qq,
                     resample_qq = resample_qq)

qq_ci <- qq_sim[, .(median = median(resample_qq),
                     lower = quantile(resample_qq, 0.025),
                     upper = quantile(resample_qq, 0.975)),
                 by = normal_qq]

ggplot(data = qq_observed,
       aes(x = normal_qq, y = sample_qq)) +
  geom_point() +
  
  # ribbon
  geom_ribbon(data = qq_ci,
              aes(ymin = lower,
                  ymax = upper,
                  y = median,
                  fill = "band"),
              fill = pal_okabe_ito[1],
              alpha = 0.3) +
  # auto
  geom_qq_line(aes(sample = sample_qq),
               color = "black",
               size = 2,
               alpha = 0.3) +
  # parametric
  geom_abline(aes(intercept = parametric_intercept,
                  slope = parametric_slope),
              color = "blue") +
  # quantile
  geom_abline(aes(intercept = quantile_intercept,
                  slope = quantile_slope),
              color = "red") +
  # robust
  geom_abline(aes(intercept = robust_intercept,
                  slope = robust_slope),
              color = "green") +
  NULL

# car plot
qqPlot(m1, id = FALSE)

# base R plot
qqnorm(m1_res)
qqline(m1_res)

# ggpubr plot
ggqqplot(data = data.table(m1_res=m1_res),
         x = "m1_res")


```

### Normal Q-Q plots

`car::qqPlot` has several important arguments to control the type of Q-Q plot. The function uses base graphics instead of ggplot2. Typically, these plots would not be published other than possibly a supplement. [Q-Q Plots and Worm Plots from Scratch](https://www.tjmahr.com/quantile-quantile-plots-from-scratch/){target="_blank"} is a good source of some of the arguments in `qqPlot`. Three important arguments are:

1. simulate. If passing a `lm` object, then the default confidence band is generated by a parametric bootstrap (`simulate = TRUE`). This band will differ somewhat each time you replot unless you set the seed with `set.seed`. Setting the argument `simulate = FALSE` returns the parametric band.
2. line. If passing a `lm` object, then the default line is a fit from a robust regression (`line = "robust"`). Setting the argument `line = "quartiles"` fits a line throught the 25th and 75th percentile (or "quartiles") quantiles.
3. id. The default identifies the index of the two points with the most extreme quartiles. Set to FALSE to hide.

The robust line is more sensitive to departures from Normality than the quartiles line.

```{r}
# defaults: robust line with bootstrap CI
set.seed(1)
qqPlot(m1, id = FALSE)
```

```{r, eval=FALSE}
# classic: standard line with parametric CI
qqPlot(m1,
       line = "quartiles",
       simulate = FALSE,
       id = FALSE)
```

`ggpubr::ggqqplot` generates a pretty, ggplot2 based Normal Q-Q plot, using the standard method for computing the line and confidence band.

```{r}
m1_residuals <- data.table(m1_residuals = residuals(m1))
m1_residuals[, studentized := m1_residuals/sd(m1_residuals)]
ggqqplot(data = m1_residuals,
                          x = "studentized")
```









<!--chapter:end:chapters/26-model-checking.Rmd-->

# Model Fitting  and Model Fit (OLS)

## Least Squares Estimation and the Decomposition of Variance

The linear models in the last chapter and for much of this book are fit to data using a method called "ordinary least squares" (OLS). This chapter explores the meaning of OLS and related statistics, including $R^2$, as well as some alternative methods for bivariate regression.

## OLS regression

```{r setup-ls, echo=FALSE, message=FALSE}
library(ggplot2)
library(readxl)
library(data.table)
library(lmodel2)
library(cowplot)

data_path <- "../data" # notebook
data_path <- "data" # bookdown
```

```{r fake-data-ma, echo=FALSE, message=FALSE, warning=FALSE}
n <- 2000
set.seed(7)
density <- rnorm(n, mean=3, sd=0.5)
# qplot(density)
beta_0 <- 600
beta_1 <- 700
fcm <- beta_0 + beta_1*density + rnorm(n, sd=700)
r <- cor(density,fcm)
fit <- lm(fcm ~ density)

# model 2 regression
fit2 <- data.table(lmodel2(fcm ~ density)$regression.results)
sse <- sum(residuals(fit)^2)
s.error <- sqrt(sse/(n-2))
ssm <- sum((fitted(fit) - mean(fcm))^2)
sst <- var(fcm)*(n-1)

# reshape the fig to show how MA slope appearance changes with aspect ratio
qplot(x=density, y=fcm) +
  #geom_abline(slope=fit2[Method=="MA", Slope], intercept=fit2[Method=="MA", Intercept]) +
  geom_abline(slope=fit2[Method=="SMA", Slope], intercept=fit2[Method=="SMA", Intercept]) +
  #geom_abline(slope=fit2[Method=="OLS", Slope], intercept=fit2[Method=="OLS", Intercept]) +
  NULL

# check
# sma or reduced major axis
sma <- sd(fcm)/sd(density)

# ma regression
fd <- data.table(fcm=fcm, density=density)
E <- eigen(cov(fd))$vectors[,1]
ma <- E[1]/E[2]
```

The fake data illustrated in the scatterplot above (Figure \@ref(fig:fake-data-ma)) were modeled to look something like the squirrel fecal cortisol metabolite data in the previous chapter. If a typical student is asked to draw a regression line through the scatter, they typically draw a line similar to that in Figure \@ref(fig:fake-data-ma). This line is not the OLS regression line but the major axis of an elipse that encloses the scatter of points--that students invariably draw this line suggests that the brain interprets the major axis of an elliptical scatter of points as a trend (This major axis line is an alternative method for estimating a slope and is known as standard major-axis regression. More about this at the end of this chapter.)

```{r fake-data-ols, echo=FALSE}
qplot(x=density, y=fcm) +
  geom_abline(slope=fit2[Method=="SMA", Slope], intercept=fit2[Method=="SMA", Intercept]) +
  geom_abline(slope=fit2[Method=="OLS", Slope], intercept=fit2[Method=="OLS", Intercept], color="red") +
  NULL


```

The OLS regression line is the red line in Figure \@ref(fig:fake-data-ols) -- the standard major axis line is left for comparison). The OLS regression line

1. passes through the bivariate mean ($\bar{x}$, $\bar{y}$) of the scatter, and
2. minimizes the sum of the squared deviations from each point to it's modeled value $\sum{(y_i - \hat{y}_i)^2}$

There are an infinite number of lines that pass through the bivariate mean (think of anchoring a line at the bivariate mean and spinning it). The OLS line is the line that minimizes the squared (vertical) deviations ("least squares").

For a bivariate regression, the slope (coefficient $b_1$ of $X$) of the OLS model fit is computed by

\begin{equation}
b_1 = \frac{\mathrm{COV}(X, Y)}{\mathrm{VAR}(X)}
\end{equation}

This equation is worth memorizing. We will generalize this into a more flexible equation in a few chapters. 

## How well does the model fit the data? $R^2$ and "variance explained"

Let's switch to real data.

1. Source: Dryad Digital Repository. https://doi.org/10.5061/dryad.056r5
2. File: "Diet-shift data.xls"

Fish require arachidonic acid (ARA) and other highyly unsaturated fatty acids in their diet and embryo and yolk-stage larvae obtain these from yolk. Fuiman and Faulk (xxx) designed an experiment to investigate if red drum (*Sciaenops ocellatus*) mothers provision the yolk with ARA from recent dietary intake or from stored sources in somatic tissues. The data below are from experiment 8. The *x*-axis is the days since a diet shift to more and less ARA ($days$) and the *y*-axis is the ARA content of the eggs ($ARA$).

```{r egg-data, echo=FALSE, message=FALSE}
folder <- "Data from Batch spawning facilitates transfer of an essential nutrient from diet to eggs in a marine fish"
filename <- "Diet-shift data.xls"
file_path <- paste(data_path, folder, filename, sep="/")
egg <- data.table(read_excel(file_path))
exp_8 <- egg[EXPERIMENT==8,]
fit.egg <- lm(ARA_EGG_MG ~ DAYSSINCE, data=exp_8)
exp_8[, yhat:=predict(fit.egg)]
gg <- ggplot(data=exp_8, aes(x=DAYSSINCE, y=ARA_EGG_MG)) +
  geom_smooth(method="lm", se=FALSE) +
  geom_point(size=2) +
  xlab("days since diet shift") +
  ylab("ARA content") +
  NULL
gg
```


The statistic $R^2$ is a measure of the fit of a model to data. The $R^2$ for the fit of the egg data is `r round(summary(fit.egg)$r.squared, 2)`. $R^2$ is the fraction of two variances $\frac{\mathrm{VAR}(Model)}{\mathrm{VAR}(Y)}$, or, the fraction of the variance of $Y$ "explained by the model." The value of $R^2$ ranges from zero (the fit cannot be any worse) to one (the fit is "pefect").

To understand $R^2$, and its computation, a bit more, let's look at three kinds of deviations.

```{r ols-deviations, echo=FALSE, fig.cap="Three kinds of deviations from a fit model. A. Deviations of the measured values from the mean. These are in the numerator of the equation of the sample variance. The dashed line is the mean ARA content. B. Deviations of the measured values from the modeled values. The sum of these deviations squared is what is minimized in an OLS fit. C. Deviations of the modeled values from the mean ARA content. The measured values are in gray, the modeled values in black"}
gg <- ggplot(data=exp_8, aes(x=DAYSSINCE, y=ARA_EGG_MG)) +
  geom_point(size=2) +
  xlab("time since diet shift (days)") +
  ylab("ARA content") +
  NULL

gg_error_deviation <- gg + 
  geom_smooth(method="lm", se=FALSE) +
  geom_linerange(aes(ymax=ARA_EGG_MG, ymin=yhat))
#gg_error_deviation

ybar <- mean(exp_8$ARA_EGG_MG)
gg_total_deviation <- gg +
  geom_hline(yintercept=ybar, linetype=2) +
  geom_linerange(aes(ymax=ARA_EGG_MG, ymin=ybar))
#gg_total_deviation

gg_model_deviation <- gg +
  geom_point(size=2, color="gray") +
  geom_point(size=2, aes(y=yhat)) +
  geom_smooth(method="lm", se=FALSE) +
  geom_hline(yintercept=ybar, linetype=2) +
  geom_linerange(aes(ymax=yhat, ymin=ybar))
#gg_model_deviation

plot_grid(gg_total_deviation, gg_error_deviation, gg_model_deviation, labels = c("A", "B", "C"))
```

Figure \@ref(fig:ols-deviations)A shows the deviations from the measured values to the mean value (dashed line). These are the deviations in the numerator of the equation to compute the variance of $ARA_EGG_MG$. Figure \@ref(fig:ols-deviations)B shows the deviations of the measured values from the modeled values. The sum of these deviations squared is what is minimized by the OLS fit. The bigger these deviations are, the worse the model fit. Figure \@ref(fig:ols-deviations)C shows the deviations of the modeled values to the mean value. The bigger these deviations are, the better the model fit.

The sums of the squares of these deviations (or "sums of squares") have names:

\begin{equation}
\mathrm{SS(total)} = \sum{(y_i - \bar{y})^2}
\end{equation}

\begin{equation}
\mathrm{SS(error)} = \sum{(y_i - \hat{y_i})^2}
\end{equation}

\begin{equation}
\mathrm{SS(model)} = \sum{(\hat{y_i} - \bar{y})^2}
\end{equation}

Again, $\mathrm{SS(total)}$ is the numerator of the equation for the sample variance. It is called "s-s-total" because $\mathrm{SS(total)} = \mathrm{SS(model)} + \mathrm{SS(error)}$. That is, the total sums of squares can be **decomposed** into two **components**: the modeled sums of squares and the error sums of squares. Given these components, it's easy to understand $R^2$

\begin{equation}
R^2 = \frac{SS(model)}{SS(total)}
\end{equation}

$R^2$ is the fraction of the total sums of squares that is due to (or "explained by") the model sums of squares. Above I said that $R^2$ is the fraction of *variance* explained by the model. Equation xxx is a ratio of variance, but the $(n-1)^{-1}$ in both the numerator and the denominator cancel out. Finally, many sources give the equation for $R^2$ as

\begin{equation}
R^2 = 1- \frac{SS(error)}{SS(total)}
\end{equation}

which is an obvious alternative given the decomposition. I prefer the former equation because it emphasizes the model fit instead of model ill-fit.



<!--chapter:end:chapters/27-model-fit.Rmd-->

# Best practices -- issues in inference

```{r best-setup, warning=FALSE, message=FALSE, echo=FALSE}
library(readxl)
library(janitor)
library(data.table)
library(MASS)
library(emmeans)
library(lmPerm, quietly=T)
library(simpleboot)
library(boot)
library(ggpubr)
library(doBy)
library(here)
here <- here::here
data_path <- "data"
```

## Multiple testing

**Multiple testing** is the practice of adjusting *p*-values (and less commonly confidence intervals) to account for the expected increase in the frequency of Type I error when there are multiple tests (typically Null Hypothesis Significance Tests). Multiple testing tends to arise in two types of situations:

1. Multiple pairwise contrasts among treatment levels (or combinations of levels) are estimated.
2. The effects of a treatment on multiple responses are estimated. This can arise if
    a. there are multiple ways of measuring the consequences of something -- for example, an injurious treatment on plant health might effect root biomass, shoot biomass, leaf number, leaf area, etc.
    b. one is exploring the consequences of an effect on many, many outcomes -- for example, the expression levels of 10,000 genes between normal and obese mice.

Despite the ubiquitous presence of multiple testing in elementary biostatistics textbooks, in the applied biology literature, and in journal guidelines, the practice of adjusting *p*-values for multiple tests is highly controversial among statisticians. My thoughts:

1. In situations like (1) above, I advocate that researchers **do not adjust p-values for multiple tests**. In general, its a best practice to only estimate contrasts for which you care about because of some *a priori* model of how the system works. If you compare all pairwise contrasts of an experiment with many treatment levels and/or combinations, expect to find some false discoveries.
2. In situations like (2a) above, I advocate that researchers **do not adjust p-values for multiple tests**.
3. In situations like (2b) above, adjusting for the **False Discovery Rate** is an interesting approach. But, recognize that tests with small *p*-values are *highly provisional* discoveries of a patterns only and not a discovery of the causal sequelae of the treatment. For that, one needs to do the hard work of designing experiments that rigorously probe a working, mechanistic model of the system.

Finally, recognize that anytime there are multiple tests, Type M errors will arise due to the vagaries of sampling. This means that in a rank-ordered list of the effects, those at the top have measured effects that are probably bigger than the true effect. An alternative to adjusted *p*-values is a **penalized regression** model that shrinks effects toward the mean effect.

### Some background
#### Family-wise error rate

The logic of multiple testing goes something like this: the more tests that a researcher does, the higher the probability that a false positive (Type I error) will occur, therefore a researcher should should adjust *p*-values so that the Type I error over the set (or "family") of tests is 5%. This adjusted Type I error rate is the "family-wise error rate". 

If a researcher carries out multiple tests *of data in which the null hypothesis is true*, what is the probability of finding at least one Type I error? This is easy to compute. If the frequency of Type I error for a single test is $\alpha$, then the probability of no Type I error is $1 - \alpha$. For two tests, the probability of no Type I error in either test is the product of the probability for each test, or $(1 - \alpha)^2$. By the same logic, for $m$ tests, the probabilty of no type I error in any of the tests is $(1 - \alpha)^m$. The probability of at least one type one error, across the $m$ tests, then, is $1 - (1 - \alpha)^m$. A table of these probabilities for different $m$ is given below. If the null is true in all tests, then at least one Type I error is more likely than not if there are 14 tests, and close to certain if there more than 50 tests. Don't skip over this paragraph -- the logic is important even if I don't advocate adjusting for multiple tests.

```{r best-type1-table, echo=FALSE}
m <- c(1, 3, 6, 10, 50, 100)
p <- 1 - (1 - 0.05)^m
knitr::kable(data.table(m = m,
                        p = p),
             digits=c(0, 2),
             caption="Probability of at least one type I error within the set of multiple tests, for data in which the null hypothesis is true. The Type I error rate for a single test is 0.05. The number of tests is m. The probability is p.")
```
#### False discovery rate

If a researcher carries out thousands of tests to "discover" new facts, and uses $p < 0.05$ as evidence of discovery, then what is the frequency of **false discoveries**?

#### p-value filter I -- Inflated effects

If a researcher caries out many tests, and ranks the effects by magnitude or *p*-value, then the effect sizes of the largest effects will be inflated. Before explaining why, let's simulate this using an experiment of allelopathic effects of the invasive garlic mustard (*Alliaria petiolata*) on gene expression in the native American ginseng (*Panax quinquefolius*). In the treated group, we have ten pots, each with an American ginseng plant grown in a container with a mustard plant. In the control group, we have ten pots, each with an American ginseng plant grown in a container with another American ginseng. I've simulated the response of 10,000 genes. The treatment has a true effect in 10% of the 10,000 genes but most effects are very small.

```{r}
set.seed(4)
p <- 10^4 # number of genes
pt <- 0.1*p # number of genes with true response to treatment
n <- 10

# sample the gene effects from an exponential distribution
theta <- .3
beta <- c(rexp(pt, rate=1/theta),
          rep(0, (p-pt))) # the set of 10,000 effects

# sample the variance of the expression level with a gamma, and set a minimum
sigma <- rgamma(p, shape=2, scale=1/4) + 0.58
# quantile(sigma, c(0.001, 0.1, 0.5, 0.9, 0.999))

Y1 <- matrix(rnorm(n*p, mean=0, sd=rep(sigma, each=n)), nrow=n)
Y2 <- matrix(rnorm(n*p, mean=rep(beta, each=n), sd=rep(sigma, each=n)), nrow=n) # check
# use n <- 10^4 to check
# apply(y2, 2, mean)[1:5]
# b[1:5]
x <- rep(c("cn","tr"), each=n)
bhat <- numeric(p)
p.value <- numeric(p)
sigma_hat <- numeric(p)
for(j in 1:p){
  fit <- lm(c(Y1[,j], Y2[, j]) ~ x)
  bhat[j] <- coef(summary(fit))["xtr", "Estimate"]
  p.value[j] <- coef(summary(fit))["xtr", "Pr(>|t|)"]
  sigma_hat[j] <- sqrt(sum(fit$residuals^2)/fit$df.residual)
}

```

```{r inflation-histogram, fig.cap="A histogram of the distribution of the 10,000 effects", echo=FALSE}
gghistogram(x="b", data=data.table(b=beta[1:pt]), bins=30)

```

```{r, echo=FALSE}
res <- data.table(effect=beta, estimate=bhat, sigma=sigma, sd=sigma_hat, p.value=p.value)
res[, "relative true effect" := effect/max(effect)]
res[effect > 0, rank:=pt-rank(effect)+1]
res <- orderBy(~p.value, data=res)
knitr::kable(res[1:10,], digits=c(2, 2, 2, 2, 7, 2, 0), caption="The top 10 genes ranked by p-value. Rank is the rank of the true effect, from large to small.")
```

The table above lists the top 10 genes ranked by *p*-value, using the logic that the genes with the smallest *p* values are the genes that we should pursue with further experiments to understand the system. Some points

1. Six of the top ten genes with biggest true effects are *not* on this list. And, in the list are three genes with true effects that have relatively low ranks based on true effect size (column \"rank\") *and* two genes that have no true effect at all. Also in this list is one gene with an estimated effect (-1.78) that is *opposite* in sign of the true effect (but look at the *p*-value!)
2. The estimate of the effect size for all top-ten genes are inflated. The average estimate for these 10 genes is `r round(mean(res[1:10, estimate]), 2)` while the average true effect for these 10 genes is `r round(mean(res[1:10, effect]), 2)` (the estimate ).
3. The sample standard deviation (sd) for all top-ten genes is less than the true standard deviation (sigma), in some cases substantially.

The consequence of an inflated estimate of the effect and a deflated estimate of the variance is a large *t* (not shown) and small *p*. What is going on is an individual gene's estimated effect and standard deviation are functions of 1) the true value and 2) a random sampling component. The random component will be symmetric, some effects will be overestimated and some underestimated. When we rank the genes by the estimate of the effect or *t* or *p*, some of the genes that have "risen to the top" will be there because of a large, positive, sampling (random) component of the effect and/or a large, negative, sampling component of the variance. Thus some genes' high rank is artificial in the sense that it is high because of a random fluke. If the experiment were re-done, these genes at the top because of a large, random component would (probably) fall back to a position closer to their expected rank (regression to the mean again).

In the example here, all genes at the top have inflated estimates of the effect because of the positive, random component. This inflation effect is a function of the signal to noise ratio, which is controled by theta and sigma in the simulation. If theta is increased (try theta=1), or if sigma is decreased, the signal to noise ratio increases (try it and look at the histogram of the new distribution of effects) and both the 1) inflation and the 2) rise to the top phenomenon decrease.

#### p-hacking


### Multiple testing -- working in R

#### Tukey HSD adjustment of all pairwise comparisons

The `adjust` argument in `emmeans::contrast()` controls the method for *p*-value adjustment. The default is "tukey".

1. "none" -- no adjustment, in general my preference.
2. "tukey" -- Tukey's HSD, the default
3. "bonferroni" -- the standard bonferroni, which is conservative
4. "fdr" -- the false discovery rate
5. "mvt" -- based on the multivariate *t* distribution and using covariance structure of the variables

The data are those from Fig. 2D of "Data from The enteric nervous system promotes intestinal health by constraining microbiota composition". There is a single factor with four treatment levels. The response is neutrophil count.

```{r best-import-enteric, echo=FALSE}
folder <- "The enteric nervous system promotes intestinal health by constraining microbiota composition"
filename <- "journal.pbio.2000689.s008.xlsx"

# figure 2D data
sheet_i <- "Figure 2"
range_i <- "F2:I24"
file_path <- here(data_path, folder, filename)
#file_path <- paste(data_path, folder, fn, sep="/")
dt_wide <- data.table(read_excel(file_path, sheet=sheet_i, range=range_i))
# clean names
dt_wide <- clean_names(dt_wide)
# get rid of "_donor"
setnames(dt_wide, old=colnames(dt_wide), new=gsub("_donor", "", colnames(dt_wide)))
# wide to long
exp2d <- na.omit(melt(dt_wide, measure.vars=colnames(dt_wide), variable.name="donor", value.name="count"))
exp2d[, donor:=factor(donor, c("wt", "gf", "sox10", "iap_mo"))]
```

No adjustment:

```{r}
m1 <- lm(count ~ donor, data=exp2d)
m1.emm <- emmeans(m1, specs="donor")
m1.pairs.none <- contrast(m1.emm, method="revpairwise", adjust="none")
summary(m1.pairs.none, infer=c(TRUE, TRUE))
```

Tukey HSD:

```{r}
m1.pairs.tukey <- contrast(m1.emm, method="revpairwise", adjust="tukey")
summary(m1.pairs.tukey, infer=c(TRUE, TRUE))
```

### False Discovery Rate

## difference in p is not different

## Inference when data are not Normal
No real data are normal, although many are pretty good approximations of a normal distribution.

I'll come back to this point, but first, let's back up. Inference in statistical models (standard errors, confidence intervals, *p*-values) are a function of the modeled distributions of the parameters (for linear models, this parameter is the conditional (or error) variance $\sigma^2$); if the data do not approximate the modeled distribution, then inferential statistics might be to liberal (standard errors are too small, confidence intervals are too narrow, Type I error is more than nominal) or to conservative (standard errors are too large, confidence intervals are too wide, Type I error is less than nominal).

Linear models assume that "the data" (specifically, the conditional response, or, equivalently, the residuals from the model) approximate a Normal distribution. Chapter xxx showed how to qualitatively assess how well residuals approximate a Normal distribution using a Q-Q plot. If the researcher concludes that the data poorly approximate a normal distribution because of outliers, the researcher can use robust methods to estimate the parameters. If the approximation is poor because the residuals suggest a skewed distribution or one with heavy or light tails, the researcher can choose among several strategies

1. continue to use the linear model; inference can be fairly robust to non-normal data, especially when the sample size is not small.
2. use a generalized linear model (GLM), which is appropriate if the conditional response approximates any of the distributions that can be modeled using GLM (Chapter xxx)
3. use bootstrap for confidence intervals and permutation test for *p*-values
4. transform the data in a way that makes the conditional response more closely approximate a normal distribution.
5. use a classic non-parametric test, which are methods that do not assume a particular distribution

This list is roughly in the order of how I would advise researchers, although the order of 1-3 is pretty arbitrary. I would rarely advise a researcher to use (4) and never advise (5). Probably the most common strategies in the biology literature are (4) and (5). The first is also common but probably more from lack of recognition of issues or because a "test of normality" failed to reject that the data are "not normal".

On this last point, do not use the *p*-value from a "test for normality" (such as a Shapiro-Wilk test) to decide between using the linear model (or t-test or ANOVA) and an alternative such as a generalized linear model (or transformation or non-parametric test). No real data is normal. Tests of normality will tend to "not reject" normality (p > 0.05) when the sample size is small and "reject" normality (p < 0.05) when the sample size is very large. But again, a "not rejected" hypothesis test does not mean the null (in this case, the data are normal) is true. More importantly, where the test for normality tends to fail to reject (encouraging a researcher to use parametric statistics) is where parametric inference performs the worst (because of small *n*) and where the test for normality tends to reject (encouraging a researcher to use non-parametric statistics) is where the parametric inference performs the best (because of large sample size) (Lumley xxx).

### Working in R
The data for demonstrating different strategies are from Fig. 4A of "Data from The enteric nervous system promotes intestinal health by constraining microbiota composition". There is a single factor with two treatment levels. The response is neutrophil count.

```{r best-import-non-normal-counts, echo=FALSE, warning=FALSE, message=FALSE, fig.cap="Distribution of the counts in the wildtype (WT) and sox10 knockout (sox10-) groups. Both groups show a strong right skew, which is common with count data."}
folder <- "The enteric nervous system promotes intestinal health by constraining microbiota composition"
filename <- "journal.pbio.2000689.s008.xlsx"

# figure 2D data
sheet_i <- "Figure 4"
range_i <- "a2:b125"
file_path <- here(data_path, folder, filename)
fig4a <- read_excel(file_path, sheet=sheet_i, range=range_i) %>%
  clean_names() %>%
  melt(variable.name="treatment", value.name="count") %>%
  data.table() %>%
  na.omit

gghistogram(x = "count", 
            bins = 30,
            fill = "treatment", 
            color = "treatment",
            add = "mean", 
            rug = FALSE,
            palette = "npg",
            #palette = c("#00AFBB", "#E7B800"),
            data=fig4a)

```


A linear model to estimate the treatment effect and 95% confidence interval.

```{r best-lm-counts}
m1 <- lm(count ~ treatment, data=fig4a)
m1_emm <- emmeans(m1, specs="treatment")
summary(contrast(m1_emm, method="revpairwise"),
        infer=c(TRUE, TRUE))
```

### Bootstrap Confidence Intervals

A bootstrap confidence interval is computed from the distribution of a statistic from many sets of re-sampled data. The basic algorithm is

1. compute the statistic for the observed data, assign this to $\theta_1$
2. resample $n$ rows of the data, with replacement. "with replacement" means to sample from the entire set of data and not the set that has yet to be sampled. $n$ is the original sample size; by resampling $n$ rows with replacement, some rows will be sampled more than once, and some rows will not be sampled at all.
3. compute the statistic for the resampled data, assign these to $\theta_{2..m}$
4. repeat 2 and 3 $m-1$ times
5. Given the distribution of $m$ estimates, compute the lower interval as the $\frac{\alpha}{2}$th percentile and the upper interval as the $1 - \frac{\alpha}{2}$th percentile. For 95% confidence intervals, these are the 2.5th and 97.5th percentiles.

Let's apply this algorithm to the data from fig4A neutrophil count data in the coefficient table above. The focal statistic in these data is the difference in the mean count for the sox10 and wild type groups (the parameter for $treatment$ in the linear model). The script below, which computes the 95% confidence intervals of this difference, resamples within **strata**, that is, within each group; it does this to preserve the original sample size within each group.

```{r best-bootstrap}
n_iter <- 5000
b1 <- numeric(5000)
inc <- 1:nrow(fig4a) # the rows for the first iteration are all rows, so this is the observed effect
for(i in 1:n_iter){
  # inc creates the index of rows to resample preserving the sample size specific to each group
  b1[i] <- coef(lm(count ~ treatment, data=fig4a[inc, ]))["treatmentsox10"]
  inc <- c(sample(which(fig4a[, treatment] == "wt"), replace=TRUE),
           sample(which(fig4a[, treatment] == "sox10"), replace=TRUE))
}
ci <- quantile(b1, c(0.025, 0.975))
c(contrast = b1[1], ci[1], ci[2])
```

The intervals calculated in step 5 are **percentile intervals**. A histogram of the the re-sampled differences helps to visualize the bootstrap (this is a pedagogical tool, not something you would want to publish).

```{r best-bootstrap-histogram, echo=FALSE, fig.cap="Distribution of the 5000 resampled estimates of the difference in means between the sox10 and wt treatment levels. The dashed lines are located at the 2.5th and 97.5th percentiles of the distribution."}
gghistogram(x = "b1", 
            bins = 30,
            rug = FALSE,
            palette = "npg",
            data=data.table(b1=b1)) +
  geom_vline(xintercept=c(ci), linetype=2) +
  xlab("mean(sox10) - mean(wt)") +
  NULL

```


#### Some R packages for bootstrap confidence intervals

Percentile intervals are known to be biased, meaning the intervals are shifted. The `boot` package computes a bias-corrected interval in addition to a percentile interval. `boot` is a very powerful bootstrap package but requires the researcher to write functions to compute the parameter of interest. `simpleboot` provides functions for common analysis that does this for you (in R speak, we say that `simpleboot` is a "wrapper" to `boot`). The function `simpleboot::two.boot` computes a `boot`-like object that returns, among other values, the distribution of $m$ statistics. The `simpleboot` object is then be fed to `boot::boot.ci` to get bias-corrected intervals.

```{r best-bca-intervals}
bs_diff <- two.boot(fig4a[treatment=="sox10", count],
                    fig4a[treatment=="wt", count],
                    mean, 
                    R=5000)
boot.ci(bs_diff, type="bca")
```

### Permutation test

A permutation test effectively computes the probability that a random assignment of a response to a particular value of *X* generates a test statistic as large or larger than the observed statistic. If this probability is small, then this "random assignment" is unlikely. From this we infer that the actual assignment matters, which implies a treatment effect.


The basic algorithm is

1. compute the test statistic for the observed data, assign this to $\theta_1$
2. permute the response
3. compute the test statistic for the permuted data, assign these to $\theta_{2..m}$
4. repeat 2 and 3 $m-1$ times
5. compute $p$ as

\begin{equation}
p_{perm} = \frac{N_{\theta_i \ge \theta_{1}}}{m}
\end{equation}

This is easily done with a **for loop** in which the observed statistic is the first value in the vector of statistics. If this is done, the minimum value in the numerator for the computation of $p_{perm}$ is 1, which insures that $p_{perm}$ is not zero.

The test statistic depends on the analysis. For the simple comparison of means, a simple test statistic is the difference in means. This is the numerator of the test statistic in a *t*-test. The test has more power if the test-statistic is scaled (Manley xxx), so a better test statistic would be *t*, which scales the difference by its standard error.

Here, I implement this algorithm. The test is two-tailed, so the absolute difference is recorded. The first value computed is the observed absolute difference.

```{r best-permutation}
set.seed(1)
n_permutations <- 5000
d <- numeric(n_permutations)

# create a new column which will contain the permuted response
# for the first iteration, this will be the observed order
fig4a[, count_perm := count]

for(i in 1:n_permutations){
  d[i] <- abs(t.test(count_perm ~ treatment, data = fig4a)$statistic)
  
  # permute the count_perm column for the next iteration
  fig4a[, count_perm := sample(count)]
}
p <- sum(d >= d[1])/n_permutations
p
```

#### Some R packages with permutation tests.

`lmPerm::lmp` generates permutation p-values for parameters of any kind of linear model. The test statistic is the sum of squares of the term scaled by the residual sum of squares of the model.

```{r best-lmp}
set.seed(2)
coef(summary(lmp(count ~ treatment, perm="Prob", Ca=0.01, 
                 data=fig4a)))
```

### Non-parametric tests

1. In general, the role of a non-parametric test is a better-behaved *p*-value, that is, one whose Type I error is well controlled. As such, non-parametric tests are more about Null-Hypothesis Statistical Testing and less (or not at all) about Estimation.
2. In general, classic non-parametric tests are only available for fairly simple experimental designs. Classic non-parametric tests include
    * Independent sample (Student's) *t* test: Mann-Whitney-Wilcoxan
    * Paired *t* test: Wilcoxan signed-rank test
    
One rarely sees non-parametric tests for more complex designs that include covariates, or multiple factors, but for these, one could 1) convert the response to ranks and fit the usual linear model, or 2) implement a permutation test that properly preserves **exchangeability**.

Permutation tests control Type I error and are powerful. That said, I would recommend a permutation test as a supplment to, and not replacement of, inference from a generalized linear model.

A non-parametric (Mann-Whitney-Wilcoxon) test of the fake data generated above

```{r best-wilcoxon}
wilcox.test(count ~ treatment, data=fig4a)
```

### Log transformations

Many response variables within biology, including count data, and almost anything that grows, are right skewed and have variances that increase with the mean. A log transform of a response variable with this kind of distribution will tend to make the residuals more approximately normal and the variance less dependent of the mean. At least two issues arise

1. if the response is count data, and the data include counts of zero, then a fudge factor has to be added to the response since log(0) doesn't exist. The typical fudge factor is to add 1 to *all* values, but this is arbitrary and results do depend on the magnitude of this fudge factor.
2. the estimates are on a log scale and do not have the units of the response. The estimates can be back-transformed by taking the exponent of a coefficient or contrast but this itself produces problems. For example, the backtransformed mean of the log-transformed response is not the mean on the origianl scale (the arithmetic mean) but the **geometric mean**. Geometric means are smaller than arithmetic means, appreciably so if the data are heavily skewed. Do we want our understanding of a system to be based on geometric means?

#### Working in R -- log transformations

If we fit a linear model to a log-transformed response then the resulting coefficients and predictions are on the **log scale**. To make interpretation of the analysies easier, we probably want to **back-transform** the coefficients or the predictions to the original scale of the response, which is called the **response scale**. 

```{r best-fit-log}
m2 <- lm(log(count + 1) ~ treatment, data=fig4a)
(m2_emm <- emmeans(m2,
                  specs="treatment",
                  type = "response"))
```

The emmeans package is amazing. Using the argument `type = "response"` not only backtransforms the means to the response scale but also substracts the 1 that was added to all values in the model.

What about the effect of treatment on count?

```{r}
summary(contrast(m2_emm, 
                 method="revpairwise",
                 type = "response"),
        infer=c(TRUE, TRUE))

```

It isn't necessary to backtransform the estimated marginal means prior to computing the contrasts as this can be done in the contrast function itself. Here, the `type = "response"` argument in the contrast function is redundant since this was done in the computation of the means. But it is transparent so I want it there.

**Don't skip this paragraph** Look at the value in the "contrast" column -- it is "sox10 / wt" and not "sox10 - wt". The backtransformed effect is a ratio instead of a difference. **A difference on the log scale is a ratio on the response scale** because of this equality

\begin{equation}
\mathrm{exp}(\mu_2-\mu_1) = \frac{\mathrm{exp}(\mu_2)}{\mathrm{exp}(\mu_1)})
\end{equation}

The interpretation is: If $b^*$ is the backtransformed effect, then, given a one unit increase in $X$, the expected value of the response increases $b^*\times$. For a categorical $X$, this means the backtransformed effect is the ratio of backtransformed means -- its what you have to multiply the mean of the reference by to get the mean of the treated group. And, because it is the response that is log-transformed, these means are not arithemetic means but geometric means. Here, this is complicated by the model -- the response is not a simple log transformation but log(response + 1). It is easy enough to get the geometric mean of the treated group -- multiply the backtransformed intercept by the backtransformed coefficient and then subtract 1 -- but because of this subtraction of 1, the interpretation of the backtransformed effect is awkward at best (recall that I told you that a linear model of a log transformed response, and especially the log of the response plus one, leads to difficulty in interpreting the effects).

```{r}
# backtransformed control mean -- a geometric mean
mu_1 <- exp(coef(m2)[1])

# backtransformed effect
b1_star <- exp(coef(m2)[2])

# product minus 1
mu_1*b1_star -1

# geometric mean of treatment group
n <- length(fig4a[treatment=="sox10", count])
exp(mean(log(fig4a[treatment=="sox10", count+1])))-1
```


Back-transformed effect
```{r best-backtransformed effect}
m1 <- lm(count ~ treatment, data=fig4a)

```


```{r best-backtransformed-estimates}
exp(coef(m2)) 
```


### Performance of parametric tests and alternatives

#### Type I error

If we are going to compute a $p$-value, we want it to be uniformly distributed "under the null". A simple way to check this is to compute Type I error. If we set $\alpha = 0.05$, then we'd expect 5% of tests of an experiment with no effect to have $p < 0.05$.

```{r, comment=FALSE, results='hide'}
# first create a matrix with a bunch of data sets, each in its own column
n <- 10
n_sets <- 4000
fake_matrix <- rbind(matrix(rnegbin(n*n_sets, mu=10, theta=1), nrow=n),
                   matrix(rnegbin(n*n_sets, mu=10, theta=1), nrow=n))
treatment <- rep(c("cn", "tr"), each=n)

tests <- c("lm", "log_lm","mww", "perm")
res_matrix <- matrix(NA, nrow=n_sets, ncol=length(tests))
colnames(res_matrix) <- tests
for(j in 1:n_sets){
  res_matrix[j, "lm"] <- coef(summary(lm(fake_matrix[,j] ~ treatment
                                 )))[2, "Pr(>|t|)"]
  res_matrix[j, "log_lm"] <- coef(summary(lm(log(fake_matrix[,j] + 1) ~ treatment
                                 )))[2, "Pr(>|t|)"]
  res_matrix[j, "mww"] <- wilcox.test(fake_matrix[,j] ~ treatment,
                                      exact=FALSE)$p.value
  res_matrix[j, "perm"] <- coef(summary(lmp(fake_matrix[,j] ~ treatment,
                                 perm="Prob", Ca=0.01)))[2, "Pr(Prob)"]
}

```

```{r}
apply(res_matrix, 2, function(x) sum(x < 0.05)/n_sets)
```

Type I error is computed for the linear model, the linear model with a log transformed responpse, Mann-Whitney-Wilcoxon, and permutation tests. All four tests are slightly conservative for data that look like that modeled. The computed Type I error of the permutation test is closest to the nominal value of 0.05.

#### Power

Power is the probability of a test to reject the null hypothesis if the null hypothesis is false (that is, if an effect exists)

\begin{equation}
\mathrm{Power} = \mathrm{Prob}(p < \alpha | mathrm{effect} \neq 0)
\end{equation}

If all we care about is a $p-value$ then we want to use a test that is most powerful. But, while power is defined using $\alpha$, we *can* care about power even if we don't consider $\alpha$ to be a very useful concept because increased power also increases the precision of an estimate (that is, narrows confidence intervals).

```{r, comment=FALSE, results='hide'}
# first create a matrix with a bunch of data sets, each in its own column
n <- 5
n_sets <- 4000
fake_matrix <- rbind(matrix(rnegbin(n*n_sets, mu=10, theta=1), nrow=n),
                   matrix(rnegbin(n*n_sets, mu=20, theta=1), nrow=n))
treatment <- rep(c("cn", "tr"), each=n)

tests <- c("lm", "log_lm","mww", "perm")
res_matrix <- matrix(NA, nrow=n_sets, ncol=length(tests))
colnames(res_matrix) <- tests
for(j in 1:n_sets){
  res_matrix[j, "lm"] <- coef(summary(lm(fake_matrix[,j] ~ treatment
                                 )))[2, "Pr(>|t|)"]
  res_matrix[j, "log_lm"] <- coef(summary(lm(log(fake_matrix[,j] + 1) ~ treatment
                                 )))[2, "Pr(>|t|)"]
  res_matrix[j, "mww"] <- wilcox.test(fake_matrix[,j] ~ treatment,
                                      exact=FALSE)$p.value
  res_matrix[j, "perm"] <- coef(summary(lmp(fake_matrix[,j] ~ treatment,
                                 perm="Prob", Ca=0.01)))[2, "Pr(Prob)"]
}

```

```{r}
apply(res_matrix, 2, function(x) sum(x < 0.05)/n_sets)
```

As above, Power is computed for the linear model, linear model with a log-transformed response, Mann-Whitney-Wilcoxan, and permutation, by simulating a "low power" experiment. The effect is huge (twice as many cells) but the power is low because the sample size is small ($n = 5$). At this sample size, and for this model of fake data, all tests have low power. The power of the log-transformed response is the largest. A problem is, this is not a test of the means but of the log transformed mean plus 1. The power of the permutation test is about 25% larger than that of the linear model and Mann-Whitney-Wilcoxan test. An advantage of this test is that it is a p-value of the mean. A good complement to this p-value would be bootstraped confidence intervals. Repeat this simulation using $n=40$ do see how the relative power among the three change in a simulation of an experiment with more power.


<!--chapter:end:chapters/28-best_practices.Rmd-->

# Part VI: More than one $X$ -- Multivariable Models {-}

<!--chapter:end:chapters/30-part-iv-multivariable.Rmd-->

# Linear models with an added covariate ("ANCOVA") {#covariates}

```{r setup-ancova, echo=FALSE, warnng=FALSE, message=FALSE}
library(here)
library(janitor)
library(readxl)
library(data.table)
library(stringr)

# analysis packages
library(nlme) # gls
library(lmerTest) # lmm
library(emmeans)
library(car) # qqplot, spreadlevel
library(lmPerm)

# graphing packages
library(ggplot2) # ggplot environment
library(ggpubr) # publication ready plots
library(ggforce) # jitter
library(ggsci) # color palettes
library(ggpubfigs) # color palettes
library(ggthemes) # pander is okabe ito
# execute this command only the first time you use ggpubfigs
# devtools::install_github("JLSteenwyk/ggpubfigs")
library(cowplot) # combine plots
library(lazyWeave) # pretty pvalues
library(broom)

library(mvtnorm)
library(harrellplot)

here <- here::here
data_folder <- "data"

# Okabe & Ito palette
pal_ito_seven <- friendly_pal("ito_seven") # ggpubfigs
pal_okabe_ito <- pal_ito_seven[c(6,5,3,7,1,2,4)] # order of Wilke
pal_okabe_ito_2 <- colorblind_pal()(8)[2:8] # ggthemes
```

In its most general sense, **Covariates** are simply the $X$ variables in a statistical model. With data from experiments, "covariates" more typically refers to continuous $X$ variables that are added to a model to increase precision of the treatment effects. In observational designs, continuous and categorical covariates might be added to a model to 1) increase predictive ability, 2) because the researcher is interested in specific conditional effects, or 3) to eliminate confounding. These are discussed in later chapters. In this chapter, a single, continuous covariate is added to a linear model. Nothing is fundamentally different if the covariate added is categorical ($Sex$ is a common categorical covariate) or if multiple covariates are added.

## Adding covariates can increases the precision of the effect of interest

I use fake data to introduce the concept of **statistical elimination** of a **covariate** in a statistical model. Here I am modeling the effect of a new drug on blood LDL-C levels. LDL is a kind of lipoprotein, which are particles in the blood that transport fats and cholesterol to and from different tissues. LDL-C is cholesterol associated with LDL particles. LDL-C is considered "bad cholesterol" because LDL is believed to transport cholesterol and other lipids to arterial walls, which is the basis for atherosclerosis.

Thirty applied biostats students are recruited and are randomly assigned to either the "placebo" treatment level or "drug" treatment level. The response is blood LDL-C concentration. The drug manufacturer wants a measure of the effect of the new drug on ldlc.

```{r ancova-ldlc, echo=FALSE, eval=TRUE}
n_iter <- 100
# seed_i <- 69 # good
# seed_i <- 48 # best
p1 <- numeric(n_iter)
p2 <- numeric(n_iter)
start_i <- 46
end_i <- 46
seed_i <- start_i
for(i in start_i:end_i){
  set.seed(seed_i)
  sigma <- 2.5
  n <- 15 # replicates per grop
  fd <- data.table(treatment = rep(c("placebo", "drug"), each=n),
                   age = rnorm(n*2, mean = 25, sd = 3))
  fd[, age := runif(n*2, min = 20, max = 30)]
  fd[, treatment := factor(treatment, c("placebo", "drug"))]
  X <- model.matrix(formula(~ age + treatment), data = fd) # the model matrix
  
  b0 <- 70 # LDL-C for the reference group and zero years old
  b1 <- 1.5 # the effect of age. Increasing fat by 20 years adds
  # 20*1.5 = 30 mg LDL-C/dL blood to their blood
  b2 <- -5 # the effect of the new drug -- lowers LDL-C by 5 mg/dL blood
  b <- c(b0, b1, b2) # but the parameters into a vector
  fd[, ldlc := X%*%b + rnorm(n, sd = sigma)] # add ldlc to the fake data.table
  m1 <- coef(summary(lm(ldlc ~ treatment, data = fd)))
  m2 <- coef(summary(lm(ldlc ~ age + treatment, data = fd)))
  p1[i] <- m1[2, "Pr(>|t|)"]
  p2[i] <- m2[3, "Pr(>|t|)"]
  seed_i <- seed_i + 1
}
max_i <- which(p1 > 0.2)
# m1
# m2
# qplot(x = age, y = ldlc, color = treatment, data = fd)


```

```{r covariate-fd-plot-no-cov, echo = FALSE, fig.cap="Effect of drug therapy on plasma LDL-C."}
m1 <- lm(ldlc ~ treatment, data = fd)

m1_emm <- emmeans(m1, specs = "treatment") %>%
  summary() %>%
  data.table()
m1_pairs <- contrast(emmeans(m1, specs = "treatment"),
                     method = "revpairwise") %>%
  summary(infer = TRUE) %>%
  data.table()
# pvalString is from package lazyWeave
m1_pairs[ , p_pretty := pvalString(p.value)]
# also create a column with "p-val: "
m1_pairs[ , pval_pretty := paste("p-val:", p_pretty)]

range_ldlc <- range(fd[, ldlc])
gg_response <- ggplot(data = fd,
                            aes(x = treatment,
                                y = ldlc,
                                color = treatment)) +
  # points
  geom_sina(alpha = 0.5) + # ggforce package
  
  # plot means and CI
  geom_errorbar(data = m1_emm,
                aes(y = emmean,
                    ymin = lower.CL,
                    ymax = upper.CL,
                    color = treatment),
                width = 0
  ) +
  
  geom_point(data = m1_emm,
             aes(y = emmean,
                 color = treatment),
             size = 3
  ) +

  # aesthetics
  # coord_cartesian(ylim = c(round(range_ldlc[1] - 5, 0),
  #                          round(range_ldlc[2] + 5, 0))) +
  ylab("Plasma LDL-C (mg/dL blood)") +
  scale_color_manual(values=pal_okabe_ito,
                     name = NULL) +
  theme_pubr() +
  theme(legend.position="none") +
  theme(axis.title.y = element_blank()) +
  
  NULL

gg_effect <- ggplot(data = m1_pairs, 
                          aes(x = estimate,
                              y = contrast)) +
  # confidence level of effect
  geom_errorbar(aes(xmin = lower.CL, 
                    xmax = upper.CL),
                width = 0, 
                color = "black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # draw a line at effect = 0
  geom_vline(xintercept = 0, linetype = 2) +
  
  # p-value. The y coordinates are set by eye
  annotate(geom = "text",
           label = m1_pairs$p_pretty,
           y = 1,
           x = 5) +
  
  # x-axis label and aesthetics
  xlab("Effect (mg/dL blood)") +
  ylab("Contrast") +
#  coord_cartesian(xlim = c(-10,30)) +
  
  scale_x_continuous(breaks = c(-5, -2.5, 0, 2.5, 5),
                     labels = c("-5", "-2.5", "0", "2.5", "p-value"),
                     position = "top") + # move to top
  scale_y_discrete(labels = "drug -\nplacebo") +

  # use ggpubr theme
  theme_pubr() +
  theme(axis.title.y = element_blank()) +
  
  NULL
# gg_effect

gg_response <- gg_response +
  coord_flip(ylim = c(95, 115))

gg1 <- plot_grid(gg_effect,
                 gg_response,
                 nrow=2,
                 align = "v",
                 axis = "lr",
                 rel_heights = c(0.5,1))
gg1
```

The linear model fit to the simulated LDL-C data is

\begin{equation}
ldlc = \beta_0 + \beta_1 treatment_{drug} + \varepsilon
(\#eq:cov-no-cov)
\end{equation}

where $treatment_{drug}$ is the dummy variable, which is set to 0 when $treatment = \mathrm{''placebo''}$ and to 1 when $treatment = \mathrm{''drug''}$.

The coefficient table is

```{r, echo=FALSE}
m1_coef <- cbind(coef(summary(m1)),
                 confint(m1))
round(m1_coef, 3)
```

The response-effect plot shows large overlap in the LDL-C response and treatment effect that is small relative to the noise. "No effect of the drug ($p = .31$)" is an incorrect interpretation of the *p*-value of the significance test of the estimate of $\beta_1$. A better interpretation is, the estimated effect is -1.6 but everything from large, negative effects to moderate positive effects are consistent with the data. 

As expert biologists, we know that LDL-C is strongly correlated with age and there is a large range in age among the Applied Bistats students. If age contributes to a large fraction of the variance in LDL-C among applied biostats students, then age-related variance *might* mask the effect of the drug. Here is a plot of LDL-C vs. age, with treatment assignment color coded. Remember, these are the exact same values of LDL-C as in figure \@ref(fig:cov-harrelplot) above.

```{r ancova-plot2, message=FALSE, echo=FALSE, fig.cap="Linear regression of $ldlc$ on age fit to the fake LDL-C data. The points are color coded by treatment."}

gg_cov <- qplot(x=age, y=ldlc, data=fd) +
  geom_smooth(method="lm", se=FALSE) +
  geom_point(aes(color = treatment)) +
  scale_color_manual(values=pal_okabe_ito,
                     name = NULL) +
  
  ylab("LDL-C (mg/dL blood)") +
  xlab("Age (years)") +
  theme_pubr() +
  theme(legend.position = "right")

gg_cov
```

The regression line is the fit of the linear model

\begin{equation}
ldlc = \beta_0 + \beta_1 age + \varepsilon
(\#eq:cov-age)
\end{equation}

The points are color-coded by treatment level but $treatment$ is not in model \@ref(eq:cov-age). The color-coding makes it clear that most of the "placebo" data points are above the line, or have positive residuals from the model, while the "drug" data points are below the line, or have negative residuals from the model. That is, **at any specific value of age, there is very small overlap of LDL-C values for drug and for placebo**.

What is happening? Age is contributing to the variance of LDL-C, and the noise in $\varepsilon$ in model \@ref(eq:cov-no-cov), and this added noise makes decreases the precision in our estimate of the effect of the new drug relative to placebo. When we view the data as in Figure \@ref(covariate-fd-plot-no-cov), age is masking the effect. If we could somehow measure the effect of the drug at a specific age, then we could get a more precise estimate of the effect. But how to do this? Here are three possible methods. The third is *the only* one you should use but the second is useful for understanding the third.

1. We could just analyze a subset of the data, that is, only the cases in which the value of age is nearly equal. This throws away perfectly good data and, consequently, greatly reduces the sample size and thus precision to estimate the effect.

2. We could first fit model \@ref(eq:cov-age) and then use the *residuals* of this fit as a new response variable to estimate the effect of drug treatment (this is what we did by eye in figure \@ref(fig:ancova-plot2)).

step 1:

\begin{equation}
ldlc = \beta_0 + \beta_1 age + \varepsilon
\end{equation}

step 2:

\begin{equation}
ldlc\_residual = \beta_0 + \beta_1 treatment_{drug} + \varepsilon
(\#eq:ancova-3)
\end{equation}

Here, I use this two-stage method because it is useful to introduce the concept of  **adjusting for a covariate** in a linear model, where the covariate here is $age$. But, in general, don't do this -- the method usually "works" pretty well if the mean of the **covariate** (the step 1 $X$ variable) is nearly the same in both treatment levels but artifacts that lead to wrong inference are introduced if the mean of the covariate is far apart. 

```{r covariate-fd-plot-cov, echo = FALSE, fig.cap="Effect of drug therapy on plasma LDL-C using residuals. Don't do this!"}
m2_step1 <- lm(ldlc ~ age, data = fd)
fd[, residuals := residuals(m2_step1)]
m2 <- lm(residuals ~ treatment, data = fd)

m2_emm <- emmeans(m2, specs = "treatment") %>%
  summary() %>%
  data.table()
m2_pairs <- contrast(emmeans(m2, specs = "treatment"),
                     method = "revpairwise") %>%
  summary(infer = TRUE) %>%
  data.table()
# pvalString is from package lazyWeave
m2_pairs[ , p_pretty := pvalString(p.value)]
# also create a column with "p-val: "
m2_pairs[ , pval_pretty := paste("p-val:", p_pretty)]
gg_response <- ggplot(data = fd,
                            aes(x = treatment,
                                y = residuals,
                                color = treatment)) +
  # points
  geom_sina(alpha = 0.5) + # ggforce package
  
  # plot means and CI
  geom_errorbar(data = m2_emm,
                aes(y = emmean,
                    ymin = lower.CL,
                    ymax = upper.CL,
                    color = treatment),
                width = 0
  ) +
  
  geom_point(data = m2_emm,
             aes(y = emmean,
                 color = treatment),
             size = 3
  ) +

  # aesthetics
  ylab("Residual LDL-c (mg/dL blood)") +
  scale_color_manual(values = pal_okabe_ito,
                     name = NULL) +
  theme_pubr() +
  theme(legend.position = "none") +
  theme(axis.title.y = element_blank()) +
  
  NULL


gg_effect <- ggplot(data = m2_pairs, 
                          aes(x = estimate,
                              y = contrast)) +
  # confidence level of effect
  geom_errorbar(aes(xmin = lower.CL, 
                    xmax = upper.CL),
                width = 0, 
                color = "black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # draw a line at effect = 0
  geom_vline(xintercept = 0, linetype = 2) +
  
  # p-value. The y coordinates are set by eye
  annotate(geom = "text",
           label = m2_pairs$p_pretty,
           y = 1,
           x = 5) +
  
  # x-axis label and aesthetics
  xlab("Effect (mg/dL blood)") +
  ylab("Contrast") +
#  coord_cartesian(xlim = c(-10,30)) +
  
  scale_x_continuous(breaks = c(-5, -2.5, 0, 2.5, 5),
                     labels = c("-5", "-2.5", "0", "2.5", "p-value"),
                     position = "top") + # move to top
  scale_y_discrete(labels = "drug -\nplacebo") +

  # use ggpubr theme
  theme_pubr() +
  theme(axis.title.y = element_blank()) +
  
  NULL

gg_response <- gg_response +
  coord_flip(ylim = c(-10, 10))


gg2 <- plot_grid(gg_effect,
          gg_response,
          nrow=2,
          align = "v",
          axis = "lr",
          rel_heights = c(0.5,1))

# gg2
plot_grid(gg2, gg1, ncol = 2, labels = "AUTO")
```

Figure \@ref(covariate-fd-plot-cov)A is an effect-response plot of the effect of treatment on the LDL-C **adjusted for age** using the two-step method. Figure \@ref(covariate-fd-plot-cov)B is the original (not adjust) effect-response plot. The **scale** of both plots are the same. This means that the response axis has the same length in both plots and the effects axis has the same length in both plots. This makes comparing the two easy. Two patterns are conspicuous

1. The **age-adjusted means** in Figure \@ref(covariate-fd-plot-cov)A are further apart (the difference is bigger) then the unadjusted means in Figure \@ref(covariate-fd-plot-cov)B. This is seen in both the response and the effects components of the plot.
2. The spread of the residual LDL-C measures within each treatment level in Figure \@ref(covariate-fd-plot-cov)A is less than the spread of the raw LDL-C measures in Figure \@ref(covariate-fd-plot-cov)B.
3. The confidence interval of the effect (difference in means) is smaller using the adjusted model (Figure \@ref(covariate-fd-plot-cov)A) than in the unadjusted model (Figure \@ref(covariate-fd-plot-cov)B)
4. The *p*-value of the effect (difference in means) is smaller using the adjusted model (Figure \@ref(covariate-fd-plot-cov)A) than in the unadjusted model (Figure \@ref(covariate-fd-plot-cov)B)

These patterns are quantified by the Estimates and SEs of the coefficient table. Again, I show both for comparison. Can you match comparisons 1-4 above with the statistic in the coefficient table?

Coefficient table of two-step, "adjusted" model (adjusting for $age$):

```{r, echo = FALSE}
m2_coef <- cbind(coef(summary(m2)),
                 confint(m2))
round(m2_coef, 3)
```

Coefficient table of unadjusted model (adjusting for $age$):

```{r echo=FALSE}
round(m1_coef, 3)
```

It is clear from the plots and the tables that this two-stage adjustment increases the precision of the estimates of the means and the differences in means by *eliminating the contribution of Age to the variance in LDL-C*.

3. The best practice for **adjusting for a covariate** (or the *statistical elimination of a covariate*) is to simply add the covariate to the linear model.

\begin{equation}
ldlc = \beta_0 + \beta_1 age + \beta_2 treatment_{drug} + \varepsilon
(\#eq:cov-cov)
\end{equation}

```{r echo=FALSE}
m3 <- lm(ldlc ~ age + treatment, data=fd)
m3_coef <- cbind(coef(summary(m3)),
                 confint(m3))
signif(m3_coef, 3)
```

Again, compare the coefficient table from the model fit without the covariate.

```{r echo=FALSE}
signif(m1_coef, 3)
```

1. The adjusted effect is larger (-4.5 vs. -1.6)
2. The adjusted SE of the difference is smaller (0.8 vs. 1.5)
3. The adjusted CIs are narrower (-6.1, -2.8 vs. -4.7, 1.6)
4. The p-value of the adjusted difference is smaller (0.000007 vs. 0.31)

A plot of the model is

```{r covariate-plot-the-model, echo=FALSE, fig.cap="Effect of drug therapy, adjusted for age, on plasma LDL-C."}
groups <- levels(fd$treatment)
age_range <- fd[, .(min_age = min(age),
                    max_age = max(age),
                    dummy = as.integer(treatment)-1), by = treatment]
line_data <- data.table(age_range)
b <- coef(m3)
line_data[, ldlc_min := b[1] + b[2]*min_age + b[3]*dummy]
line_data[, ldlc_max := b[1] + b[2]*max_age + b[3]*dummy]

gg_response <- ggplot(data = fd,
                aes(x = age,
                    y = ldlc,
                    color = treatment)) +
  geom_point() +
  
  # draw regression lines for each group
  geom_segment(data = line_data,
               aes(x = min_age,
                   y = ldlc_min,
                   xend = max_age,
                   yend = ldlc_max)) +
  scale_color_manual(values=pal_okabe_ito,
                     name = NULL) +
  ylab("LDL-C (mg/dL blood)") +
  xlab("Age (years)") +
  theme_pubr() +
  theme(legend.position="top") +

  NULL

# gg_response

m3_emm <- emmeans(m3, specs = "treatment")
m3_pairs <- contrast(m3_emm, method = "revpairwise") %>%
  summary(infer = TRUE) %>%
  data.table()
# pvalString is from package lazyWeave
m3_pairs[ , p_pretty := pvalString(p.value)]
# also create a column with "p-val: "
m3_pairs[ , pval_pretty := paste("p-val:", p_pretty)]

gg_effect <- ggplot(data = m3_pairs, 
                          aes(x = contrast,
                              y = estimate)) +
  # confidence level of effect
  geom_errorbar(aes(ymin = lower.CL, 
                    ymax = upper.CL),
                width = 0, 
                color = "black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # draw a line at effect = 0
  geom_hline(yintercept = 0, linetype = 2) +
  
  # p-value. The y coordinates are set by eye
  annotate(geom = "text",
           label = m3_pairs$pval_pretty,
           x = 1,
           y = -2) +
  
  # x-axis label and aesthetics
  ylab("Effect (mg/dL blood)") +
  xlab("Contrast") +
  coord_cartesian(ylim = c(-6,2)) +
  scale_y_continuous(position="left") +
  
  theme_pubr() +
  # theme(axis.title.x = element_blank()) +
  
  NULL
# gg_effect

gg <- plot_grid(gg_response,
                gg_effect,
                 ncol=2,
                 align = "h",
                 axis = "tb",
                 rel_widths = c(1,0.35))
gg
```

## Understanding a linear model with an added covariate -- heart necrosis data

[1-Deoxydihydroceramide causes anoxic death by impairing chaperonin-mediated protein folding](https://www.nature.com/articles/s42255-019-0123-y)

[Source data](https://www.nature.com/articles/s42255-019-0123-y#Sec38)

In this article, the researchers are investigating the effects of specific sphingolipids on hypoxia (low O2) in the heart. This hypoxia results in necrosis (death) of heart tissue. The researchers are specifically looking at the sphingolipid 1-deoxydihydroceramide (DoxDHCer), which is derived from 1-deoxysphinganine6 (DoxSa).

In the experiment for Figure 4h, the researchers measured the effect of three treatments on necrosis.

1. Vehicle -- there is no "cardioprotection" from the hypoxia-producing sphingolipids. This is the "control". We expect more necrotic area in this group.
2 Myriocin -- a drug that inhibits the enzyme that iniates sphingolipid production. This drug should provide protection from the hypoxia-producing sphingolipids. We expect less necrotic area (or, cardioprotection) in this group.
3. Myriocin + DoxSa -- DoxSa is the specific sphingolipid that the researchers believe cause the hypoxia/necrosis. The drug should inhibit the production of sphingolipids but the added DoxSa should reverses the protective effect of the drug. If there is a reversal of the protection, then this supports the hypothesis that DoxSa is the sphingolipid causing the hypoxia/necrosis.

The response ($Y$) variable is $area\_of\_necrosis$ -- the measured "area" of the necrotic tissue (area is used here in the sense of "region" and not in the sense of length times width). The covariate is $area\_at\_risk$ -- the area of heart tissue that is susceptible to necrosis.

```{r covariate-fig4h-import, echo = FALSE}
data_from <- "1-Deoxydihydroceramide causes anoxic death by impairing chaperonin-mediated protein folding"
file_name <- "42255_2019_123_MOESM7_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)
  
sheet_i <- "Figure 4h"
fig4h_1 <- read_excel(file_path,
                         sheet = sheet_i,
                         range = "A5:B19",
                         col_names = TRUE) %>%
  clean_names() %>%
  data.table()
fig4h_2 <- read_excel(file_path,
                         sheet = sheet_i,
                         range = "D5:E17",
                         col_names = TRUE) %>%
  clean_names() %>%
  data.table()
fig4h_3 <- read_excel(file_path,
                         sheet = sheet_i,
                         range = "G5:H15",
                         col_names = TRUE) %>%
  clean_names() %>%
  data.table()

fig4h <- rbind(data.table(treatment = "Vehicle", fig4h_1),
                    data.table(treatment = "Myriocin", fig4h_2),
                    data.table(treatment = "Myriocin + DoxSa", fig4h_3))

treatment_levels <- c("Vehicle", "Myriocin", "Myriocin + DoxSa")
fig4h[, treatment := factor(treatment, levels = treatment_levels)]

```

### Fit the model

The verbal model with added covariate is

$area\_of\_necrosis \sim area\_at\_risk + treatment\;\;\;(\mathcal{M}_1)$.

which I'll name model $\mathcal{M}_1$. To understand the model coefficients, it helps to expand model $\mathcal{M}_1$ into the full linear model.

\begin{equation}
area\_of\_necrosis = \beta_0 + \beta_1 area\_at\_risk + \beta_2 treatment_{myriosin} + \beta_3 treatment_{Myriocin\;+\;DoxSa} + \varepsilon
(\#eq:cov-fig4h-additive)
\end{equation}

The model includes parameters for the effects of area at risk ($\beta_1$), the myriosin treatment ($\beta_2$) and the myriosin + DoxSa treatment ($\beta_3$). I explain the interpretation of these effects in "Interpretation of the model coefficients" below.

```{r covariate-fig4h-m1, echo=TRUE}
fig4h_m1 <- lm(area_of_necrosis ~ area_at_risk + treatment,
               data = fig4h)
```

### Plot the model

```{r fig4h-m1-plot-model, warning = TRUE, echo = FALSE, fig.cap = "The effect of treatment on area of necrosis, adjusted for area of risk."}

m1 <- lm(area_of_necrosis ~ area_at_risk + treatment, data = fig4h)
b <- coef(m1)
m1_emm <- emmeans(m1, specs = "treatment")
m1_pairs <- contrast(m1_emm,
                     method = "revpairwise",
                     adjust = "none") %>%
  summary(infer = TRUE) %>%
  data.table()
# pvalString is from package lazyWeave
m1_pairs[ , p_pretty := pvalString(p.value)]
# also create a column with "p-val: "
m1_pairs[ , pval_pretty := paste("p = ", p_pretty)]

groups <- levels(fig4h$treatment)
line_data <- fig4h[, .(min_x = min(area_at_risk),
                       max_x = max(area_at_risk)), by = treatment]
line_data[, dummy1 := ifelse(treatment == "Myriocin", 1, 0)]
line_data[, dummy2 := ifelse(treatment == "Myriocin + DoxSa", 1, 0)]

line_data[, y_min := b[1] + b[2]*min_x + b[3]*dummy1 + b[4]*dummy2]
line_data[, y_max := b[1] + b[2]*max_x + b[3]*dummy1 + b[4]*dummy2]

gg_response <- ggplot(data = fig4h,
                      aes(x = area_at_risk,
                          y = area_of_necrosis,
                          color = treatment)) +
  # points
  geom_point(size = 2) +
  
  geom_segment(data = line_data,
               aes(x = min_x,
                   y = y_min,
                   xend = max_x,
                   yend = y_max)) +
  scale_color_manual(values = pal_okabe_ito,
                     name = NULL) +
  ylab("Area of necrosis (mg)") +
  xlab("Area at risk (mg)") +
  theme_pubr() +
  theme(legend.position = "left") +

  NULL
#gg_response

contrast_order <- m1_pairs[, contrast]
m1_pairs[, contrast := factor(contrast, contrast_order)]

gg_effect <- ggplot(data = m1_pairs, 
                    aes(y = contrast,
                        x = estimate)) +
  # confidence level of effect
  geom_errorbar(aes(xmin = lower.CL, 
                    xmax = upper.CL),
                width = 0, 
                color = "black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # draw a line at effect = 0
  geom_vline(xintercept = 0, linetype = 2) +
  
  # p-value. The y coordinates are set by eye
  annotate(geom = "text",
           label = m1_pairs$pval_pretty,
           y = 1:3,
           x = 4.5) +
  
  # x-axis label and aesthetics
  xlab("Effect (mg)") +
  ylab("Contrast") +
  coord_cartesian(xlim = c(-8,5.5)) +
  scale_x_continuous(position="top") +
  
  theme_pubr() +
  # theme(axis.title.x = element_blank()) +
  
  NULL
# gg_effect

# there is a bug causing "Arial" postscript font error. I can suppress this but it fails to compile into book
#set_null_device("png")
gg <- plot_grid(gg_effect,
                gg_response,
                nrow=2,
                align = "v",
                axis = "rl",
                rel_heights = c(0.35, 1))
gg
```

### Interpretation of the model coefficients 

```{r covariate-fig4h-m1-coef, echo=TRUE}
fig4h_m1_coef <- cbind(coef(summary(fig4h_m1)),
      confint(fig4h_m1))
signif(fig4h_m1_coef, digits = 3)
```

1. The estimate in the "(Intercept)" row ($b_0$) is the expected value of the reference (here, this is the Vehicle group) when $area\_at\_risk=0$. Using Figure \@ref(fig:fig4h-m1-plot-model), this can be visualized as the value of $Y$ where the Vehicle line (the regression line for the reference group) crosses the *y*-axis (at $X = 0$). This is not a meaningful estimate since the area at risk in all hearts is above zero.
2. The estimate in the "area_at_risk" row ($b_1$) is a common slope for all three regression lines. Some might refer to this slope as the "effect" of $area\_at\_risk$ on $area\_of\_necrosis$ but I recommend against using this causal language because the $area\_at\_risk$ is not randomly assigned -- its purpose in *this* model (but not necessarily all linear models with an added covariate) is not for interpretation but for improving inference in the expimental treatment factor.
3. The estimate in the "treatmentMyriocin" row ($b_2$) is the effect of $Myriocin$ on $area\_of\_necrosis$ **conditional on** (or **adjusted for**) $area\_at\_risk$. A frequent phrase is "the estimate controlling for area at risk" but avoid this because "control" implies experimenter intervention and this is not true for $area\_at\_risk$. The value is the difference in the elevation of the regression line for the reference group and that for the Myriocin group. This difference is equal to the difference in conditional expectations at a specific value of the covariate.

\begin{equation}
b_2 = \mathrm{E}(area\_of\_necrosis|treatment = "Myriocin", X = x) - \mathrm{E}(area\_of\_necrosis|treatment = "Vehicle", X = x)
\end{equation}

4. The estimate in the "treatmentMyriocin + DoxSa" row ($b_3$) is the effect of $Myriocin + DoxSa$ on $area\_of\_necrosis$ **conditional on** (or **adjusted for**) $area\_at\_risk$. It's interpretation is similar to that for $b_2$.

How do we summarize these interpretations given the motivating hypothesis? The results do not support the hypothesis. If DoxSa is one of the sphingolipids inducing hypoxia/necrosis, then we'd expect the Myriocin + DoxSa line to be elevated above the Myriocin line. But the coefficient table "tests" this prediction only indirectly. The explicit statistic for this prediction is the (Myriocin + DoxSa) - Myriocin contrast in the contrast table below.

### Everything adds up

Remember that everything adds up in a linear model. A regression line is a line of expected values (the expectation of $Y$ conditional on $X$). The point on the Vehicle line at $area\_of\_risk = 30$ is $b_0 + b_1 \cdot 30$. The point on the "Myriocin" line at $area\_of\_risk = 30$ is $b_0 + b_1 \cdot 30 + b_2$. And, the point on the "Myriocin + DoxSa" line at $area\_of\_risk = 30$ is $b_0 + b_1 \cdot 30 + b_3$. Understanding how the components of the linear model add up gives you phenomenonal cosmic power in statistical analysis.

### Interpretation of the estimated marginal means

```{r}
fig4h_m1_emm <- emmeans(fig4h_m1, specs = "treatment")
fig4h_m1_emm
```
The values in the column "emmean" are the expected values of each group when ("conditional on") $area\_at\_risk$ is equal to the mean $area\_at\_risk$.

```{r echo=FALSE, eval=FALSE}
new_dat <- data.table(treatment = levels(fig4h[, treatment]),
                      area_at_risk = mean(fig4h[, area_at_risk]))
predict(fig4h_m1,
        new_dat,
        se.fit = TRUE,
        interval = "confidence")$fit
```

### Interpretation of the contrasts

```{r}
fig4h_m1_pairs <- contrast(fig4h_m1_emm,
                           method = "revpairwise",
                           adjust = "none") %>%
  summary(infer = TRUE)
fig4h_m1_pairs
```

The values in the column "estimate" are the differences between the estimated marginal means in the estimated marginal means table. The first two contrasts are equal to the coefficients $b_2$ and $b_3$ in the coefficient table.

### Adding the covariate improves inference

Compare the effects in Model $\mathcal{M}_1$ (with the added covariate) to the effects in a linear model without the added covariate.

```{r, echo = FALSE}

fig4h_m0 <- lm(area_of_necrosis ~ treatment, data = fig4h)
fig4h_m0_coeff <- cbind(coef(summary(fig4h_m0)),
                        confint(fig4h_m0))
signif(fig4h_m0_coeff, 3)
```

Inference from these two models is very different. From the model adjusting for area at risk (Model $\mathcal{M}_1$), we would conclude "Unexpectedly, the estimated Myriocin + DoxSa effect (-3.2 mg, 95% CI: -5.9, -0.4) is effectively as big as the Myriocin alone effect." By contrast, using the model without the covariate, we would conclude "Relative to the Myriocin effect, the estimated Myriocin + DoxSa effect is small (.29 mg), and there is large uncertainty in its direction and magnitude (95% CI: -4.1, 4.7)." (Note that the authors published a different conclusion to either of these. I cannot recover the results leading to their conclusion using the methods published by the authors.)

## Understanding interaction effects with covariates
### Fit the model

When we add a continuous covariate to a model we sometimes want to model the **interaction** with the categorical factor variable. An interaction effect represents how the effect of one variable changes given the level of a second variable. The verbal model with an added interaction effect is

$area\_of\_necrosis \sim area\_at\_risk + treatment + area\_at\_risk \times treatment\;\;\;(\mathcal{M}_2)$

which I'll refer to as Model $\mathcal{M}_2$. To understand the model coefficients, it helps to expand this into the full linear model.

\begin{align}
area\_of\_necrosis = \beta_0 &+ \beta_1 area\_at\_risk + \beta_2 treatment_{myriosin} + \beta_3 treatment_{Myriocin\;+\;DoxSa}\\
&+ \beta_4 area\_at\_risk \cdot treatment_{myriosin}\\
&+ \beta_5 area\_at\_risk \cdot treatment_{Myriocin\;+\;DoxSa}\\
&+ \varepsilon
(\#eq:cov-fig4h-interaction)
\end{align}

In addition to the effects of the myriosin ($\beta_2$) and the myriosin + DoxSa ($\beta_3$) treatments of Model $\mathcal{M}_1$, Model $\mathcal{M}_2$ includes coefficients for two **interaction effects**, the interaction between myriosin and area at risk ($\beta_4$) and the interaction between myriosin + DoxSa and area at risk ($\beta_4$). The parameters $\beta_1$ and $\beta_2$ are **additive effects** because they are linear (additive) "in the $X$" variables. The parameters $\beta_3$ and $\beta_4$ are **non-additive effects** because they are not linear "in the $X$" variables -- these effects are coefficients of an $X$ that is the product of two variables. You can see this in the model equation but I explain this further below.

For the goals of the researchers in the heart necrosis study, we specifically would not want to model the interaction. Nevertheless, we would want to plot the interaction effect during the "Step 2: examine the data" phase of analysis to be sure that an additive model is a reasonable model given the data.

```{r}
# see working in R to understand this model formula
fig4h_m2 <- lm(area_of_necrosis ~ area_at_risk*treatment,
               data = fig4h)
```

```{r echo=FALSE, eval=FALSE}
# what model is rstatix::emmeans_test evaluating?
# answer the additive model but the output states non-additive.

emmeans_test(data = fig4h,
    area_of_necrosis ~ treatment, covariate = area_at_risk,
    p.adjust.method = "none"
    )

contrast(emmeans(fig4h_m1, specs="treatment"),
         method = "revpairwise",
         adjust = "none")
contrast(emmeans(fig4h_m2, specs="treatment"),
         method = "revpairwise",
         adjust = "none")

```

### Plot the model with interaction effect

```{r fig4h-m2-plot-model, warning = TRUE, echo = FALSE, fig.cap = "The effect of treatment on area of necrosis, adjusted for area of risk."}
m2 <- lm(area_of_necrosis ~ area_at_risk*treatment, data = fig4h)
b <- coef(m2)

m2_coef <- cbind(coef(summary(m2)),
                 confint(m2))
m2_coef <- data.table(coefficient = row.names(m2_coef),
                      m2_coef)
setnames(m2_coef,
         old = c("Estimate", "Pr(>|t|)", "2.5 %", "97.5 %"),
         new = c("estimate", "p_value", "lower.CL", "upper.CL"))
# pvalString is from package lazyWeave
m2_coef[ , p_pretty := pvalString(p_value)]
# also create a column with "p-val: "
m2_coef[ , pval_pretty := paste("p = ", p_pretty)]

groups <- levels(fig4h$treatment)
line_data <- fig4h[, .(min_x = min(area_at_risk),
                       max_x = max(area_at_risk)), by = treatment]
line_data[, dummy1 := ifelse(treatment == "Myriocin", 1, 0)]
line_data[, dummy2 := ifelse(treatment == "Myriocin + DoxSa", 1, 0)]

line_data[, y_min := b[1] + b[2]*min_x + b[3]*dummy1 + b[4]*dummy2 +
            b[5]*dummy1*min_x + b[6]*dummy2*min_x]
line_data[, y_max := b[1] + b[2]*max_x + b[3]*dummy1 + b[4]*dummy2 +
            b[5]*dummy1*max_x + b[6]*dummy2*max_x]

gg_response <- ggplot(data = fig4h,
                      aes(x = area_at_risk,
                          y = area_of_necrosis,
                          color = treatment)) +
  # points
  geom_point(size = 2) +
  
  geom_segment(data = line_data,
               aes(x = min_x,
                   y = y_min,
                   xend = max_x,
                   yend = y_max)) +
  scale_color_manual(values = pal_ito_seven) +
  ylab("Area of necrosis (mg)") +
  xlab("Area at risk (mg)") +
  theme_pubr() +
  theme(legend.position="bottom") +
  
  NULL
#gg_response

contrast_order <- m2_coef[, coefficient]
m2_coef[, term := factor(coefficient, contrast_order)]

gg_effect <- ggplot(data = m2_coef, 
                    aes(y = coefficient,
                        x = estimate)) +
  # confidence level of effect
  geom_errorbar(aes(xmin = lower.CL, 
                    xmax = upper.CL),
                width = 0, 
                color = "black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # draw a line at effect = 0
  geom_vline(xintercept = 0, linetype = 2) +
  
  # p-value. The y coordinates are set by eye
  annotate(geom = "text",
           label = m2_coef$pval_pretty,
           y = 1:6,
           x = 4.5) +
  
  # x-axis label and aesthetics
  xlab("value") +
  ylab("Coefficient") +
#  coord_cartesian(xlim = c(-8,5.5)) +
  scale_x_continuous(position="top") +
  
  theme_pubr() +
  # theme(axis.title.x = element_blank()) +
  
  NULL
# gg_effect

gg <- plot_grid(gg_effect,
                gg_response,
                nrow=2,
                align = "v",
                axis = "rl",
                rel_heights = c(0.6, 1)
                )
gg_response
```

### Interpretation of the model coefficients 

```{r covariate-fig4h-m2-coef, echo=TRUE}
fig4h_m2_coef <- cbind(coef(summary(fig4h_m2)),
      confint(fig4h_m2))
round(fig4h_m2_coef[,c(1,2)], digits = 2)
```
1. The estimate in the "(Intercept)" row ($b_0$) is the expected value of the reference (here, this is the Vehicle group) when $area\_at\_risk=0$. This is the same interpretation of the intercept in the additive model (Model $\mathcal{M}_1$) but the value is different. This is because the slope coefficient ($b_1$) estimated in Model $\mathcal{M}_1$ is a common slope (to all three groups) estimated by pooling data from all three groups. The slope coefficient in the non-additive model (Model $\mathcal{M}_2$) is computed from just the reference data. Because the slope differs between Model $\mathcal{M}_1$ and Model $\mathcal{M}_2$, the intercept differs (think about why they are necessarily related). As in Model $\mathcal{M}_1$, the value and inferential statistics are meaningful for this parameterization but there are slight modifications of the model that can make this meaningful.
2. The estimate in the "area_at_risk" row ($b_1$) is the slope of the regression line for the reference (Vehicle) group, that is, the slope conditional on $treatment = \mathrm{''Vehicle''}$
3. The estimate in the "treatmentMyriocin" row ($b_2$) is the effect of $Myriocin$ on $area\_of\_necrosis$ **conditional on** $area\_at\_risk = 0$. Note the difference in interpretation with $b_2$ in Model $\mathcal{M}_1$. The difference is because of the interaction. The effect of $Myriocin$ on $area\_of\_necrosis$ is no longer constant for all values of $area\_at\_risk$. This is easily seen in Figure \@ref(fig:fig4h-m2-plot-model), where the effect of Myriocin (the vertical distance between the Vehicle and Myriocin line) is very small at small values of area at risk but very large at large areas at risk.
4. The estimate in the "treatmentMyriocin + DoxSa" row ($b_3$) is the effect of $Myriocin + DoxSa$ on $area\_of\_necrosis$ **conditional on** $area\_at\_risk = 0$. It's interpretation is similar to that for $b_2$.
5. The estimate in the "area_at_risk:treatmentMyriocin" row ($b_4$) is the **interaction effect** between area at risk and Myriocin. It's value is the difference between the slope of the regression line through the Myriocin points in Figure \@ref(fig:fig4h-m2-plot-model) and the slope of the regression line through the reference (Vehicle) points in Figure \@ref(fig:fig4h-m2-plot-model). Consequently, the slope of the regression line through the Myriocin points is $b_1 + b_4$.
6. The estimate in the "area_at_risk:treatmentMyriocin + DoxSa" row ($b_5$) is the **interaction effect** between area at risk and Myriocin + DoxSa It's value is the difference between the slope of the regression line through the Myriocin + DoxSa points in Figure \@ref(fig:fig4h-m2-plot-model) and the slope of the regression line through the reference (Vehicle) data in Figure \@ref(fig:fig4h-m2-plot-model). Consequently, the slope of the regression line through the Myriocin + DoxSa points is $b_1 + b_5$.

Interaction effects are *differences in slopes* (this is also true with interactions between two factor variables, even though we usually describe this as "differences in differences") and these difference are easily seen by inspection of a plot similar to that in Figure \@ref(fig:fig4h-m2-plot-model). The bigger the interaction, the less parallel the regression line.

### What is the effect of a treatment, if interactions are modeled? -- it depends.

A very useful way to remember what an interaction *is* is "it depends". What is the effect of Myriocin on area at risk? With an interaction term in the linear model, the answer is "it depends". It depends on the level of $area\_at\_risk$. At small values of area at risk, the effect of Myriocin is very small. At large values of area at risk, the effect of Myriocin is large. Likewise, what is the effect of area at risk on necrosis? It depends (if interactions are modeled). It depends on the level of treatment, for example the effect in the Myriocin group is smaller (a smaller slope) than the effect in the Vehicle group. "It depends" is the hallmark of interactions -- the effect of one variable on the outcome depends on the level of a second variable.

### Which model do we use, $\mathcal{M}_1$ or $\mathcal{M}_2$?

The answer to this question depends partly on our goals. If we are explicitly interested in measuring an interaction effect (perhaps some theory predicts an positive interaction) then we would necesarily add interaction effects to the model. But if we are interested in estimating a treatment effect conditional on a covariate, then we don't include the interaction. Remember that a purpose of initial explorations of data ("Step 2 -- examine the data") is to help decide which model to specify in the model formula. If an initial plot shows large interactions between treatment levels and the covariate, then adjusting for the covariate won't work, or at least won't work without additional complexity to the interpretation. "Large interactions" of course raises the question, how large is too larege to ignore interactions? Many textbooks in biostatistics recommend using an interaction *p*-value to make this decision. I disagree. There is no objective answer to the question. We lose information with all models -- that is what we do in science. It is up to the researcher to be transparent with all decisions. If an additive (no interactions) model is used to condition on (adjust for) a covariate, the researcher should use the supplement to plot the model with interactions as evidence of why this decision is reasonable.

## Understanding ANCOVA tables


## Working in R
### Importing the heart necrosis data

```{r, echo = TRUE}
data_folder <- "data"
data_from <- "1-Deoxydihydroceramide causes anoxic death by impairing chaperonin-mediated protein folding"
file_name <- "42255_2019_123_MOESM7_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)
  
sheet_i <- "Figure 4h"
fig4h_1 <- read_excel(file_path,
                         sheet = sheet_i,
                         range = "A5:B19",
                         col_names = TRUE) %>%
  clean_names() %>%
  data.table()
fig4h_2 <- read_excel(file_path,
                         sheet = sheet_i,
                         range = "D5:E17",
                         col_names = TRUE) %>%
  clean_names() %>%
  data.table()
fig4h_3 <- read_excel(file_path,
                         sheet = sheet_i,
                         range = "G5:H15",
                         col_names = TRUE) %>%
  clean_names() %>%
  data.table()

fig4h <- rbind(data.table(treatment = "Vehicle", fig4h_1),
                    data.table(treatment = "Myriocin", fig4h_2),
                    data.table(treatment = "Myriocin + DoxSa", fig4h_3))

treatment_levels <- c("Vehicle", "Myriocin", "Myriocin + DoxSa")
fig4h[, treatment := factor(treatment, levels = treatment_levels)]

```

### Fitting the model
For an additive model, add the covariate to the model formula using the `+` operator. The order of the variables doesn't matter (the right-hand-side of the model formula could be `treatment + area_at_risk`).

```{r}
m1 <- lm(area_of_necrosis ~ area_at_risk + treatment, data = fig4h)
```

For an nonadditive model with interactions, add the covariate to the model formula using the `*` operator.  `~ area_at_risk * treatment` is a shortcut to the full model formula, which is `1 + area_at_risk + treatment + area_at_risk * treatment`. R expands the short formula to the full formula automatically. Again, the order of the variables doesn't matter (the right-hand-side of the model formula could be `treatment * area_at_risk`).

```{r}
m2 <- lm(area_of_necrosis ~ area_at_risk * treatment, data = fig4h)
```

### Using the emmeans function

```{r}
m1_emm <- emmeans(m1,
        specs = c("treatment"))
m1_emm
```
```{r}
m1_emm <- emmeans(m1,
        specs = c("area_at_risk", "treatment"))
m1_emm

```

```{r}
# recenter emms at the mean of the vehicle group

vehicle_mean_x <- mean(fig4h[treatment == "Vehicle",
                         area_at_risk])
m1_emm_2 <- emmeans(m1,
                    specs = c("area_at_risk","treatment"),
                    at = list(area_at_risk = vehicle_mean_x))
m1_emm_2
```
```{r}
fig4h[, .(mean = mean(area_of_necrosis)),
      by = treatment]
```

### ANCOVA tables

If you are going to report statistics from an ANCOVA table, be sure that what you report is what you think you are reporting -- it is very easy to get this wrong in R. One way to easily get a table using Type II or III sums of squares is to set the contrast option of the factor variable to `contr.sum` (this tells R to use deviation coding to form the model matrix for the linear model) and then use the function `Anova` from the car package (not base R `anova`).

1. It is safest to set this with the `lm` function itself.
2. The coefficients of the linear model will have a different interpretation than that given above. Either understand the new interpretation (not given here) or fit two models, one without the reset contrasts and one with the reset contrasts.
3. The estimated marginal mean table and the contrast table are the same, regardless of how contrasts are set in the linear model.
4. Many websites show code that changes the default coding in the R session. I strongly caution against this. If you don't change it back, then you are likely to misinterpret coefficients in later analyses (see point 2).

For additive models, use Type II sum of squares (Type III will give same results), 

```{r echo = FALSE}
# for coefficients
m1 <- lm(area_of_necrosis ~ treatment + area_at_risk,
         data = fig4h)
# for anova
m1_aov <- lm(area_of_necrosis ~ treatment + area_at_risk,
         data = fig4h,
         contrasts = list(treatment = contr.sum))
```

```{r}
Anova(m1_aov, type = 2) # be sure to use the correct model!
```
For non-additive models with interactions

1. The statistics for the interaction effect are the same, regardless of the sum of squares used.
2. The interpretation of the main effects is problematic, at best. This is discussed more in the ANOVA chapter.
3. Some statisticians recommend using Type III sums of squares. The effect of $treatment* has a very specific interpretation with Type III -- it is the effect when the covariate = 0. This can be biologically relevant if the covariate data is re-centered.

```{r}
# for coefficients
m2 <- lm(area_of_necrosis ~ treatment * area_at_risk,
         data = fig4h)
# for anova
m2_aov <- lm(area_of_necrosis ~ treatment * area_at_risk,
         data = fig4h,
         contrasts = list(treatment = contr.sum))
```

```{r}
Anova(m2_aov, type = 2) # be sure to use the correct model!
```
```{r}
Anova(m2_aov, type = 3) # be sure to use the correct model!

```

### Plotting the model
#### Using ggpubr

ggpubr will only plot the model with interactions.

```{r message=FALSE}
ggscatter(data = fig4h,
          x = "area_at_risk",
          y = "area_of_necrosis",
          color = "treatment",
          add = "reg.line",
          ylab = "Area of necrosis (mg)",
          xlab = "Area at risk (mg)",
          palette = pal_okabe_ito
)


```

#### ggplot2

```{r, message = FALSE}
# pal_batlow <- scico(n, alpha = NULL, begin = 0, end = 1, direction = 1, palette = "batlow")

m1 <- lm(area_of_necrosis ~ area_at_risk + treatment,
         data = fig4h)

gg <- ggplot(fig4h, 
             aes(x = area_at_risk,
                 y = area_of_necrosis,
                 color = treatment)) +
  geom_point(aes(color=treatment)) +
  geom_smooth(method = "lm",
              mapping = aes(y = predict(m1, fig4h))) +
  scale_color_manual(values = pal_okabe_ito,
                     name = NULL) +
  # scale_color_manual(values = pal_batlow,
  #   name = NULL) +
  ylab("Area of necrosis (mg)") +
  xlab("Area at risk (mg)") +
  theme_pander() +
  theme(legend.position="bottom") +
  NULL
gg
```


#### A response-effects plot using ggplot2

```{r}
m1 <- lm(area_of_necrosis ~ area_at_risk + treatment, data = fig4h)
b <- coef(m1)

groups <- levels(fig4h$treatment)
line_data <- fig4h[, .(min_x = min(area_at_risk),
                       max_x = max(area_at_risk)), by = treatment]
line_data[, dummy1 := ifelse(treatment == "Myriocin", 1, 0)]
line_data[, dummy2 := ifelse(treatment == "Myriocin + DoxSa", 1, 0)]

line_data[, y_min := b[1] + b[2]*min_x + b[3]*dummy1 + b[4]*dummy2]
line_data[, y_max := b[1] + b[2]*max_x + b[3]*dummy1 + b[4]*dummy2]

gg_response <- ggplot(data = fig4h,
                      aes(x = area_at_risk,
                          y = area_of_necrosis,
                          color = treatment)) +
  # points
  geom_point(size = 2) +
  
  geom_segment(data = line_data,
               aes(x = min_x,
                   y = y_min,
                   xend = max_x,
                   yend = y_max)) +
  scale_color_manual(values = pal_okabe_ito,
                     name = NULL) +
  ylab("Area of necrosis (mg)") +
  xlab("Area at risk (mg)") +
  theme_pubr() +
  theme(legend.position="bottom") +

  NULL
#gg_response
```

```{r, warning=FALSE, message=FALSE}
m1_emm <- emmeans(m1, specs = "treatment")
m1_pairs <- contrast(m1_emm,
                     method = "revpairwise",
                     adjust = "none") %>%
  summary(infer = TRUE) %>%
  data.table()
# pvalString is from package lazyWeave
m1_pairs[ , p_pretty := pvalString(p.value)]
# also create a column with "p-val: "
m1_pairs[ , pval_pretty := paste("p = ", p_pretty)]

contrast_order <- m1_pairs[, contrast]
m1_pairs[, contrast := factor(contrast, contrast_order)]

gg_effect <- ggplot(data = m1_pairs, 
                    aes(y = contrast,
                        x = estimate)) +
  # confidence level of effect
  geom_errorbar(aes(xmin = lower.CL, 
                    xmax = upper.CL),
                width = 0, 
                color = "black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # draw a line at effect = 0
  geom_vline(xintercept = 0, linetype = 2) +
  
  # p-value. The y coordinates are set by eye
  annotate(geom = "text",
           label = m1_pairs$pval_pretty,
           y = 1:3,
           x = 4.5) +
  
  # x-axis label and aesthetics
  xlab("Effect (mg)") +
  ylab("Contrast") +
  coord_cartesian(xlim = c(-8,5.5)) +
  scale_x_continuous(position="top") +
  
  theme_pubr() +
  # theme(axis.title.x = element_blank()) +
  
  NULL
# gg_effect
```


```{r, warning=TRUE, message=FALSE}
gg <- plot_grid(gg_effect,
                gg_response,
                nrow=2,
                align = "v",
                axis = "rl",
                rel_heights = c(0.35, 1)
                )
gg
```

## Best practices
### Do not use a ratio of part:whole as a response variable -- instead add the denominator as a covariate
### Do not use change from baseline as a response variable -- instead add the baseline measure as a covariate

### Do not "test for balance" of baseline measures

A test of the null hypothesis of no difference in mean at baseline is a "test for balance." Researchers frequently test for balance at baseline and use the *p*-value of the test to decide the next step: 1) if $p > 0.05$, conclude that the pre-treatment means "do not differ" and use something like a simple *t* test of the post-treatment means, 2) if $p < 0.05$, then use the change score, or the percent change, as the response in a simple *t*-test, or 3) if $p < 0.05$, then use use a linear model with the pre-treatment value as a covariate. Here, and in general, hypothesis tests used to decide which of several ways to proceed do not make sense. First, a null-hypothesis significance test cannot tell you that there is "no difference" -- this is not what null-hypothesis tests do. Second, any $p$-value after the initial test isn't strictly valid as it does not take into account this decision step, but this is minor. Third, **it doesn't matter**; there will always be some difference in the actual means of the initial measures and, consequently, the conditional expectation of the final measures, or change in measures, or percent change will be dependent on this initial difference. So, if one has initial measures, one should use an linear model that adjusts for baseline measures to estimate the treatment effect in pre-post designs. And, if one isn't planning on taking an initial measure, then maybe you should, because the initial measure used in a linear model allows a better estimate of the treatment effect, as discussed above in [Adding covariates can increases the precision of the effect of interest].

## Best practices 2: Use a covariate instead of normalizing a response

<!--chapter:end:chapters/32-covariates.Rmd-->

---
title: "Models for two (or more) categorical X variables -- Factorial designs"
author: "Jeffrey A. Walker"
date: "11/13/2020"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 3
    toc_float: true
    fig_caption: yes
    number_sections: true
    code_folding: show
    includes:
      before_body: copyright.html
---
# Linear models with two categorical $X$ -- Factorial designs ("two-way ANOVA") {#factorial}

```{r factorial-setup, echo=FALSE, warning=FALSE, message=FALSE}
library(here)
library(janitor)
library(readxl)
library(data.table)
library(stringr)
library(dplyr)

# analysis packages
library(nlme) # gls
library(lmerTest) # lmm
library(emmeans)
library(car) # qqplot, spreadlevel
library(lmPerm)

# graphing packages
library(ggplot2) # ggplot environment
library(ggpubr) # publication ready plots
library(ggforce) # jitter
library(ggsci) # color palettes
library(ggpubfigs) # color palettes
library(ggthemes) # themes and palettes including colorblind
# execute this command only the first time you use ggpubfigs
# devtools::install_github("JLSteenwyk/ggpubfigs")
library(cowplot) # combine plots
library(lazyWeave) # pretty pvalues
library(broom)
library(kableExtra) #tables


here <- here::here
data_folder <- "data"

# Okabe & Ito palette
pal_ito_seven <- friendly_pal("ito_seven") # ggpubfigs
pal_okabe_ito <- colorblind_pal()(8)[2:8] # ggthemes

```

## Factorial experiments 
A factorial experiment is one in which there are two or more factor variables (categorical $X$) that are **crossed**, resulting in a group for each combination of the levels of each factor. Factorial experiments are used to estimate the **interaction** effect between factors. Interaction effects were introduced in Chapter xxx, Models with a added covariate. Remember the key to interactions -- two factors interact when the effect of one factor depends on the level of the other factor. Interactions are ubiquitous, although sometimes they are small enough to ignore with little to no loss of understanding.

## Example 1: Maize defense response

This chapter uses data from an experiment measuring the effect of two defense signals on the defense response in Maize plants. In response to herbivory from insects, maize, and other plants, release multiple, chemical signals into the air (chemicals that evaporate into the air are known as volatile compounds). These chemicals signal the plant, and neighboring plants, to secrete anti-herbivory hormones, including abcisic acid and jasmonic acid. The researchers investigated the effects of two volatile compounts, (Z)‐3‐hexenyl acetate (HAC) and Indole, on the defense response both each without the other and in combination.

The example data come from Figure 1a, which is the effect of HAC and Indole on tissue concentrations of the hormone abcisic acid (ABA). The design is fully crossed with two factors, each with two levels: $hac$, with levels "HAC-" and "HAC+", and $indole$, with levels ("Indole-" and "Indole+").  A good way to visualize the treatment combinations in a crossed design is with a $m \times p$ table showing all combinations of the $m$ levels of factor 1 ($hac$) against the $p$ levels of factor 2 ($indole$).

```{r factorial-2way-table, echo = FALSE, message=FALSE, warning=FALSE}
df <- data.frame("HAC-" = c("Control", "Indole"),
                 "HAC+" = c("HAC", "HAC+Indole"))
row.names(df) <- c("Indole-", "Indole+")
knitr::kable(df, col.names = c("HAC-", "HAC+"))
```

```{r factorial-import-ecology, echo=FALSE, warning=FALSE, message=FALSE}

data_from <- "Integration of two herbivore-induced plant volatiles results in synergistic effects on plant defense and resistance"
file_name <- "Data for Dryad.xlsx"
file_path <- here(data_folder, data_from, file_name)

fig1 <- read_excel(file_path,
                     sheet = "Fig. 1",
                     range = "A3:K23", # 1 blank column
                     col_names = TRUE) %>%
  data.table() %>%
  clean_names()

fig1[treat == "Control", c("hac", "indole") := list("HAC-", "Indole-")]
fig1[treat == "HAC", c("hac", "indole") := list("HAC+", "Indole-")]
fig1[treat == "Indole", c("hac", "indole") := list("HAC-", "Indole+")]
fig1[treat == "HAC+Indole", c("hac", "indole") := list("HAC+", "Indole+")]

treat_levels <- c("Control", "HAC", "Indole", "HAC+Indole")
fig1[, treat := factor(treat, levels = treat_levels)]
fig1[, hac := factor(hac)] # levels in correct order
fig1[, indole := factor(indole)] # levels in correct order
# View(fig1)
```
### Analysis

#### Examine the data

```{r}
qplot(x = treat, y = aba, data = fig1)
```

Too few points for box plot. control variance is small. No obvious implausible points. fit with lm but recognize small n warning for any inference.

#### Fit the model
```{r factorial-fit-model}
m1 <- lm(aba ~ hac * indole, data = fig1)
```

#### Model check

```{r}
qqPlot(m1, id = FALSE)
```

A plausible sample from a Normal distribution.

```{r}
spreadLevelPlot(m1)
```
A bit of increase in variance with mean. 

#### Inference from the model

```{r}
m1_coef <- cbind(coef(summary(m1)),
                 confint(m1))
knitr::kable(m1_coef, digits = 3)
```


```{r}
m1_emm <- emmeans(m1, specs = c("hac", "indole"))
knitr::kable(m1_emm, digits = 2)
```

```{r}
m1_pairs <- contrast(m1_emm,
                     method = "revpairwise",
                     simple = "each",
                     combine = TRUE,
                     adjust = "none") %>%
  summary(infer = TRUE)

knitr::kable(m1_pairs, digits = 2)
```
```{r}
m1_interaction <- contrast(m1_emm,
                           interaction = c("revpairwise"),
                           by = NULL) %>%
  summary(infer = TRUE)

m1_interaction %>%
  kable(digits = c(1,1,1,2,0,2,1,2,3)) %>%
  kable_styling()
```

#### Plot the model

xxx if interested in synergy plot the coefficients! this is what we want. When would we want to plot all simple effects?

```{r factorial-aba-prepare-plot, echo=FALSE}

m1_emm_dt <- summary(m1_emm) %>%
  data.table
m1_emm_dt[, treat := levels(fig1$treat)]
m1_pairs_dt <- data.table(m1_pairs)

# create contrast column with unique value each row
within_group <- c("Indole-", "Indole+", "HAC-", "HAC+")
m1_pairs_dt[, contrast := paste(within_group, contrast, sep =": ")]

# add interaction row
m1_pairs_dt <- rbind(m1_pairs_dt,
                     data.table(contrast = "Interaction",
                                estimate = m1_coef[4, "Estimate"],
                                lower.CL = m1_coef[4, "2.5 %"],
                                upper.CL = m1_coef[4, "97.5 %"],
                                p.value = m1_coef[4, "Pr(>|t|)"]),
                     fill = TRUE)

# pvalString is from package lazyWeave
m1_pairs_dt[ , p_pretty := pvalString(p.value)]
# also create a column with "p-val: "
m1_pairs_dt[ , pval_pretty := paste("p-val:", p_pretty)]
```


```{r factorial-aba-build-plot, echo=FALSE}
# response plot using treat column
gg_response <- ggplot(data = fig1,
                            aes(x = aba,
                                y = treat,
                                color = hac,
                                shape = indole)) +
  # points
  geom_jitter(alpha = 0.5,
            height = 0.4) +
  
  # plot means and CI
  geom_errorbar(data = m1_emm_dt,
                aes(x = emmean,
                    xmin = lower.CL,
                    xmax = upper.CL,
                    color = hac),
                width = 0,
  ) +
  
  geom_point(data = m1_emm_dt,
             aes(x = emmean,
                 color = hac),
             size = 3
  ) +

  # aesthetics
  xlab("ABA (ng per g FW)") +
  scale_color_manual(values=pal_okabe_ito,
                     name = NULL) +
  theme_pubr() +
  theme(legend.position="none") +
  theme(axis.title.y=element_blank()) +
  
  NULL

# gg_effect

contrast_order <- m1_pairs_dt[, contrast]
m1_pairs_dt[, contrast := factor(contrast, contrast_order)]
contrast_labels <- c("HAC - Control",
                     "HAC+Indole - Indole",
                     "Indole - Control",
                     "HAC+Indole - HAC",
                     "Interaction")

gg_effect <- ggplot(data = m1_pairs_dt, 
                    aes(x = estimate,
                        y = contrast)) +
  # confidence level of effect
  geom_errorbar(aes(xmin=lower.CL, 
                    xmax=upper.CL),
                width=0, 
                color="black") +
  # estimate of effect
  geom_point(size = 3) +
  
  # draw a line at effect = 0
  geom_vline(xintercept=0, linetype = 2) +
  # draw a line to sep interaction from simple effects
  geom_vline(xintercept=4.5, color = "gray") +
  
  # p-value. The y coordinates are set by eye
  annotate(geom = "text",
           label = m1_pairs_dt$p_pretty,
           y = 1:5,
           x = 40,
           size = 3) +
  
  # aesthetics
  
  scale_y_discrete(labels = contrast_labels) +
  scale_x_continuous(position = "top") + # move to top

  xlab("Effect (ng per g FW)") +
  ylab("Contrast") +
  coord_cartesian(xlim = c(-10,40)) +
  
  # use ggpubr theme
  theme_pubr() +
  theme(axis.title.y=element_blank()) +
  
  NULL

#gg_effect
```

```{r factorial-aba-plot, echo=FALSE}
plot_grid(gg_effect,
          gg_response,
          nrow=2,
          align = "v",
          axis = "lr",
          rel_heights = c(0.5,1))
```

### A factorial model adds an interaction effect to the coefficients table

A factorial design allows a researcher to estimate the interaction between two factors. An interaction is an effect between two variables in a linear model. The interaction effect is the coefficient $beta_3$ in the linear model

\begin{equation}
aba = \beta_0 + \beta_1 hac_{HAC^+} + \beta_2 indole_{Indole^+} + \beta_3 hac_{HAC^+}:indole_{Indole^+} +\varepsilon
(\#eq:factorial-full)
\end{equation}

$hac_{HAC^+}$ and $indole_{Indole^+}$ are dummy-coded **indicator variables** indicating group membership. $hac_{HAC^+} : indole_{Indole^+}$ is a dummy-coded variable for the interaction between $hac$ and $indole$. The value of this variable is the product of the two indicator variables in the interaction ($hac_{HAC^+}$ and $indole_{Indole^+}$), which can be verified with the model matrix (which here, is computed from the subset of the data that includeds only the first two rows of each treatment combination)

```{r factorial-dummy, echo=FALSE}
m1_mm <- model.matrix(m1)
first_two <- c(1:2, 6:7, 11:12, 16:17)
knitr::kable(m1_mm[first_two,])
```

The coefficient table is

```{r factorial-coef, echo=FALSE}
knitr::kable(m1_coef, digits=c(2,2,1,3, 2, 2), caption="Coefficient table of the factorial model")
```

1. The Intercept ($b_0$) is the mean of the reference (HAC-/Indole-) group, and so the mean of the upper left cell ("Control") in Table \@ref(tab:factorial-coef).
2. The hacHAC+ coefficient ($b_1$) is the estimate of the added HAC effect relative to the reference, and so is the mean of the lower left cell ("HAC") minus the mean of the upper left cell ("Control"). Another way of stating this is, it is **the effect of HAC when Indole is at its reference level**.
3. The indoleIndole+ coefficient ($b_2$) is the estimate of the added Indole effect relative to the reference, and so is the mean of the upper right cell ("Indole") minus the mean of the upper left cell ("Control"). Another way of stating this is, **it is the effect of Indole when HAC is at its reference level**.
4. The hacHAC+:indoleIndole+ coefficient ($b_3$) is the estimate of the **interaction effect**. If we added the HAC effect ($b_1$) and the Indole effect ($b_2$) to the Control mean ($b_0$), we would get the expected mean of the "HAC+Indole" group **if the effects were additive**. The hacHAC+:indoleIndole+ effect ($b_3$) is the additional bit that we need to get from this additive expectation to the modeled HAC+Indole mean (Figure \@ref(factorial-what-are-coefficients-plot)). It is the difference between the mean of the bottom right cell ("HAC+Indole") and the sum of the coefficients of the other three cells ($b0$, $b1$, $b2$).

An interaction is a **non-additive effect**. Think about this. Adding HAC alone increases ABA concentration by 13.3 ng per g FW. Adding Indole alone increases ABA concentration by 8.8 ng per g FW. If these effects were purely additive, then adding both HAC and Indole to the Control mean should result in a mean of 19.2 + 13.3 + 8.8 = 41.3 ng per g FW in the HAC+Indole group. The modeled mean is 54.1 ng per g FW. The difference observed - additive is 54.1 - 41.3 = 12.8 ng per g FW. Compare this to the interaction coefficient in the coefficient table.

```{r factorial-what-are-coefficients-plot, echo=FALSE, fig.cap="Meaning of coefficients in factorial model. b0 (blue line) is the mean of the reference (Control). b1 (orange line) is the /HAC effect. Numerically, it is the mean of the HAC group minus the mean of the reference. b2 (green line) is the Indole effect. Numerically it is the mean of the Indole group minus the mean of the reference. The expected mean of the HAC+Indole group if HAC and Indole were additive is b0 + b1 + b2 (gray circle). b3 (purple line) is the interaction effect. Numerically, it is the observed mean of the HAC+Indole group minus the expected additive mean (gray circle)"}
m1 <- lm(aba ~ hac * indole, data = fig1)
m1_emm <- emmeans(m1, specs=c("hac", "indole"))%>%
  summary() %>%
  data.table()

treatment_levels <- c("Control", "HAC", "Indole", "HAC+Indole")
m1_emm[, treatment := factor(treatment_levels, treatment_levels)]

b <- coef(m1)
b0 <- b[1]
b1 <- b[2]
b2 <- b[3]
b3 <- b[4]


gg_add <- ggplot(data = m1_emm,
                 aes(x = treatment,
                     y = emmean,
                     color = treatment)) +
  geom_point(size = 4, alpha = 0.5) +
  coord_cartesian(ylim = c(0, 60)) +

  geom_point(aes(x = 4,
                 y = b0 + b1 + b2),
             size = 4,
             alpha = 0.5) +
  
  geom_segment(aes(x = c(1, 2, 3, 4),
                   y = c(0,emmean[1],emmean[1],emmean[1]),
                   xend = c(1, 2, 3, 4),
                   yend = emmean),
               size = 1,
               show.legend = FALSE) +
 
  geom_segment(aes(x = c(4),
                   y = c(emmean[1]),
                   xend = c(4),
                   yend = c(emmean[2])),
               size = 1,
               color = pal_ito_seven[2]) +
  geom_segment(aes(x = c(4),
                   y = c(emmean[2]),
                   xend = c(4),
                   yend = c(emmean[2] + (emmean[3]-emmean[1]))),
               size = 1,
               color = pal_ito_seven[3]) +
  
  annotate(geom = "text",
           label="b[0]",
           parse=TRUE,
           x = 1.15,
           y = b0/2) +
  annotate(geom = "text",
           label="b[1]",
           parse=TRUE,
           x = 2.15,
           y = b0 + b1/2) +
  annotate(geom = "text",
           label="b[2]",
           parse=TRUE,
           x = 3.15,
           y = b0 + b2/2) +
  annotate(geom = "text",
           label="b[3]",
           parse=TRUE,
           x = 4.15,
           y = b0 + b1 + b2 + b2/2) +

  annotate(geom = "text",
           label="interaction effect",
           x = 2.55,
           y = 51.5,
           size = 4) +
  geom_segment(aes(x = 3.1,
                   y = 51,
                   xend = 3.9,
                   yend = b0+b1+b2+b3/2),
               arrow=arrow(length = unit(0.05, "npc"),
                           ends="last",type="open"),
               color="black") +

  annotate(geom = "text",
           label="additive expectation",
           x = 2.55,
           y = 44.5,
           size = 4) +
  geom_segment(aes(x = 3.2,
                   y = 44,
                   xend = 3.9,
                   yend = b0+b1+b2+0.3),
               arrow=arrow(length = unit(0.05, "npc"),
                           ends="last",type="open"),
               color="black") +
  
  scale_color_manual(values=pal_ito_seven,
                     name = "Group mean") +
  
  ylab("ABA (ng per g FW)") +
  theme_minimal_hgrid() +
  
  theme(axis.title.x=element_blank()) +
  scale_y_continuous(
    # don't expand y scale at the lower end
    expand = expansion(mult = c(0, 0.05)))  +

  NULL
  
gg_add
```

### The biological interpretation of an interaction effect

The biological reasons causing interaction effects are highly variable but lets consider how interactions *might* arise in the context of the ABA defense response to HAC and Indole. Additive effects (no interaction) may occur when combined treatments act independently of each other. This might occur in the Maize ABA response if the signaling path from HAC reception to ABA secretion and Indole reception to ABA secretion occur in different cells or by different signaling pathways and activity in either pathway has no influence on the other. Positive, or **synergistic** interaction effects may occur when combined treatments augment each other's ability to affect the response. This could occur in the Maize ABA response if an active signaling path from one of the volatile compound to ABA secretion makes the signaling path from the other compound to ABA secretion more sensitive to that compound. Negative, or **antagonistic** interaction effects may occur when combined treatments interfere with each other's ability to affect the response. This could occur in the Maize ABA response if an active signaling path from one of the volatile compound to ABA secretion inhibits the signaling path from the other compound to ABA secretion. Negative interaction effects could also occur if there is a simple threshold response and all it takes is either the HAC or the Indole signaling pathway to activate the response.

### Conditional and marginal means

```{r factorial-conditional-means, echo=FALSE, message=FALSE}
options(knitr.kable.NA = '')
ycols <- c("group", names(m1_emm))

m1_emm_dt <- emmeans(m1, specs=c("hac", "indole")) %>%
  summary() %>%
  data.table()

m1_emm_hac <- emmeans(m1, specs=c("hac")) %>%
  summary() %>%
  data.table()

m1_emm_indole <- emmeans(m1, specs=c("indole")) %>%
  summary() %>%
  data.table()


hac_0 <- c(m1_emm_dt[hac == "HAC-", emmean],
           m1_emm_hac[hac == "HAC-", emmean])
hac_1 <- c(m1_emm_dt[hac == "HAC+", emmean],
           m1_emm_hac[hac == "HAC+", emmean])
indole_means <- c(m1_emm_indole[, emmean], NA)

means_table <- data.frame("HAC-" = hac_0,
                          "HAC+" = hac_1,
                          "mean" = indole_means)
row.names(means_table) <- c("Indole-", "Indole+","mean")


bgc1 <- pal_okabe_ito[1]
bgc2 <- pal_okabe_ito[2]

knitr::kable(means_table, 
             digits = 3,
             col.names = c("HAC-", "HAC+", "mean"),
             caption = "Conditional (white background) and marginal (color background) means from full factorial model fit to ABA data") %>%
  kable_styling(full_width = FALSE) %>%
  # row_spec(row = 0,
  #          bold = F) %>%
  column_spec(column = 1,
              bold = c(T, T, T)) %>%
  column_spec(column = 4,
              background = c(bgc1, bgc1, "white")) %>%
  column_spec(column = 3,
              background = c("white", "white", bgc2)) %>%
  column_spec(column = 2, 
              background = c("white", "white", bgc2))
 

```

The conditional means from model \@ref(eq:factorial-full) are shown in the upper left $2 \times 2$ block (white background) of Table \@ref(tab:factorial-conditional-means). These means are conditional on the level of $hac$ and $indole$. The values in the last row and column are the **marginal means**, which are the means of the associated row or column cells (these values are in the margins of the table). More generally, *marginal* refers to a statistic averaged across multiple levels of another variable. The marginal means of $indole$ (orange background) are the means of the `Indole+` and `Indole-` rows. The marginal means of $hac$ (blue background) are the means of the `HAC+` and `HAC-` columns.

### Simple (conditional) effects

In a factorial experiment with crossed A and B factors, there is an effect of factor A (relative to the reference, or another level of factor A) for each of the *p* levels of factor B. And, there is an effect of factor B (relative to the reference, or another level of factor B) for each of the *m* levels of factor A.

For the maize defense response experiment, there is an Indole effect at `hac = HAC-` and a different effect at `hac = HAC+`. Similarly, there is a HAC effect at `indole = Indole-` and a different effect at `indole = Indole+`. These effects, or **contrasts** (differences in modeled means), are called the **simple effects**. I prefer **conditional effects**, since the value of the effect is conditional on the level factor of the other factor.

```{r, echo=FALSE}
m1_emm <- emmeans(m1, specs=c("hac", "indole"))
m1_pairs <- contrast(m1_emm,
                        method = "revpairwise",
                        adjust = "none",
                        simple = "each",
                        combine = TRUE)
knitr::kable(m1_pairs, 
             digits=c(1,1,1,1,2,1,1,5),
             caption = "Conditional (or \"simple\") effects of the factorial model.") %>%
  kable_styling()
```

The first two rows are the conditional effects of $hac$ in each of the levels of $Indole$. The last two rows are the conditional effects of $Indole$ in each of the levels of $hac$.

```{r factorial-simple-effects, echo = FALSE}
options(knitr.kable.NA = '')
ycols <- c("group", names(m1_emm))

m1_emm <- emmeans(m1, specs=c("hac", "indole"))
m1_pairs_dt <- contrast(m1_emm,
                        method = "revpairwise",
                        adjust = "none",
                        simple = "each",
                        combine = TRUE) %>%
  summary() %>%
  data.table()

hac_0 <- c(m1_emm_dt[hac == "HAC-", emmean],
           m1_pairs_dt[hac == "HAC-", estimate])
hac_1 <- c(m1_emm_dt[hac == "HAC+", emmean],
           m1_pairs_dt[hac == "HAC+", estimate])
hac_diffs <- c(m1_pairs_dt[1:2, estimate], NA)

simple_table <- data.frame("HAC-" = hac_0,
                          "HAC+" = hac_1,
                          "simple" = hac_diffs)
row.names(simple_table) <- c("Indole-", "Indole+","simple")



bgc1 <- pal_okabe_ito[1]
bgc2 <- pal_okabe_ito[2]

knitr::kable(simple_table, 
             digits = 3,
             col.names = c("HAC-", "HAC+", "simple"),
             caption = "Conditonal (\"simple\") effects of HAC (orange background) and Indole (blue background) on ABA concentration. The conditional means of each combination of the factor levels are in the cells with the white background. The simple effect is the difference in the means of the associated row or column.") %>%
  kable_styling(full_width = FALSE) %>%
  # row_spec(row = 0,
  #          bold = F) %>%
  column_spec(column = 1,
              bold = c(T, T, T)) %>%
  column_spec(column = 4,
              background = c(bgc1, bgc1, "white")) %>%
  column_spec(column = 3,
              background = c("white", "white", bgc2)) %>%
  column_spec(column = 2, 
              background = c("white", "white", bgc2))
 


```

One way to understand conditional effects is by using the $m \times p$ table of treatment combination means (Table \@ref(factorial-simple-effects)). The values in the right-side column (orange) are the conditional effects of HAC at each level of Indole. These values are the difference of the means in the associated row. For example, the conditional effect of $hac$ when `indole = Indole-` is 13.297 (first value in orange column). The values in the bottom row (blue) are the conditional effects of Indole at each level of HAC. These values are the difference of the means in the associated column. For example, the conditional effect of $indole$ when `hac = HAC+` is 21.594 (second value in blue row). Note that the first conditional effect for each factor has a corresponding row in the table of coefficients of the fit model.

### Marginal effects

The average of the conditional effects for a factor are the **marginal effects**, or the **main effects** in ANOVA terminology.

```{r factorial-main, echo=FALSE, warning=FALSE, message=FALSE}

m1_hac_dt <- emmeans(m1, specs = "hac") %>%
  contrast(
    method = "revpairwise") %>%
  summary() %>%
  data.table()

m1_indole_dt <- emmeans(m1, specs = "indole") %>%
  contrast(
    method = "revpairwise") %>%
  summary() %>%
  data.table()

marginal_table <- copy(simple_table)

marginal_table <- rbind(marginal_table, marginal =
                        c(NA, NA, m1_hac_dt[1, estimate]))
marginal_table <- cbind(marginal_table,
                        marginal = c(NA, NA, m1_indole_dt[1, estimate], NA))

orange_color <- pal_okabe_ito[1]
blue_color <- pal_okabe_ito[2]
gray_color <- pal_jco(alpha = 0.6)(9)[3]

knitr::kable(marginal_table, 
             digits = 3,
             col.names = c("HAC-", "HAC+", "simple", "marginal"),
             caption = "Marginal effects of HAC (orange) and Indole (blue) on ABA concentration. Simple effects are in grey cells. The conditional means of each combination of the factor levels are in the white cells.") %>%
  kable_styling(full_width = FALSE) %>%
  # row_spec(row = 0,
  #          bold = F) %>%
  column_spec(column = 1,
              bold = c(T, T, T,T)) %>%
  column_spec(column = 2, 
              background = c("white", "white", gray_color, "white")) %>%
  column_spec(column = 3,
              background = c("white", "white", gray_color, "white")) %>%
  column_spec(column = 4,
              background = c(gray_color, gray_color, "white", orange_color)) %>%
  column_spec(column = 5,
              background = c("white", "white", blue_color, "white"))

```

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
Be careful with reporting marginal effects. In this example, the simple effects of HAC are very different for the `Indole-` and `Indole+` treatment levels and the mean of these two effects may not be especially relevant to the question that we are pursuing. If the simple effects of a factor are similar, then it may make more sense to report the marginal effect.
</div> 

### The additive model

Marginal effects can be useful for summarizing a general trend, but, like any average, might not be especially meaningful if there is large heterogeneity of the simple effects, which occurs when the interaction effect is large. The maize ABA response is a good example of marginal effects that would be highly misleading to present without further comment.

If an interaction effect is small, then it can be useful to estimate the effects of the two factors as if the interaction were equal to zero.

\begin{equation}
aba = \beta_0 + \beta_1 hac_{HAC^+} + \beta_2 indole_{Indole^+} + \varepsilon
(\#eq:factorial-reduced)
\end{equation}

This is a **reduced model** because one of the terms has been removed from the model. This particular reduced model is often referred to as the **additive model**, since it excludes the interaction term, which is a *product* of other terms. In R, this model is

```{r}
m2 <- lm(aba ~ hac + indole, data = fig1)
```

The model coefficients of the additive model are

```{r factorial-reduced-model, echo=FALSE}
m2_coef <- cbind(coef(summary(m2)),
        confint(m2))
knitr::kable(m2_coef, 
             digits=c(2,2,1,3),
             caption = "Model coefficients of additive model.") %>%
  kable_styling()

```

The conditional effects of the reduced model are

```{r, echo=FALSE}
m2_emm <- emmeans(m2, specs=c("hac", "indole"))
m2_pairs <- contrast(m2_emm,
                 method = "revpairwise",
                 adjust = "none",
                 simple = "each",
                 combine = TRUE) %>%
  summary(infer = TRUE)


knitr::kable(m2_pairs,
             digits = c(1,1,1,5,2,1,2,2,2,5),
             caption = "Conditional effects of additive model.") %>%
  kable_styling()

```

The table shows that, in an additive model, all conditional effects for one factor are the same for each level of the other factor. This makes sense. If the model fit is additive, the interaction effect is set to zero by the model and there cannot be differences in conditional effects among the contrasts at each of the levels of the other factor (otherwise, there would be an interaction). Also note that the conditional effects xxx. Probably a better way of thinking about this is, it doesn't make sense to compute or discuss conditional effects in an additive model. Instead, an additive model automatically computes marginal effects.

Compare the table of marginal effects of the additive model to the table of marginal effects of the full model. The estimates are the same but the *t*-values and *p*-values differ because of different degrees of freedom (the full model estimates one more parameter, the interaction effect). The estimate is the same only if the design is balanced, which means that each combination of treatment levels has the same sample size *n*.

### Reduce models for the right reason

Unless one factor truly has no effect, there will always be an interaction. As stated above, interactions are ubiquitous. If an interaction is small, it can make sense to drop the interaction term and re-fit an additive model to estimate marginal effects in order to present a simplified picture of what is going on, with the recognition that these estimates are smoothing over the heterogenity in conditional (simple) effects that truly exist.

Aided and abetted by statistics textbooks for biologists, there is a long history of researchers dropping an interaction effect because the interaction $p>0.05$. Don't do this. It doesn't make any sense.

1. The $p$-value is an arbitrary dichotomization of a continuous variable. Would it make sense to behave differently if the interaction were $p=0.051$ vs. $p=0.049$, given that these two p-values are effectively identical?
2. A $p$-value is not evidence that an effect is zero, or "doesn't exist", or even that an effect is "trivially small". This is because $p$-values are a function of measurement error, sampling error, and sample size, in addition to effect size.

## Working in R
### Model formula

A full-factorial model with two factors is specified in the model formula as `y ~ A*B` where $texttt{A}$ is the first factor, and $texttt{B}$ is the second factor. R expands this formula to `y ~ 1 + A + B + A:B` where the colon indicates an interaction (multiplicative) effect.

```{r }
m1 <- lm(aba ~ hac * indole, data = fig1)
m1_coef <- coef(summary(m1))
  
m1_coef %>%
  kable(digits = c(1,2,2,3))
```

The additive model is specified by the formula `y ~ A + B`

```{r urchin-additive}
m2 <- lm(aba ~ hac + indole, data = fig1)
m2_coef <- coef(summary(m2))
  
m2_coef %>%
  kable(digits = c(1,2,2,5))
```


<!--chapter:end:chapters/34-factorial.Rmd-->

# ANOVA Tables

```{r anova-setup, echo=FALSE, warning=FALSE, message=FALSE}
library(here)
library(reshape2)
library(readxl)
library(ggplot2)
library(emmeans) # pooled CIs
library(data.table)
library(doBy)
library(harrellplot)
library(car)
library(broom)
library(cowplot)

data_path <- "data" # bookdown

source_path <- here::here("R", "gg_interaction.R")
source(source_path)

options(knitr.kable.NA = '')

```

Treatment effects are most often analyzed using ANOVA, which is short for "Analysis of Variance". This is somewhat of an odd name for a method to test for treatments effects - what do differences in means have to do with an analyis of variance? The name makes sense in light of the decomposition of the total variance into a model variance and the residual variance (chapter xxx). If there are differences among the means, then the total variance is increased because of variation among groups.

The engine underneath modern ANOVA is a linear model. If the model has a single categorical factor, the ANOVA is **one-way**. If the model has two categorical factors it is a two-way ANOVA. If the model has a single categorical factor and one continuous factor it is an ANCOVA, short for **analysis of covariance** (next chapter). More complex experimental designs classically analyzed with ANOVA are nested, split-plot, latin-square, and many others.

## Summary of usage

If you choose to report an ANOVA, also report the effects and their uncertainty in some way, either the model coefficients or contrasts. 

1. ANOVA generates a table with one row for each term in the linear model. A term is a factor or a covariate or an interaction. For a two-way factorial ANOVA, these terms are the two main effects and the interaction effect.
2. The ANOVA generates an $F$ and $p$-value for the whole model and for each term in the ANOVA table.
3. The $p$-value of an interaction term is often used as a decision rule to interpret the main effects. If $p \le 0.05$ then do not interpret the main effects but instead examine the condition ("simple") effects. If $p > 0.05$, then interpret the main effects. Regardless, this sort of decision rule is itself controversial, and for good reason. 
4. If the main effects are to be interpreted, some statisticians advocate re-fitting the model without the interaction effect, others advocate interpreting the main effects with the interaction term in the model. This only matters if the design is unbalanced (see below). 
5. Regardles of any decision, always plot the data using a Harrell plot or interaction plot to understand and communicate the magnitude and pattern of interaction.
6. For factors with more than two levels, the $p$-value is often used as a decision rule to dissect the factor with post-hoc tests, such as Tukey HSD.
7. A design is balanced if all the cells have the same number of replicates. A design is unbalanced if one or more of the cells has a different number of replicates. Unbalanced designs make it necessary to make decisions, none of which are perfect, and all of which are controversial. Some statisticians have even advocated randomly excluding data until the design is back in balance. Don't do this.
8. There are multiple ways to decompose the sum of squares. I highlight the major three: Type I (sequential), Type II (partial sequential), and Type III. Most statistics software and introductory statistics books default to Type III and, consequently, many researchers are unaware that Types I and II exist. R's default is Type I, and this can make a difference if the design is unbalanced. This is *not* a rare error in publications.
9. Because R defaults to Type I sum of squares, the $p$-value of a factor depends on the order of the factors in the model if the design is unbalanced. This is a feature, not a bug.
10. ANOVA based on type II sum of squares do not depend on factor order if the design is unbalanced, but it does assume that the interaction is zero.
11. ANOVA based on type III sum of squares do not depend on order if the design is unbalanced and does not assume the interaction is zero.
12. If the design is balanced, Type I, II, and III sum of squares generate the same ANOVA table. And the ANOVA table of just the main effects is the same as the ANOVA table that includes the interaction term. None of this is true when the design is unbalanced, However, the decision to use type II or type III is very controversial.

## Example: a one-way ANOVA using the vole data
The vole data has a single factor ("treatment") with three levels ("control", "vitamin_E", "vitamin_C"). In statistics textbooks that emphasize hypothesis testing, the "Which test should I use" flowchart would guide a researcher given this design to a single classification, or one-way ANOVA, since a t-test can only compare two levels but an ANOVA can compare more than two levels. There are better ways to think about what ANOVA is doing, but okay.

Here is an ANOVA table of the vole data:

```{r anova-vole, echo=FALSE}
folder <- "Deleterious consequences of antioxidant supplementation on lifespan in a wild-derived mammal"
filename <- "RSBL-2013-0432 vole data.xlsx"
file_path <- paste(data_path, folder, filename, sep="/")
vole_wide <- data.table(read_excel(file_path, sheet="COLD VOLES LIFESPAN", range="a2:d98"))
setnames(vole_wide, old=colnames(vole_wide), new=c("lifespan", "control", "vitamin_E", "vitamin_C"))
vole <- na.omit(melt(vole_wide, measure.vars=c("control", "vitamin_E", "vitamin_C"), variable.name="treatment"))
```

```{r anova-vole-table, echo=FALSE}
fit <- lm(lifespan ~ treatment, data=vole)
knitr::kable(anova(fit), digits=c(1,0,1,2,3))
```

I'll explain all the parts of the ANOVA table later, but for now, focus on the $p$-value, which is that most researchers want out of the table. What null hypothesis does this $p$-value test? The p-value gives the probability of the observed $F$ or larger $F$, if the null were true. The null hypothesis models the data as if they were sampled from a single, normal distribution and randomly assigned to different groups. Thus the null hypotheis includes the equality of the means among factor levels. In the vole data, the single treatment factor has three levels and a small $p$-value could occur because of a difference in means between the vitamin_E treatment and control, or between the vitamin_C treatment and control, or between the two vitamin treatments. The $p$-value or ANOVA table doesn't indicate what is different, only that the observed $F$ is unexpectedly large if the null were true. As a consequence, researchers typically interpret a low $p$-value in an ANOVA table as evidence of "an effect" of the term but have to use additional tools to dissect this effect. The typical additional tools are either **planned comparisons**, which are contrasts among a subset of a priori identified treatment levels (or groups of levels) or unplanned comparisons ("post-hoc" tests) among all pairs of levels.

The $p$-value in the ANOVA table acts as a decision rule: if $p < 0.05$ then it is okay to further dissect the factor with planned comparisons or post-hoc tests because the significant $p$ "protects" the type I error of further comparisons. I'm not fond of using $p$-values for these sorts of decision rules.

## Example: a two-way ANOVA using the urchin data

```{r anova-urchin, echo=FALSE}
folder <- "Data from Temperature and CO2 additively regulate physiology, morphology and genomic responses of larval sea urchins"
filename <- "urchin.txt"
file_path <- paste(data_path, folder, filename, sep="/")
urchin <- fread(file_path)
urchin[, CO2:=factor(ifelse(CO2==400,"CO2-", "CO2+"))]
urchin[, Temp:=factor(ifelse(Temp==13, "Temp-", "Temp+"))]
urchin <- na.omit(urchin[, .SD, .SDcols=c("Resp", "Temp", "CO2")])
```

Let's use the urchin data from the previous chapter xxx to explore the ANOVA table, which is what is typically reported. The experiment has two factors ($Temp$ and $CO2$), each with two levels. Here is the linear model

\begin{equation}
Resp = \beta_0 + \beta_1 Temp + \beta_2 CO2 + \beta_3 TempCO2 + \varepsilon
\end{equation}

In order to understand factorial ANOVA (or any ANOVA with multiple factors), it is useful to know the difference between **conditional means** and **marginal means**

```{r anova-urchin-means, echo=FALSE, message=FALSE, warning=FALSE}
con3 <- list(Temp=contr.sum, CO2=contr.sum)
fit3 <- lm(Resp ~ Temp*CO2, data=urchin, contrasts=con3)
conditional_means <- summary(emmeans(fit3, specs=c("Temp", "CO2")))
temp.marginal_means <- summary(emmeans(fit3, specs=c("Temp")))
co2.marginal_means <- summary(emmeans(fit3, specs=c("CO2")))
grand_mean <- mean(urchin$Resp)
# put into table
means_table <- matrix(NA, nrow=3, ncol=3)
means_table[1,1] <- conditional_means[1,3] #temp-, CO2-
means_table[2,1] <- conditional_means[2,3]
means_table[1,2] <- conditional_means[3,3]
means_table[2,2] <- conditional_means[4,3]
means_table[1,3] <- temp.marginal_means[1,2]
means_table[2,3] <- temp.marginal_means[2,2]
means_table[3,1] <- co2.marginal_means[1,2]
means_table[3,2] <- co2.marginal_means[2,2]
means_table[3,3] <- grand_mean
row.names(means_table) <- c("Temp-","Temp+","CO2-mm")
colnames(means_table) <- c("CO2-","CO2+","Temp-mm")
#knitr::kable(means_table, digits=c(1,1,1))
round(means_table, 3)
```

In the table above, the upper, left $2 \times 2$ grid of cells are the conditional means, which are the means of each group, where a group is a specific combination of factor levels. The first two values of the third row are the marginal means for CO2. The first (10.488) is the mean of the two means when CO2=CO2-. This can be written as $\mathrm{E}(Resp|CO2-)$. The second (8.829) is the mean of the two means when CO2=CO2+, or $\mathrm{E}(Resp|CO2+)$. The first two elements of the third column are the marginal means for Temp. These are $\mathrm{E}(Resp|Temp-)$ and $\mathrm{E}(Resp|Temp+)$. The bottom right value (9.659) is the grand mean.

A **conditional effect** is a difference between conditional means. For example the conditional effect of $Temp$ *conditional on* CO2=CO2- is $12.743-8.233$. A **marginal effect** is a difference in marginal means within a factor, for example the marginal effect of $Temp$ is $11.243 - 8.075$.

Here is the ANOVA table of the urchin data

```{r anova-table, echo=FALSE}
con3 <- list(Temp=contr.sum, CO2=contr.sum)
fit3 <- lm(Resp ~ Temp*CO2, data=urchin, contrasts=con3)
urchin_anova_table <- anova(fit3)
knitr::kable(urchin_anova_table[1:4,], digits=c(1,1,1,1,4))
```

This ANOVA table uses what are called Type 3 sum of squares, which is *NOT* the default in R but is the default in many other statistics software and is, therefore, the *only* type of ANOVA that many researchers know (and, many researchers are unaware that there are multiple types of ANOVA table). Understanding these differences is important, at least if one is reporting ANOVA tables. I'll return to the importance of this later.

### How to read an ANOVA table

An ANOVA table has a row for each term in the underlying linear model -- each of these adds a component of variance to the total, and a row for the residual variance (this residual variance row is frequently excluded from the published table). The urchin model has three terms (one level of $Temp$, one level of $CO2$, and one interaction). The statistics for each term are

1. **Degrees of Freedom** (df) -- If the term is a factor, the df will equal the number of levels ($k$) for the factor minus 1. Think of it this way: the contribution of the variance due to a factor is a function of the variability of the $k$ level means around the grand mean. How many degrees of independent variation do these level means have, given that we know the grand mean? The answer is $k-1$ -- once the values of $k-1$ level means are written down, the $k$th level mean has no freedom to vary; its value has to be $k\bar{\bar{Y}} - \sum_i^{k-1}{Y_i}$. For an interaction term, the df is the product of the df of each of the factors in the interaction.
2. **Sum of Squares** -- the sum of squared differences between the modeled value and the grand mean. In addition to a sum of squares for each term, a **residual mean square** is computed as the sum of squared differences between the measured and modeled values.
3. **Mean Square** -- The sum of squares divided by the df (this is a "mean" with df acting as the number of things that were summed).
4. **F-ratio** -- the Mean Square of the term dived by the residual mean square.
5. **p-value** -- the p-value for the F-ratio. F is compared to an F-distribution, which is a distribution of F-ratios under the null.

#### Each row in the table tests a null hypothesis

The row for each term in an ANOVA table tests a null hypothesis. In order to understand the null hypotheses, I need to define a few more terms

For the ANOVA table above, which uses Type 3 sum of squares, the probabilities are

1. Temp -- $p = \mathrm{prob}(F \ge F_o|CO2, Temp:CO2)$. The null is no difference in means conditional on the level of CO2 and Temp:CO2. This is equivalent to no difference between the grand mean and the marginal mean of Temp+, or

\begin{equation}
b_1 = \overline{\overline{Resp}} - \mathrm{E}(Resp|Temp^+)
\end{equation}

2. CO2-- $p = \mathrm{prob}(F \ge F_o|Temp, Temp:CO2)$. The null is no difference in means conditional on the level of Temp and Temp:CO2. This is equivalent to no difference between the grand mean and the marginal mean of CO2+, or

\begin{equation}
b_2 = \overline{\overline{Resp}} - \mathrm{E}(Resp|CO2^+)
\end{equation}

3. Temp:CO2 -- $p = \mathrm{prob}(F \ge F_o|Temp, CO2)$. The null is no difference in means conditional on the level of Temp and CO2. This is equivalent to the difference between the conditional mean of Temp+/CO2+ and the expected conditional mean of Temp+/CO2+ if there were no interaction.

\begin{equation}
b_3 = \mathrm{E}(Resp|Temp^+, CO2^+) - (\overline{\overline{Resp}} - b_1 - b_2)
\end{equation}

As noted in the equations, these three differences are the coefficients of the linear model behind the ANOVA. Here is the coefficient table

```{r anova-coef-effect, echo=FALSE}
knitr::kable(coef(summary(fit3)), digits=c(2,2,1,5))
```

In ANOVA with type 3 sum of squares, the dummy variables are coded using effect coding, which differs from the dummy coding introduced in chapter xxx. The consequence is that the **grand mean** (the mean of $Resp$ across all values) is now the "reference" value. The intercept in this table, then, is the grand mean. The coefficients are *differences from the grand mean*, as described above.

Use the table of conditional and marginal effects above to check that the coefficients equal the differences in the equations above. Also not that the $p$-values for the effects in the coefficient table equals the $p$-values in the ANOVA table.

It is important to note that this table differs from the coefficient table with dummy coding because that reference is the mean of Temp-/CO2- and not the grand mean.

```{r anova-coef-dummy, echo=FALSE}
fit1 <- lm(Resp ~ Temp*CO2, data=urchin)
knitr::kable(coef(summary(fit1)), digits=c(2,2,1,5))
```

Importantly, note that $p$-values for $b_1$ (the Temp effect) and $b_2$ differ between the two tables. This is because the $t$-value tests different hypotheses! In the coefficient table with effect coding (that behind the ANOVA with type 3 sums of squares), the $p$-value tests marginal effects and so is a function of both marginal means within a factor. By contrast, in the coefficient table with dummy coding, the $p$-value tests conditional effects, and so is only a function of the conditional means when the other factor is at its reference level (right? The coefficient $b_1$ in the dummy coded coefficient table is the effect of only increasing $Temp$ -- $CO2$ is left at its reference level). For the interaction effect, the coefficient differs between the effects coded model and the dummy coded model (because different reference means) but the $p$-value ultimately tests the same hypothesis (non-additive effects of the factors) and so the $t$ and $p$ values are the same.

#### What to do after ANOVA?

Researchers frequently report ANOVA statistics ($F$ and $p$ values) for factorial models in a way that suggests that they misunderstand the hypotheses tested. It probably doesn't help that there is a long-standing debate among statisticians about the most sensible strategy for interpreting factorial ANOVA results. And it doesn't help that the default ANOVA table in R can suggest a very different interpretation than the default ANOVA table in some other software packages.

Here are three strategies for interpreting a factorial ANOVA table that uses Type III sum of squares. All strategies use $p$-values to make a series of decision rules. In the first strategy, which is a type of model simplification or model selection, a researcher starts with the interactions at the bottom of the ANOVA table and works up, eliminating terms with $p > 0.05$ and re-fitting the reduced model before interpreting main effects. In the second strategy, the researcher uses the original ANOVA table that includes all terms to interpret main effects.

**Strategy 1**

1. is interaction *p* < 0.05?
    a. if yes, then do NOT test main effects. Show a graph to show pattern of conditional effects. Test conditional effects if this is of interest.
    b. if no, then refit model without the interaction and test main effects -- This now is equivalent to ANOVA using Type II sum of squares.
        2. is main effect *p* < 0.05$?
            a. if yes, then keep in model
            b. if no, then refit model without that main effect

**Strategy 2**

2. is interaction *p* < 0.05?
    a. if yes, then do NOT test main effects. Show a graph to show pattern of conditional effects. Test conditional effects if this is of interest.
    b. if no, then use the same table as the test of the main effects. This is interpreting the main effects with the interaction term in the model. This is the logic of ANOVA using type III sum of squares.
    
**Strategy 3**

3. is interaction *p* < 0.05?
    a. if yes, then look at interaction plot to determine if it makes sense test main effects. For example, if CO2+ had obviously lower $Resp$ at both levels of $Temp$, even if one was much lower (ie. interactaction), then some people would say that the test of the main effect is meaningful. Test conditional effects if this is of interest.
    b. if no, then use the same table as the test of the main effects
 
In general, statisticians advise against strategy 3 (interpreting main effects in the presence of interaction) -- its not wrong, its just that a main effect has an awkward interpretation if there is an interaction. Of course this is true if there is *any* interaction term in the model, not just a statistically significant term. The controversy is more, if the interaction $p$ is not significant, then do we implement stategy 1 (refit model excluding interaction to test main effects) or strategy 2 (use full factorial anova table to test main effects).

```{r urchin-fake-data, echo=FALSE}
n <- 6
temp_levels <- rep(c("Temp-", "Temp+"), each=n)
co2_levels <- c("CO2-", "CO2+")
x <- expand.grid(Temp=temp_levels, CO2=co2_levels)

fit <- lm(Resp ~ Temp*CO2, data=urchin)
sigma <- summary(fit)$sigma # sqrt(sd(fit$residuals)^2*23/20)
X <- model.matrix(fit)
b <- coef(fit)
fd <- copy(na.omit(urchin[, .SD, .SDcols=c("Resp", "Temp", "CO2")]))
done <- FALSE
i <- 0
while(done==FALSE){
  i <- i+1
  set.seed(i)
  fd[, Resp:=X%*%b + rnorm(n*4, mean=0, sd=sigma)]
  b.fake <- coefficients(summary(lm(Resp ~ Temp*CO2, data=fd)))
  if(b.fake["TempTemp+:CO2CO2+", "Pr(>|t|)"] < 0.05){
    done <- TRUE
  }
}
fit3fd <- lm(Resp ~ Temp*CO2, data=fd, contrasts=con3)
#emm.fit3 <- emmeans(fit3, specs=c("Temp", "CO2"))
#contrast(emm.fit3, method="revpairwise", adjust="none")[c(1,2,5,6),]
knitr::kable(anova(fit3fd)[1:3,], digits=c(1,1,1,1,4))
```

then one shouldn't report the ANOVA results using something like "Temperature had a significant effect on metabolism ($F_{1,20} = 14.5$, $p=0.001$). There was no effect of CO2 on metabolism ($F_{1,20} = 4.1$, $p=0.26$)". There was a significant interaction effect between Temperature and CO2 on metabolism ($F_{1,20} = 14.8$, $p=0.04$)". If one accepts that the small interaction $p$-value is evidence of an interaction effect then this interpretation of the main factors makes no sense, as the first two results imply that the interaction effect is zero (or, that there is a constant effect of $Temp$ or $CO2$ across both levels of the other factor), which is then contradicted by the third result.

More specifically, if one is using a $p$-value to guide decision making, then a significant interaction $p$ indicates that there is no single "main" effect of a factor. Instead, the effect of $Temp$ is conditional on the level of $CO2$, and the effect of $CO2$ is conditional on the level of $Temp$. This is easily seen in the interaction plot, where the effect of $Temp$ is large when $CO2$ is high but much smaller when $CO2$ is low. Indeed, the effect of $Temp$ at the low CO2 is 0.16.

Instead of interpreting the factors as constant effects, A better strategy is to compare the **conditional effects**, that is, the effects of $Temp$ within each level of $CO2$ and the effects of $CO2$ within each level of $Temp$ (conditional effects are sometimes called the "simple effects").

The controversy arises in what to do after an ANOVA if the interaction effect has a non-significant $p$-value. At this point, I am punting instead of explaining the basis for the controversy, because ultimately I think the major problem with both strategies is the use of null hypothesis significance testing to make analysis decisions.

In fact, the entire reason that I use the urchin data as the example for factorial ANOVA is because it beautifully illustrates the absurdity of the interaction $p$-value decision rule. Why should we interpret the results of the ANOVA when the interaction $p$ is 0.079 differently than when the interaction $p$ is 0.04? Remember, the $p$-value is a "sample" statistic (in the sense that it is entirely a function of the sampled data) and in conditions of low power (which is likely, but not necessarily, true for the urchin data given n=6), a $p$-value is highly variable.

There are several problems with this approach. 1) a $p$-value is not evidence of "no effect", 2) the power to test interaction effects is small relative to that for the main effects (this is a general rule, not something specific to these data), 3) the interaction SS accounts for about 7.2$\%$ of the total SS, which doesn't seem inconsequential, and 4) the interaction $p$-value is small enough to raise a red flag, and, most importantly, 5) the confidence interval of the interaction effect indicates that the large, negative values of the interaction are *as consistent with the data* as trivially small values (or a value of zero). But the CI is not in an ANOVA table and many researchers fail to report it. These five points suggest that this experiment be replicated, with a larger sample size, to get a better estimate of the interaction effect. The problem of course is that experiments are rarely replicated, except in biomedical research.

The absurdity of the $p$-value decision rule strategy for interpretation of effects after an ANOVA is highlighted by comparing the forest plot of model coefficients of the real and fake urchin data. It would be absurd to use an ANOVA table to interpret these patterns as radically different (one without an interaction and constant main effects, the other with an interactioni and conditional effects).

```{r urchin-anova-ggplot, echo=FALSE, fig.cap="Forest plots (the upper part of a Harrell plot) of the actual and fake urchin data. A) Real urchin data. The interaction effect is not significant ($p=0.079$). B) Fake urchin data. The interaction effect is significant ($p=0.04$)."}
# gg <- gg_interaction_plot(x=c("Temp", "CO2"), y="Resp", data=urchin)
# gg.fd <- gg_interaction_plot(x=c("Temp", "CO2"), y="Resp", data=fd)
urchin1 <- copy(na.omit(urchin[, .SD, .SDcols=c("Resp", "Temp", "CO2")]))
gg <- harrellplot(x="Temp", y="Resp", g="CO2", data=urchin1, contrasts.method="coefficients")$gg_contrasts
gg.fd <- harrellplot(x="Temp", y="Resp", g="CO2", data=fd, contrasts.method="coefficients")$gg_contrasts
plot_grid(gg, gg.fd, nrow=2, labels="AUTO")
```

### How to read ANOVA results reported in the text

ANOVA results are often reported in the text of a results section, using something like "Temperature had a significant effect on metabolism ($F_{1,20} = 14.5$, $p=0.001$). There was no effect of CO2 on metabolism ($F_{1,20} = 4.1$, $p=0.26$)". The subscripts of the $F$ statistic are the numerator and denominator degrees of freedom (df) of the $F$-value (These df are a column in the ANOVA table. The denomintor df may not appear in the table if it is the residual df and the row for the residual term was not reported). Sometimes I find the reported df are not consistent with the description of the design and analysis, which means the data were not analyzed as stated.

### Better practice -- estimates and their uncertainty

As emphasized in the previous chapter, the decision to include or exclude an interaction effect in the model should not be based on a $p$-value but on the goals of the model.

1. If the goal is the interaction (because a scientific model predicts one, or because this is biology and everything is conditional), then estimate the interaction effect (as a coefficient of the model!) and its uncertainty, including a CI and $p$-value. There is no controversy on how to estimate this effect and its uncertainty. The coefficient will be different between dummy and effect coded models but this is okay because they have different specific interpretations but the same general interpretation. Use a Harrel plot with the coefficients (including the interaction coefficient) to show this estimate and uncertainty.

2. If the goal is to estimate constant main effects, then exclude the interaction effect from the model and report the main effects (again, as coefficients from the model or contrasts if other pairwise effects are desired) with their uncertainty. Use an interaction plot (or bottom part of the harrell plot) to justify forcing the interaction to zero (for example the interaction effect adds little to the total sum of squares or the interpretation of a single main effect or two (or more) conditional effects would be the same. Use a Harrel plot that excludes the interaction term to show these main effects and uncertainty.

3. And if a researcher is interested in the effects of the factors but there is strong evidence for a non-trivial interaction, then report the conditional effects (as contrasts) with their uncertainty. Use a Harrel plot that includes the interaction term to show these conditional effects and uncertainty. If there is an obvious interaction, it probably doesn't make sense to interpret the main effects, contrary to what some people argue. If there is a positive effect of factor A across all levels of factor B, we don't really need a $p$-value to test that the average of these positive effects is significant. This doesn't add value to the plot and any conditional effects that are reported.

Notice that an ANOVA table has no role in this recommendation.

## Unbalanced designs

My recommendation above is to not bother with ANOVA, but to simply compute the contrasts of interest using the linear model. But if you really want to use ANOVA, you should be aware that **if the design is unbalanced, factor order matters in the default R anova function** and that I routinely find published ANOVA tables that report statistics ($F$ and $p$ values) that are not what the authors think they are.

An **unbalanced** design is one in which the number of replicates differs among the cell. The urchin data is balanced because there are six replicates in each cell. If the respirometer broke before taking the respiratory measures of the final tank, the design would be unbalanced, one of the cells would have only five replicates.

Let's look at the effect of row order on the statistics of the urchin data using R's default anova function.

```{r anova-order-balanced, echo=FALSE}
fit1a <- lm(Resp ~ Temp*CO2, data=urchin)
fit1b <- lm(Resp ~ CO2*Temp, data=urchin)
knitr::kable(anova(fit1a)[1:3,], digits=c(1,2,2,2,5))
knitr::kable(anova(fit1b)[1:3,], digits=c(1,2,2,2,5))
```
Now let's unbalance the data, by removing three random replicates (these may be both in one cell or spread across cells. First, here is the number of replicates in each cell:

```{r anova-unbalanced, echo=FALSE}
set.seed(2)
urchin.missing <- copy(na.omit(urchin[, .SD, .SDcols=c("Resp", "Temp", "CO2")]))
N <- nrow(urchin.missing)
inc <- sample(1:N, (N-3))
urchin.missing <- urchin.missing[inc, ]
table(urchin.missing$Temp, urchin.missing$CO2)
fit1a <- lm(Resp ~ Temp*CO2, data=urchin.missing)
fit1b <- lm(Resp ~ CO2*Temp, data=urchin.missing)
#anova(fit1a)
#anova(fit1b)

```

And here are the two tables with the order of Temp and CO2 reversed in the model

```{r anova-unbalanced-tables, echo=FALSE}
knitr::kable(anova(fit1a)[1:3,], digits=c(1,2,2,2,5))
knitr::kable(anova(fit1b)[1:3,], digits=c(1,2,2,2,5))
```

Several observations are important.

1. the statistics for the last row, which is the interaction, does not change.
2. if these data were analyzed in the software package JMP, or SAS, or SSPS then **order wouldn't matter**. Here is what the tables would look like

```{r anova-unbalanced-type3, echo=FALSE}
fit1a <- lm(Resp ~ Temp*CO2, data=urchin.missing, contrasts=con3)
fit1b <- lm(Resp ~ CO2*Temp, data=urchin.missing, contrasts=con3)
knitr::kable(Anova(fit1a, type="3")[2:4,], digits=c(2,1,2,5))
knitr::kable(Anova(fit1b, type="3")[2:4,], digits=c(2,1,2,5))
```

3. Order does not change the statistics in the coefficient table, even for unbalanced data:

```{r anova-unbalanced-coeffs, echo=FALSE}
knitr::kable(coef(summary(fit1a)), digits=c(2,3,3,4))
knitr::kable(coef(summary(fit1b)), digits=c(2,3,3,4))
```

### What is going on in unbalanced ANOVA? -- Type I, II, III sum of squares

**Type I sum of squares**.  Here is the (default) ANOVA table using Type I sum of squares for the urchin data with the three missing rows. 

```{r anova-urchin-typeI, echo=FALSE}
fit1 <- lm(Resp ~ Temp*CO2, data=urchin.missing)
knitr::kable(anova(fit1)[1:4,], digits=c(0,3,3,1,4))
```

The default coding of dummy variables in R's `lm` function is dummy coding, which is the coding used for Type I or **Sequential Sum of Squares**. The hypothesis tested by each row in the ANOVA table using Type I sum of squares is the effect of that row's term conditional on all terms before it in the model (or above it in the table) and ignoring all terms after it in the model (or below it in the table).

1. The hypothesis tested by the $p$-value for $Temp$ is the same as if $Temp$ were the only term in the model (other than the intercept). That is, the means are estimated for each level of $Temp$ ignoring the fact that half the replicates within each level of $Temp$ experienced low $CO2$ and half experienced high $CO2$

2. The hypothesis tested by the $p$-value for $CO2$ is conditional on $Temp$. That is, the difference in metabolism between $CO2+$ and $CO2-$ when $Temp$ is "held constant" (or for all cases where $Temp$ takes the same value). This is equivalent to the hypothesis that the difference in the marginal means of CO2 is zero.

3. The hypothesis tested by the $p$-value for the interaction is conditional on all other terms and nothing is ignored.

```{r anova-urchin-typeI-part2, echo=FALSE, eval=FALSE}
fit1 <- lm(Resp ~ Temp+CO2, data=urchin)
anova(fit1)
coef(summary(fit1))
b2 <- 11.2425-8.07

```

**Type II sum of squares**.  Here is the ANOVA table using Type II sum of squares for the urchin data with missing values. The interaction term is excluded from the linear model, because type II sum of squares are used to estimate main effects ignoring the interaction (so this would make sense only if a plot of the effects suggested a small interaction relative to the main effects). The sum of squares for the main effects would be the same if the interaction were included but the residual df, and thus the F and P-values would differ.

```{r anova-urchin-typeII, echo=FALSE}
fit1b <- lm(Resp ~ CO2+Temp, data=urchin.missing)
part1 <- anova(fit1b)[2,]
fit1a <- lm(Resp ~ Temp+CO2, data=urchin.missing)
part2 <- anova(fit1a)[2:3,]
type2 <- rbind(part1, part2)
knitr::kable(type2, digits=c(0,3,3,1,4))
```

The hypothesis tested by each row in the ANOVA table using Type II sum of squares is the effect of that row's term conditional on all terms *at the same level or below* but ignoring all terms at a higher level in the model (or below it in the table). For example, the hypothesis test for a factor is conditioned on other factors but ignores interaction terms among the factors. Consequently, these hypotheses tested are


1. The hypothesis tested by the $p$-value for $Temp$ is conditional on $CO2$. This is the same hypothesis that would occur using Type I sum of squares but placing $Temp$ second in the model, after $CO2$ (and it is in fact how I computed it for the table).

2. The hypothesis tested by the $p$-value for $CO2$ is conditional on $Temp$. This is exactly the hypothesis for $CO2$ using the Type I sum of squares above.

**Type III sum of squares**.  Here is the ANOVA table using Type III sum of squares for the urchin data for missing data. The interaction term is excluded from the linear model, and advocates of using Type III sum of squares explicitly want this in the model.

```{r anova-urchin-typeIII, echo=FALSE}
fit3 <- lm(Resp ~ Temp*CO2, data=urchin.missing, contrasts=con3)
knitr::kable(Anova(fit3, type="3")[2:5,], digits=c(3,1,3,4))
```

The hypothesis tested by each row in the ANOVA table using Type III sum of squares is the effect of that row's term conditional on all terms in the model.

1. The hypothesis tested by the $p$-value for $Temp$ is conditional on $CO2$ and $Temp:CO2$.

2. The hypothesis tested by the $p$-value for $CO2$ is conditional on $Temp$ and $Temp:CO2$.

3. The hypothesis tested by the $p$-value for $Temp:CO2$ is conditional on $Temp$ and $CO2$. This is the same for Type I sum of squares (and Type II, if the interaction term were included)

### Back to interpretation of main effects

### The anova tables for Type I, II, and III sum of squares are the same if the design is balanced.

## Working in R

### Type I sum of squares in R
The base R function `anova()` computes the ANOVA table using Type I sum of squares for any fit model object, such as that returned by `lm`. Here is a script for the urchin data. I first create unbalanced data by deleting the first row that is the control row.

```{r anova-urchin-typeI-script}
cn_rows <- which(urchin[, Temp]=="Temp-" & urchin[, CO2]=="CO2-") # gives the rows of the controls
urchin_unbalanced <- urchin[-cn_rows[1],] # deletes the row that is in first element of cn_rows
urchin.t1 <- lm(Resp ~ Temp*CO2, data=urchin_unbalanced)
anova(urchin.t1)
```

### Type II and III Sum of Squares

Type II sum of squares can be computed manually simply by fitting the model twice, once with the factors ordered one way and then with the factors ordered the opposite way. The car package has the function `Anova` that specifically outputs Type II and Type III ANOVA tables.

Type II sum of squares can be fit with the interaction in the model, and this generates the Type II sum of squares for the main terms but the residual is wrong for the $F$-ratio because it is the residual from the full model and Type II assumes the interaction effect is zero. So, if one wants an ANOVA table with a $F$ and $p$ that reflect this, then the interaction should be dropped from the model.

```{r anova-urchin-typeII-script}
urchin.t2 <- lm(Resp ~ Temp*CO2, data=urchin_unbalanced)
Anova(urchin.t2, type="2")

urchin.t2 <- lm(Resp ~ Temp + CO2, data=urchin_unbalanced)
Anova(urchin.t2, type="2")
```

To get type III sum of squares, we need to specify effects coding for the model matrix. The safest way to do this is something like this

```{r anova-urchin-typeIII-script}
con3 <- list(Temp=contr.sum, CO2=contr.sum) # change the contrasts coding for the model matrix
urchin.t3 <- lm(Resp ~ Temp*CO2, data=urchin_unbalanced, contrasts=con3)
Anova(urchin.t3, type="3")
```


<!--chapter:end:chapters/42-anova.Rmd-->

# Predictive Models

```{r sm-setup, echo=FALSE, message=FALSE}
library(ggplot2)
library(ggpubr)
library(cowplot)
library(readxl)
library(emmeans)
library(data.table)
library(lmerTest)
library(harrellplot)
library(nlme)

bookdown_compile <- TRUE
if(bookdown_compile==TRUE){
  data_path <- "data" # bookdown
  source("R/clean_label.R") # bookdown
}else{
  data_path <- "../data" # notebook
  source("../R/clean_label.R")
}

```

This chapter focusses on modeling **observational data** with multiple $X$ variables, both continous and categorical. The classical analysis of multiple $X$ variables is **multiple regression**, sometimes called **multivariable regression** and occassionally, but incorrectly, called **multivariate regression** -- "multivariate" refers to multiple $Y$ variables.

The models in this chapter have the structure

\begin{equation}
Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + \beta_3 X_3 + ... \beta_p X_p + \varepsilon 
\end{equation}
%
where $p$ is the number of $X$ variables or **predictors** in the model. This equation is easily generalized to both generalized linear models, linear mixed models, and generalized linear mixed models.

## Overfitting

When a model is fit to data, the model coefficients are estimates of the parameters that "generated the data". The value of an estimate is partly a function of the signal (the parameter) and partly a function of the noise, which is unique to the sample. At a low signal to noise ratio a model is mostly fitting the noise. A measure of how well the model "fits" the data is $R^2$, which is

\begin{equation}
R^2 <- 1 - \frac{SS_{residual}}{SS_{total}}
\end{equation}

As $X$ variables are added to a model, the $R^2$ necessarily increases. Part of this increase is due to added signal, but part is due to added noise. If the added noise is more than the added signal, then the model fit -- that is the parameter estimates -- increasingly reflects the noise unique to the sample rather the signal common to every sample. This is the basis of **overfitting**.

To demonstrate overfitting, I fit completely random $X$ variables to the lifespans for the control voles. 

```{r sm-overfit, echo=FALSE}
folder <- "Deleterious consequences of antioxidant supplementation on lifespan in a wild-derived mammal"
filename <- "RSBL-2013-0432 vole data.xlsx"
file_path <- paste(data_path, folder, filename, sep="/")
vole <- data.table(read_excel(file_path, sheet="COLD VOLES LIFESPAN", range="a2:d98"))
setnames(vole, old=colnames(vole), new=c("lifespan", "control", "vitamin_E", "vitamin_C"))
vole <- vole[control=="1", lifespan]
n <- length(vole)
p <- 50
set.seed(1)
X <- matrix(rnorm(n*p), nrow=n)
colnames(X) <- paste("X", 1:p, sep="_")
fit1 <- lm(vole~X[,1])
fit10 <- lm(vole~X[,1:10])
fit50 <- lm(vole~X)
```


Think about it this way: if I create fake data in there are ten $X$ variables that are correlewhich $Y$ is a simple column of random, normal variables that are not a function of 


## Model building vs. Variable selection vs. Model selection

### Stepwise regression

### Cross-validation

### Penalization

#### AIC
#### LASSO

## Shrinkage



<!--chapter:end:chapters/46-predictive-models.Rmd-->

# Part VII -- Expanding the Linear Model {-}

<!--chapter:end:chapters/50-part-v-expanding-linear-model.Rmd-->

# Models with random factors -- Blocking and pseudoreplication {#lmm}

## Random effects

```{r lmm-setup, echo=FALSE, message=FALSE, warning=FALSE}
library(here)
library(readxl)
library(janitor)
library(forcats)
library(broom)
library(broom.mixed)
library(ggplot2)
library(ggpubr)
library(ggsci)
library(cowplot)
library(data.table)
library(emmeans)
library(lmerTest)
library(nlme)
library(doBy)

here <- here::here
data_path <- "data"
```

```{r fake-data, echo=FALSE}

all_treatments_per_batch <- function(sigma=1,
                                     sigma_b0=1,
                                     sigma_b1=1,
                                     beta_0=10,
                                     beta_1=1,
                                     n_block=6,
                                     n_subsamp=10, Y_label="Y", trt_levels=c("Cn","Tr"), block_label="block", confound=FALSE){
  # sigma - conditional error sd
  # sigma_b0 - sd of random intercepts
  # sigma_b1 - sd of random slope
  if(sigma_b0==0){sigma_b0 <- 1e-10}
  if(sigma_b1==0){sigma_b1 <- 1e-10}
  fake_data <- data.table(NULL)
  # this is a slow, inefficient way to build the data but should be easy to follow
  for(j in 1:n_block){
    #  random intercept for each plot
    beta_0_j <- rnorm(1, mean=0, sd=sigma_b0)
    beta_1_j <- rnorm(1, mean=0, sd=sigma_b1)
    # the response in the control and treated groups
    y1 <- beta_0 + beta_0_j + (beta_1 + beta_1_j)*0 + rnorm(n_subsamp, mean=0, sd=sigma)
    y2 <- beta_0 + beta_0_j + (beta_1 + beta_1_j)*1 + rnorm(n_subsamp, mean=0, sd=sigma)
    # combine into a data.table and rbind to end of fake_data
    fake_data <- rbind(fake_data,
                       data.table(Treatment=rep(trt_levels, each=n_subsamp),
                                  subsample=rep(1:n_subsamp, 2),
                                  block=letters[j],
                                  Y=c(y1, y2),
                                  beta_0j=beta_0_j+beta_0,
                                  beta_1j=beta_1_j+beta_1)
    )
  }
  if(confound==TRUE){
    # this creates data with 1 treatment level per block so block and treatment are confounding
    ex <- c(which(fake_data[,Treatment]==trt_levels[2] & 
                   fake_data[,block] <= letters[floor(n_block/2)]),
            which(fake_data[,Treatment]==trt_levels[1] & 
                   fake_data[,block] > letters[floor(n_block/2)]))
    fake_data <- fake_data[ex]
  }
  fake_data[, block:=factor(block)]
  fake_data[, Treatment:=factor(Treatment, trt_levels)]
  setnames(fake_data, old=c("Y", "block"), new=c(Y_label, block_label))
  return(fake_data)
}
```

Researchers often collect data in batches, for example

1. An ecologist interested in the effects of insectivorous birds on tree seedling performance in a forest stake out ten 1 m$^2$ plots and use a wire-mesh cage to cover half of each plot ^[Giffard, B., Corcket, E., Barbaro, L., & Jactel, H. (2012). Bird predation enhances tree seedling resistance to insect herbivores in contrasting forest habitats. Oecologia, 168(2), 415-424]. The cage allows insect herbivores into the seedlings inside but excludes insectivorous birds that eat the insects from the seedlings. In every plot, five seedlings are planted within the exclosure and five outside of the exclosure. At the end of the experiment, the total leaf mass is measured on each seedling. Small, uncontrolled, environmental factors (including soil factors and density of insectivorous birds) will differ between plots but will be common to all seedlings within a plot and we would expect a common response to this uncontrolled variation on top of the differential response to each treatment. As a consequence, the ten measures of leaf mass within a plot are not independent.
2. A nutrition researcher wants to compare the effect of glucose vs. fructose on glucose metabolism in humans. Ten individuals are recruited. Each individual has blood insulin measured 60 minutes after a noon meal over six successive days. The meal alternates between high glucose and high fructose on each day. Each individual has three measures under high glucose treatment and three measures under high fructose treatment. Small, uncontrolled, environmental factors (including metabolic variation, other meals, activity levels) will differ between the individuals but be common within an individual and we would expect a common response to this uncontrolled variation on top of the differential response to each treatment. As a consequence, the six measures of insulin within an individual are not independent.
3. An ecologist wants to measure the effect of an invasive plant on the reproduction of a native plant. They stake-out ten, 2 m$^2$ plots in a forest and divide each plot into four quadrants, with each quadrant assigned a different treatment: control, activated carbon (a procedural control), extract from the invasive plant's leaves, and both activated carbon and extract from the invasive plant's leaves. The response is seedling count. Small, uncontrolled, environmental factors (including soil, drainage, and light) will differ between plots but will be common to all four quadrants within a plot and we would expect a common response to this uncontrolled variation on top of the differential response to each treatment. As a consequence, the four sets of counts within a plot are not independent.
4. A physiologist has skeletal muscle cells growing in 5 control cultures, and 5 treated cultures. The $Y$ variable is cell diameter, which is measured in 10 cells per culture. Small, uncontrolled, environmental factors (including chemical) will differ between cultures but will be common to all cells within a culture and we would expect a common response to this uncontrolled variation on top of the differential response to each treatment. As a consequence, the ten measures of diameter within a culture are not independent.
5. A behavioral biologist wants to measure the effect of a predator fish on the preferred feeding location (open water or vegetation) of a prey fish. Ten tanks are set up with equal amounts of vegetated and unvegetated area. One-third of each tank is screened off to house a predator fish, which are added to five of the tanks. Ten prey fish are added to each tank. The response is minutes spent foraging in the open water as a fraction of total time foraging, which is measured in each fish in each tank. Small, uncontrolled, environmental factors (including temperature, water chemistry, light, and fish behavior) will differ between the tanks but be common within tanks and we would expect a common response to this uncontrolled variation on top of the differential response to each treatment. As a consequence, the ten measures of foraging of each fish within a tank are not independent.
6. A microbiologist wants to measure the effect of the microbiome on autism-spectrum-disorder(ASD)-like behavior in mice^[Sharon, G., Cruz, N.J., Kang, D.W., Gandal, M.J., Wang, B., Kim, Y.M., Zink, E.M., Casey, C.P., Taylor, B.C., Lane, C.J. and Bramer, L.M., 2019. Human Gut Microbiota from Autism Spectrum Disorder Promote Behavioral Symptoms in Mice. Cell, 177(6), pp.1600-1618.]. Intestinal microbobial communities from five humans with ASD and five humans without ASD are transferred to germ-free mice via fecal transplant. Each human donor is used to colonize three mice. The response is time socializing in a direct social interaction. Uncontrolled features of each microbial community (species composition and proportions) will differ among human donors but will be the same within human donors and we would expect a common response to this uncontrolled variation in addition to any differential response to ASD-associated microbiota. As a consequence, the measures of behavior in the three mice within a donor group are not independent.

The batches -- plots in experiment 1, individuals in experiment 2, plots in experiment 3, cultures in experiment 4, tanks in experiment 5, and mice in experiment 6 -- are the experimental units, meaning that it is at this level that the experimenter is controlling the conditions. In each of these studies, there is systematic variation at two levels: among treatments due to treatment effects and among batches due to **batch effects**. This among-batch variation is the **random effect**. An assumption of modeling random effects is that the batches (plots/individuals/cultures/tanks/donor) are a random sample of the batches that could have been sampled. This is often not strictly true as batches are often **convenience samples**.

The multiple measures within a batch are **subsamples** but are often called **repeated measures** if the batch is an individual (experiment 2 is an example). If multiple measures within a treatment level within a batch (that is, within a $batch \times treatment$ combination) are taken over time, the data are **longitudinal**. Not infrequently, subsamples within a treatment within a batch are called "replicates", but this can be confusing because the treatments are replicated at the level of the batch and not at the level of the subsamples within a treatment by batch combination. The batches are the independent experimental units. The subsamples within a batch are not replicates.

The variation among batches/lack of independence within batches has different consequences on the uncertainty of the estimate of a treatment effect. The batches in experiments 1-3 are similar in that each contains all treatment levels. In these, the researcher is interested in the treatment effect but not the variation due to differences among the batches. The batches are nuissance factors that add additional variance to the response, with the consequence that estimates of treatment effects are less precise, unless the variance due to the batches is explicitly modeled. In experiments like 1-3, the batches are known as **blocks**. Including block structure in the design is known as **blocking**. Adding a blocking factor to a statistical mode is used to increase the precision of an estimated treatment effect.

Experiments 1 and 2 are examples of a **randomized complete block with subsampling** design. "Complete" means that each block has all treatment levels or combinations of levels if there is more than one factor. *The subsampling is not replication*. The replicates are the blocks, because *it was at this level that treatment assignment was randomized*. Experiment 3 is an example of a **randomized complete block** design. The blocks are complete but there is only one measure of the response per treatment.

The batches in experiments 4-6 are similar in that treatment is randomized *to* batch, so each batch contains only a single treatment level. In these **segregated** experimental designs, the variation among batches that arises from non-treatment related differences among batches **confounds** the variation among batches due to a true treatment effect. An extreme example of this would be experiment 4 (muscle cell cultures) with only a single culture with control conditions and a single culture with treatment conditions. Imagine 1) the true effect of the treatment is zero and 2) an unmeasured growth factor that happens to be more concentrated in the treatment culture at the beginning of the experiment. At the end of the experiment the cells in the treatment culture have an average diameter twice that of that in the control culture. The researcher is fooled into thinking that the treatment caused the increased growth. Again, the replicates are at the level of the cultures, because it was at this level that treatment assignment was randomized. This means the researcher has a single replicate (or, $n=1$) in each treatment level, regardless of the number of cells that are measured within each culture. A statistical analysis that uses the subsampling within a replicate as the sample size is an example of **pseudoreplication** (Hurlbert 1984 xxx).

## Random effects in statistical models

In all of the above examples, the researcher is interested in the treatment effect but not the variation due to differences among the blocks. The blocks are nuissance factors that add additional variance to the response, with the consequence that estimates of treatment effects are less precise, unless the variance due to the blocks is explicitly modeled. Including block structure in the design and in the statistical model is known as **blocking**. A natural way to think about the block factor is as a **random effect**, meaning that plots in experiment 1 or the mice in experiment 3 are simply random samples from a population of plots or mice. Modeling this using the residual-error specification looks like 

\begin{equation}
y_{ij} = (\beta_{0} + \beta_{0j}) + (\beta_{1} + \beta_{1j}) x_i + \varepsilon_i 
(\#eq:lmm-spec1)
\end{equation}

where $i$ indexes the observation and $j$ indexes the block (culture, plot, mouse, etc). The intercept parameter $\beta_{0j}$ is a **random intercept** and the slope parameter $\beta_{1j}$ is a **random slope**. The intercept for observation *i* (that is, its expected value when $X=0$) has a **fixed** component ($\beta_0$) that is common to all observations and a random component ($\beta_{0j}$) that is common within a block but differs among blocks (see table below). In the above equation, I've used parentheses to show how these combine into the random intercept that is unique for each block. Similarly, the random slope (treatment effect) has a fixed part ($\beta_1$) that is common to all observations and a random component ($\beta_{1j}$) that is common within a block but differs among blocks (see table below). Again, these are collected within a pair of parentheses in the equation above.

```{r lmm-table, echo=FALSE, message=FALSE, warning=FALSE}
n <- 5 # number of measures per treatment within each block
k <- 6 # number of blocks
set.seed(3)
fd <- all_treatments_per_batch(sigma=1, sigma_b0=1, sigma_b1=0.1, beta_0=10, beta_1=1, n_block=k, n_subsamp=n)
fit <- lmer(Y~Treatment + (Treatment|block), data=fd)
coef_table <- coef(fit)$block
b <- apply(coef_table,2, mean)
fixed <- matrix(rep(b,each=k), nrow=k, ncol=2)
random <- coef_table-fixed
b_table <- data.table(block=unique(fd$block),
                      "$b_0$"=fixed[,1],
                      "$b_{0j}$"=random[,1],
                      "$b_1$"=fixed[,2],
                      "$b_{1j}$"=random[,2]
                      )

b_table <- data.table(block=1:k, 
                         "$b_0$"="$b_0$", 
                         "$b_{0j}$"=paste("$b_{0,j=",1:6,"}$", sep=""),
                         "$b_1$"="$b_1$", 
                         "$b_{1j}$"=paste("$b_{1,j=",1:6,"}$", sep="")
                        )
knitr::kable(b_table, caption = "The linear mixed model specified above estimates a fixed intercept and fixed slope (treatment effect) that are common to all observations and a random intercept and random slope for each block, each of which is common among observations within a block but differ among observations in different blocks.")
```

Linear mixed models are called "mixed models" because they are a mix of fixed and random factors. Another useful way to specify this model is to think about it hierarchically, using

\begin{align}
y_{ij} &= \beta_{0j} + \beta_{1j}x_i + \varepsilon_i \\
\varepsilon_i &\sim N(0, \sigma) \\
\beta_{0j} &= \beta_{0} + N(0, \sigma_{0}) \\
\beta_{1j} &= \beta_{1} + N(0, \sigma_{1})
(\#eq:lmm-spec2)
\end{align}

The first line states that the response is a function of a block-specific intercept and a block specific slope plus some error that is unique to each observation. The third and fourth lines state that these block-specific effects are themselves a function of a common effect and a random term that is unique to each block. That is, we have a hierarchical or multi-level structure to the model. Line 1 is the top level and the effects that are specified in line 1 are a function of effects at a second, lower level, which are specified in lines 3 and 4. Because of this structure, linear mixed models are sometimes called hierarchical or multi-level models.

Finally, it's useful to think how to specify a linear mixed model using the random-draw specification, as this leads naturally to generalized linear mixed models, or GLMMs.

\begin{align}
y_{ij} &\sim N(\mu_{ij}, \sigma) \\
\mu_{ij} &=\beta_{0j} + \beta_{1j}x_i \\
\beta_{0j} &\sim N(\beta_0, \sigma_0) \\
\beta_{1j} &\sim N(\beta_1, \sigma_1)
(\#eq:lmm-spec3)
\end{align}


## Linear mixed models are flexible

The linear mixed model in Equation \@ref(eq:lmm-spec1) specifies both a random intercept and a random slope but a researcher might limit the random effect to the intercept only, or less commonly, the slope only. Excluding the random slope from Equation \@ref(eq:lmm-spec1) results in the model

\begin{equation}
y_{ij} = (\beta_{0} + \beta_{0j}) + \beta_{1}x_i + \varepsilon_i 
(\#eq:lmm-spec1b)
\end{equation}

We might use a random-intercept-only model if we think that features of the block would effect the mean response among blocks but not effect the difference in treatment level (or treatment effect) among blocks. For example, differences in the immune systems among the individual mice in experiment 3 might effect growth in both the wild-type and engineered strains of staph but won't effect the difference in growth between wild-type and engineered strains from one mouse to another.

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**Not more than you should know** -- For more complex mixed models, matrix algebra makes the specification of the model much more manageable than the scalar algebra in \@ref(lmm-spec1b). 

\begin{equation}
\mathbf{y} = \mathbf{X}\boldsymbol{\beta} + \mathbf{Zu} + \boldsymbol{\varepsilon}
\end{equation}

where $\mathbf{y}$ is the vector of the response, $\mathbf{X}\boldsymbol{\beta}$ is the linear predictor of fixed effects and $\mathbf{Zu}$ is the linear predictor of random effects. $\mathbf{X}$ is the model matrix for the fixed effects and $\boldsymbol{\beta}$ is the vector of fixed-effect terms (the fixed part of the intercept ($\beta_0$), including the fixed-effect coefficients for each of the 
</div> 

## Blocking
### Visualing variation due to blocks

To visualize random effects due to block, Let's create fake data that look something like experiment 1, with a single factor with two treatment levels, $k=10$ blocks, and $n=3$ measures for each treatment level within each block. This is a randomized complete block design with subsampling and has a total of $N=2 \times k \times n$ measures of $Y$ (and rows of the data.table).

```{r lmm1-blocking-fig, warning=FALSE, message=FALSE, echo=FALSE, fig.cap="Visualizing random effects. A) The response in the two treatment levels. B) The same data but separated by block. The blue line is at the control mean and the yellow line is at the treated mean. The black dots are the mean response within a block."}
n <- 3 # number of measures per treatment within each block
k <- 10 # number of blocks
set.seed(5)
fd <- all_treatments_per_batch(sigma=1, sigma_b0=1, sigma_b1=0.1, beta_0=10, beta_1=1, n_block=k, n_subsamp=n)
m1 <- lm(Y~Treatment, data=fd)
m2 <- lm(Y~Treatment, data=fd[, .(Y=mean(Y)), by=.(block, Treatment)])
m3 <- lmer(Y~Treatment + (Treatment|block), data=fd)
m4 <- lmer(Y~Treatment + (1|block), data=fd)

jco <- pal_jco()(2)
gg1 <- ggerrorplot(x="Treatment", 
            y="Y", 
            color="Treatment",
            data=fd, 
            desc_stat = "mean_sd", 
            palette = "jco",
            position = position_dodge(0.3), 
            add=c("jitter")
             ) +
  geom_hline(yintercept = mean(fd[Treatment=="Cn", Y]), color=jco[1]) +
  geom_hline(yintercept = mean(fd[Treatment=="Tr", Y]), color=jco[2])

gg2 <- ggerrorplot(x="block", 
            y="Y", 
            color="Treatment",
            data=fd, 
            desc_stat = "mean_sd", 
            palette = "jco",
            position = position_dodge(0.3), 
            add=c("jitter")
             ) +
  geom_hline(yintercept = mean(fd[Treatment=="Cn", Y]), color=jco[1]) +
  geom_hline(yintercept = mean(fd[Treatment=="Tr", Y]), color=jco[2])

plot_grid(gg1, gg2, rel_widths=c(1,1.5), labels="AUTO")

fd[, resid.lm:=residuals(m1)]
fd[, resid.lmm:=residuals(m3)]

gg3 <- ggerrorplot(x="Treatment", 
            y="resid.lm", 
            color="Treatment",
            data=fd, 
            desc_stat = "mean_sd", 
            palette = "jco",
            position = position_dodge(0.3), 
            add=c("jitter")
             ) +
  scale_y_continuous(limits=c(-2.5, 2.5))

gg4 <- ggerrorplot(x="Treatment", 
            y="resid.lmm", 
            color="Treatment",
            data=fd, 
            desc_stat = "mean_sd", 
            palette = "jco",
            position = position_dodge(0.3), 
            add=c("jitter")
             ) +
  scale_y_continuous(limits=c(-2.5, 2.5))

#plot_grid(gg3, gg4, labels="AUTO")


```

Figure \@ref(fig:lmm1-blocking-fig)A shows the response as a function of treatment. The responses are nicely symmetric around the treatment means (the blue and yellow lines). A linear model (and generalized linear models, more generally) assumes that a response, conditional on the $X$, are independent. Figure \@ref(fig:lmm1-blocking-fig)B shows how this assumption is violated for the simulated data. That pattern of residuals within a block around the treatment means does not look at all random. Instead, there is a distinct pattern within a block for the points to cluster either below the treatment means or above the treatment means. In blocks a, b, g, and i, all or most of the responses are below their treatment mean (for example in block b, all the yellow points are below the yellow line and two of three blue points are below the blue line). In blocks d, e, f, and j, all or most of the responses are above their treatment mean (for example, in block e, all three yellow points are above the yellow line and all three blue points are above the blue line). In other words, the responses within a block covary together. For a linear model, this is known as **correlated error**.

### Blocking increases precision of point estimates

Block effects are differences in expected mean response among blocks due to unmeasured factors that are shared within blocks but not among blocks. A classical linear model fails to model this component of the total variance in the response, and as a consequence, this block-specific variance is part of the error variance. One way to think about this is by moving the random intercept and random slope components of equation \@ref(eq:lmm-spec1) to the right and combining it with the observation-specific (or "conditional") error ($\varepsilon_i$)

\begin{equation}
y_{ij} = \beta_{0} + \beta_{1} x_i + (\beta_{0j} + \beta_{1j} + \varepsilon_i)
(\#eq:lmm-spec1b)
\end{equation}

A linear mixed model estimates the random effects parameters, so the residual from observation $i$ is $\varepsilon_i$. A linear model does not estimate the random effects parameters, so the residual of observation $i$ from a linear model is $\beta_{0j} + \beta_{1j} + \varepsilon_i$. Consequently, the error variance from a linear model is larger than the error variance from a linear mixed model fit to the same data. The consequence of this on inference depends on the variance of the random effects relative to the variance of the observation-specific error and on the subsample size. If the variance due to random effects is relatively big and subsample size is relatively small, then a linear mixed model estimates treatment effects with much more precision (and *p*-values will be smaller).

This increased precision is seen in the coefficient table of three models fit to the fake data in Figure \@ref(fig:lmm1-blocking-fig).

A linear model fit to all data
```{r, echo=FALSE}
knitr::kable(tidy(m1, conf.int=TRUE)[2,], 
             digits = c(0, 5, 5, 2, 3, 3, 3))
```

A linear model fit to the means of each treatment level with each block
```{r, echo=FALSE}
knitr::kable(tidy(m2, conf.int=TRUE)[2,], 
             digits = c(0, 5, 5, 2, 3, 3, 3))
```

A linear mixed model with both a random intercept and random slope fit to all data
```{r, echo=FALSE}
knitr::kable(tidy(m3, conf.int=TRUE)[2, 3:10], 
             digits = c(0, 5, 5, 2, 2, 3, 3, 3))
```

```{r, echo=FALSE}
perc.m1 <- 1 - coef(summary(m3))["TreatmentTr","Std. Error"]/coef(summary(m1))["TreatmentTr","Std. Error"]
perc.m2 <- 1 - coef(summary(m3))["TreatmentTr","Std. Error"]/coef(summary(m2))["TreatmentTr","Std. Error"]
```

Note that the estimates of treatment effect do not differ among models. What does differ is the estimated standard error of the treatment effect, which is `r round(perc.m1*100, 0)`% smaller in the linear mixed model relative to that in the linear model, and `r round(perc.m2*100, 0)`% smaller in the linear mixed model relative to that in the linear model fit to the block means. This difference in SE propogates to the confidence intervals and *p*-values.

Let's explore this a wee bit more systematically using a simple, Monte Carlo simulation experiment. 5000 fake data sets were generated. Each data set simulated an experiment with a single treatment factor with two levels ("control" and "treatment"). The design is a randomized complete block with subsampling. There are 8 blocks and 3 subsamples in each treatment level per block. The treatment effect ($\beta_1$) is 1. The observation-specific (or "within-block") variance ($\sigma^2_{varepsilon}$) is 1. The ("among-block") variance of the random intercept ($\sigma^2_{\beta_0j}$) is 1 -- the same as the the within-block variance. The variance variance of the random slope ($\sigma^2_{\beta_1j}$), which is due to a variable response of the treatment among blocks, is $0.1^2$).

The following three models were fit to all 5000 data sets 

1. a linear model fit to all data, ignoring blocks (violating the independence assumption)
2. a linear model fit to the treatment means of the blocks (valid, but throwing out data)
3. a linear mixed model that models the random intercept^[a random slope is not modeled because many of the models specifying a random slope fail to converge, which is expected given the relatively small variance of the random slope].

From each fit, I saved the 95% confidence interval of the treatment effect and the *p*-value. The median width of the confidence interval and the power, which is the relative frequency of *p*-values less than 0.05. The simulation was re-run using the same parameters except setting the treatment effect ($\beta_1$) to 0. In this new simulation, the relative frequency of *p*-values less than 0.05 is the Type I error.

```{r lmm-power, echo=FALSE}
fn <- "lmm_ci-QcSv.txt"
file_path <- here("output", "chapter_lmm", fn)
ci=fread(file_path)

fn <- "lmm_prob-QcSv.txt"
file_path <- here("output", "chapter_lmm", fn)
prob1=fread(file_path)

fn <- "lmm_prob-PCna.txt"
file_path <- here("output", "chapter_lmm", fn)
prob0=fread(file_path)

knitr::kable(data.table(
  Model = c("1. lm", "2. lm of means", "3. lmm b0j"),
  "CI width" = apply(ci, 2, median),
  Power = apply(prob1, 2, function(x) sum(x<0.05)/nrow(prob1)),
  "Type I" = apply(prob0, 2, function(x) sum(x<0.05)/nrow(prob0))),
  digits=c(0,2,2,3))

```

For data similar to that simulated, the linear mixed model using a blocking factor has much more power than the linear model ignoring block (and ignoring the lack of independence) or the linear model comparing the treatment level means of the blocks. This lost power in the two linear models is due to the conservative Type I error rate (extreme in the case of the linear model of the group means). One take-home lesson here is, don't throw away perfectly good data if the design of the experiment includes a blocking factor!

## Pseudoreplication
### Visualizing pseduoreplication
Subsamples from batches are not replicates. Inference from a model fit to subsampled observations without modeling the batches is called pseudoreplication, a term famously coined by Hurlbert (1984). For data from a randomized complete block design, ignoring the batches in the model will typically result in larger standard errors, wider confidence intervals, and too conservative *p*-values. In a segregated experiment, with only a single treatment level per batch (like that in experiments 4-6 above), ignoring the lack of independence in the model has the opposite effect. Standard errors are too small. Confidence intervals are too narrow. *P*-values are too liberal. Type I error is inflated. Let's visualize a pseudoreplicated data set like this.

```{r lmm1-pseudoreplication-fig, echo=FALSE, fig.cap="A data set in which treatment and batch are confounded because there is only one treatment level per batch."}
n <- 10 # number of measures per treatment within each block
k <- 6 # number of blocks
set.seed(19)
fd <- all_treatments_per_batch(sigma=1, sigma_b0=1, sigma_b1=0.1, beta_0=10, beta_1=1, n_block=k, n_subsamp=n, confound=TRUE)
m3 <- lmer(Y~Treatment + (Treatment|block), data=fd)

jco <- pal_jco()(2)
gg1 <- ggerrorplot(x="Treatment", 
            y="Y", 
            color="Treatment",
            data=fd, 
            desc_stat = "mean_sd", 
            palette = "jco",
            position = position_dodge(0.3), 
            add=c("jitter")
             ) +
  geom_hline(yintercept = mean(fd[Treatment=="Cn", Y]), color=jco[1]) +
  geom_hline(yintercept = mean(fd[Treatment=="Tr", Y]), color=jco[2])

gg2 <- ggerrorplot(x="block", 
            y="Y", 
            color="Treatment",
            data=fd, 
            desc_stat = "mean_sd", 
            palette = "jco",
            position = position_dodge(0.3), 
            add=c("jitter")
             ) +
  geom_hline(yintercept = mean(fd[Treatment=="Cn", Y]), color=jco[1]) +
  geom_hline(yintercept = mean(fd[Treatment=="Tr", Y]), color=jco[2]) +
  xlab("batch") +
  NULL

plot_grid(gg1, gg2, rel_widths=c(1,1.5), labels="AUTO")

```

This consequence of pseudoreplication when batch and treatment are confounded is seen in the coefficient table of four models fit to the fake data in Figure \@ref(fig:lmm1-pseudoreplication-fig).

**A linear model fit to all data (pseudoreplication)**

```{r, echo=FALSE}
m1 <- lm(Y~Treatment, data=fd)
knitr::kable(tidy(m1, conf.int=TRUE)[2,], 
             digits = c(0, 5, 5, 2, 3, 3, 3))
```

**A linear model fit the batch means**

```{r, echo=FALSE}
m2 <- lm(Y~Treatment, data=fd[, .(Y=mean(Y)), by=.(block, Treatment)])
knitr::kable(tidy(m2, conf.int=TRUE)[2,], 
             digits = c(0, 5, 5, 2, 3, 3, 3))

```

**A linear mixed model with both random intercept and random slope.**
```{r, echo=FALSE}
m3 <- lmer(Y~Treatment + (Treatment|block), data=fd)
knitr::kable(tidy(m3, conf.int=TRUE)[2, 3:10], 
             digits = c(0, 5, 5, 2, 2, 3, 3, 3))

```

**A linear mixed model with random intercept**

```{r, echo=FALSE}
m4 <- lmer(Y~Treatment + (1|block), data=fd)
knitr::kable(tidy(m4, conf.int=TRUE)[2, 3:10], 
             digits = c(0, 5, 5, 2, 2, 3, 3, 3))
# emmeans(m4, specs = "Treatment",
#            lmer.df = "kenward-roger") %>%
#   contrast(method = "revpairwise") %>%
#   summary(infer = TRUE)
# 
# emmeans(m4, specs = "Treatment",
#            lmer.df = "satterthwaite") %>%
#   contrast(method = "revpairwise") %>%
#   summary(infer = TRUE)
# 
# emmeans(m4, specs = "Treatment",
#            lmer.df = "asymptotic") %>%
#   contrast(method = "revpairwise") %>%
#   summary(infer = TRUE)

```


As with the blocked design, all models compute the same estimate of the treatment effect. But in contrast to the blocked design, in this confounded design, the standard error of the treatment effect is smaller in the linear model of all data than that of the linear mixed models. This increased precision is an illusion because the model fails to account for the lack of independence within the batches.

One interesting result to note is the equivalence of the standard error, test statistic, *p*-value, and confidence intervals of the linear model on the batch means and the linear mixed model with a random intercept (but no random slope) only. The two are equivalent. Murtaugh (xxx) has argued that with this kind of design, it is much simpler to the analyst, and to your audience, to simply use the linear model on the batch means. This raises the question, why bother with subsampling within batch? One answer is, subsampling increases the precision of the batch mean, and therefore, the precision of the treatment effect. That said, *the precision of a treatment effect is increased more efficiently by adding more replicates (more batches), not more subsamples*.

Let's explore the consequence of pseudoreplication with a confounded with a Monte Carlo simulation experiment. 5000 fake data sets were generated. Each data set simulated an experiment with a single treatment factor with two levels ("control" and "treatment"). The design is treatment level randomized to batch (only one treatment level per batch). There are 8 batches and 3 subsamples in each batch. The treatment effect ($\beta_1$) is zero. The observation-specific (or "within-block") variance ($\sigma^2_{varepsilon}$) is 1. The ("among-block") variance of the random intercept ($\sigma^2_{\beta_0j}$) is 1 -- the same as the the within-block variance. The variance variance of the random slope ($\sigma^2_{\beta_1j}$), which is due to a variable response of the treatment among blocks, is $0.1^2$).

Because the effect is zero, the frequency of *p*-values less than 0.05 is the type I error. The same three models fit to the simulated blocked data are fit to these data. I don't simulate an effect in order to compute power (at that effect size) because the increased power in the linear model of all data is, again, an illusion. It only comes at the cost of strongly elevated Type I error.

```{r lmm-pseudoreplication-sim-table, echo=FALSE}
fn <- "lmm_ci-TzSa.txt"
file_path <- here("output", "chapter_lmm", fn)
ci=fread(file_path)

fn <- "lmm_prob-TzSa.txt"
file_path <- here("output", "chapter_lmm", fn)
prob0=fread(file_path)

knitr::kable(data.table(
  Model = c("1. lm", "2. lm of means", "3. lmm b0j"),
  "CI width" = apply(ci, 2, median),
  "Type I" = apply(prob0, 2, function(x) sum(x<0.05)/nrow(prob0))),
  digits=c(0,2,2,3))

```

When treatment level is randomized *to* batch, the type I error rate of a linear model fit to all data (and ignoring the lack of independence) is highly inflated. Don't do this.

## Mapping NHST to estimation: A paired t-test is a special case of a linear mixed model

Specifically, a **paired t-test** is equivalent to a linear mixed model with a single factor with two treatment levels, $k$ blocks, and a single measure of each treatment level within each block. A good example is the wild type vs. engineered staph count in mice in experiment 3 above. A linear mixed model is much more flexible than a paired *t*-test because it allows a researcher to add treatment levels, additional factors, and covariates to the model. In addition, a linear mixed model can handle missing data.

Here is fake data similar in design to experiment 3 with a single factor with two treatment levels and both levels applied to the same experimental unit.

```{r lmm-paired-t}
set.seed(2)
n <- 10 # number of mice (blocks)
x <- rep(c("WT","Tr"), each=n) # treatments
id <- rep(letters[1:n], 2) # block id
y <- c(rnorm(n, mean=10), rnorm(n, mean=11))
fake_data <- data.table(Y=y, X=x, ID=id)
```

The *t*-test *p*-value is

```{r}
t.test(Y~X, data=fake_data, paired=TRUE)$p.value
```

and the coefficient table of the fixed effect in the linear mixed model is 

```{r}
coef(summary(lme(Y~X, random = ~1|ID, correlation=corCompSymm(form=~1|ID), data=fake_data)))
```

```{r lmer t-test, echo=FALSE, eval=FALSE}
# lmer only equivalent to paired t-test if it doesn't collapse to fixed model
# set.seed(1) - different from paired t
# set.seed(2) - equal paired t

fit <- lmer(Y~X + (1|ID), data=fake_data)
coef(summary(fit))
coef(fit)
```

```{r lmm-ttest-missing, echo=FALSE, eval=FALSE}
# now delete two random observations from the treated level
NA_rows <- sample(1:n, size=2)
fake_data[, Y2:=Y] # copy Y column
fake_data[(NA_rows+10), Y2:=NA] # replace the two obs in Y2 with NA
coef(summary(lme(Y2~X, random = ~1|ID, correlation=corCompSymm(form=~1|ID), 
                 data=fake_data,
                 na.action=na.exclude)))
# cannot run t-test with NA so re-run by deleting matching ID
fake_data[, Y3:=Y2] # copy Y2 column
fake_data[NA_rows, Y3:=NA] # replace the two obs in Y3 with NA
t.test(Y3~X, data=fake_data, paired=TRUE, na.action=na.omit)$p.value
```

## Advanced topic -- Linear mixed models shrink coefficients by partial pooling

In experiment 1 above, there are 10 sites (maybe different woodlots). In each plot, five seedlings are planted inside a cage and five outside the cage. The cage excludes insectivorous birds but not herbivorous insects. The researchers are investigating how birds affect plant growth indirectly -- by eating insects that feed on the plants. The response is total leaf area in each seedling.

```{r lmm-make-partial-pooling-data, echo=FALSE}
fake_data <- all_treatments_per_batch(
  beta_0=10, # mean leaf area inside cage
  beta_1=1, # effect of birds, i.e. increases leaf area 
  sigma=1, # Cohen's d is 1 -- a large effect
  sigma_b0=1, # variation in mean among blocks
  sigma_b1=0.1, # variation in treatment effect among blocks
  n_block=10, # number of plots
  n_subsamp=5, # number of seedlings ber block x treatment level
  Y_label = "leaf_area",
  trt_levels = c("no bird","bird"),
  block_label = "site"
  )
```

Let's say we want to know the treatment effect in each of these sites. There are several ways of estimating this.

1. Fit $k$ separate models, one for each site. The intercept (control mean) and slope (treatment effect) parameters for each site are estimated independently from all other sites. Consequently, the model parameters are computed using **no pooling**. For the estimation of the $\beta$ terms, this is equivalent to a single, factorial linear model with $Site$ modeled as a **fixed effect** (this is not true for the estimate of the standard errors of these terms since these are computed from the residual sum of squares of the model. For balanced data, all of the "intercept" or "slope" terms will have the same SE in the factorial analysis but differ among the $k$ independent analyses).

2. Fit a linear model to all the data combined as if these were from a single site, and assign the intercept and treatment effect paramters to all sites. The model parameters are computed using **complete pooling**.

3. Fit a linear mixed model to all the data, using site as a random factor to estimate both random intercepts and slopes. Similar to the no-pooling analysis, there will be different intercept (control mean) and slope (treatment effect) estimates for each site, but unlike the no-pooling analysis, these estimates are computed by combining information from the other sites. The information used to estimate parameters in a linear mixed model is somewhere in between no pooling and complete pooling and is sometimes called **partial pooling**.

The consequence of partial pooling in a linear mixed model is that site intercepts (control means) are pulled toward the single intercept in the complete-pooling analysis and the site slopes (treatment effects) are pulled toward the single slope in the complete-pooling analysis. This has the consequence that the **differences** in parameter estimates among sites are shrunk toward zero. A consequence of this shrinkage is that the variance of the intercept estimates or of the slope estimates is smaller than that in the no-pooling analysis. Figure \@ref(fig:lmm-partialpooling) shows this shrinkage effect using fake data simulating the seedling experiment.

```{r lmm-partialpooling, echo=FALSE, fig.cap="Shrinkage estimates of the treatment effect in a linear mixed model. The grey line is the estimate using complete pooling (so there is only one estimate which is assigned to each site). In general, the partial-pooling (linear mixed model) estimates (yellow) are generally closer to the complete pooling estimate than the no-pooling (separate linear models) estimates (blue). More specifically, if the no-pooling estimate is far from the complete pooling estimate, the partial pooling estimate is *much* closer to the complete pooling estimate. The consequence of partial pooling is that the differences among the estimates are shrunk toward zero."}
no_pooling <- lmList(leaf_area ~ Treatment | site, data=fake_data)
complete_pooling <- lm(leaf_area ~ Treatment, data=fake_data)
partial_pooling <- lmer(leaf_area ~ Treatment + (Treatment|site), data=fake_data)

intercept<- data.table(
  site=unique(fake_data$site),
  no=coef(no_pooling)[,"(Intercept)"],
  complete=coef(complete_pooling)["(Intercept)"],
  partial=coef(partial_pooling)$site[,"(Intercept)"]
)

slope<- data.table(
  site=unique(fake_data$site),
  beta_1=unique(fake_data$beta_1),
  no=coef(no_pooling)[,"Treatmentbird"],
  complete=coef(complete_pooling)["Treatmentbird"],
  partial=coef(partial_pooling)$site[,"Treatmentbird"]
)

gg1 <- ggscatter(x="site", 
            y=c("no", "partial"), 
            merge=TRUE,
            data=intercept, 
            palette = "jco",
            position = position_dodge(0.3)) +
  geom_hline(yintercept = intercept[1, complete], color="gray")
gg2 <- ggscatter(x="site", 
            y=c("no", "partial"), 
            merge=TRUE,
            data=slope, 
            palette = "jco",
            position = position_dodge(0.1)) +
  geom_hline(yintercept = slope[1, complete], color="gray")

gg2

mse_no <- mean((slope$no-slope$beta_1)^2)
mse_partial <- mean((slope$partial-slope$beta_1)^2)

```

The linear mixed model estimates of the treatment effects for each site are a type of **shrinkage estimate** and a linear mixed model is one kind of **shrinkage estimator**. Shrinkage estimates have fascinating properties:

1. the variance of shrinkage estimates is less than that of ordinary least squares estimates (no-pooling, or using the block as a fixed factor)
2. shrinkage estimates are **biased** but the OLS estimates are not. This means that the expected value of a coefficient from the linear mixed model *does not equal* the true (parameter) value! Or, more formally, $\mathrm{E}(b_j) \ne \beta_j$.
3. the **mean square error** of shrinkage estimates will be smaller than that for OLS estimates.

The first property was discussed above and shown in Figure \@ref(fig:lmm-partialpooling). The second property raises the question, if we want to estimate the treatment effects within each site, why would we ever want to use $Site$ as a random instead of fixed effect? The answer is the third property, which can be summarized as, "if we were to replicate the experiment many times, the shrinkage estimates will be, on average, less wrong (or closer to the true value) than the OLS estimates, where "wrong" is the absolute deviation from the true value."

When shrinkage estimators were first discovered, the third property surprised stasticians. The third property has profound consequences. Consider a scenario where researchers want to compare the performance of a new expression vector to that of an existing expression vector on protein production using *E. coli*. The researchers have ten different *E. coli* strains and are interested in strain-specific effects because they will choose the three strains with the largest effects for further testing. The researchers measure the response of each strain five times.

```{r lmm-fish-passage, echo=FALSE}
set.seed(10)
fake_data <- all_treatments_per_batch(
  beta_0=10^6, # mean nest count 
  beta_1=10^5, # effect of culvert
  sigma=10^5, # Cohen's d is 1 -- a large effect
  sigma_b0=10^5, # variation in mean among blocks
  sigma_b1=5*10^4, # variation in treatment effect among blocks
  n_block=10, # number of streams
  n_subsamp=3, # number of days sampled
  Y_label = "Antibody",
  trt_levels = c("CN","S+"),
  block_label = "Strain"
  )

no_pooling <- lmList(Antibody ~ Treatment | Strain, data=fake_data)
partial_pooling <- lmer(Antibody ~ Treatment + (Treatment|Strain), data=fake_data)
antibody_table <- data.table(
  Strain=unique(fake_data[, Strain]),
  beta_1j=unique(fake_data[, beta_1j])/10^5,
  b_1j_fixed=coef(no_pooling)[,"TreatmentS+"]/10^5,
  b_1j_random=coef(partial_pooling)$Strain[,"TreatmentS+"]/10^5
)
# antibody_table <- orderBy(~ -b_1j_fixed, data=antibody_table)

setnames(antibody_table, old=colnames(antibody_table), 
         new=c("Strain", "$\\beta_{1j}$", "fixed $b_{1j}$", "random $b_{1j}$"
))
knitr::kable(antibody_table, digits=c(NA, 2, 2, 2, 2, 2), caption="Effect of new expression vector on protein production in ten strains of *E. coli* using a fixed effect factorial model and linear mixed model.")


```

The table above shows the true strain-specific effect and both the fixed (OLS) and random (LMM) effect estimates. The largest OLS estimate is 70% larger than the true effect and the strain with the largest true effect is not among the top three biggest OLS estimates (its ranking is 9/10). By contrast, the LMM estimates are closer to the true effects and the top strain is among the three largest LMM estimates.

These results are specific to these fake data but more generally,
1) the largest OLS estimates are inflated (larger error from the true effect), relative to the largest LMM estimates
2) overall, the LMM estimates will be closer than the OLS estimates to the true effects

To understand this, rank order the treatment effects for each strain. An individual strain's position in this rank is the sum of the true effect for that strain and some random error. Because OLS, relative to shrinkage estimates, have greater variance in the estimate (that is, the random error component is bigger), the biggest effects estimated by OLS are more likely to be big because of the error component, compared to shrinkage estimates.

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**Not more than you want to know** -- Shrinkage estimators are not only useful when we are interested in block-specific effects but are also useful for estimating effects when there are **multiple responses**. For example, consider a researcher interested in measuring the effects of some exercise treatment on gene expression in adipose cells. The researcher measures expression levels in 10,000 genes. Given the typical content in undergraduate biostatics courses, a researcher would probably model these responses using 10,000 *t*-tests, or equivalently, 10,000 separate linear models. If the tests were ranked by $p$-value or absolute effect size, many of the genes with largest absolute effect would be there because of a large error component and many of the largest effects would be massively overinflated. Re-imagining the design as a single, linear mixed model with each gene modeled as a block would lead to a rank order in which the biggest measured effects more closely approximate the true effects.
</div> 

## Working in R

The major function for working with linear mixed models is `lmer()` from the lme4 package. An older, and still sometimes used and useful function is `lme()` from the nlme package. The authors of the lme4 package argue that the df in a linear mixed model are too approximate for a useful $p$-value and, consequently, the `lme` function does not return a $p$-value. Many biological researchers want a $p$-value and typically use the `lmerTest` package to get this.

**Specifying a linear mixed model using `lme`**. The random factor is in the column "block". To conform to some packages that use lme4 objects, any variable used to model a random effect should be converted to type `factor`.

1. add a random intercept using `y ~ treatment + (1|block)`. This adds a random intercept for each level of treatment.
2. add a random slope and intercept using  `y ~ treatment + (treatment|block)`. This adds a random intercept for each level of treatment and a random slope for each level of treatment.

A message that might appear is "boundary (singular) fit: see ?isSingular". This does not mean there is a problem with the fit model.

A warning that "Model failed to converge with 1 negative eigenvalue" does mean there is a problem. A solution is to simplify the model by, for example, removing a random slope.

### coral data

**Source** Zill, J. A., Gil, M. A., & Osenberg, C. W. (2017). When environmental factors become stressors: interactive effects of vermetid gastropods and sedimentation on corals. Biology letters, 13(3), 20160957.

**Dryad source** https://datadryad.org/resource/doi:10.5061/dryad.p59n8

**file name** "VermetidSedimentData_ZillGilOsenberg_DRYAD.xlsx"

```{r coral-import}
folder <- "Data from When environmental factors become stressors- interactive effects of vermetid gastropods and sedimentation on corals"
fn <- "VermetidSedimentData_ZillGilOsenberg_DRYAD.xlsx"
sheet_i <- "Coral Growth Rate Data"
file_path <- here(data_path, folder, fn)
coral <- read_excel(file_path, sheet=sheet_i) %>%
  clean_names() %>%
  data.table()
coral[, vermetids:=factor(vermetids)]
coral[, sediment:=factor(sediment)]

# recode levels of factors since 0 and 1
coral[, vermetids := fct_recode(vermetids,
                                absent = "0",
                                present = "1")]
coral[, sediment := fct_recode(sediment,
                                control = "0",
                                addition = "1")]

```
#### Fitting models

```{r coral-fit, warning=FALSE, message=FALSE}
# to reproduce the results
# observation 2 should be excluded from the analysis
inc <- c(1, 3:nrow(coral))

# random intercept only
m1 <- lmer(growth_rate ~ vermetids*sediment + (1|block), data=coral[inc])
# random intercept and slope
m2 <- lmer(growth_rate ~ vermetids*sediment + (vermetids|block) + (sediment|block), data=coral[inc])
# to include the interaction as a random effect we'd need subsampling within each factorial treatment combination
```

The conditional effects of m1 are

```{r}
# results using lmer fit
fit.emm <- emmeans(m1, specs=c("vermetids", "sediment"))
summary(contrast(fit.emm, 
                 method = "revpairwise", 
                 simple = "each",
                 combine = TRUE,
                 adjust="none"),
        infer=c(TRUE, TRUE))
```

There is no "correct" way to compute the degrees of freedom for inferential statistics (SE, CIs, *p*-values). Two common choices are "Kenward-Roger" and "Satterthwaite". There is little empirical reason to vastly prefer one over the other (they each seem to perform a wee bit bitter under different conditions).

In `emmeans`, the Kenward-Roger degrees of freedom are the default. For Satterthwaite, use the `lmer.df` argument:

```{r}
fit.emm <- emmeans(m1, 
                   specs=c("vermetids", "sediment"),
                   lmer.df = "satterthwaite")
summary(contrast(fit.emm, 
                 method = "revpairwise", 
                 simple = "each",
                 combine = TRUE,
                 adjust="none"),
        infer=c(TRUE, TRUE))
```

If you want to compute the coefficient table or an ANOVA, lmer does not output test statistics. To get test statistics, you have to load the library "lmerTest" (which automatically loads "lme4"). With `lmerTest`, the Satterthwaite degrees of freedom are the default. For Kenward-Roger, use the `ddf` argument in either `summary()` (for the coefficients) or `anova()` (for ANOVA).

```{r}
coef(summary(m1)) # default is Satterthwaite
```

```{r}
coef(summary(m1, ddf="Kenward-Roger"))
```

Compare the output from emmeans and `coef(summary())` using the different methods for computing the df.

<!--chapter:end:chapters/55-lmm01-blocking.Rmd-->

---
title: "Models for longitudinal experiments -- I. pre-post designs"
author: "Jeffrey A. Walker"
date: "11/22/2020"
output:
  html_document:
    toc: true
    toc_float:
      toc_collapsed: true
    toc_depth: 3
    number_sections: true
    theme: lumen
    code_folding: hide
    fig_caption: true
---
# Linear models for longitudinal experiments -- I. pre-post designs {#pre-post}

```{r pre-post-setup, echo=FALSE, warning=FALSE, message=FALSE}

library(here)
library(janitor)
library(readxl)
library(data.table)

# analysis packages
library(nlme) #gls
library(lmerTest)
library(emmeans)
library(car) #qqPlot, SpreadLevelPlot
library(afex) # ANOVA

# graphing packages
library(ggplot2) # ggplot environment
library(ggpubr) # publication ready plots
library(ggforce) # jitter
library(ggsci) # color palettes
library(ggpubfigs) # color palettes
library(ggthemes) # themes and palettes including colorblind
# execute this command only the first time you use ggpubfigs
# devtools::install_github("JLSteenwyk/ggpubfigs")
library(cowplot) # combine plots
library(lazyWeave) # pretty pvalues
library(broom)
library(kableExtra) #tables


here <- here::here
data_folder <- "data"

# Okabe & Ito palette
pal_ito_seven <- friendly_pal("ito_seven") # ggpubfigs
pal_okabe_ito <- colorblind_pal()(8)[2:8] # ggthemes
```

In a Longitudinal experiment, the response is measured in the same set of individuals over multiple time points. For example, in a glucose tolerance test (GTT), plasma glucose levels are measured at 0, 15, 30, 60, and 120 minutes (or some variation of this) after glucose has been given to the individual. The initial measure (time 0) is the **baseline** measure and subsequent measures are post-baseline measures. In a pre-post experiment, there is a single post-baseline measure. This chapter is focuses on pre-post designs. The next chapter generalizes this information to longitudinal experiments with more than one post-baseline measure. For this chapter, I'll use $pre_i$ and $post_i$ for the baseline and post-baseline measures of the response variable of indiviual $i$. The change from baseline for individual $i$ is $change_i = post_i - pre_i$. The change from baseline is often called a **change score**.

Imagine two versions of an experiment where researchers are interested in the effect of different intestinal microbiomes on body weight.

1. In the intial experiment of the study, the researchers harvest intestinal microbiomes from five mice fed on chow and from five mice fed on a high-fat diet (HFD) and transplant these into ten, male C57bl/6 mouse randomly assigned to donor (this is done by fecal transplant). We have five replicates of the non-obese microbiome and five of the obese microbiome. Weight is measured at baseline (the day of the transplant) and four weeks later. The expected difference at baseline is zero, because the mice are sampled from the same population and the treatment was randomly assigned.
2. Additional experiments identified the potential role of inflammation-induced obesity in response to microbiome-derived lipopolysaccharide (LPS). In the second experiment, the researchers transplant intestinal microbiota from ten HFD mice into five male, wild-type C57bl/6 mice and five male, C57bl/6 mice in which the LPS receptor (CD14) has been knocked-out. Weight is measured at baseline (the day of the transplant) and four weeks later. The expected difference at baseline is not zero. It is not zero because the mice in each treatment are sampled from different populations (different genotypes). Even if we *think* that genotype shouldn't have an effect on weight at baseline, the expected difference is not zero because the treatment is not randomized at baseline but prior to baseline.

This distinction is important because the best practice linear model depends on it.

## Best practice models

**If the treatment is randomized at baseline to individuals sampled from the same population**

1. Model $\mathcal{M}_1$ -- "ANCOVA model"
* model formula: `post ~ pre + treatment`
* a linear model that estimates the effect as the coefficient of treatment adjusted for the baseline measure, with $post$ as the response.
* The response variable can either be the post-baseline measure or the change-score -- the estimate of the treatment effect is the same.
* **This is the best practice model if the expected difference at baseline is zero**.

**If the treatment is randomized prior to baseline (the groups are not composed of individuals sampled from the same population at baseline)**

2. Model $\mathcal{M}_2$ -- "change-score model"
* model formula: `change ~ treatment`
* a linear model that estimates the effect as the coefficient of $treatment$, with the change score as the response. This is not the same as Model $\mathcal{M}_1$ as there is no covariate in the model.
* **This is the best practice model if the expected difference at baseline is not zero**.

These two models present vary different ways of thinking about what the treatment effect is. In the ANCOVA model, we think of the treatment effect as the difference between treated and control means at the post-baseline time. Because the post-baseline measure of individual *i* is partly determined by it's value at baseline, variation in baseline values partly masks the effect of treatment. By adding baseline value as a covariate in the model, this masking effect of initial variation is eliminated.

In the change-score model, we think of the treatment effect as the mean difference in the change from baseline. The response changes from baseline to post-baseline in all individuals and we can compute the mean change for both treated ($\overline{post_{tr} - pre_{tr}}$) and control ($\overline{post_{cn} - pre_{cn}}$) levels. The treatment effect is the difference in these means.

\begin{align}
\mathrm{effect} &= \overline{change}_{tr} - \overline{change}_{cn}\\
&= (\overline{post_{tr} - pre_{tr}}) - (\overline{post_{cn} - pre_{cn}})
\end{align}

In the change-score model, the treatment effect is a difference of differences. Differences of differences are interactions and an alternative way to estimate the same effect is with the interaction coefficient ($\beta_3$) of the factorial linear model

\begin{equation}
weight = \beta_0 + \beta_1 treatment + \beta_2 time + \beta_3 treatment:time + \varepsilon
\end{equation}

## Common alternatives that are not recommended

3. $\mathcal{M}_3$ -- "post model"
* model formula: `post ~ treatment`
* a linear model that estimates the effect as the coefficient of treatment, with $post$ as the response.
* this model is valid but is not taking advantage of the increased precision of $\mathcal{M}_1$
4. $\mathcal{M}_4$ -- "interaction model"
* model formula: `y ~ treatment*time`
* a factorial fixed-effects model that estimates the effect as the coefficient of the interaction $treatment:time$. The response is the pre and post values stacked into a single column $y$.
* this will give the same estimate as Model $\mathcal{M}_2$ but the standard error of the effect will incorrect because of the unmodeled correlated error due to multiple (pre and post) measures taken from the same individual. **Never use this for pre-post or longitudinal models**.
5. $\mathcal{M}_4$ -- "repeated measures model"
* doesn't estimate the fixed effect but does compute a *p*-value of this effect as the $treatment \times time$ interaction term of the ANOVA table. This is effectively equivalent to change-score model $\mathcal{M}_2$.
* makes highly contstraining assumptions 
* A repeated measures model is more common when there are multiple post-baseline measures 

## Advanced models
6. $\mathcal{M}_6$ -- "linear mixed model"
* model formula: `y ~ treatment*time + (1|id)`
* a factorial linear mixed model that estimates the effect as the coefficient of the interaction $treatment:time$. The response is the pre and post values stacked into a single column $y$.
* the correlated error is implicitly modeled using the random factor $id$.
* This will give the same estimate as Model $\mathcal{M}_2$ and the same confidence interval and p-value *if* the Satterthwaite degrees of freedom are used.
7. $\mathcal{M}_7$ -- "fixed effects with correlated error model"
* model formula: `y ~ treatment*time`
* linear model with fixed effects and correlated error due to repeated measures within individuals.
* This looks like the fixed effect model but the error is explicitly modeled to allow 1) correlation due to id and 2) heterogeneity due to time. The advantage of explicitly modeling the correlated error is the flexibility in the model of the error.
* estimates the effect as the coefficient of the interaction $treatment:time$. The response is the pre and post values stacked into a single column $y$.
* This is an alternative to the change-score model, but, as in the change-score model, **use this only if the expected difference at baseline is not zero**.
8. $\mathcal{M}_8$ -- "constrained LDA"
* model formula: `y ~ time + treatment:time`
* As in a $\mathcal{M}_7$, this is linear model with fixed effects and correlated error due to repeated measures within individuals. Unlike $\mathcal{M}_7$, the treatment levels are constrained to have the same baseline mean, which is why there is no coefficient for $treatment$. This model is known as Constrained Longitudinal Data Analysis (cLDA). The estimated treatment effect is effectively equal to that of the ANCOVA model ($\mathcal{M}_1$) but, compared to inference from the ANCOVA model, the SE is smaller, the CIs are narrower, and the p-value is smaller. This increased precision and power, however, comes at a cost to increased Type I error.
* This is an interesting alternative to the ANCOVA model.

## Understanding the alternative models

The data used to explore the alternative models for analyzing pre-post data are from Figure 3F of Reed et al. 2020.

[Source article: Reed, M.D., Yim, Y.S., Wimmer, R.D. et al. IL-17a promotes sociability in mouse models of neurodevelopmental disorders. Nature 577, 249–253 (2020). https://doi.org/10.1038/s41586-019-1843-6](https://www.nature.com/articles/s41586-019-1843-6)

[Source data](https://www.nature.com/articles/s41586-019-1843-6#Sec37)

```{r pre-post-eg1-import, echo=FALSE, message=FALSE}
data_from <- "IL-17a promotes sociability in mouse models of neurodevelopmental disorders"
file_name <- "41586_2019_1843_MOESM3_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)

# response - fmr1
# vehicle
fig3f_vehicle <- read_excel(file_path,
                     sheet = "Fig 3f",
                     range = "DA2:DI3", # 1 blank column
                     col_names = FALSE) %>%
  data.table() %>%
  transpose() %>%
  setnames(old = c("V1", "V2"), new = c("pre", "post"))

fig3f_il17 <- read_excel(file_path,
                     sheet = "Fig 3f",
                     range = "DL2:DV3", # 1 blank column
                     col_names = FALSE) %>%
  data.table() %>%
  transpose() %>%
  setnames(old = c("V1", "V2"), new = c("pre", "post"))

treatment_levels <- c("Vehicle", "IL-17")
fig3f <- rbind(data.table(treatment = treatment_levels[1], fig3f_vehicle),
               data.table(treatment = treatment_levels[2], fig3f_il17))

fig3f[, treatment := factor(treatment, levels = treatment_levels)]
fig3f[, id := paste0("mouse_", .I)]
fig3f[, background := "fmr1"] # which mouse group?
fig3f[, change := post - pre]

```

The data is from an experiment estimating the effect of xxx on a sociability in mice. For some of the models, we need a wide version of the data.table and for other models, we need a long version. In the wide version, values of $pre$ and $post$ are in separate columns and the data.table includes the column $change$. In the long version, values of $pre$ and $post$ are stacked into a single column, and the column $time$ (with values "pre" and "post") is added to identify the time period of the measurement. There is no $change$ column because it is not needed in the analyses using the data in long format.

```{r pre-post-eg1-data-wide, echo=FALSE}
fig3f[1:4, .SD, , .SDcols = c("treatment", "id", "pre", "post", "change")] %>%
  knitr::kable(digits = c(1,2,2,1),
             caption = "First four rows of sociability data in wide format") %>%
  kable_styling()
```


```{r pre-post-eg1-melt, echo=FALSE}
fig3f_long <- melt(fig3f,
                   id.vars = c("treatment", "id", "change"),
                   measure.vars = c("pre", "post"),
                   variable.name = "time",
                   value.name = "sociability")
fig3f_long[, time := factor(time,
                            levels = c("pre", "post"))]
fig3f_long[, id := factor(id)]

fig3f_long[1:4, .SD, , .SDcols = c("treatment", "time", "id", "sociability")] %>%
  knitr::kable(digits = c(1,2,2,1),
             caption = "First four rows of sociability data in long format.") %>%
  kable_styling()
```

### (M1) Linear model with the baseline measure as the covariate (ANCOVA model)

Model $mathcal{M}_1$ is a linear model with the baseline variable added as a covariate. This is almost universally referred to as the "ANCOVA model".

\begin{equation}
\mathcal{M}_1\;\;\;post = \beta_0 + \beta_1 pre + \beta_2 treatment + \varepsilon
\end{equation}

The model formula is 

`post ~ pre + treatment`

```{r pre-post-ancova, echo=FALSE}
m1 <- lm(post ~ pre + treatment, data = fig3f)
m1_coef <- cbind(coef(summary(m1)),
                 confint(m1))
knitr::kable(m1_coef,
             digits = c(1,2,1,3,1,1),
             caption = "Coefficient table of Model M1") %>%
  kable_styling()
```

The third row (`treatmentIL-17`) of coefficient table \@ref(tab:longitudinal-ancova) of Model $\mathcal{M}_1$ contains the statistics of interest. The estimate is the treatment effect adjusted for the baseline value, or "what the treatment effect would be if all the baseline values were equal". Adding the covariate increases precision, which is why the ANCOVA model is preferred to a simple analysis of the post-baseline values (the "post model").

```{r pre-post-longitudinal-ancova, echo=FALSE, eval=FALSE}
lmeModel = lmer(dep ~ cv + group*trial + trial:cv + (1|subj), data=long_sas)

fig3f_m1 <- gls(glucose ~ -1 + glucose_0 + time_f + treatment:time_f,
          data = fig3f,
          weights = varIdent(form= ~ 1 | time_f),
          correlation= corSymm(form=~ 1| id))


```

### (M2) Linear model of the change score (change-score model)

Model $\mathcal{M}_2$ is a linear model with the change score used as the response and treatment factor as the only $X$ variable. This is equivalent to a Student's *t*-test of the change scores.

\begin{equation}
\mathcal{M}_2\;\;\;change = \beta_0 + \beta_1 treatment + \varepsilon
\end{equation}

The model formula is 

`change ~ treatment`

```{r pre-post-change-score, echo=FALSE}
m2 <- lm(change ~ treatment, data = fig3f)
m2_coef <- cbind(coef(summary(m2)),
                 confint(m2))
knitr::kable(m2_coef,
             digits = c(1,2,1,3,1,1),
             caption = "Coefficient table of Model M2") %>%
  kable_styling()
```

The second row (`treatmentIL-17`) of coefficient table \@ref(tab:longitudinal-change-score) of Model $\mathcal{M}_2$ contains the treatment effect, which is the simple difference between the mean "IL-17" and "Vehicle" change scores.

### (M3) Linear model of post-baseline values without the baseline as a covariate (post model)

Model $\mathcal{M}_3$ is a linear model with the post-baseline measure only. This is equivalent to a Student's *t*-test of the post-baseline measures.

\begin{equation}
\mathcal{M}_3\;\;\;post = \beta_0 + \beta_1 treatment + \varepsilon
\end{equation}

The model formula is 

`post ~ treatment`

```{r pre-post-post, echo=FALSE}
m3 <- lm(post ~ treatment, data = fig3f)
m3_coef <- cbind(coef(summary(m3)),
                 confint(m3))
knitr::kable(m3_coef,
             digits = c(1,2,1,3,1,1),
             caption = "Coefficient table of Model M3") %>%
  kable_styling()
```

* The second row (`treatmentIL-17`) of coefficient table \@ref(tab:longitudinal-post) of Model $\mathcal{M}_3$ contains the treatment effect, which is the simple difference between the "IL-17" and "Vehicle" means.
* If the experiment includes baseline measures **and** the expected difference at baseline is zero, the ANCOVA model ($\mathcal{M}_1$) should be used because the addition of the baseline measure as a covariate increases precision.
*  If the experiment includes baseline measures **and** the expected difference at baseline is not zero, the change-score model ($\mathcal{M}_2$) should be used.

### (M4) Linear model with factorial fixed effects (fixed-effects model)

Model $\mathcal{M}_4$ is a linear model with $treatment$ and $time$ crossed.

\begin{equation}
\mathcal{M}_4\;\;\;sociability = \beta_0 + \beta_1 treatment + \beta_2 time + \beta_3 treatment:time + \varepsilon
\end{equation}

The model formula is 

`post ~ treatment*time`

```{r pre-post-fixed-effects, echo=FALSE}
m4 <- lm(sociability ~ treatment*time,
          data = fig3f_long)
m4_coef <- cbind(coef(summary(m4)),
                 confint(m4))
m4_coef %>%
  knitr::kable(digits = c(1,2,1,1,3,1),
               caption = "Coefficient table of Model M4") %>%
  kable_styling()
```

### (M5) Repeated measures ANOVA

```{r pre-post-rmanova, message=FALSE, echo = FALSE}
m5 <- aov_4(sociability ~ treatment*time + (time|id), data = fig3f_long)
m5$anova_table %>%
  knitr::kable(digits = c(1,1,1,2,2,3),
             caption = "ANOVA table of Model M5.") %>%
  kable_styling()
```

### (M6) Linear mixed model

Model $\mathcal{M}_6$ is the linear mixed model

\begin{equation}
\mathcal{M}_6\;\;\;sociability = (\beta_0 + \beta_{0j}) + \beta_1 treatment + \beta_2 time + \beta_3 treatment:time + \varepsilon
\end{equation}

where $\beta_{0j}$ is the random intercept for mouse $j$. The model formula is 

`sociability ~ treatment*time + (1|id)`


```{r pre-post-lmm, echo=FALSE, message=FALSE}
m6 <- lmer(sociability ~ treatment*time + (1|id), data = fig3f_long)
m6_coef <- cbind(coef(summary(m6)),
                  confint(m6)[3:6,])
m6_coef %>%
  knitr::kable(digits = c(1,2,1,1,3,1),
             caption = "Coefficient table of Model M6") %>%
  kable_styling()

```

The estimate of the treatment effect is not the 2nd row (`treatmentIL-17`) but the interaction row (`treatmentIL-17:timepost`) of coefficient table \@ref(tab:longitudinal-lmm) of Model $\mathcal{M}_6$. The coefficient of `treatmentIL-17` ($b_1$) is an estimate of the effect of treatment at baseline. The coefficient of `timepost` ($b_2$) is the "slope" of the model -- it is the estimate of the effect of the treatment on the reference ("Vehicle") group. The coefficient of the interaction ($b_3$) is an estimate of the difference in slope between the Vehicle and IL-17 groups, that is, the difference in the response to treatment. This is the treatment effect.

The inferential statistics for the treatment effect (the interaction effect!) for the LMM ($\mathcal{M}_6$) are the same as those for the change-score model ($\mathcal{M}_3) *except* for the confidence intervals, which are too narrow in the coefficient table of the LMM. A better method for computing the confidence limit of the treatment effect for the LMM is to compute the interaction contrast of the model. The confidence interval of the contrast computed with the Satterthwaite degrees of freedom are identicle to the confidence intervals from the change-score model.

```{r pre-post-lmm-contrast, echo=FALSE}
m6_emm <- emmeans(m6,
                    specs = c("treatment", "time"),
                    lmer.df = "Satterthwaite")
m6_pairs <- contrast(m6_emm, interaction = c("trt.vs.ctrl"), by = NULL) %>%
  summary(infer = TRUE)
# do not use profile or Wald intervals of coefficients, which are 
# similar/equal to lmer.df = "asymptotic"
knitr::kable(m6_pairs,
             digits = c(1,1,1,2,0,1,1,1,3),
             caption = "Contrasts table of Model M6.") %>%
  kable_styling()
```

```{r compare-m6-ci, echo=FALSE, eval=FALSE}
confint(m6, method = "profile")[6,]
confint(m6, method = "Wald")[6,]
confint(m6, method = "boot")[6,]
m6_pairs
```

### (M7) Linear model with correlated error

```{r pre-post-lmCorErr, echo=FALSE}
m7 <- gls(sociability ~ treatment*time,
          data = fig3f_long,
          weights = varIdent(form= ~ 1 | time),
          correlation= corSymm(form=~ 1| id))
m7_coef <- cbind(coef(summary(m7)),
                 confint(m7))
m7_coef %>%
  knitr::kable(digits = c(1,2,1,1,3,1),
               caption = "Coefficient table of Model M7") %>%
  kable_styling()
```

```{r pre-post-lmCorErr-contrast, echo=FALSE, message=FALSE}
m7_emm <- emmeans(m7,
                  specs = c("treatment", "time"),
                  mode = "satterthwaite")
m7_pairs <- contrast(m7_emm, interaction = c("trt.vs.ctrl"), by = NULL) %>%
  summary(infer = TRUE)
# do not use profile or Wald intervals of coefficients, which are 
# similar/equal to lmer.df = "asymptotic"
knitr::kable(m7_pairs,
             digits = c(1,1,1,2,0,1,1,1,3),
             caption = "Contrasts table of Model M7.") %>%
  kable_styling()
```

### (M8) Constrained fixed effects model with correlated error (cLDA model)

Model $mathcal{M}_7$, the Constrained Longitudinal Data Analysis (cLDA) is a linear model with correlated error. This is similar to a linear mixed model

\begin{align}
\mathcal{M}_7\;\;\;sociability &= \beta_0 + \beta_1 time + \beta_2 treatment:time + \varepsilon_j\\
$\varepsilon_j \sim N(0, \sigma^2R)$
\end{align}

The model formula is

`sociability ~ time + treatment*time`

```{r pre-post-clda-functions, echo=FALSE, eval=TRUE}
clda1 <- function(y,
                  dt_long){
  model_matrix <- model.matrix(~ treatment*time,
                          data = dt_long)
  # remove intercept column and effect of treatment at baseline
  X <- model_matrix[, 3:4] # only for 2 levels
  X <- clean_names(data.table(X))
  dt_long_x <- cbind(dt_long, X)
  form <- formula(paste0(y, " ~ ", paste(colnames(X), collapse = " + ")))
  fit <- gls(form,
            data = dt_long_x,
            weights = varIdent(form= ~ 1 | time),
            correlation= corSymm(form=~ 1| id))
  return(fit)
}

clda2 <- function(y,
                  dt_long){
  dt_clda <- copy(dt_long)
  time0 <- levels(dt_clda$time)[1]
  tr <- levels(dt_clda$treatment)[2] # again only 2 levels
  
  dt_clda[, time.treatment := ifelse(time != time0 &
                                       treatment==tr,
                                     "interaction",
                                     "control")]
  dt_clda[, time.treatment := factor(time.treatment,
                                     c("control","interaction"))]
  
  form <- formula(paste0(y, " ~ time + time.treatment"))
  fit2 <- gls(form,
            data = dt_clda,
            weights = varIdent(form= ~ 1 | time),
            correlation= corSymm(form=~ 1| id))
  
  return(fit)
}

m8 <- clda1(y = "sociability",
            fig3f_long)
m8_coef <- cbind(coef(summary(m8)),
                 confint(m8))
m8_coef %>%
  knitr::kable(digits = c(1,2,1,4,1,1),
             caption = "Coefficient table of Model M8") %>%
  kable_styling()

```

### Comparison table

```{r comparison-table, echo=FALSE}
m1_row <- cbind(Model = "ANCOVA", data.table(m1_coef)[3])
setnames(m1_row, old = c("Std. Error", "t value"), new = c("SE", "t"))

m2_row <- cbind(Model = "change-score", data.table(m2_coef)[2])
setnames(m2_row, old = c("Std. Error", "t value"), new = c("SE", "t"))

m3_row <- cbind(Model = "post", data.table(m3_coef)[2])
setnames(m3_row, old = c("Std. Error", "t value"), new = c("SE", "t"))

m4_row <- cbind(Model = "factorial", data.table(m4_coef)[4])
setnames(m4_row, old = c("Std. Error", "t value"), new = c("SE", "t"))

m5_row <- data.table(Model = "RM-ANOVA",
                     Estimate = as.numeric(NA),
                     SE = as.numeric(NA),
                     t = as.numeric(NA),
                     "Pr(>|t|)" = anova(m5)[3, "Pr(>F)"],
                     "2.5 %" = as.numeric(NA),
                     "97.5 %" = as.numeric(NA))

m6_row <- cbind(Model = "LMM-coef", data.table(m6_coef[,-3])[4])
setnames(m6_row, old = c("Std. Error", "t value"), new = c("SE", "t"))
m6_row2 <- cbind(Model = "LMM-contrast", data.table(m6_pairs))
m6_row2 <- m6_row2[, .SD, .SDcols = c("Model", "estimate", "SE", "t.ratio", "p.value", "lower.CL", "upper.CL")]
setnames(m6_row2, 
         old = c("estimate", "t.ratio", "p.value", "lower.CL", "upper.CL"),
         new = c("Estimate", "t", "Pr(>|t|)", "2.5 %", "97.5 %"))

m7_row <- cbind(Model = "CorErr-coef", data.table(m7_coef)[4])
setnames(m7_row, 
         old = c("Value", "Std.Error", "t-value", "p-value"),
         new = c("Estimate", "SE", "t", "Pr(>|t|)"))
m7_row2 <- cbind(Model = "CorErr-contrast", data.table(m7_pairs))
m7_row2 <- m7_row2[, .SD, .SDcols = c("Model", "estimate", "SE", "t.ratio", "p.value", "lower.CL", "upper.CL")]
setnames(m7_row2, 
         old = c("estimate", "t.ratio", "p.value", "lower.CL", "upper.CL"),
         new = c("Estimate", "t", "Pr(>|t|)", "2.5 %", "97.5 %"))

m8_row <- cbind(Model = "cLDA", data.table(m8_coef)[3])
setnames(m8_row,
         old = c("Value", "Std.Error", "t-value", "p-value"),
         new = c("Estimate", "SE", "t", "Pr(>|t|)"))

model_comparison <- rbind(m1_row, m2_row, m3_row, m4_row, m5_row, m6_row, m6_row2, m7_row, m7_row2, m8_row)
model_comparison %>%
  knitr::kable(caption = "Estimated treatment effects from Models 1-8") %>%
  kable_styling()
```
## Example 1 -- a single post-baseline measure (pre-post design)
#### Fit the model

```{r pre-post-eg1-m1, echo=TRUE}
m1 <- lm(post ~ pre + treatment, data = fig3f)
```

#### Model checking

```{r pre-post-eg1-qq, echo=TRUE}
set.seed(1)
qqPlot(m1, id = FALSE)
```

The residuals are within the bounds reasonably expected of a sample from a normal distribution

```{r pre-post-eg1-spreadlevel, echo=TRUE}
spreadLevelPlot(m1)
```

The residuals do not show conspicuous heterogeneity.

#### Inference from the model

```{r pre-post-eg1-inference, echo=TRUE}
m1_coef <- cbind(coef(summary(m1)),
                 confint(m1))
knitr::kable(m1_coef,
             digits = c(1,2,1,3,1,1),
             caption = "Coefficient table of model m1") %>%
  kable_styling()
```
Sociability in IL-17 treated *fmr1* mice is 12.7 (95% CI: 3.5, 21.9, $p = 0.01$) percentage points higher than that in Control mice.

#### Plot the model

```{r pre-post-eg1-adjust, echo=FALSE}
# adjusted response
new_dt <- copy(fig3f)
new_dt[, pre := mean(pre)]
fig3f[, post_adj := predict(m1, newdata = new_dt) +
         residuals(m1)]
# check against emmeans
check_it <- FALSE
if(check_it == TRUE){
  fig3f[, .(mean = mean(post_adj)), by = treatment] # check!
}
```

```{r pre-post-eg1-prep-gg, echo=FALSE}
m1_emm <- emmeans(m1, specs = "treatment")
m1_pairs <- contrast(m1_emm, method = "revpairwise") %>%
  summary(infer = TRUE) # this is no more info than in coef table

m1_emm_dt <- summary(m1_emm) %>%
  data.table()
m1_pairs_dt <- data.table(m1_pairs)
```

```{r pre-post-eg1-response, echo=FALSE}
gg_response <- ggplot(data = fig3f,
                      aes(y = treatment,
                          x = post_adj,
                          color = treatment)) +
  geom_jitter(height = 0.25) +
  
  geom_point(data = m1_emm_dt,
             aes(x = emmean),
             size = 3) +
  
  geom_errorbar(data = m1_emm_dt,
                aes(x = emmean,
                    xmin = lower.CL,
                    xmax = upper.CL),
                width = 0.1) +
  xlab("Adjusted sociability (%)") +
  theme_pubr() +
  scale_color_manual(values=pal_okabe_ito) +
  theme(axis.title.y=element_blank()) +
  theme(legend.position="none") +

  
  NULL
# gg_response
```

```{r pre-post-eg1-effect, echo=FALSE}
gg_effect <- ggplot(data = m1_pairs_dt,
                      aes(y = contrast,
                          x = estimate)) +

  geom_point(size = 3) +
  
  geom_errorbar(aes(xmin = lower.CL,
                    xmax = upper.CL),
                width = 0.1) +
  
  xlab("Difference in adjusted sociability (%)") +

  # draw a line at effect = 0
  geom_vline(xintercept=0, linetype = 2) +
  coord_cartesian(xlim = c(-8,28)) +

  theme_pubr() +
  theme(axis.title.y=element_blank()) +
  scale_x_continuous(position = "top") + # move to top
  
  NULL
#gg_effect
```

```{r pre-post-eg1-plot, echo=FALSE}
plot_grid(gg_effect,
          gg_response,
          nrow=2,
          align = "v",
          axis = "lr",
          rel_heights = c(0.5,1))

```

## Working in R

**(M1) Linear model with the baseline measure as the covariate (ANCOVA model)**

```{r pre-post-ancova-show, echo=TRUE}
m1 <- lm(post ~ pre + treatment, data = fig3f)
m1_coef <- cbind(coef(summary(m1)),
                 confint(m1))
knitr::kable(m1_coef,
             digits = c(1,2,1,3,1,1),
             caption = "Coefficient table of Model M1") %>%
  kable_styling()
```


**(M2) Linear model of the change score (change-score model)**

```{r pre-post-change-score-show, echo=TRUE}
m2 <- lm(change ~ treatment, data = fig3f)
m2_coef <- cbind(coef(summary(m2)),
                 confint(m2))
knitr::kable(m2_coef,
             digits = c(1,2,1,3,1,1),
             caption = "Coefficient table of Model M2") %>%
  kable_styling()
```

**(M5) Repeated measures ANOVA**

use `aov_4` from the afex package, which allos input of a `lmer` model formula. Note that `time` is coded here as a random slope. This model would fail in `lmer` because there is only one time:id measure.

```{r pre-post-rmanova-show, message=FALSE, echo = FALSE}
m5 <- aov_4(sociability ~ treatment*time + (time|id), data = fig3f_long)
m5$anova_table %>%
  knitr::kable(digits = c(1,1,1,2,2,3),
             caption = "ANOVA table of Model M5.") %>%
  kable_styling()
```

**(M6) Linear mixed model**

```{r pre-post-lmm-show, echo=TRUE, message=FALSE}
m6 <- lmer(sociability ~ treatment*time + (1|id), data = fig3f_long)
m6_coef <- cbind(coef(summary(m6)),
                  confint(m6)[3:6,])
m6_coef %>%
  knitr::kable(digits = c(1,2,1,1,3,1),
             caption = "Coefficient table of Model M6") %>%
  kable_styling()

```

Use `lmer.df = "Satterthwaite"` as an argument in `emmeans` (not `contrast`!) to get the satterthwaite df for the interaction contrast.

```{r pre-post-lmm-contrast-show, echo=TRUE}
m6_emm <- emmeans(m6,
                    specs = c("treatment", "time"),
                    lmer.df = "Satterthwaite")
m6_pairs <- contrast(m6_emm, interaction = c("trt.vs.ctrl"), by = NULL) %>%
  summary(infer = TRUE)
# do not use profile or Wald intervals of coefficients, which are 
# similar/equal to lmer.df = "asymptotic"
knitr::kable(m6_pairs,
             digits = c(1,1,1,2,0,1,1,1,3),
             caption = "Contrasts table of Model M6.") %>%
  kable_styling()
```

**(M7) Linear model with correlated error**

```{r pre-post-lmCorErr-show, echo=TRUE}
m7 <- gls(sociability ~ treatment*time,
          data = fig3f_long,
          weights = varIdent(form= ~ 1 | time),
          correlation= corSymm(form=~ 1| id))
m7_coef <- cbind(coef(summary(m7)),
                 confint(m7))
m7_coef %>%
  knitr::kable(digits = c(1,2,1,1,3,1),
               caption = "Coefficient table of Model M7") %>%
  kable_styling()

```

Use `mode = "satterthwaite"` as an argument in `emmeans` (not `contrast`!) to get the satterthwaite df for the interaction contrast.

```{r pre-post-lmCorErr-contrast-show, echo=TRUE, message = FALSE}
m7_emm <- emmeans(m7,
                  specs = c("treatment", "time"),
                  mode = "satterthwaite")
m7_pairs <- contrast(m7_emm, interaction = c("trt.vs.ctrl"), by = NULL) %>%
  summary(infer = TRUE)
knitr::kable(m7_pairs,
             digits = c(1,1,1,2,0,1,1,1,3),
             caption = "Contrasts table of Model M7.") %>%
  kable_styling()
```

**(M8) Constrained fixed effects model with correlated error (cLDA model)**

**cLDA function**

```{r pre-post-clda-functions-show, echo=TRUE}

clda <- function(data,
                  y_col,
                  treatment_col = "treatment",
                  time_col = "time",
                  id_col = "id"){
# warning - not debugged for missing
  
  dt <- setDT(data)
  n_groups <- length(unique(dt[, get(treatment_col)]))
  n_times <- length(unique(dt[, get(time_col)]))
  
  # create model matrix with no treatment effect at baseline
  # part 1 - get the model matrix
  form <- formula(paste0(" ~ ", treatment_col," * ", time_col))
  model_matrix <- model.matrix(form, data = dt)
  
  # part 2 - exclude columns
  treat_cols <- 2:n_groups
  exc <- c(1, treat_cols)
  
  # constrained model matrix
  X <- model_matrix[, -exc] %>%
    data.table() %>%
    clean_names()
  dt <- cbind(dt, X)
  rhs <- paste0(names(X), collapse = " + ")
  
  model_form <- formula(paste0(y_col, " ~ ", rhs))
  weight_form <- formula(paste0("~ 1 | ", time_col))
  cor_form <- formula(paste0("~ 1 | ", id_col))
  fit <- gls(model_form,
             data = dt,
             weights = varIdent(form = weight_form),
             correlation= corSymm(form = cor_form))
  return(fit)
}
```

```{r pre-post-clda-show, echo=TRUE}
m8 <- clda(fig3f_long,
           y_col = "sociability",
           treatment_col = "treatment",
           time_col = "time",
           id_col = "id")
m8_coef <- cbind(coef(summary(m8)),
                 confint(m8))
m8_coef %>%
  knitr::kable(digits = c(1,2,1,4,1,1),
             caption = "Coefficient table of Model M8") %>%
  kable_styling()
```

## Hidden code
### Import and wrangle mouse sociability data

```{r pre-post-eg1-import-show, echo=TRUE, eval=FALSE, message=FALSE}
data_from <- "IL-17a promotes sociability in mouse models of neurodevelopmental disorders"
file_name <- "41586_2019_1843_MOESM3_ESM.xlsx"
file_path <- here(data_folder, data_from, file_name)

# response - fmr1
# vehicle
fig3f_vehicle <- read_excel(file_path,
                     sheet = "Fig 3f",
                     range = "DA2:DI3", # 1 blank column
                     col_names = FALSE) %>%
  data.table() %>%
  transpose() %>%
  setnames(old = c("V1", "V2"), new = c("pre", "post"))

fig3f_il17 <- read_excel(file_path,
                     sheet = "Fig 3f",
                     range = "DL2:DV3", # 1 blank column
                     col_names = FALSE) %>%
  data.table() %>%
  transpose() %>%
  setnames(old = c("V1", "V2"), new = c("pre", "post"))

treatment_levels <- c("Vehicle", "IL-17")
fig3f <- rbind(data.table(treatment = treatment_levels[1], fig3f_vehicle),
               data.table(treatment = treatment_levels[2], fig3f_il17))

fig3f[, treatment := factor(treatment, levels = treatment_levels)]
fig3f[, id := paste0("mouse_", .I)] # add id for each mouse
fig3f[, background := "fmr1"] # which mouse group?
fig3f[, change := post - pre] # change score

# melt to long
fig3f_long <- melt(fig3f,
                   id.vars = c("treatment", "id", "change"),
                   measure.vars = c("pre", "post"),
                   variable.name = "time",
                   value.name = "sociability")
fig3f_long[, time := factor(time,
                            levels = c("pre", "post"))]
fig3f_long[, id := factor(id)]
```




<!--chapter:end:chapters/57-longitudinal.Rmd-->

# Linear models for count data -- Generalized Linear Models I {#counts}

```{r glm1-setup, echo=FALSE, message=FALSE}
library(ggplot2)
library(ggpubr)
library(cowplot)
library(readxl)
library(emmeans)
library(data.table)

library(MASS) # negative binomial
#library(vcd)
library(DHARMa) # diagnostic plots
#library(COUNT) # dispersion stats

bookdown_compile <- TRUE

if(bookdown_compile==TRUE){
  data_path <- "data" # bookdown
  source("R/clean_label.R") # bookdown
}else{
  data_path <- "../data" # notebook
  source("../R/clean_label.R")
}



```

Biologists frequently count stuff, and design experiments to estimate the effects of different factors on these counts. For example, the effects of environmental mercury on clutch size in a bird, the effects of warming on parasite load in a fish, or the effect of exercise on RNA expression.

Count data differ from data with normal error in many ways, including 1) counts are discrete, and can be zero or positive integers only, 2) counts tend to bunch up on the small side of the range, creating a distribution with a positive skew, 3) a sample of counts can have an abundance of zeros, and 4) the variance of counts increases with the mean (see Figure \@ref(fig:glm1-plot1) for some of these properties). Some count data can be approximated by a normal distribution and reasonably modeled with a linear model but more often, count data are modeled with **Poisson distribution** or **negative binomial distribution** using a **generalized linear model** (GLM). Poisson and negative binomial distributions are **discrete probability distributions** with two important properties: 1) the distribution contains only zero and positive integers and 2) the variance is a function of the mean. Back before modern computing and fast processors, count data were often analyzed by either **transforming** the response or by **non-parametric hypothesis tests**. One reason to prefer a statistical modeling approach with a GLM is that we can get interpretable parameter estimates. By contrast, both the analysis of transformed data and non-parametric hypothesis tests are really tools for computing "correct" $p$-values.

```{r glm1-plot1, echo=FALSE, message=FALSE, warning=FALSE, fig.cap="Histogram of the count of a trematode parasite larvae in Control vs. Infected fish. Fish in the Infected treatment are infected with a tapeworm."}

folder <- "Data from Experimental parasite community ecology- intraspecific variation in a large tapeworm affects community assembly"
file <- "Lab_exp.csv"
file_path <- paste(data_path, folder, file, sep="/")
worm <- fread(file_path)
worm[, Treatment:=factor(Treatment, c("Control", "Uninfected", "Infected LG", "Infected HG"))]

gghistogram(data=worm[Treatment %in% c("Control", "Infected HG")], x = "Diplo_intensity",
   color="Treatment", fill="Treatment",
   add = "mean", rug = FALSE,
   palette = c("#00AFBB", "#E7B800"))

```

## The generalized linear model

As outlined in section [Assumptions for inference with statistical models] in Chapter 1, a common way that biological researchers think about a response variable is

\begin{align}
y_i &= \beta_0 + \beta_1 x_i + \varepsilon_i \\
\varepsilon &\sim N(0, \sigma)
(\#eq:spec1)
\end{align}

That is, we can think of a response as the sum of some systematic part and "random errror", which is a random draw from a normal distribution with mean zero and variance $\sigma^2$. This way of thinking about the generation of the response is useful for linear models, and model checking linear models, but is not useful for generalized linear models or model checking generalized liner models. For example, if we want to model the number of parasites that have infected a fish using a Poisson distribution, the following is the **wrong** way to think about the statistical model

\begin{align}
y_i &= \beta_0 + \beta_1 x_i + \varepsilon_i\\
\varepsilon &\sim Poisson(\lambda)
(\#eq:spec1-poisson)
\end{align}

That is, we should not think of a count as the sum of a systematic part and a random draw from a Poisson distribution. Why? Because it is the counts (or the counts conditional on $X$) that are poisson distributed, not the residuals from the fit model.

Thinking about the distribution of count data using model \@ref(eq:spec1-poisson) leads to absurd consequences. For example, if we set the mean of the Poisson "error" to zero (like with a normal distribution), then the error term for every observation would *have to* be zero (because the only way to get a mean of zero with non-negative integers is if every value is zero). Or, if the study is modeling the effect of a treatment on the counts (that is, the $X$ are dummy variables) then $\beta_0$ is the expected mean count of the control (or reference) group. But if we add non-zero Poisson error to this, then the mean of the control group would be larger than $\beta_0$. This doesn't make sense. And finally, equation \@ref(eq:spec1-poisson) generates a continuous response, instead of an integer, because $\beta_0$ and $\beta_1$ are continuous.

A better way to think about the data generation for a linear model, because this naturally leads to the *correct* way to think about data generation for a generalized linear model, is

\begin{align}
y_i &\sim N(\mu_i, \sigma)\\
\mathrm{E}(Y|X) &= \mu\\
\mu_i &= \beta_0 + \beta_1 x_i
(\#eq:lm-spec2)
\end{align}

That is, a response is a random draw from a normal distribution with mean $mu$ (not zero!) and variance $\sigma^2$. Line 1 is the stochastic part of this specification. Line 3 is the systematic part.

The specification of a generalized linear model has both stochastic and systematic parts but adds a third part, which is a **link function** connecting the stochastic and systematic parts.

1. **The stochastic part**, which is a probability distribution from the exponential family (this is sometimes called the "random part")
\begin{equation}
y_i \sim \mathrm{Prob}(\mu_i)
\end{equation}
2. **the systematic part**, which is a linear predictor (I like to think about this as the deterministic part)
\begin{equation}
\eta=\mathbf{X}\boldsymbol{\beta}
\end{equation}
3. a **link function** connecting the two parts
\begin{equation}
\eta_i = g(\mu_i)
\end{equation}

$\mu$ (the Greek symbol mu) is the conditional mean (or expectation $\mathrm{E}(Y|X)$) of the response on the **response scale** and $\eta$ (the Greek symbol eta) is the conditional mean of the response on the **link scale**. A GLM models the response with a distribution specified in the stochastic part. The probability distributions introduced in this chapter are the Poisson and Negative Binomial. The natural link function for the Poisson and Negative Binomial is the "log link", $\eta = log(\mu)$. More generally, while each distribution has a natural (or, "canonical") link function, one can use alternatives. Given this definition of a generalized linear model, a linear model is a GLM with a normal distribution and an Identity link ($\eta = \mu$).

When modeling counts using the Poisson or negative binomial distributions with a log link, the link scale is linear, and so the effects are additive on the link scale, while the response scale is nonlinear (it is the exponent of the link scale), and so the effects are multiplicative on the response scale. If this doesn't make sense now, an example is worked out below. The inverse of the link function backtransforms the parameters from the link scale back to the response scale. So, for example, a prediction on the response sale is $\mathrm{exp}(\hat{\eta})$ and a coefficient on the response scale is $\mathrm{exp}(b_j)$.

## Count data example -- number of trematode worm larvae in eyes of threespine stickleback fish

The example is an experiment measuring the effect of the parasitic tapeworm *Schistocephalus solidus* infection on the susceptibility of infection from a second parasite, the trematode *Diplostomum pseudospathaceum*, in the threespine stickleback fish *Gasterosteus aculeatus*^[Benesh, D. P., & Kalbe, M. (2016). Experimental parasite community ecology: intraspecific variation in a large tapeworm affects community assembly. Journal of Animal Ecology, 85(4), 1004-1013]. The treatment levels are "Control" (unexposed to the tapeworm), "Uninfected" " (exposed to the tapeworm but uninfected), "Infected LG" (exposed and infected with the low growth population of the tapeworm), and "Infected HG" (exposed and infected with the high growth population of tapeworm). The response is the number of trematode larvae counted in the eyes (right and left combined) of the fish. A histogram of the counts is shown in Figure \@ref(fig:glm1-plot1) for the control and Infected HG treatment levels.

### Modeling strategy

<div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**NHST blues** -- Students are often encouraged by textbooks, colleagues, or the literature to start the analysis by first "testing" assumptions with hypothesis tests -- for example using a Shaprio-Wilks test of normality as a decision rule to decide if to use a parametric test such as a $t$-test or ANOVA if the null hypothesis of normality is not rejected, or a non-parametric test such as a Mann-Whitney U test if the null hypothesis of normality is rejected. I advise against this, because 1) this pre-test filtering automatically invalidates the $p$-value of the hypothesis test as it does not adjust for the filtering procedure, 2) real data are only approximately normal and as $n$ increses, a normality test will reject any real dataset, and 3) hypothesis tests are pretty robust to non-normality anyway.
</div> 

Instead of testing assumptions of a model using formal hypothesis tests *before* fitting the model, a better strategy is to 1) fit a model, and then do 2) **model checking** using **diagnostic plots**, diagnostic statistics, and simulation.

With these data, a researcher would typically fit a GLM with a Poisson or negative binomial distribution and log link. Here, I start with a linear model to illustrate the interpretation of diagnostic plots with non-normal data. I use the "linear model" specification (equation \@ref(eq:spec1)) because the diganostic plots for model checking a linear model use the residuals of the fit model.

\begin{align}
Diplo\_intensity_i &= \beta_0 + \beta_1 Uninfected_i + \beta_2 Infected\_LG_i + \beta_3 Infected\_HG_i + \varepsilon_i \\
\varepsilon &\sim N(0, \sigma)
(\#eq:worm-lm)
\end{align}

### Checking the model I -- a Normal Q-Q plot

Figure \@ref(fig:glm1-plot2)A shows a histogram of the residuals from the fit linear model. The plot shows that the residuals are clumped at the negative end of the range, which suggests that a model with a normally distributed conditional outcome (or normal error) is not well approximated.

```{r glm1-plot2, echo=FALSE, message=FALSE, warning=FALSE, fig.cap="Diagnostic plots of stickleback parasite data. A) Distribution of the residuals of the fit linear model. B) Normal Q-Q plot of the residuals of the fit linear model."}
fit.lm <- lm(Diplo_intensity ~ Treatment, data=worm, na="na.exclude")
worm[, Diplo_intensity_residual:=residuals(fit.lm)]
gg1 <- gghistogram(data=worm, x = "Diplo_intensity_residual",
   add = "mean", rug = FALSE,
   color = c("#00AFBB"), fill = c("#00AFBB"))
gg2 <- ggqqplot(data=worm, x = "Diplo_intensity_residual", title="Normal Q-Q")
plot_grid(gg1, gg2, labels="AUTO")
```

A better way to investigate this is with the **Normal Q-Q** plot in Figure \@ref(fig:glm1-plot2)B, which plots the sample quantiles for a variable against their theoretical quantiles. If the conditional outcome approximates a normal distribution, the points should roughly follow the line. Instead, for the worm data, the points are above the line at both ends. At the left (negative) end, this means that we aren't seeing the most negative values that would be expected (the observed values are more positive than the theoretical values). Remembering that this plot is of residuals, if we think about this as counts, this means that our smallest counts are not as small as we would expect given the mean and a normal distribution. This shouldn't be surprising -- the counts range down to zero and counts cannot be below zero. At the positive end, the sample values are again more positive than the theoretical values. Thinking about this as counts, this means that are largest counts are larger than expected given the mean and a normal distribution. This pattern is exactly what we'd expect of count data, or at least count data that borders zero.

 <div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**Intuition Pump** -- Let's construct a Normal Q-Q plot. A **quantile** (or percentile) of a vector of numbers is the value of the point at a specified percentage rank. The median is the 50% quantile. The 95% confidence intervals are at the 2.5% and 97.5% quantiles. In a Normal Q-Q plot, we want to plot the quantiles of the residuals against a set of theoretical quantiles.

1. To get the observed quantiles, rank the residuals of the fit linear model from most negative to most positive -- these are your quantiles! For example, if you have $n=145$ residuals, then the 73rd point is the 50% quantile.
2. A theoretical quantile from the normal distribution can be constructed using the `qnorm` function which returns the normal quantiles for a specified vector of percents. Alternatively, one could randomly sample $n$ points using `rnorm`. These of course will be sampled quantiles so will only approximate the expected theoretical quantiles, but I add this here because we use this method below.

Now simply plot the observed against theoretical quantiles. Often, the **standardized** quantiles are plotted. A standardized variable has a mean of zero and a standard deviation of one and is computed by 1) centering the vector at zero by subtracting the mean from every value, and 2) dividing each value by the standard deviation of the vector. Recognize that because a standard deviation is a function of deviations from the mean, it doesn't matter which of these operations is done first. A standardized theoretical quantile is specified by `qnorm(p, mean = 0, sd = 1)`, which is the default.

Below, I've plotted the standardized observed and theoretical quantiles against the vector of percents (from 0 to 100%). This plot also nicely shows how the residuals of the worm data deviate from that expected if these had a normal distribution. The plot nicely shows that the most negative observed quintiles are not as negative as expected given a normal distribution, which again makes sense because this would imply negative counts since the mean is close to zero. And it nicely shows that the most positive observed quantiles are more positive than expected given a normal distribution, again this makes sense in right skewed count data. Finally, the plot nicely shows that the median is less positive than that expected given a normal distribution, which is at the mean (a right skew tends to pull the mean to the right of the median).

```{r glm1-qqplot-sim, echo=FALSE}
Observed <- scale(sort(worm[, Diplo_intensity_residual]))
n <- length(Observed)
q <- n+1
x <- seq(1/q, 1-1/q,by=1/q)
Theoretical <- qnorm(x)
qq_data <- data.table(X=rep(x,2),
                      Data=rep(c("Observed", "Theoretical"), each=n), 
                      Quantile=c(Observed, Theoretical))
gg <- ggplot(data=qq_data, aes(x=X, y=Quantile, color=Data)) +
  geom_point() +
  NULL
gg
#qplot(x=Theoretical, y=Observed) + geom_abline(slope=1, intercept=0)
```
</div> 

### Checking the model II -- scale-location plot for checking homoskedasticity

```{r glm1-worm-scale-location, echo=FALSE}
worm[, scale_y:= sqrt(abs(scale(Diplo_intensity_residual)))]
worm[, fitted_y:=fitted(fit.lm)]
ggscatter(data=worm, 
          x="fitted_y",
          y="scale_y",
          title="Scale-Location",
          xlab="Fitted value",
          ylab="sqrt(standardized residual)",
          add="reg.line",
          conf.int = TRUE,
          add.params = list(color = "blue",
                            fill = "lightgray")
          )
```

A linear model also assumes the error has constant variance (that is, the error variance is not a function of the value of $X$), or homoskedasticity. The fit model can be checked for homoskedasticity using a scale-location plot, which is a scatterplot of the positive square-root of the standardized residuals against the fitted values^[fitted values are the predicted values, $\hat{Y}$]. If the residuals approximate a normal distribution, then a regression line through the scatter should be close to horizontal. The regression line in the scale-location plot of the fit of the linear model to the worm data shows a distinct increase in the "scale" (the square root of the standardized residuals) with increased fitted value, which is expected of data that are lognormally, Poisson, or negative binomially distributed.

### Two distributions for count data -- Poisson and Negative Binomial

The pattern in the normal Q-Q plot in Figure \@ref(fig:glm1-plot2)B should discourage one from modeling the data with a normal distribution and instead model the data with an alternative distribution using a Generalized Linear Model. There is no unique mapping between how data are generated and a specific distribution, so this decision is not as easy as thinking about the data generation mechanism and then simply choosing the "correct" distribution. Section 4.5 in Bolker (xxx) is an excellent summary of how to think about the generating processes for different distributions in the context of ecological data. Since the response in the worm data are counts, we need to choose a distribution that generates integer values, such as the Poisson or the negative binomial.

1. Poisson -- A Poisson distribution is the probability distribution of the number of occurrences of some thing (an egg, a parasite, or a specific mRNA transcript) generated by a process that generates the thing at a constant rate per unit effort (duration or space). This constant rate is $\lambda$, which is the expectation, so $\mathrm{E}(Y) = \mu = \lambda$. Because the rate per effort is constant, *the variance of a Poisson variable equals the mean*, $\sigma^2 = \mu = \lambda$. Figure \@ref(fig:glm1-poisson) shows three samples from a Poisson distribution with $\lambda$ set to 1, 5, and 10. The plots show that, as the mean count ($\lambda$) moves away from zero, a Poisson distribution 1) becomes less skewed and more closely approximates a normal distribution and 2) has an increasingly low probability of including zero (less than 1% zeros when the mean is 5).

A Poisson distribution, then, is useful for count data in which the conditional variance is close to the conditional mean. Very often, biological count data are not well approximated by a Poisson distribution because the variance is either less than the mean, an example of **underdispersion**^[the variance is less than that expected by the probability model], or greater than the mean, an example of **overdispersion**^[the variance is greater than that expected by the probability model]. A useful distribution for count data with overdispersion is the negative binomial.


```{r glm1-poisson-ci, echo=FALSE, eval=FALSE}
# interesting results - essentially normal does better than poisson
n <- 20
exp_b0 <- 1000
exp_b1 <- 1000
x <- rep(c("a", "b"), each=n)
niter <- 1000
b1 <- matrix(NA, nrow=niter, ncol=2)
b1.glm <- matrix(NA, nrow=niter, ncol=2)
for(iter in 1:niter){
  y <- c(rpois(n, lambda=exp_b0), rpois(n, lambda=(exp_b0+exp_b1)))
  fit <- lm(y~x)
  b1[iter,] <- confint(fit)[2,]
  fit.glm <- glm(y~x, family="poisson")
  b1.glm[iter,] <- exp(confint(fit.glm)[2,] + coef(fit.glm)[1])  - exp(coef(fit.glm)[1])
  #b1.glm2[iter,] <- exp(confint(fit.glm))[2,]*exp(coef(fit.glm)[1]) - exp(coef(fit.glm)[1])
}
sum((exp_b1 > b1[,1] & exp_b1 < b1[, 2]))/niter*100
sum((exp_b1 > b1.glm[,1] & exp_b1 < b1.glm[, 2]))/niter*100
```


```{r glm1-poisson, echo=FALSE}
n_samp <- 10^4
rp1 <- data.frame(y=rpois(n=n_samp, lambda=1))
rp5 <- data.frame(y=rpois(n=n_samp, lambda=5))
rp10 <- data.frame(y=rpois(n=n_samp, lambda=10))
fake_data_wide <- data.table(lambda.eq.1=rp1[,1], lambda.eq.5=rp5[,1], lambda.eq.10=rp10[,1])
fake_data_wide <- data.table("lambda = 1"=rp1[,1], "lambda = 5" =rp5[,1], "lambda = 10" =rp10[,1])
fake_data <- data.table(lambda=rep(c("1", "5", "10"), each=n_samp), count=c(lambda.eq.1=rp1[,1], lambda.eq.5=rp5[,1], lambda.eq.10=rp10[,1]))

zeros <- fake_data[, .(zeros=length(which(count==0))/n_samp*100), by=.(lambda)]
# gg <- gghistogram(data=fake_data,
#                   x = "count",
#                   add = "mean",
#                   binwidth=1,
#                   color="lambda",
#                   fill = "lambda",
#                   palette = rep("#00AFBB",3),
#                 )
# facet(gg, facet.by = "lambda")

gghistogram(data=fake_data_wide,
            x = c("lambda = 1", "lambda = 5", "lambda = 10"),
            add = "mean",
            combine=TRUE,
            binwidth=1,
            color = c("#00AFBB"), fill = c("#00AFBB"),
            xlab = "count",
            ylab =FALSE
                )

```

2. Negative Binomial -- The negative binomial distribution is a discrete probability distribution of the number of successes that occur before a specified number of failures $k$ given a probability $p$ of success. This isn't a very useful way of thinking about modeling count data in biology. What is useful is that the Negative Binomial distribution can be used simply as way of modeling an "overdispersed" Poisson process. The mean of a negative binomial variable is $\mu = k\frac{p}{1-p}$ and the variance is $\sigma^2 = \mu + \mu^2/k$. As a method for modeling an overdispersed Poisson variable, $k$ functions as a **dispersion parameter* controlling the amount of overdispersion and can be any real, positive value (not simply a positive integer), including values less than 1.

```{r glm1-test2, echo=FALSE, eval=FALSE}
n_samp <- 10^4
rp1 <- data.frame(y=rnbinom(n=n_samp, mu=10, size=0.9))
rp5 <- data.frame(y=rnbinom(n=n_samp, mu=10, size=10))
rp10 <- data.frame(y=rnbinom(n=n_samp, mu=100, size=0.9))
fake_data_wide <- data.table(lambda.eq.1=rp1[,1], lambda.eq.5=rp5[,1], lambda.eq.10=rp10[,1])
fake_data_wide <- data.table("mu=10, k=0.9"=rp1[,1], "mu=10, k=10" =rp5[,1], "mu=100, k=0.9" =rp10[,1])

gghistogram(data=fake_data_wide,
            x = c("mu=10, k=0.9", "mu=10, k=10", "mu=100, k=0.9"),
            add = "mean",
            combine=TRUE,
            binwidth=1,
            color = c("#00AFBB"), fill = c("#00AFBB"),
            xlab = "count",
            ylab =FALSE,
            scales="free"
                )

```

### Fitting a GLM with a Poisson distribution to the worm data
Let's fit a GLM with a Poisson distribution to the worm data. The model is

\begin{align}
Diplo\_intensity_i &\sim Poisson(\mu_i)\\
\mathrm{E}({Diplo\_intensity|Treatment}) &= \mu\\
\mu_i &= \mathrm{exp}(\eta_i)\\
\eta_i &= \beta_0 + \beta_1 Uninfected_i + \beta_2 Infected\_LG_i + \beta_3 Infected\_HG_i
\end{align}

1. The first line of the model is the stochastic part stating the response is modeled as a random Poisson variable with mean and variance $\mu$ (the rate parameter $\lambda$ of the Poisson distribution).
2. The second line states the $\mu$ is the conditional mean or conditional expectation
3. The third line connects the conditional mean on the link scale ($\eta$) with the conditional mean on the response scale ($\mu$)
4. The fourth line is the linear predictor, and includes three dummy variables.

Remember that the conditional mean is the expected/predicted/fitted/modeled value when $X=x_i$.

### Model checking fits to count data

we use the fit model to check
1. the overall similarity of observed and theoretical distributions
2. if the observed distribution is over or under dispersed
3. if there more zeros than expected by the theoretical distribution. If so, the observed distribution is **zero-inflated**

#### Model checking a GLM I -- the quantile residual Q-Q plot
A quantile-quantile (Q-Q) plot is used to check overall similarity of the observed distribution with the distribution that would be expected under the model. An alternative to a Normal Q-Q plot for a GLM fit is a quantile residual Q-Q plot of observed vs. expected **quantile residuals**. The basic algorithm for this is

1. Use the model parameters to simulate $p$ fake values of the response for each row of the data. This will be a $n \times p$ matrix of fake data where each column is a new, random sample of a population with parameters equal to that estimated for the observed data. For the Poisson, the parameter for each observation will be $\hat{\mu}_i$, the modeled value of observation $i$. For the negative binomial, the parameters will be $\hat{\mu}_i$ and the dispersion parameter $k$, which is the same for all observations.
2. For each observation (each row of the matrix of fake data), compute the fraction of simulated values smaller than the observed value of the response variable for that row. This fraction is the observed **quantile residual**, which ranges in value from 0 to 1. If the true data are distribitued as that specified by the model, then quantile residuals will have a uniform distribution.
3. Sort the observed quantile residuals from smallest to largest and plot against theoretical quantile residuals from a uniform distribution. One could transform the quantile residuals to standard, normal residuals and then plot using a traditional Normal Q-Q plot but this step isn't necessary (if reported, a Normal Q-Q plot of transformed quantile residuals might confuse readers who failed to read the fine print).

 <div style="background-color:#cccccc; text-align:left; vertical-align: middle; padding:20px 47px;">
**Misconceivable** -- A common misconception is that if the distribution of the response approximates a Poisson distribution, then the residuals of a GLM fit with a Poisson distribution should be normally distributed, which could then be checked with a Normal Q-Q plot, and homoskedastic, which could be checked with a scale-location plot. Neither of these is true because a GLM does not transform the data and, in fact, the model definition does not specify anything about the distribution of an "error" term -- there is no $\varepsilon$ in the model defintion above! This is why thinking about the definition of a linear model by specifying an error term with a normal distribution can be confusing and lead to misconceptions when learning GLMs.
</div> 

```{r glm1-worm-poisson-qr, echo=FALSE, message=FALSE}
n_sim <- 250
fit.pois <- glm(Diplo_intensity ~ Treatment, family="poisson", data=worm)

# DHARMa package diagnostic plots
plot_it <- FALSE
if(plot_it==TRUE){
  simulationOutput <- simulateResiduals(fittedModel = fit.pois, n = n_sim)
  plot(simulationOutput, asFactor = F)
  simulationOutput.refit <- simulateResiduals(fittedModel = fit.pois, n = n_sim, refit=TRUE)
  testDispersion(simulationOutput.refit)
}

# do this by hand
plot_it <- TRUE
n_row <- nrow(worm)

if(plot_it==TRUE){
  fake_data <- cbind(worm[, Diplo_intensity], simulate(fit.pois, nsim = n_sim, seed = 101))
  
  # do fake data by hand
  fake_data_2 <- matrix(NA, nrow=n_row, ncol=n_sim)
  for(j in 1:n_sim){
    fake_data_2[,j] <- rpois(n_row, lambda=fitted(fit.pois))
  }
  fake_data_2 <- cbind(worm[, Diplo_intensity], fake_data_2)
  # this works

  # Build uniform QQ plot
  quantile_residuals <- numeric(n_row)
  for(i in 1:n_row){
    quantile_residuals[i] <- (rank(fake_data[i,])[1] - 1)/n_sim
  }
  Observed <- sort(quantile_residuals)
  q <- n_row+1
  x <- seq(1/q, 1-1/q,by=1/q)
  Theoretical <- qunif(x)
  ggscatter(data=data.frame(Theoretical=Theoretical, Observed=Observed),
         x="Theoretical", 
         y="Observed",
         title = "Quantile Residual Q-Q Plot"
  ) +
    geom_abline(slope=1, intercept=0) +
    NULL
  
}

# convert to normal with box-mueller
# u1 <- quantile_residuals
# u2 <- runif(n_row)
# z <- sqrt(-2*log(u1))*cos(2*pi*u2)
# worm[, fitted_pois:=fitted(fit.pois)]
# worm[, scale_pois:=sqrt(abs(z))]
# ggscatter(data=worm, 
#           x="fitted_pois",
#           y="scale_pois",
#           title="Scale-Location",
#           xlab="Fitted value",
#           ylab="sqrt(standardized residual)",
#           add="reg.line",
#           conf.int = TRUE,
#           add.params = list(color = "blue",
#                             fill = "lightgray")
#           )

```

The Q-Q plot using quantile residuals with a Poisson distribution indicates that the counts of *Diplostomum* larvae in the eyes of the threespine stickleback are not well approximated by a Poisson distribution -- there are too many observed values near the ends of the expected tails, indicating the expected values are not spread out enough. This pattern emerges because the observed counts are overdispersed compared to a Poisson distribution.

#### Model checking a GLM II -- a dispersion plot

If observed counts are Poisson distributed, then the **Pearson residuals** ($r_i$) and the residual degrees of freedom of the fit model ($df$) can be used to compute a dispersion statistic

\begin{equation}
\frac{\sum{r_i}}{df}
\end{equation}

that has an expected value of 1. Instead of a formal hypothesis test of this statistic, I use a simulation approach and ask, "if the observed counts are Poisson distributed, what is the expected frequency distribution of this dispersion statistic?" and then use simulation to generate this expected distribution. The algorithm for this is

1. For each observation $i$, generate a random Poisson count using $\hat{\mu}$ as the parameter.
2. Fit the model and compute the dispersion statistic.
3. Repeat 1 and 2 $N_{iter}$ times.

```{r glm1-worm-poisson-dispersion, echo=FALSE}
pr <- sum(residuals(fit.pois, type="pearson")^2)  # Pearson Chi^2
df <- fit.pois$df.residual
observed_dispersion_statistic <- pr/df # dispersion statistic

set.seed(1)
niter <- 1000
pr.fake <- numeric(n_sim)
fake_data <- simulate(fit.pois, nsim = niter)
for(j in 1:niter){
  fit.pois.fake <- glm(fake_data[,j] ~ Treatment, family="poisson", data=worm)
  pr.fake[j] <- sum(residuals(fit.pois.fake, type="pearson")^2)  #PearsonChi2
}

mean_dispersion_statistic <- mean(pr.fake/df)

```

The plot below shows a histogram of the dispersion statistic computed for 1000 simulations of the worm data. The observed dispersion statistic is `r round(observed_dispersion_statistic, 1)`. The expected value is 1.0. The mean of the simulated values is `r round(mean_dispersion_statistic, 1)`. 

```{r glm1-worm-poisson-dispersion-plot, echo=FALSE, fig.cap="Observed vs. expected dispersion statistic. The observed statistic marked by the red line. The histogram of expected statistics are from 1000 simulations of the observed data."}

# plot simulation resulsts
qplot(pr.fake/df, binwidth=0.5/20) +
  xlab("dispersion statistic") +
  geom_vline(aes(xintercept=observed_dispersion_statistic), color="red") +
  ggtitle(" Observed vs. expected dispersion statistic") +
  NULL
```


### Fitting a GLM with a Negative Binomial distribution to the worm data

The model is

\begin{align}
Diplo\_intensity &\sim NB(\mu, k)\\
\mathrm{E}({Diplo\_intensity|Treatment}) &= \mu\\
\mu &= \mathrm{exp}(\eta)\\
\eta &= \beta_0 + \beta_1 Uninfected + \beta_2 Infected\_LG + \beta_3 Infected\_HG
\end{align}

This model specifies a negative binomial distribution but otherwise is just like that above specifying a Poisson distribution.

#### Model checking

```{r glm1-worm-nb, echo=FALSE}
n_sim <- 250
fit.nb <- glm.nb(Diplo_intensity ~ Treatment, data=worm)

plot_it <- FALSE
if(plot_it==TRUE){
  simulationOutput <- simulateResiduals(fittedModel = fit.nb, n = n_sim)
  plot(simulationOutput, asFactor = F)
  # testUniformity(simulationOutput = simulationOutput)
  # testDispersion(simulationOutput)
}

# do this by hand
plot_it <- TRUE
if(plot_it==TRUE){
  fake_data_nb <- cbind(worm[, Diplo_intensity], simulate(fit.nb, nsim = n_sim, seed = 101))

  # do fake data by hand
  set.seed(101)
  fake_data <- matrix(NA, nrow=n_row, ncol=n_sim)
  for(j in 1:n_sim){
    fake_data[,j] <- rnegbin(n_row, mu=fitted.values(fit.nb), theta=fit.nb$theta)
  }
  fake_data <- cbind(worm[, Diplo_intensity], fake_data)
  # this works

  n_row <- nrow(worm)
  quantile_residuals <- numeric(n_row)
  for(i in 1:n_row){
    quantile_residuals[i] <- (rank(fake_data[i,])[1] - 1)/n_sim
  }
  Observed <- sort(quantile_residuals)
  q <- n_row+1
  x <- seq(1/q, 1-1/q,by=1/q)
  Theoretical <- qunif(x)
  ggscatter(data=data.frame(Theoretical=Theoretical, Observed=Observed),
         x="Theoretical", 
         y="Observed",
         title = "Quantile Residual Q-Q Plot"
  ) +
    geom_abline(slope=1, intercept=0) +
    NULL
  
}

```

A quantile residual Q-Q plot of the GLM model fit with negative binomial distribution is illustrated above. This looks pretty good.

#### Model means and coefficients

In a Generalized Linear Model of counts using either a Poisson or negative binomial distribution, modeled means, coefficients, and contrasts can be reported either on the link or response scale. Remember, the response scale is a count, while the link scale is a log(count).

The modeled means on the link scale are

```{r glm1-worm-means-link, echo=FALSE}
# cell means link space
emmeans(fit.nb, specs="Treatment")
```

While the means on response scale are
```{r glm1-worm-means-response, echo=FALSE}
emmeans(fit.nb, specs="Treatment", type="response")
```

1. A mean on the response scale is simply the exponent of the mean on the link scale. For example, the mean of the Control treatment level on the response scale is $\mathrm{exp}(1.821408) = 6.180555$.

2. The CIs on the link scale are symmetric around the mean but those on the response scale are not. This is a feature, not a bug. Remember that counts are right skewed which means a CI will have a wider right than left interval. Check this!

3. If a plot includes a 1 SE error bar on the response scale, this is technically correct but it encourages the practice of computing CIs using the 2*SE rule of thumb. This rule breaks down for count data with right skewed distributions.

4. Plotting the response scale CIs is both technically correct and makes the 2*SE rule of thumb unnecessary.

```{r glm1-worm-nb-ci, echo=FALSE, eval=FALSE}
# link scale
link <- data.table(summary(emmeans(fit.nb, specs="Treatment")))
link[Treatment=="Control",asymp.UCL] - link[Treatment=="Control", emmean]
link[Treatment=="Control",asymp.LCL] - link[Treatment=="Control", emmean]
response <- data.table(summary(emmeans(fit.nb, specs="Treatment", type="response")))
response[Treatment=="Control",asymp.UCL] - response[Treatment=="Control", response]
response[Treatment=="Control",asymp.LCL] - response[Treatment=="Control", response]

```

The model coefficients on the link scale are

```{r glm1-worm-nb-coeffs-link, echo=FALSE, message=FALSE}
#cbind(coefficients(summary(fit.nb)), confint(fit.nb))
emm <- emmeans(fit.nb, specs="Treatment")
summary(contrast(emm, method="trt.vs.ctrl", adjust="none"), infer=c(TRUE, TRUE))
```

Backtransforming the coefficients (but not the intercept) to the response scale (using $\mathrm{exp}{b_j}$) results in a **response ratio**. 

```{r glm1-worm-nb-coeffs-response, echo=FALSE}
emm <- emmeans(fit.nb, specs="Treatment")
summary(contrast(emm, method="trt.vs.ctrl", adjust="none", type="response"), infer=c(TRUE, TRUE))
```

1. Note how the emmeans package reports the name of the term as the ratio of the coefficient term to the intercept term (the reference treatment level). Why are the coefficients tranformed to ratios on the response scale? Remember that a coefficient is a difference in conditional means and that $\mathrm{exp}(B-A) = \frac{\mathrm{exp}(B)}{\mathrm{exp}(A)}$. For a dummy variable as here (say "Infected HG"), the response ratio is

\begin{equation}
RR_{Infected\_HG} = \frac{\overline{Infected\_HG}}{\overline{Control}}
\end{equation}

which give us the **relative effect** of Infected_HG compared to the Control. Relative effects could be reported as a response ratio in a table, or in the text it could be reported as a percent "Infected HG fish had 71.5% (95%CI: 38.9% - 111.8%) more *Diplostomum* larvae than Control fish." Where do these percents come from? The percent effect is $100(RR_j - 1)$ larger than the reference mean if the $RR_j > 1$ or $100(1 - RR_j)$ smaller than the reference mean if the $RR_j < 1$.

2. Backtransforming the intercept does not generate a ratio since the intercept on the link scale is not a difference. For the worm analysis, the intercept on the link scale is the mean count of the control group on the link scale and the backtransformed intercept is the mean count of the control group on the response scale.

3. Effects on the response scale are not additive but multiplicative! So, for example, the mean of the Infected HG treatment level on the response scale is $\overline{Control}*RR_{Infected\_HG}$ (remember that with a linear model the mean would be $b_{Control} + b_{Infected\_HG}$). Check and see if this works.

```{r glm1-worm-percent, echo=FALSE, eval=FALSE}

emm.response <- data.table(summary(emmeans(fit.nb, specs="Treatment", type="response")))
coef.response <- exp(coef(fit.nb))

# notice that exp(b0) = control mean
# so Uninfected is 27.2% smaller than the control
# this number is 100(1-b1)
100*(1-coef.response[2])
# does it work? This should equal the Uninfected mean
emm.response[Treatment=="Control", response]*coef.response[2]
```

## Working in R

**Fitting a GLM to count data**. The poisson family is specified with the base R `glm()` function. For negative binomial, use `glm.nb` from the MASS package

```{r, eval=FALSE}
# poisson - less likely to fit to real biological data well because of overdispersion
fit <- glm(y ~ treatment, family = "poisson", data = dt)

# two alternatives to overdispersed poisson fit
# quasipoisson
fit <- glm(y ~ treatment, family = "quasipoisson", data=dt)
# negative binomial - more likely to fit to real biological data well
# note that "family" is not an argument since this function is used only to fit a negative binomial distribution!
fit <- glm.nb(y ~ treatment, data = dt)
```

**Fitting a GLM to a continuous conditional response with right skew**. The Gamma family is specified with the base R `glm()` function. 

```{r, eval=FALSE}
fit <- glm(y ~ treatment, family = Gamma(link = "log"), data = dt)
```

**Fitting a GLM to a binary (success or failure, presence or absence, survived or died) response**

The binomial family is specified with base R `glm()` function.

```{r, eval=FALSE}
# if the data includes a 0 or 1 for every observation of y
fit <- glm(y ~ treatment, family = "binomial", data = dt)

# if the data includes the frequency of success AND there is a measure of the total n
dt[ , failure := n - success]
fit <- glm(cbind(success, failure) ~ treatment, family = "binomial", data = dt)
```

**Fitting Generalized Linear Mixed Models**
Generalized linear mixed models are fit with `glmer` from the lmer package.

```{r, eval=FALSE}
# random intercept of factor "id"
fit <- glmer(y ~ treatment + (1|id), family = "poisson", data = dt)

# random intercept and slope of factor "id"
fit <- glmer(y ~ treatment + (treatment|id), family = Gamma(link = "log"), data = dt)

# Again, negative binomial uses a special function
fit <- glmer.nb(y ~ treatment + (treatment|id), data = dt)

```

Another good package for GLMMs is `glmmTMB` from the glmmTMB package
```{r, eval=FALSE}
# negative binomial
fit <- glmmTMB(y ~ treatment + (1|id), family="nbinom2", data = dt)

# nbinom1, the mean variance relationship is that of quasipoisson
fit <- glmmTMB(y ~ treatment + (1|id), family="nbinom1", data = dt)

```


### Fitting a GLM to count data
Source publication: Benesh, D. P., & Kalbe, M. (2016). Experimental parasite community ecology: intraspecific variation in a large tapeworm affects community assembly. Journal of Animal Ecology, 85(4), 1004-1013.

Source data URL: https://datadryad.org/resource/doi:10.5061/dryad.bq8j8

Source file: "Lab_exp.csv"

Poisson fit. A quantile residual Q-Q plot can be generated using the package DHARMa

```{r glm1-working-in-r}
fit.pois <- glm(Diplo_intensity ~ Treatment, family="poisson", data=worm)

# from the DHARMa package
  n_sim <- 250
  simulationOutput <- simulateResiduals(fittedModel = fit.pois, n = n_sim)
  plot(simulationOutput, asFactor = F)

```

A plot of the dispersion statistic can be generated using the object returned by the `SimulateOutput` function but with `refit = TRUE`, which refits a model each iteration. This refitting isn't necessary if only the quantiile residuals are needed. The Dharma package does not divide the sum of squared Pearson residuals by the residual degrees of freedom and so the expected value of the statistic is $df$.

```{r}
# from the DHARMa package
  n_sim <- 250
  simulationOutput <- simulateResiduals(fittedModel=fit.pois, n=n_sim, refit=TRUE)
  testDispersion(simulationOutput)
```


Negative binomial fit.

```{r glm-working-in-r-2}
fit.nb <- glm.nb(Diplo_intensity ~ Treatment, data=worm)
# from the DHARMa package
  simulationOutput <- simulateResiduals(fittedModel = fit.nb, n = n_sim)
  plot(simulationOutput, asFactor = F)

# link scale
emm <- emmeans(fit.nb, specs="Treatment")
emm
summary(contrast(emm, method="trt.vs.ctrl", adjust="none"), infer=c(TRUE, TRUE))

# response scale
emm.response <- emmeans(fit.nb, specs="Treatment", type="response")
summary(contrast(emm, method="trt.vs.ctrl", adjust="none", type="response"), infer=c(TRUE, TRUE))
```

### Fitting a generalized linear mixed model (GLMM) to count data

### Fitting a generalized linear model to continouus data


## Problems

Analyze the data that went into Fig 6B of Tena, A., Pekas, A., Cano, D., Wäckers, F. L., & Urbaneja, A. (2015). Sugar provisioning maximizes the biocontrol service of parasitoids. Journal of Applied Ecology, 52(3), 795-804.

1. Compute contrasts and CIs among all pairs of all three treatment levels
2. Make a better plot like 6b including 1) use the modeled mean instead of the simple group mean and 2) use the modeled CI of the mean instead of the SE computed within each group independently.

source URL: https://datadryad.org/resource/doi:10.5061/dryad.bj001

source file: "4_Parastism_Fig_6.csv"

```{r, echo=FALSE, eval=FALSE}
# nice results

folder <- "Data from Sugar provisioning maximizes the biocontrol service of parasitoids"
file <- "4_Parastism_Fig_6.csv"
file_path <- paste(data_path, folder, file, sep="/")
sugar <- fread(file_path)
sugar[, Treatment:=factor(Treatment, c("Control", "Releases", "Releases_sugar"))]
ggerrorplot(x="Treatment", y="Parasitized", desc_stat="mean_ci", data=sugar)

fit.nb <- glm.nb(Parasitized ~ Treatment, data=sugar)
emm <- summary(emmeans(fit.nb, specs="Treatment", type="response"))
gg <- ggplot(data=emm, aes(x=Treatment, y=response)) +
  geom_point(size=2) +
  geom_errorbar(aes(ymin=asymp.LCL, ymax=asymp.UCL), width=.2)
gg

fit.lm <- lm(Parasitized ~ Treatment, data=sugar)
emmeans(fit.lm, specs="Treatment")
emm

# fake data
niter <- 5000
mu <- predict(fit.nb, type="response")
n <- length(mu)
lm.conf <- matrix(nrow=niter, ncol=2)
glm.conf <- matrix(nrow=niter, ncol=2)

mu1 <- coef(fit.lm)[1]
for(iter in 1:niter){
  fake_count <- rnegbin(n, mu, theta=fit.nb$theta)
  fake.lm <-  lm(fake_count ~ Treatment, data=sugar)
  lm.conf[iter,] <- confint(fake.lm)[1,]
  
  fake.nb <- glm.nb(fake_count ~ Treatment, data=sugar)
  glm.conf[iter,] <- exp(confint(fake.nb)[1,])
}
sum(mu1 > lm.conf[,1] & mu1 < lm.conf[,2])/niter*100  # coverage is 99.1%
sum(mu1 > glm.conf[,1] & mu1 < glm.conf[,2])/niter*100 # coverage is 94.5%

```



<!--chapter:end:chapters/60-glm01-counts.Rmd-->

# Linear models with heterogenous variance {#gls}


<!--chapter:end:chapters/70-gls.Rmd-->

# Part V: Expanding the Linear Model -- Generalized Linear Models and Multilevel (Linear Mixed) Models {-}

<!--chapter:end:chapters/80-supplements.Rmd-->

# Appendix 1: Getting Started with R {-}
## Get your computer ready

### Start here

Watch this video. The links for installing R and R studio are in the next sections.

[Andy Field's Installing R and RStudio](https://www.youtube.com/watch?v=ZvPFKfNHBNQ){target="_blank"}

### Install R
R is the core software. It runs under the hood. You never see it. To use R, you need another piece of software that provides a **user interface**. The software we will use for this is R Studio.

[Download R for your OS](https://cran.r-project.org){target="_blank"}

### Install R Studio
R Studio is a slick (very slick) **graphical user interface** (GUI) for developing R projects.

[Download R Studio Desktop](https://www.rstudio.com/products/rstudio/download/){target="_blank"}

#### Additional resources for installing R and R Studio**

[On Windows](https://medium.com/@GalarnykMichael/install-r-and-rstudio-on-windows-5f503f708027){target="_blank"}

[On a Mac](https://medium.com/@GalarnykMichael/install-r-and-rstudio-on-mac-e911606ce4f4){target="_blank"}

### Install R Markdown

In this class, we will write code to analyze data using R Markdown. R markdown is a version of Markdown. Markdown is tool for creating a document containing text (like microsoft Word), images, tables, and code that can be output to the three modern output formats: html (web pages), pdf (reports and documents), and microsoft word (okay, this isn't modern but it is widely used).

R Markdown can output pdf files. The mechanism for this is to first create a LaTeX ("la-tek") file. LaTeX is an amazing tool for creating professional pdf documents. You do not need PDF output for BIO 414/513. The directions for installing R Markdown include directions for installing LaTeX. This is optional, for this class, but I encourage you to do it.

[Directions for installing R Markdown](https://bookdown.org/yihui/rmarkdown/installation.html#installation
){target="_blank"}

### (optional) Alternative LaTeX installations

[On Windows](https://medium.com/@sorenlind/create-pdf-reports-using-r-r-markdown-latex-and-knitr-on-windows-10-952b0c48bfa9){target="_blank"}

[On a Mac](https://medium.com/@sorenlind/create-pdf-reports-using-r-r-markdown-latex-and-knitr-on-macos-high-sierra-e7b5705c9fd){target="_blank"}

## Start learning R Studio

[R Studio Essentials, Programming Part 1 (Writing code in RStudio)](https://rstudio.com/resources/webinars/programming-part-1-writing-code-in-rstudio/){target="_blank"}

[Getting Started with R Markdown](https://www.rstudio.com/resources/webinars/getting-started-with-r-markdown/){target="_blank"}

Andy Field's [RStudio basics of R Markdown](https://www.youtube.com/watch?v=u4ZdvYXjsIo){target="_blank"}



[Data Visualisation chapter from *R for Data Science*](http://r4ds.had.co.nz/data-visualisation.html)

[Graphics for communication chapter from *R for Data Science*](http://r4ds.had.co.nz/graphics-for-communication.html)

Youtube: [An Introduction to The data.table Package](https://www.youtube.com/watch?v=pc1ARG6kbAM)

Coursera: [The data.table Package](https://www.coursera.org/learn/data-cleaning/lecture/trMZ7/the-data-table-package)


<!--chapter:end:chapters/92-R_resources.Rmd-->

# Appendix 2: Online Resources for Getting Started with Statistical Modeling in R {-}

Roughly, in order from most elementary to most advanced

[Learning Statistics with R](https://https://learningstatisticswithr-bookdown.netlify.com) by Danielle Navarro and adapted to Bookdown (for web viewing) by Emily Kothe.

[Statististical Thinking for the 21st Century](http://statsthinking21.org) by Russell A. Poldrack

[Regression Models for Data Science in R](https://leanpub.com/regmods) by Brian Caffo

[Broadening Your Statistical Horizons: Generalized Linear Models and Multilevel Models by J. Legler and P. Roback](https://bookdown.org/roback/bookdown-bysh/)

[Modern Statistics for Modern Biology](https://www.huber.embl.de/msmb/index.html)

[The Art of Data Science by Roger D. Peng and Elizabeth Matsui](https://bookdown.org/rdpeng/artofdatascience/)

<!--chapter:end:chapters/93-Getting_started_with_linear_modeling.Rmd-->

